<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.43">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>5&nbsp; Session 5: RNA-seq part 1 – WEHI R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./further_reading.html" rel="next">
<link href="./session_4.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ec15196d257545f34024b28906e3f52e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-24312aa3480259c3e15096488a9fa9d6.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session_5.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Session 5: RNA-seq part 1</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">WEHI R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Session 3: Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_5.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Session 5: RNA-seq part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this session:</h2>
   
  <ul>
  <li><a href="#loading-in-read-count-data-and-annotation" id="toc-loading-in-read-count-data-and-annotation" class="nav-link active" data-scroll-target="#loading-in-read-count-data-and-annotation"><span class="header-section-number">5.1</span> Loading in read count data and annotation</a></li>
  <li><a href="#a-brief-detour-to-factors" id="toc-a-brief-detour-to-factors" class="nav-link" data-scroll-target="#a-brief-detour-to-factors"><span class="header-section-number">5.2</span> A brief detour to factors</a></li>
  <li><a href="#creating-the-dgelist-object" id="toc-creating-the-dgelist-object" class="nav-link" data-scroll-target="#creating-the-dgelist-object"><span class="header-section-number">5.3</span> Creating the DGEList object</a></li>
  <li><a href="#alternative-construction-of-the-dgelist-object" id="toc-alternative-construction-of-the-dgelist-object" class="nav-link" data-scroll-target="#alternative-construction-of-the-dgelist-object"><span class="header-section-number">5.4</span> Alternative construction of the DGEList object</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering"><span class="header-section-number">5.5</span> Filtering</a></li>
  <li><a href="#library-size-normalisation" id="toc-library-size-normalisation" class="nav-link" data-scroll-target="#library-size-normalisation"><span class="header-section-number">5.6</span> Library-size normalisation</a></li>
  <li><a href="#visualising-the-effect-of-normalisation" id="toc-visualising-the-effect-of-normalisation" class="nav-link" data-scroll-target="#visualising-the-effect-of-normalisation"><span class="header-section-number">5.7</span> Visualising the effect of normalisation</a></li>
  <li><a href="#mds-plots" id="toc-mds-plots" class="nav-link" data-scroll-target="#mds-plots"><span class="header-section-number">5.8</span> MDS plots</a></li>
  <li><a href="#mds-plot-using-ggplot2" id="toc-mds-plot-using-ggplot2" class="nav-link" data-scroll-target="#mds-plot-using-ggplot2"><span class="header-section-number">5.9</span> MDS plot using ggplot2</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">6</span> Summary</a>
  <ul class="collapse">
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">6.1</span> References</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Session 5: RNA-seq part 1</span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this session we will run through the basic steps for analysing a simple RNA-seq experiment using the <a href="https://f1000research.com/articles/5-1408">limma-voom workflow</a>. This includes:</p>
<ul>
<li>filtering out lowly expressed genes</li>
<li>normalisation</li>
<li>creating a multidimensional scaling (MDS) plot</li>
<li>creating a design matrix</li>
<li>fitting gene-wise linear models (with empirical Bayes moderation to more accurately estimate gene-wise variability)</li>
<li>performing statistical testing for differential expression The aim of this session is to give you experience with a real-world RNA-seq analysis, and making extensive use of an external library. We will not cover the statistics in any depth. The focus is on the practical aspects of using the <code>limma</code> package to perform RNA-seq analysis. The <code>limma</code> package is a widely used package for analysing microarray and RNA-seq data, and it provides a range of functions for normalisation, filtering, statistical testing and visualisation. The aim of this session is to give you experience with a real-world RNA-seq analysis, and making extensive use of an external library. We will not cover the statistics in any depth. Instead, the goal is to understand how to construct data structures required for specific packages and how to use the functions in those packages to perform the analysis.</li>
</ul>
<p>Much of the materials here are explained in greater detail in the limma user’s guide. You can view this by typing <code>help("limma")</code> and following the links.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Constructing a <code>DGEList</code> object to use with the <code>edgeR</code> package</li>
<li>Performing filtering of lowly expressed genes</li>
<li>Normalising RNA-seq data to account for library size and RNA composition</li>
<li>Performing exploratory data analysis using MDS plots</li>
<li>Pulling data out of <code>edgeR</code> package objects and re-organising it into a tidy format for use with <code>ggplot2</code>.</li>
</ul>
</div>
</div>
<section id="loading-in-read-count-data-and-annotation" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="loading-in-read-count-data-and-annotation"><span class="header-section-number">5.1</span> Loading in read count data and annotation</h2>
<p>The data we are looking at comes from three cell populations (basal, luminal progenitor (LP) and mature luminal (ML)) sorted from the mammary glands of female virgin mice, each profiled in triplicate. These samples have been sequenced using Illumina RNAseq platforms and we will be using the <code>edgeR</code> package to analyse the data.</p>
<p>To use the <code>edgeR</code> package we first need to construct a <code>DGEList</code> object. This object contains 3 key pieces of data:</p>
<ul>
<li><code>counts</code>: the main data of this object, a matrix of count values with samples along the columns and features/genes along the rows.</li>
<li><code>samples</code>: a data frame containing annotation for the samples. The rows in this table describe the corresponding column of the counts data.</li>
<li><code>genes</code>: a data frame containing annotation for the genes in the counts matrix. The rows in this table describe the corresponding row in the counts matrix.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="Data objects">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Data objects
</div>
</div>
<div class="callout-body-container callout-body">
<p>Data objects in general are a way to store related pieces of information together. Often they provide features to maintain the relationships between the pieces of information. For example, in a <code>DGEList</code> object, the <code>counts</code> are stored as a matrix while <code>samples</code> and <code>genes</code> are stored as data frames. When you subset the <code>DGEList</code> object, the rows and columns of the <code>counts</code> matrix are subset along with the corresponding rows of the <code>samples</code> and <code>genes</code> data frames. This prevents errors that can occur if the user tries to manually subset all three pieces of information separately.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">parse_factor</span>(</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Basal"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>samplenames <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"10_6_5_11"</span>, <span class="st">"9_6_5_11"</span>, <span class="st">"purep53"</span>, <span class="st">"JMS8-2"</span>, <span class="st">"JMS8-3"</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"JMS8-4"</span>, <span class="st">"JMS8-5"</span>, <span class="st">"JMS9-P7c"</span>, <span class="st">"JMS9-P8c"</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-brief-detour-to-factors" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="a-brief-detour-to-factors"><span class="header-section-number">5.2</span> A brief detour to factors</h2>
<p>Not that when we declared the <code>group</code> variable, we used factors. Factors are a special data type that is used to encode categorical variables. They have multiple useful properties in comparison to regular character vectors: - They allow you to specify the order of the levels. - They are stored as integers but displayed as labels. - They encode all the valid levels of the factor, even if they are not present in the data.</p>
<p>Specifying the order of the levels is useful because it allows you to re-arrange labels when used in plots. In this data we would have the “Basal” group first, followed by “LP” and then “ML”. Using <code>parse_factor()</code> also allows you to check that the values are all valid levels, for example if one of the samples was labelled “Bassal” instead of “Basal”, it would throw an error. You can read the R for Data Science <a href="https://r4ds.had.co.nz/factors.html">chapter on factors</a> for more information, as well as the <a href="https://forcats.tidyverse.org">forcats package</a> for many useful functions for working with factors.</p>
</section>
<section id="creating-the-dgelist-object" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="creating-the-dgelist-object"><span class="header-section-number">5.3</span> Creating the DGEList object</h2>
<p>We will create a <code>DGEList</code> object following the <code>RNAseq123</code> workflow. We use the <code>readDGE()</code> function to read in count data from files, and provide sample information to the <code>readDGE()</code> function and adding in the gene annotation information afterwards.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># vector of file names</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">dir</span>(<span class="at">path =</span> <span class="st">"data/counts"</span>, <span class="at">pattern =</span> <span class="st">"GSM*"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create DGEList object</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">readDGE</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  files,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"data/counts"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">columns =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> group,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> samplenames</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co"># add gene annotation information</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>gene_anno <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/Ses3_geneAnnot.tsv"</span>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>genes <span class="ot">&lt;-</span> gene_anno</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="alternative-construction-of-the-dgelist-object" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="alternative-construction-of-the-dgelist-object"><span class="header-section-number">5.4</span> Alternative construction of the DGEList object</h2>
<p>A more common way to create a <code>DGEList</code> object is to read in the count data as a matrix and then create the <code>DGEList</code> object using the <code>DGEList()</code> function. We can pull apart our existing <code>DGEList</code> object and recreate it using this method.</p>
<div class=".callout-info">
<p>What you can do with data objects is determined by package that the object comes from. You will need to read the package documentation to find out how to access data from the object. Here <code>edgeR</code> informs us that the <code>DGEList</code> has data that can be accessed as if it was a list.</p>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> dge<span class="sc">$</span>counts</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> dge<span class="sc">$</span>genes</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># create DGEList object</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> counts, <span class="at">samples =</span> samples, <span class="at">genes =</span> genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have a data object that can be used for downstream analysis.</p>
</section>
<section id="filtering" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="filtering"><span class="header-section-number">5.5</span> Filtering</h2>
<p>The first step of our analysis is to filter out lowly expressed genes. There are two main problems with low abundant genes:</p>
<ul>
<li>Technical variation is more problematic for low abundance genes. This variation is thought to be due to two factors; insufficient mixing and low sampling fraction<span class="citation" data-cites="mcintyre2011"><sup><a href="#ref-mcintyre2011" role="doc-biblioref">1</a></sup></span>.
<ul>
<li>Insufficient mixing of solutions during library preparation can result in uneven distribution of reads.</li>
<li>RNA sequencing can be thought of as sampling. Measurement errors will occur simply due to the random nature of the sampling process. This problem affects lowly abundant RNA species more because the relative error for small count values is larger than it would be for more highly abundant RNA species.</li>
</ul></li>
<li>Genes that are very lowly expressed do not produce sufficient information to be useful for biological interpretation. For example, it is very hard to believe the biological significance of genes that have counts ranging from 0 to 3 across samples even if come up as statistically significant.</li>
</ul>
<p>Removing these highly variable, lowly expressed genes increases your ‘power’ to detect differentially expressed genes<span class="citation" data-cites="bourgon2010"><sup><a href="#ref-bourgon2010" role="doc-biblioref">2</a></sup></span>, where ‘power’ is your ability to detect true positives. In testing for differential expression, a statistical test is conducted for each gene. When a high number of statistical tests are performed, a portion of them will be significant purely due to random chance. A common procedure to control for the number of false positive is to perform ‘multiple testing correction’ on the p-values. This adjusts the p-value in a way that reduces the number of false positives but comes at the cost of reduced power to detect true positives. If we filter out uninteresting, lowly expressed genes, we need to perform fewer statistical tests and reduce the impact that multiple testing adjustment has on detection power.</p>
<p>The <code>edgeR</code> provides the <code>filterByExpr()</code> function to automate gene filtering. By default, it aims to keep genes with a count of 10 or more, in at least as many samples as the smallest experimental group. In our experiment, there are 3 phenotype groups each with 3 samples. Therefore we retain only genes that have 10 or more counts in 3 or more samples. The actual filtering is done on counts per million, prevent bias against samples with small library sizes. This complex procedure is the reason why the package provides a function to perform the filtering for you.</p>
<p>The output of this function is a vector of logicals, indicating which genes (rows) should be kept and which filtered.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(keep)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>keep
FALSE  TRUE 
10555 16624 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">proportions</span>(<span class="fu">table</span>(keep))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>keep
    FALSE      TRUE 
0.3883513 0.6116487 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep, , keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 16624     9</code></pre>
</div>
</div>
<p>We can see that we now have 16624 genes. We started with 27179 genes - meaning that ~40% of genes have been filtered out.</p>
</section>
<section id="library-size-normalisation" class="level2" data-number="5.6">
<h2 data-number="5.6" class="anchored" data-anchor-id="library-size-normalisation"><span class="header-section-number">5.6</span> Library-size normalisation</h2>
<p>After filtering, our next step is to normalise the data. Normalisation refers to the process of adjusting the data to reduce or eliminate systematic bias. This allows the data to be meaningfully compared across samples or experimental groups.</p>
<p>There are two main factors that need to be normalised for in RNA-seq:</p>
<ul>
<li>Sequencing depth/library size - technically, sequencing a sample to half the depth will give, on average, half the number of reads mapping to each gene<span class="citation" data-cites="robinson2010"><sup><a href="#ref-robinson2010" role="doc-biblioref">3</a></sup></span>.</li>
<li>RNA composition - if a large number of genes are unique to, or highly expressed in, only one experimental condition, the sequencing capacity available for the remaining genes in that sample is decreased. For example, if there are only five genes being studied in two experimental groups, if one gene is particularly high in group A, then with limited sequencing depth, that gene will reduce the counts of the remaining four genes. The effect of this is that the remaining four genes appear under-expressed in group A compared to group B when the true amount of gene product is actually equal for these 4 genes<span class="citation" data-cites="robinson2010"><sup><a href="#ref-robinson2010" role="doc-biblioref">3</a></sup></span>.</li>
</ul>
<p>Sequencing depth is accounted for by calculating the counts per million (cpm). This metric is calculated by:</p>
<ol type="1">
<li>taking the library size (sum of all counts for a sample),</li>
<li>dividing this by 1,000,000 to get the ‘per million’ scaling factor,</li>
<li>then dividing all read counts for each gene in that sample by the ‘per million’ scaling factor</li>
</ol>
<p>RNA composition can be accounted for by using more sophisticated normalisation methodologies. We will use ‘trimmed mean of M-values’ (TMM), which estimates relative RNA levels from RNA-seq data<span class="citation" data-cites="robinson2010"><sup><a href="#ref-robinson2010" role="doc-biblioref">3</a></sup></span>. Under the assumption that most genes are not differentially expressed, TMM calculates a library size scaling factor for each library (sample). This is done using the following steps:</p>
<ol type="1">
<li>calculate the gene expression log fold changes and absolute expression values for pair-wise samples (selecting one sample from the experiment as a reference)</li>
<li>remove the genes with the highest and lowest fold changes and absolute expression values</li>
<li>take a weighted mean of the remaining genes (where the weight is the inverse of the approximate asymptotic variances). This gives the normalisation factor for each library (sample)</li>
</ol>
<p>Subsequent steps in this analysis will use log-cpm values, calculated using the normalisation factors, which scales each library size.</p>
<p>We can calculate the normalisation factors, specifying that we want to use the <code>"TMM"</code> method:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function calculates the normalisation factors for each library (sample) and puts this information in the <code>samples</code> data frame. Note that it takes dge (our <code>DGEList</code> object as input) and returns a <code>DGEList</code> object as well.</p>
<p>Let’s take a look at our normalisation factors:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          group lib.size norm.factors                    files lib.size.1
10_6_5_11    LP 32857304    0.8943956 GSM1545535_10_6_5_11.txt   32863052
9_6_5_11     ML 35328624    1.0250186  GSM1545536_9_6_5_11.txt   35335491
purep53   Basal 57147943    1.0459005   GSM1545538_purep53.txt   57160817
JMS8-2    Basal 51356800    1.0458455    GSM1545539_JMS8-2.txt   51368625
JMS8-3       ML 75782871    1.0162707    GSM1545540_JMS8-3.txt   75795034
JMS8-4       LP 60506774    0.9217132    GSM1545541_JMS8-4.txt   60517657
JMS8-5    Basal 55073018    0.9961959    GSM1545542_JMS8-5.txt   55086324
JMS9-P7c     ML 21305254    1.0861026  GSM1545544_JMS9-P7c.txt   21311068
JMS9-P8c     LP 19955335    0.9839203  GSM1545545_JMS9-P8c.txt   19958838
          norm.factors.1
10_6_5_11              1
9_6_5_11               1
purep53                1
JMS8-2                 1
JMS8-3                 1
JMS8-4                 1
JMS8-5                 1
JMS9-P7c               1
JMS9-P8c               1</code></pre>
</div>
</div>
<p>These normalisation factors are all close to 1 for all samples, suggesting minimal difference in RNA composition between samples.</p>
</section>
<section id="visualising-the-effect-of-normalisation" class="level2" data-number="5.7">
<h2 data-number="5.7" class="anchored" data-anchor-id="visualising-the-effect-of-normalisation"><span class="header-section-number">5.7</span> Visualising the effect of normalisation</h2>
<p>To visualise the effect of TMM normalisation, we can plot the log-counts as a boxplot, and observe the effect of applying the normalisation. To create a boxplot of the log-counts, we can use <code>log(dge$counts + 0.5)</code> to create the log-count matrix. The addition of 0.5 is to avoid taking the log of zero. Then in order to use <code>ggplot2</code> for plotting, we must convert the matrix to a data frame. We can use the <code>as_tibble(rownames = "gene")</code> function to convert the matrix to a data frame where the rownames are converted to a column called “gene”. We can then use the <code>pivot_longer()</code> function to convert the data frame from wide format to long format, where each row represents a single observation. This is necessary for <code>ggplot2</code> to plot the data correctly.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">log</span>(dge<span class="sc">$</span>counts <span class="sc">+</span> <span class="fl">0.5</span>), <span class="at">rownames =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We can compare this to the cpm values which when calculated using the <code>cpm()</code> function, automatically applies normalisation factors if they are present.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">cpm</span>(dge<span class="sc">$</span>counts, <span class="at">log =</span> <span class="cn">TRUE</span>), <span class="at">rownames =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric),</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We see that by performing normalisation on our data, the gene expression values of each sample now have similar medians and quantiles. This indicates that the relative expression values of each sample can be more meaningfully compared.</p>
</section>
<section id="mds-plots" class="level2" data-number="5.8">
<h2 data-number="5.8" class="anchored" data-anchor-id="mds-plots"><span class="header-section-number">5.8</span> MDS plots</h2>
<p>Before we perform statistical tests, it’s useful to perform some exploratory visual analysis to get an overall idea of how our data is behaving.</p>
<p>MDS is a way to visualise distances between sets of data points (samples in our case). It is a dimensionality reduction technique, similar to principal components analysis (PCA). We treat gene expression in samples as if they were coordinates in a high-dimensional coordinate system, then we can find “distances” between samples as we do between points in space. Then the goal of the algorithm is to find a representation in lower dimensional space such that points that the distance of two objects from each other in high dimensional space is preserved in lower dimensions.</p>
<p>The <code>plotMDS()</code> from <code>limma</code> creates an MDS plot from a <code>DGEList</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Each point on the plot represents one sample and is ‘labelled’ using the sample name. The distances between each sample in the resulting plot can be interpreted as the typical log2-fold-change between the samples, for the most differentially expressed genes.</p>
<p>We can change the labelling to use the name of the group the sample belongs to instead:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge, <span class="at">labels =</span> group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>This shows us that the phenotype groups tend to cluster together, meaning that the gene expression profiles are similar for samples within a phenotype group. The ‘Basal’ type samples quite close together while the ‘LP’ (luminal progenitor) and ‘ML’ (mature luminal) type samples are further apart, signifying that their expression profiles are more variable.</p>
</section>
<section id="mds-plot-using-ggplot2" class="level2" data-number="5.9">
<h2 data-number="5.9" class="anchored" data-anchor-id="mds-plot-using-ggplot2"><span class="header-section-number">5.9</span> MDS plot using ggplot2</h2>
<p>For more customisability and better consistency with the style of other plots, it’d be nice to be able to draw the MDS plot using <code>ggplot2</code>. In order to do this we would need the MDS coordinates calculated by <code>plotMDS()</code>. Luckily the documentation of <code>plotMDS()</code> shows that it returns multiple computed values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>mds_result <span class="ot">&lt;-</span> <span class="fu">plotMDS</span>(dge, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>mds_result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class MDS
$eigen.values
[1]  6.844419e+01  1.154413e+01  3.790044e+00  2.994025e+00  2.021565e+00
[6]  1.613173e+00  1.202805e+00  1.094072e+00 -7.018692e-16

$eigen.vectors
            [,1]        [,2]        [,3]         [,4]        [,5]        [,6]
 [1,] -0.2461859  0.54036454  0.12401524  0.677678569  0.20623170  0.07294476
 [2,] -0.2851986 -0.33460855  0.67471878  0.007007206 -0.46099085  0.15319914
 [3,]  0.4690174 -0.04145263  0.05041410  0.056945694 -0.03695319 -0.03494949
 [4,]  0.4711165 -0.04907588  0.05443024  0.161230059 -0.03439028 -0.55065413
 [5,] -0.2626081 -0.18747095  0.18504793 -0.355202862  0.66319443 -0.33464804
 [6,] -0.1568565  0.41423458 -0.09091514 -0.468642654  0.04620796  0.26492304
 [7,]  0.4670465  0.01138476 -0.03115306 -0.161357159  0.07822632  0.56941932
 [8,] -0.2405667 -0.57831836 -0.56474831  0.302172046  0.07568690  0.20422089
 [9,] -0.2157647  0.22494250 -0.40180977 -0.219830899 -0.53721300 -0.34445549
             [,7]         [,8]       [,9]
 [1,] -0.11742920 -0.004177629 -0.3333333
 [2,]  0.04463732 -0.048147665 -0.3333333
 [3,]  0.03859114  0.810760474 -0.3333333
 [4,]  0.33482254 -0.468041790 -0.3333333
 [5,] -0.26519650  0.047152480 -0.3333333
 [6,]  0.62647876 -0.002850499 -0.3333333
 [7,] -0.44130343 -0.344258233 -0.3333333
 [8,]  0.19689394 -0.010674134 -0.3333333
 [9,] -0.41749458  0.020236996 -0.3333333

$var.explained
[1] 0.73830888 0.12452679 0.04088328 0.03229660 0.02180666 0.01740133 0.01297468
[8] 0.01180178 0.00000000

$dim.plot
[1] 1 2

$distance.matrix.squared
           10_6_5_11   9_6_5_11   purep53    JMS8-2     JMS8-3     JMS8-4
10_6_5_11 -18.127013  -5.738791  16.10192  16.03204  -5.792252 -8.3908112
9_6_5_11   -5.738791 -18.115187  17.75982  17.85109 -11.197198 -2.5513014
purep53    16.101924  17.759816 -31.64208 -29.63783  16.733543 10.6453322
JMS8-2     16.032036  17.851094 -29.63783 -32.34830  16.749432 11.0446228
JMS8-3     -5.792252 -11.197198  16.73354  16.74943 -13.580395 -4.1528811
JMS8-4     -8.390811  -2.551301  10.64533  11.04462  -4.152881 -9.8867331
JMS8-5     15.954448  18.363197 -29.18038 -28.91308  16.698190  9.6067888
JMS9-P7c   -1.642914 -10.966083  15.03939  15.00405  -9.571938  0.3388923
JMS9-P8c   -8.396628  -5.405548  14.18030  14.21798  -5.886502 -6.6539093
              JMS8-5    JMS9-P7c   JMS9-P8c
10_6_5_11  15.954448  -1.6429143  -8.396628
9_6_5_11   18.363197 -10.9660831  -5.405548
purep53   -29.180384  15.0393893  14.180296
JMS8-2    -28.913082  15.0040512  14.217981
JMS8-3     16.698190  -9.5719377  -5.886502
JMS8-4      9.606789   0.3388923  -6.653909
JMS8-5    -31.824715  15.4926837  13.802874
JMS9-P7c   15.492684 -18.8595492  -4.834532
JMS9-P8c   13.802874  -4.8345322 -11.024033

$top
[1] 500

$gene.selection
[1] "pairwise"

$axislabel
[1] "Leading logFC dim"

$x
[1] -2.036721 -2.359477  3.880228  3.897594 -2.172583 -1.297689  3.863923
[8] -1.990232 -1.785043

$y
[1]  1.83597803 -1.13688798 -0.14084219 -0.16674343 -0.63696359  1.40743060
[7]  0.03868159 -1.96493242  0.76427939</code></pre>
</div>
</div>
<p>We see that there’s are <code>x</code> and <code>y</code> values in the list returned by the <code>plotMDS()</code> function, we can put these into a table and see if they match up to the positions plotted by the function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>mds_tibble <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> mds_result<span class="sc">$</span>x,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> mds_result<span class="sc">$</span>y</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>The positions of the points and the range of the scales seem to match, so we want to add in more metadata to help use plot. We can try to recreate the MDS plot using just <code>ggplot2</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mds_tibble <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> mds_result<span class="sc">$</span>x,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> mds_result<span class="sc">$</span>y,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colnames</span>(dge),</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>group</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>dim1_var_explained <span class="ot">&lt;-</span> <span class="fu">round</span>(mds_result<span class="sc">$</span>var.explained[<span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>dim2_var_explained <span class="ot">&lt;-</span> <span class="fu">round</span>(mds_result<span class="sc">$</span>var.explained[<span class="dv">2</span>] <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> group)) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Now that we have our data in a nice tidy format we can use with ggplot, it’s easy to create variations of the plot. For example we can draw points instead of group labels, and use colour to identify the groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively we can also use the labels to identify the individual samples while colouring them by their group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample)) <span class="sc">+</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We see that some labels are very hard to read due to the overlapping, so we can use the <code>ggrepel</code> package to fix this. The <code>ggrepel</code> package creates labels that repel each other in order to avoid overlap. It’s a good idea to use it in conjunction with <code>geom_point()</code> in order to keep track of exact coordinate of the data point.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample)) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_5_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="summary" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Summary</h1>
<p>Today we started on the early steps of the RNA-seq analysis workflow.</p>
<ul>
<li>We learned how to create a <code>DGEList</code> object, demonstrating how to create data objects required by specific packages.</li>
<li>We learned how to use edgeR’s <code>filterByExpr()</code> function to filter out lowly expressed genes.</li>
<li>We learned how to normalise the data using TMM normalisation.</li>
<li>We learned how to create MDS plots using both <code>plotMDS()</code> and <code>ggplot2</code>.</li>
<li>We learned how to use <code>ggrepel</code> to create non-overlapping labels in <code>ggplot2</code> plots.</li>
</ul>
<p>This provides a good foundation for organising data to satisfy the requirements of specific packages, and how to pull data out of the objects created by those packages. The data we pulled out can then be organised into a tidy format to leverage the power of all the <code>tidyverse</code> functions that we have learned so far.</p>
<p>In the next session we will continue the RNA-seq analysis workflow and complete our differential expression analysis.</p>
<section id="references" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="references"><span class="header-section-number">6.1</span> References</h2>


<!-- -->

<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-mcintyre2011" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">McIntyre LM, Lopiano KK, Morse AM, Amin V, Oberg AL, Young LJ, et al. RNA-seq: technical variability and sampling. BMC Genomics [Internet]. 2011 Jun 6;12(1). Available from: <a href="http://dx.doi.org/10.1186/1471-2164-12-293">http://dx.doi.org/10.1186/1471-2164-12-293</a></div>
</div>
<div id="ref-bourgon2010" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Bourgon R, Gentleman R, Huber W. Independent filtering increases detection power for high-throughput experiments. Proceedings of the National Academy of Sciences [Internet]. 2010 May 11;107(21):9546–51. Available from: <a href="http://dx.doi.org/10.1073/pnas.0914005107">http://dx.doi.org/10.1073/pnas.0914005107</a></div>
</div>
<div id="ref-robinson2010" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Robinson MD, Oshlack A. A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology [Internet]. 2010 Mar 2;11(3). Available from: <a href="http://dx.doi.org/10.1186/gb-2010-11-3-r25">http://dx.doi.org/10.1186/gb-2010-11-3-r25</a></div>
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session_4.html" class="pagination-link" aria-label="Session 4: Putting it all together">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./further_reading.html" class="pagination-link" aria-label="Further reading">
        <span class="nav-page-text">Further reading</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb24" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this session:"</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session 5: RNA-seq part 1</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, include=FALSE}</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="in">library(magrittr)</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="in">library(knitr)</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="in">library(kableExtra)</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="in">options(readr.show_col_types = FALSE)</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>In this session we will run through the basic steps for analysing a simple RNA-seq experiment using the <span class="co">[</span><span class="ot">limma-voom workflow</span><span class="co">](https://f1000research.com/articles/5-1408)</span>. This includes:</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>filtering out lowly expressed genes</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>normalisation</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>creating a multidimensional scaling (MDS) plot</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>creating a design matrix</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>fitting gene-wise linear models (with empirical Bayes moderation to more accurately estimate gene-wise variability)</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>performing statistical testing for differential expression</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>The aim of this session is to give you experience with a real-world RNA-seq analysis, and making extensive use of an external library. We will not cover the statistics in any depth. The focus is on the practical aspects of using the <span class="in">`limma`</span> package to perform RNA-seq analysis. The <span class="in">`limma`</span> package is a widely used package for analysing microarray and RNA-seq data, and it provides a range of functions for normalisation, filtering, statistical testing and visualisation.</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>The aim of this session is to give you experience with a real-world RNA-seq analysis, and making extensive use of an external library. We will not cover the statistics in any depth. Instead, the goal is to understand how to construct data structures required for specific packages and how to use the functions in those packages to perform the analysis.</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>Much of the materials here are explained in greater detail in the limma user's guide. You can view this by typing <span class="in">`help("limma")`</span> and following the links.</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Constructing a <span class="in">`DGEList`</span> object to use with the <span class="in">`edgeR`</span> package</span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Performing filtering of lowly expressed genes</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Normalising RNA-seq data to account for library size and RNA composition</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Performing exploratory data analysis using MDS plots</span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Pulling data out of <span class="in">`edgeR`</span> package objects and re-organising it into a tidy format for use with <span class="in">`ggplot2`</span>.</span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a><span class="fu">## Loading in read count data and annotation</span></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a>The data we are looking at comes from three cell populations (basal, luminal progenitor (LP) and mature luminal (ML)) sorted from the mammary glands of female virgin mice, each profiled in triplicate. These samples have been sequenced using Illumina RNAseq platforms and we will be using the <span class="in">`edgeR`</span> package to analyse the data.</span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a>To use the <span class="in">`edgeR`</span> package we first need to construct a <span class="in">`DGEList`</span> object. This object contains 3 key pieces of data:</span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`counts`</span>: the main data of this object, a matrix of count values with samples along the columns and features/genes along the rows.</span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`samples`</span>: a data frame containing annotation for the samples. The rows in this table describe the corresponding column of the counts data.</span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`genes`</span>: a data frame containing annotation for the genes in the counts matrix. The rows in this table describe the corresponding row in the counts matrix.</span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Data objects"}</span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a>Data objects in general are a way to store related pieces of information together. Often they provide features to maintain the relationships between the pieces of information. For example, in a <span class="in">`DGEList`</span> object, the <span class="in">`counts`</span> are stored as a matrix while <span class="in">`samples`</span> and <span class="in">`genes`</span> are stored as data frames. When you subset the <span class="in">`DGEList`</span> object, the rows and columns of the <span class="in">`counts`</span> matrix are subset along with the corresponding rows of the <span class="in">`samples`</span> and <span class="in">`genes`</span> data frames. This prevents errors that can occur if the user tries to manually subset all three pieces of information separately.</span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, warning=FALSE, message=FALSE}</span></span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a><span class="in"># load required packages</span></span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a><span class="in">library(tidyverse)</span></span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a><span class="in">library(edgeR)</span></span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a><span class="in">group &lt;- parse_factor(</span></span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a><span class="in">  c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"),</span></span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a><span class="in">  levels = c("Basal", "LP", "ML")</span></span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb24-66"><a href="#cb24-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-67"><a href="#cb24-67" aria-hidden="true" tabindex="-1"></a><span class="in">samplenames &lt;- c(</span></span>
<span id="cb24-68"><a href="#cb24-68" aria-hidden="true" tabindex="-1"></a><span class="in">  "10_6_5_11", "9_6_5_11", "purep53", "JMS8-2", "JMS8-3",</span></span>
<span id="cb24-69"><a href="#cb24-69" aria-hidden="true" tabindex="-1"></a><span class="in">  "JMS8-4", "JMS8-5", "JMS9-P7c", "JMS9-P8c"</span></span>
<span id="cb24-70"><a href="#cb24-70" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb24-71"><a href="#cb24-71" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-72"><a href="#cb24-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-73"><a href="#cb24-73" aria-hidden="true" tabindex="-1"></a><span class="fu">## A brief detour to factors</span></span>
<span id="cb24-74"><a href="#cb24-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-75"><a href="#cb24-75" aria-hidden="true" tabindex="-1"></a>Not that when we declared the <span class="in">`group`</span> variable, we used factors. Factors are a special data type that is used to encode categorical variables. They have multiple useful properties in comparison to regular character vectors:</span>
<span id="cb24-76"><a href="#cb24-76" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>They allow you to specify the order of the levels.</span>
<span id="cb24-77"><a href="#cb24-77" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>They are stored as integers but displayed as labels.</span>
<span id="cb24-78"><a href="#cb24-78" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>They encode all the valid levels of the factor, even if they are not present in the data.</span>
<span id="cb24-79"><a href="#cb24-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-80"><a href="#cb24-80" aria-hidden="true" tabindex="-1"></a>Specifying the order of the levels is useful because it allows you to re-arrange labels when used in plots. In this data we would have the "Basal" group first, followed by "LP" and then "ML". Using <span class="in">`parse_factor()`</span> also allows you to check that the values are all valid levels, for example if one of the samples was labelled "Bassal" instead of "Basal", it would throw an error. You can read the R for Data Science <span class="co">[</span><span class="ot">chapter on factors</span><span class="co">](https://r4ds.had.co.nz/factors.html)</span> for more information, as well as the <span class="co">[</span><span class="ot">forcats package</span><span class="co">](https://forcats.tidyverse.org)</span> for many useful functions for working with factors.</span>
<span id="cb24-81"><a href="#cb24-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-82"><a href="#cb24-82" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating the DGEList object</span></span>
<span id="cb24-83"><a href="#cb24-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-84"><a href="#cb24-84" aria-hidden="true" tabindex="-1"></a>We will create a <span class="in">`DGEList`</span> object following the <span class="in">`RNAseq123`</span> workflow. We use the <span class="in">`readDGE()`</span> function to read in count data from files, and provide sample information to the <span class="in">`readDGE()`</span> function and adding in the gene annotation information afterwards.</span>
<span id="cb24-85"><a href="#cb24-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-86"><a href="#cb24-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, warning=FALSE, message=FALSE}</span></span>
<span id="cb24-87"><a href="#cb24-87" aria-hidden="true" tabindex="-1"></a><span class="in"># vector of file names</span></span>
<span id="cb24-88"><a href="#cb24-88" aria-hidden="true" tabindex="-1"></a><span class="in">files &lt;- dir(path = "data/counts", pattern = "GSM*")</span></span>
<span id="cb24-89"><a href="#cb24-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-90"><a href="#cb24-90" aria-hidden="true" tabindex="-1"></a><span class="in"># create DGEList object</span></span>
<span id="cb24-91"><a href="#cb24-91" aria-hidden="true" tabindex="-1"></a><span class="in">dge &lt;- readDGE(</span></span>
<span id="cb24-92"><a href="#cb24-92" aria-hidden="true" tabindex="-1"></a><span class="in">  files,</span></span>
<span id="cb24-93"><a href="#cb24-93" aria-hidden="true" tabindex="-1"></a><span class="in">  path = "data/counts",</span></span>
<span id="cb24-94"><a href="#cb24-94" aria-hidden="true" tabindex="-1"></a><span class="in">  columns = c(1, 3),</span></span>
<span id="cb24-95"><a href="#cb24-95" aria-hidden="true" tabindex="-1"></a><span class="in">  group = group,</span></span>
<span id="cb24-96"><a href="#cb24-96" aria-hidden="true" tabindex="-1"></a><span class="in">  labels = samplenames</span></span>
<span id="cb24-97"><a href="#cb24-97" aria-hidden="true" tabindex="-1"></a><span class="in">)</span></span>
<span id="cb24-98"><a href="#cb24-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-99"><a href="#cb24-99" aria-hidden="true" tabindex="-1"></a><span class="in"># add gene annotation information</span></span>
<span id="cb24-100"><a href="#cb24-100" aria-hidden="true" tabindex="-1"></a><span class="in">gene_anno &lt;- read_tsv("data/Ses3_geneAnnot.tsv")</span></span>
<span id="cb24-101"><a href="#cb24-101" aria-hidden="true" tabindex="-1"></a><span class="in">dge$genes &lt;- gene_anno</span></span>
<span id="cb24-102"><a href="#cb24-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-103"><a href="#cb24-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-104"><a href="#cb24-104" aria-hidden="true" tabindex="-1"></a><span class="fu">## Alternative construction of the DGEList object</span></span>
<span id="cb24-105"><a href="#cb24-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-106"><a href="#cb24-106" aria-hidden="true" tabindex="-1"></a>A more common way to create a <span class="in">`DGEList`</span> object is to read in the count data as a matrix and then create the <span class="in">`DGEList`</span> object using the <span class="in">`DGEList()`</span> function. We can pull apart our existing <span class="in">`DGEList`</span> object and recreate it using this method.</span>
<span id="cb24-107"><a href="#cb24-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-108"><a href="#cb24-108" aria-hidden="true" tabindex="-1"></a>::: .callout-info</span>
<span id="cb24-109"><a href="#cb24-109" aria-hidden="true" tabindex="-1"></a>What you can do with data objects is determined by package that the object comes from. You will need to read the package documentation to find out how to access data from the object. Here <span class="in">`edgeR`</span> informs us that the <span class="in">`DGEList`</span> has data that can be accessed as if it was a list.</span>
<span id="cb24-110"><a href="#cb24-110" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb24-111"><a href="#cb24-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-114"><a href="#cb24-114" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-115"><a href="#cb24-115" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> dge<span class="sc">$</span>counts</span>
<span id="cb24-116"><a href="#cb24-116" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples</span>
<span id="cb24-117"><a href="#cb24-117" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> dge<span class="sc">$</span>genes</span>
<span id="cb24-118"><a href="#cb24-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-119"><a href="#cb24-119" aria-hidden="true" tabindex="-1"></a><span class="co"># create DGEList object</span></span>
<span id="cb24-120"><a href="#cb24-120" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> counts, <span class="at">samples =</span> samples, <span class="at">genes =</span> genes)</span>
<span id="cb24-121"><a href="#cb24-121" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-122"><a href="#cb24-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-123"><a href="#cb24-123" aria-hidden="true" tabindex="-1"></a>Now we have a data object that can be used for downstream analysis.</span>
<span id="cb24-124"><a href="#cb24-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-125"><a href="#cb24-125" aria-hidden="true" tabindex="-1"></a><span class="fu">## Filtering</span></span>
<span id="cb24-126"><a href="#cb24-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-127"><a href="#cb24-127" aria-hidden="true" tabindex="-1"></a>The first step of our analysis is to filter out lowly expressed genes. There are two main problems with low abundant genes:</span>
<span id="cb24-128"><a href="#cb24-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-129"><a href="#cb24-129" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Technical variation is more problematic for low abundance genes. This variation is thought to be due to two factors; insufficient mixing and low sampling fraction <span class="co">[</span><span class="ot">@mcintyre2011</span><span class="co">]</span>.</span>
<span id="cb24-130"><a href="#cb24-130" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Insufficient mixing of solutions during library preparation can result in uneven distribution of reads.</span>
<span id="cb24-131"><a href="#cb24-131" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>RNA sequencing can be thought of as sampling. Measurement errors will occur simply due to the random nature of the sampling process. This problem affects lowly abundant RNA species more because the relative error for small count values is larger than it would be for more highly abundant RNA species.</span>
<span id="cb24-132"><a href="#cb24-132" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Genes that are very lowly expressed do not produce sufficient information to be useful for biological interpretation. For example, it is very hard to believe the biological significance of genes that have counts ranging from 0 to 3 across samples even if come up as statistically significant.</span>
<span id="cb24-133"><a href="#cb24-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-134"><a href="#cb24-134" aria-hidden="true" tabindex="-1"></a>Removing these highly variable, lowly expressed genes increases your 'power' to detect differentially expressed genes <span class="co">[</span><span class="ot">@bourgon2010</span><span class="co">]</span>, where 'power' is your ability to detect true positives. In testing for differential expression, a statistical test is conducted for each gene. When a high number of statistical tests are performed, a portion of them will be significant purely due to random chance. A common procedure to control for the number of false positive is to perform 'multiple testing correction' on the p-values. This adjusts the p-value in a way that reduces the number of false positives but comes at the cost of reduced power to detect true positives. If we filter out uninteresting, lowly expressed genes, we need to perform fewer statistical tests and reduce the impact that multiple testing adjustment has on detection power.</span>
<span id="cb24-135"><a href="#cb24-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-136"><a href="#cb24-136" aria-hidden="true" tabindex="-1"></a>The <span class="in">`edgeR`</span> provides the <span class="in">`filterByExpr()`</span> function to automate gene filtering. By default, it aims to keep genes with a count of 10 or more, in at least as many samples as the smallest experimental group. In our experiment, there are 3 phenotype groups each with 3 samples. Therefore we retain only genes that have 10 or more counts in 3 or more samples. The actual filtering is done on counts per million, prevent bias against samples with small library sizes. This complex procedure is the reason why the package provides a function to perform the filtering for you.</span>
<span id="cb24-137"><a href="#cb24-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-138"><a href="#cb24-138" aria-hidden="true" tabindex="-1"></a>The output of this function is a vector of logicals, indicating which genes (rows) should be kept and which filtered.</span>
<span id="cb24-139"><a href="#cb24-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-142"><a href="#cb24-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-143"><a href="#cb24-143" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge)</span>
<span id="cb24-144"><a href="#cb24-144" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(keep)</span>
<span id="cb24-145"><a href="#cb24-145" aria-hidden="true" tabindex="-1"></a><span class="fu">proportions</span>(<span class="fu">table</span>(keep))</span>
<span id="cb24-146"><a href="#cb24-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-147"><a href="#cb24-147" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep, , keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb24-148"><a href="#cb24-148" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge<span class="sc">$</span>counts)</span>
<span id="cb24-149"><a href="#cb24-149" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-150"><a href="#cb24-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-151"><a href="#cb24-151" aria-hidden="true" tabindex="-1"></a>We can see that we now have 16624 genes. We started with 27179 genes - meaning that ~40% of genes have been filtered out.</span>
<span id="cb24-152"><a href="#cb24-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-153"><a href="#cb24-153" aria-hidden="true" tabindex="-1"></a><span class="fu">## Library-size normalisation</span></span>
<span id="cb24-154"><a href="#cb24-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-155"><a href="#cb24-155" aria-hidden="true" tabindex="-1"></a>After filtering, our next step is to normalise the data. Normalisation refers to the process of adjusting the data to reduce or eliminate systematic bias. This allows the data to be meaningfully compared across samples or experimental groups.</span>
<span id="cb24-156"><a href="#cb24-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-157"><a href="#cb24-157" aria-hidden="true" tabindex="-1"></a>There are two main factors that need to be normalised for in RNA-seq:</span>
<span id="cb24-158"><a href="#cb24-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-159"><a href="#cb24-159" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Sequencing depth/library size - technically, sequencing a sample to half the depth will give, on average, half the number of reads mapping to each gene <span class="co">[</span><span class="ot">@robinson2010</span><span class="co">]</span>.</span>
<span id="cb24-160"><a href="#cb24-160" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>RNA composition - if a large number of genes are unique to, or highly expressed in, only one experimental condition, the sequencing capacity available for the remaining genes in that sample is decreased. For example, if there are only five genes being studied in two experimental groups, if one gene is particularly high in group A, then with limited sequencing depth, that gene will reduce the counts of the remaining four genes. The effect of this is that the remaining four genes appear under-expressed in group A compared to group B when the true amount of gene product is actually equal for these 4 genes <span class="co">[</span><span class="ot">@robinson2010</span><span class="co">]</span>.</span>
<span id="cb24-161"><a href="#cb24-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-162"><a href="#cb24-162" aria-hidden="true" tabindex="-1"></a>Sequencing depth is accounted for by calculating the counts per million (cpm). This metric is calculated by:</span>
<span id="cb24-163"><a href="#cb24-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-164"><a href="#cb24-164" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>taking the library size (sum of all counts for a sample),</span>
<span id="cb24-165"><a href="#cb24-165" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>dividing this by 1,000,000 to get the 'per million' scaling factor,</span>
<span id="cb24-166"><a href="#cb24-166" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>then dividing all read counts for each gene in that sample by the 'per million' scaling factor</span>
<span id="cb24-167"><a href="#cb24-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-168"><a href="#cb24-168" aria-hidden="true" tabindex="-1"></a>RNA composition can be accounted for by using more sophisticated normalisation methodologies. We will use 'trimmed mean of M-values' (TMM), which estimates relative RNA levels from RNA-seq data <span class="co">[</span><span class="ot">@robinson2010</span><span class="co">]</span>. Under the assumption that most genes are not differentially expressed, TMM calculates a library size scaling factor for each library (sample). This is done using the following steps:</span>
<span id="cb24-169"><a href="#cb24-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-170"><a href="#cb24-170" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>calculate the gene expression log fold changes and absolute expression values for pair-wise samples (selecting one sample from the experiment as a reference)</span>
<span id="cb24-171"><a href="#cb24-171" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>remove the genes with the highest and lowest fold changes and absolute expression values</span>
<span id="cb24-172"><a href="#cb24-172" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>take a weighted mean of the remaining genes (where the weight is the inverse of the approximate asymptotic variances). This gives the normalisation factor for each library (sample)</span>
<span id="cb24-173"><a href="#cb24-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-174"><a href="#cb24-174" aria-hidden="true" tabindex="-1"></a>Subsequent steps in this analysis will use log-cpm values, calculated using the normalisation factors, which scales each library size.</span>
<span id="cb24-175"><a href="#cb24-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-176"><a href="#cb24-176" aria-hidden="true" tabindex="-1"></a>We can calculate the normalisation factors, specifying that we want to use the <span class="in">`"TMM"`</span> method:</span>
<span id="cb24-177"><a href="#cb24-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-180"><a href="#cb24-180" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-181"><a href="#cb24-181" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span>
<span id="cb24-182"><a href="#cb24-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-183"><a href="#cb24-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-184"><a href="#cb24-184" aria-hidden="true" tabindex="-1"></a>This function calculates the normalisation factors for each library (sample) and puts this information in the <span class="in">`samples`</span> data frame. Note that it takes dge (our <span class="in">`DGEList`</span> object as input) and returns a <span class="in">`DGEList`</span> object as well.</span>
<span id="cb24-185"><a href="#cb24-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-186"><a href="#cb24-186" aria-hidden="true" tabindex="-1"></a>Let's take a look at our normalisation factors:</span>
<span id="cb24-187"><a href="#cb24-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-190"><a href="#cb24-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-191"><a href="#cb24-191" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>samples</span>
<span id="cb24-192"><a href="#cb24-192" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-193"><a href="#cb24-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-194"><a href="#cb24-194" aria-hidden="true" tabindex="-1"></a>These normalisation factors are all close to 1 for all samples, suggesting minimal difference in RNA composition between samples.</span>
<span id="cb24-195"><a href="#cb24-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-196"><a href="#cb24-196" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualising the effect of normalisation</span></span>
<span id="cb24-197"><a href="#cb24-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-198"><a href="#cb24-198" aria-hidden="true" tabindex="-1"></a>To visualise the effect of TMM normalisation, we can plot the log-counts as a boxplot, and observe the effect of applying the normalisation. To create a boxplot of the log-counts, we can use <span class="in">`log(dge$counts + 0.5)`</span> to create the log-count matrix. The addition of 0.5 is to avoid taking the log of zero. Then in order to use <span class="in">`ggplot2`</span> for plotting, we must convert the matrix to a data frame. We can use the <span class="in">`as_tibble(rownames = "gene")`</span> function to convert the matrix to a data frame where the rownames are converted to a column called "gene". We can then use the <span class="in">`pivot_longer()`</span> function to convert the data frame from wide format to long format, where each row represents a single observation. This is necessary for <span class="in">`ggplot2`</span> to plot the data correctly.</span>
<span id="cb24-199"><a href="#cb24-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-202"><a href="#cb24-202" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-203"><a href="#cb24-203" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">log</span>(dge<span class="sc">$</span>counts <span class="sc">+</span> <span class="fl">0.5</span>), <span class="at">rownames =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-204"><a href="#cb24-204" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb24-205"><a href="#cb24-205" aria-hidden="true" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric),</span>
<span id="cb24-206"><a href="#cb24-206" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb24-207"><a href="#cb24-207" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb24-208"><a href="#cb24-208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb24-209"><a href="#cb24-209" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb24-210"><a href="#cb24-210" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-211"><a href="#cb24-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-212"><a href="#cb24-212" aria-hidden="true" tabindex="-1"></a>We can compare this to the cpm values which when calculated using the <span class="in">`cpm()`</span> function, automatically applies normalisation factors if they are present.</span>
<span id="cb24-213"><a href="#cb24-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-216"><a href="#cb24-216" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-217"><a href="#cb24-217" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">cpm</span>(dge<span class="sc">$</span>counts, <span class="at">log =</span> <span class="cn">TRUE</span>), <span class="at">rownames =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-218"><a href="#cb24-218" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb24-219"><a href="#cb24-219" aria-hidden="true" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric),</span>
<span id="cb24-220"><a href="#cb24-220" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb24-221"><a href="#cb24-221" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb24-222"><a href="#cb24-222" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb24-223"><a href="#cb24-223" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb24-224"><a href="#cb24-224" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-225"><a href="#cb24-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-226"><a href="#cb24-226" aria-hidden="true" tabindex="-1"></a>We see that by performing normalisation on our data, the gene expression values of each sample now have similar medians and quantiles. This indicates that the relative expression values of each sample can be more meaningfully compared.</span>
<span id="cb24-227"><a href="#cb24-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-228"><a href="#cb24-228" aria-hidden="true" tabindex="-1"></a><span class="fu">## MDS plots</span></span>
<span id="cb24-229"><a href="#cb24-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-230"><a href="#cb24-230" aria-hidden="true" tabindex="-1"></a>Before we perform statistical tests, it's useful to perform some exploratory visual analysis to get an overall idea of how our data is behaving.</span>
<span id="cb24-231"><a href="#cb24-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-232"><a href="#cb24-232" aria-hidden="true" tabindex="-1"></a>MDS is a way to visualise distances between sets of data points (samples in our case). It is a dimensionality reduction technique, similar to principal components analysis (PCA). We treat gene expression in samples as if they were coordinates in a high-dimensional coordinate system, then we can find "distances" between samples as we do between points in space. Then the goal of the algorithm is to find a representation in lower dimensional space such that points that the distance of two objects from each other in high dimensional space is preserved in lower dimensions.</span>
<span id="cb24-233"><a href="#cb24-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-234"><a href="#cb24-234" aria-hidden="true" tabindex="-1"></a>The <span class="in">`plotMDS()`</span> from <span class="in">`limma`</span> creates an MDS plot from a <span class="in">`DGEList`</span> object.</span>
<span id="cb24-235"><a href="#cb24-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-238"><a href="#cb24-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-239"><a href="#cb24-239" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge)</span>
<span id="cb24-240"><a href="#cb24-240" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-241"><a href="#cb24-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-242"><a href="#cb24-242" aria-hidden="true" tabindex="-1"></a>Each point on the plot represents one sample and is 'labelled' using the sample name. The distances between each sample in the resulting plot can be interpreted as the typical log2-fold-change between the samples, for the most differentially expressed genes.</span>
<span id="cb24-243"><a href="#cb24-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-244"><a href="#cb24-244" aria-hidden="true" tabindex="-1"></a>We can change the labelling to use the name of the group the sample belongs to instead:</span>
<span id="cb24-245"><a href="#cb24-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-248"><a href="#cb24-248" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-249"><a href="#cb24-249" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dge, <span class="at">labels =</span> group)</span>
<span id="cb24-250"><a href="#cb24-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-251"><a href="#cb24-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-252"><a href="#cb24-252" aria-hidden="true" tabindex="-1"></a>This shows us that the phenotype groups tend to cluster together, meaning that the gene expression profiles are similar for samples within a phenotype group. The 'Basal' type samples quite close together while the 'LP' (luminal progenitor) and 'ML' (mature luminal) type samples are further apart, signifying that their expression profiles are more variable.</span>
<span id="cb24-253"><a href="#cb24-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-254"><a href="#cb24-254" aria-hidden="true" tabindex="-1"></a><span class="fu">## MDS plot using ggplot2</span></span>
<span id="cb24-255"><a href="#cb24-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-256"><a href="#cb24-256" aria-hidden="true" tabindex="-1"></a>For more customisability and better consistency with the style of other plots, it'd be nice to be able to draw the MDS plot using <span class="in">`ggplot2`</span>. In order to do this we would need the MDS coordinates calculated by <span class="in">`plotMDS()`</span>. Luckily the documentation of <span class="in">`plotMDS()`</span> shows that it returns multiple computed values.</span>
<span id="cb24-257"><a href="#cb24-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-260"><a href="#cb24-260" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-261"><a href="#cb24-261" aria-hidden="true" tabindex="-1"></a>mds_result <span class="ot">&lt;-</span> <span class="fu">plotMDS</span>(dge, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-262"><a href="#cb24-262" aria-hidden="true" tabindex="-1"></a>mds_result</span>
<span id="cb24-263"><a href="#cb24-263" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-264"><a href="#cb24-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-265"><a href="#cb24-265" aria-hidden="true" tabindex="-1"></a>We see that there's are <span class="in">`x`</span> and <span class="in">`y`</span> values in the list returned by the <span class="in">`plotMDS()`</span> function, we can put these into a table and see if they match up to the positions plotted by the function.</span>
<span id="cb24-266"><a href="#cb24-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-269"><a href="#cb24-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-270"><a href="#cb24-270" aria-hidden="true" tabindex="-1"></a>mds_tibble <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb24-271"><a href="#cb24-271" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> mds_result<span class="sc">$</span>x,</span>
<span id="cb24-272"><a href="#cb24-272" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> mds_result<span class="sc">$</span>y</span>
<span id="cb24-273"><a href="#cb24-273" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-274"><a href="#cb24-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-275"><a href="#cb24-275" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb24-276"><a href="#cb24-276" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb24-277"><a href="#cb24-277" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-278"><a href="#cb24-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-279"><a href="#cb24-279" aria-hidden="true" tabindex="-1"></a>The positions of the points and the range of the scales seem to match, so we want to add in more metadata to help use plot. We can try to recreate the MDS plot using just <span class="in">`ggplot2`</span>.</span>
<span id="cb24-280"><a href="#cb24-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-283"><a href="#cb24-283" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-284"><a href="#cb24-284" aria-hidden="true" tabindex="-1"></a>mds_tibble <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb24-285"><a href="#cb24-285" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> mds_result<span class="sc">$</span>x,</span>
<span id="cb24-286"><a href="#cb24-286" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> mds_result<span class="sc">$</span>y,</span>
<span id="cb24-287"><a href="#cb24-287" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colnames</span>(dge),</span>
<span id="cb24-288"><a href="#cb24-288" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>group</span>
<span id="cb24-289"><a href="#cb24-289" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-290"><a href="#cb24-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-291"><a href="#cb24-291" aria-hidden="true" tabindex="-1"></a>dim1_var_explained <span class="ot">&lt;-</span> <span class="fu">round</span>(mds_result<span class="sc">$</span>var.explained[<span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb24-292"><a href="#cb24-292" aria-hidden="true" tabindex="-1"></a>dim2_var_explained <span class="ot">&lt;-</span> <span class="fu">round</span>(mds_result<span class="sc">$</span>var.explained[<span class="dv">2</span>] <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb24-293"><a href="#cb24-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-294"><a href="#cb24-294" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb24-295"><a href="#cb24-295" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> group)) <span class="sc">+</span></span>
<span id="cb24-296"><a href="#cb24-296" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb24-297"><a href="#cb24-297" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb24-298"><a href="#cb24-298" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb24-299"><a href="#cb24-299" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-300"><a href="#cb24-300" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-301"><a href="#cb24-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-302"><a href="#cb24-302" aria-hidden="true" tabindex="-1"></a>Now that we have our data in a nice tidy format we can use with ggplot, it's easy to create variations of the plot. For example we can draw points instead of group labels, and use colour to identify the groups.</span>
<span id="cb24-303"><a href="#cb24-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-306"><a href="#cb24-306" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-307"><a href="#cb24-307" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb24-308"><a href="#cb24-308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb24-309"><a href="#cb24-309" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb24-310"><a href="#cb24-310" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb24-311"><a href="#cb24-311" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb24-312"><a href="#cb24-312" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-313"><a href="#cb24-313" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-314"><a href="#cb24-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-315"><a href="#cb24-315" aria-hidden="true" tabindex="-1"></a>Alternatively we can also use the labels to identify the individual samples while colouring them by their group.</span>
<span id="cb24-316"><a href="#cb24-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-319"><a href="#cb24-319" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-320"><a href="#cb24-320" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb24-321"><a href="#cb24-321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample)) <span class="sc">+</span></span>
<span id="cb24-322"><a href="#cb24-322" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb24-323"><a href="#cb24-323" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb24-324"><a href="#cb24-324" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb24-325"><a href="#cb24-325" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-326"><a href="#cb24-326" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-327"><a href="#cb24-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-328"><a href="#cb24-328" aria-hidden="true" tabindex="-1"></a>We see that some labels are very hard to read due to the overlapping, so we can use the <span class="in">`ggrepel`</span> package to fix this. The <span class="in">`ggrepel`</span> package creates labels that repel each other in order to avoid overlap. It's a good idea to use it in conjunction with <span class="in">`geom_point()`</span> in order to keep track of exact coordinate of the data point.</span>
<span id="cb24-329"><a href="#cb24-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-332"><a href="#cb24-332" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb24-333"><a href="#cb24-333" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb24-334"><a href="#cb24-334" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_tibble, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">col =</span> group)) <span class="sc">+</span></span>
<span id="cb24-335"><a href="#cb24-335" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb24-336"><a href="#cb24-336" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample)) <span class="sc">+</span></span>
<span id="cb24-337"><a href="#cb24-337" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb24-338"><a href="#cb24-338" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 1 "</span>, <span class="st">"("</span>, dim1_var_explained, <span class="st">"%)"</span>),</span>
<span id="cb24-339"><a href="#cb24-339" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"Leading logFC dim 2 "</span>, <span class="st">"("</span>, dim2_var_explained, <span class="st">"%)"</span>)</span>
<span id="cb24-340"><a href="#cb24-340" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-341"><a href="#cb24-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb24-342"><a href="#cb24-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-343"><a href="#cb24-343" aria-hidden="true" tabindex="-1"></a><span class="fu"># Summary</span></span>
<span id="cb24-344"><a href="#cb24-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-345"><a href="#cb24-345" aria-hidden="true" tabindex="-1"></a>Today we started on the early steps of the RNA-seq analysis workflow.</span>
<span id="cb24-346"><a href="#cb24-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-347"><a href="#cb24-347" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We learned how to create a <span class="in">`DGEList`</span> object, demonstrating how to create data objects required by specific packages.</span>
<span id="cb24-348"><a href="#cb24-348" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We learned how to use edgeR's <span class="in">`filterByExpr()`</span> function to filter out lowly expressed genes.</span>
<span id="cb24-349"><a href="#cb24-349" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We learned how to normalise the data using TMM normalisation.</span>
<span id="cb24-350"><a href="#cb24-350" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We learned how to create MDS plots using both <span class="in">`plotMDS()`</span> and <span class="in">`ggplot2`</span>.</span>
<span id="cb24-351"><a href="#cb24-351" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We learned how to use <span class="in">`ggrepel`</span> to create non-overlapping labels in <span class="in">`ggplot2`</span> plots.</span>
<span id="cb24-352"><a href="#cb24-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-353"><a href="#cb24-353" aria-hidden="true" tabindex="-1"></a>This provides a good foundation for organising data to satisfy the requirements of specific packages, and how to pull data out of the objects created by those packages. The data we pulled out can then be organised into a tidy format to leverage the power of all the <span class="in">`tidyverse`</span> functions that we have learned so far.</span>
<span id="cb24-354"><a href="#cb24-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-355"><a href="#cb24-355" aria-hidden="true" tabindex="-1"></a>In the next session we will continue the RNA-seq analysis workflow and complete our differential expression analysis.</span>
<span id="cb24-356"><a href="#cb24-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-357"><a href="#cb24-357" aria-hidden="true" tabindex="-1"></a><span class="fu">## References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>