<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Session 2: Working with data – WEHI R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session_1.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-063905fb60d21bb612e30333561bc9be.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-b70e54b6d14b88657c83e2b1b0a40c26.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script src="site_libs/quarto-contrib/kbd/kbd.js"></script>
<link href="site_libs/quarto-contrib/kbd/kbd.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session_2.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">WEHI R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this session:</h2>
   
  <ul>
  <li><a href="#sec-pipes" id="toc-sec-pipes" class="nav-link active" data-scroll-target="#sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</a></li>
  <li><a href="#sec-dataManip" id="toc-sec-dataManip" class="nav-link" data-scroll-target="#sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</a>
  <ul class="collapse">
  <li><a href="#sec-sorting" id="toc-sec-sorting" class="nav-link" data-scroll-target="#sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</a></li>
  <li><a href="#sec-filter" id="toc-sec-filter" class="nav-link" data-scroll-target="#sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</a></li>
  <li><a href="#sec-missing" id="toc-sec-missing" class="nav-link" data-scroll-target="#sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</a></li>
  <li><a href="#sec-select" id="toc-sec-select" class="nav-link" data-scroll-target="#sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</a></li>
  <li><a href="#sec-mutate" id="toc-sec-mutate" class="nav-link" data-scroll-target="#sec-mutate"><span class="header-section-number">2.2.5</span> Modifying data</a></li>
  <li><a href="#sec-summarise" id="toc-sec-summarise" class="nav-link" data-scroll-target="#sec-summarise"><span class="header-section-number">2.2.6</span> Summarising data</a></li>
  </ul></li>
  <li><a href="#sec-grouping" id="toc-sec-grouping" class="nav-link" data-scroll-target="#sec-grouping"><span class="header-section-number">2.3</span> Grouping</a></li>
  <li><a href="#sec-reshaping" id="toc-sec-reshaping" class="nav-link" data-scroll-target="#sec-reshaping"><span class="header-section-number">2.4</span> Reshaping and combining data</a>
  <ul class="collapse">
  <li><a href="#sec-pivot" id="toc-sec-pivot" class="nav-link" data-scroll-target="#sec-pivot"><span class="header-section-number">2.4.1</span> Reshaping data with pivot functions</a></li>
  <li><a href="#sec-join" id="toc-sec-join" class="nav-link" data-scroll-target="#sec-join"><span class="header-section-number">2.4.2</span> Combining data with join functions</a></li>
  </ul></li>
  <li><a href="#sec-saving" id="toc-sec-saving" class="nav-link" data-scroll-target="#sec-saving"><span class="header-section-number">2.5</span> Saving data to a file</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">2.6</span> Summary</a>
  <ul class="collapse">
  <li><a href="#practice-questions" id="toc-practice-questions" class="nav-link" data-scroll-target="#practice-questions"><span class="header-section-number">2.6.1</span> Practice questions</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-session02" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this session we will learn how to manipulate and summarise data using the <code>dplyr</code> package.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this session, learners should be able to:</p>
<ol type="1">
<li><p>Use the pipe (<code>%&gt;%</code>) to chain multiple functions together</p></li>
<li><p>Design chains of dplyr functions to manipulate data frames</p></li>
<li><p>Apply grouping for more complex analysis of data</p></li>
<li><p>Understand how pivot and join functions can be used to reshape and combine entire data frames</p></li>
<li><p>Recall how to save data frames to a file</p></li>
</ol>
</div>
</div>
<p>Make sure you have <code>dplyr</code> package loaded by running <code>library(dplyr)</code>, otherwise you will not be able to run the code in this session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load dplyr</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
</div>
<p>You’ll see some output from the <code>dplyr</code> package when you load it, which is just telling you that some of <code>dplyr</code>’s functions have the same name as other functions in R.</p>
<section id="sec-pipes" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</h2>
<p>Pipes are a powerful feature of the <code>tidyverse</code> that allow you to chain multiple functions together. Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</p>
<p>For example, take the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>What do you think this code does? It calculates the mean of <code>my_vector</code>, rounds the result to the nearest whole number, and then converts the result to a character. But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</p>
<p>Instead, we can use the pipe operator (<code>%&gt;%</code>) to chain these functions together in a more readable way:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>See how the code reads naturally from left to right? You can think of the pipe as being like the phrase “and then”. Here, we’re telling R: “Take <code>my_vector</code>, and then calculate the mean, and then round the result, and then convert it to a character.”</p>
<p>You’ll notice that we didn’t need to specify the input to each function. That’s because the pipe automatically passes the output of the previous function as the first input to the next function. We can still specify additional arguments to each function if we need to. For example, if we wanted to round the mean to 2 decimal places, we could do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>R is clever enough to know that the first argument to <code>round()</code> is still the output of <code>mean()</code>, even though we’ve now specified the <code>digits</code> argument.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Plenty of pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Plenty of pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>There is another style of pipe in R, called the ‘base R pipe’ <code>|&gt;</code>, which is available in R version 4.1.0 and later. The base R pipe works in a similar way to the <code>magrittr</code> pipe (<code>%&gt;%</code>) that we use in this course, but it is not as flexible. We recommend using the <code>magrittr</code> pipe for now.</p>
<p>Fun fact: the <code>magrittr</code> package is named after the <a href="https://en.wikipedia.org/wiki/The_Treachery_of_Images">artist René Magritte, who made a famous painting of a pipe</a>.</p>
</div>
</div>
<p>To type the pipe operator more easily, you can use the keyboard shortcut <kbd aria-hidden="true">Cmd-shift-M</kbd><span class="visually-hidden">Cmd-shift-M</span> (although once you get used to it, you might find it easier to type <code>%&gt;%</code> manually).</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What is NOT a valid way to re-write the following code using the pipe operator: <code>round(sqrt(sum(1:10)), 1)</code>. If you’re not sure, try running the different options in the console to see which one gives the same answer.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is the output of the following code? <code>letters %&gt;% head() %&gt;% toupper()</code> Try to guess it before copy-pasting into R.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>"a" "b" "c" "d" "e" "f"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>"A" "B" "C" "D" "E" "F"</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The invalid option is <code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code>. This is because the <code>sqrt()</code> function only takes one argument, so you can’t specify <code>1</code> as an argument in addition to what is being piped in from <code>sum(1:10)</code>. Note that some options used the pipe to send <code>1:10</code> to <code>sum()</code> (like <code>1:10 %&gt;% sum()</code>), and others just used <code>sum(1:10)</code> directly. Both are valid ways to use the pipe, it’s just a matter of personal preference.</li>
<li>The output of the code <code>letters %&gt;% head() %&gt;% toupper()</code> is <code>"A" "B" "C" "D" "E" "F"</code>. The <code>letters</code> vector contains the lowercase alphabet, and the <code>head()</code> function returns the first 6 elements of the vector. Finally, the <code>toupper()</code> function then converts these elements to uppercase.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-dataManip" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</h2>
<p>To really see the power of the pipe, we will use it together with the <code>dplyr</code> package that provides a set of functions to easily filter, sort, select, and summarise data frames. These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</p>
<p>For example, even though we haven’t covered the <code>dplyr</code> functions yet, you can probably guess what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code filters the <code>mousezempic_dosage_data</code> data frame to only include data from cage 3E, then pulls out the <code>weight_lost_g</code> column, and finally calculates the mean of the values in that column. The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <code>pull()</code>) are specified in the brackets (like <code>round(digits = 2)</code> from the previous example).</p>
<p>We also used the enter key after each pipe <code>%&gt;%</code> to break up the code into multiple lines to make it easier to read. This isn’t required, but is a popular style in the R community, so all the code examples in this session will follow this format.</p>
<p>We will now introduce some of the most commonly used <code>dplyr</code> functions for manipulating data frames. To showcase these, we will use the <code>mousezempic_dosage_data</code> that we practiced reading in last session. This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in session 1</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"~/Desktop/R_course/data/mousezempic_dosage_data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 344 Columns: 9
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): mouse_strain, cage_number, replicate, sex
dbl (5): weight_lost_g, drug_dose_g, tail_length_mm, initial_weight_g, id_num

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>Before we start, let’s use what we learned in the previous session to take a look at <code>mousezempic_dosage_data</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>You might also like to use <code>View()</code> to open the data in a separate window and get a closer look.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Using RStudio autocomplete">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Using RStudio autocomplete
</div>
</div>
<div class="callout-body-container callout-body">
<p>Although it’s great to give our data a descriptive name like <code>mousezempic_dosage_data</code>, it can be a bit of a pain to type out every time. Luckily, RStudio has a handy autocomplete feature that can solve this problem. Just start typing the name of the object, and you’ll see it will popup:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F01_rstudio_autocomplete.png" class="img-fluid figure-img"></p>
<figcaption>RStudio autocomplete</figcaption>
</figure>
</div>
<p>You can then press <kbd aria-hidden="true">Tab</kbd><span class="visually-hidden">Tab</span> to autocomplete it. If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</p>
<p>Try using autocomplete this session to save yourself some typing!</p>
</div>
</div>
<section id="sec-sorting" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</h3>
<p>Often, one of the first things you might want to do with a dataset is sort it. In <code>dplyr</code>, this is called ‘arranging’ and is done with the <code>arrange()</code> function.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F02_arrange.png" class="img-fluid figure-img"></p>
<figcaption>Arrange orders rows by their values in one or more columns</figcaption>
</figure>
</div>
<p>By default, <code>arrange()</code> sorts in ascending order (smallest values first). For example, let’s sort the <code>mousezempic_dosage_data</code> data frame by the <code>weight_lost_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 CD-1         3E                   2.85 rep2      female     0.00181
 3 CD-1         3E                   2.85 rep2      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         2B                   2.9  rep2      female     0.00178
 6 CD-1         1A                   2.9  rep3      female     0.00188
 7 BALB C       2B                   2.9  rep1      female     0.00187
 8 CD-1         3E                   2.92 rep3      female     0.00193
 9 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
10 CD-1         2B                   3    rep1      female     0.00185
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</p>
<p>Sometimes you might want to sort in descending order instead (largest values first). You can do this by putting the <code>desc()</code> function around your column name, inside <code>arrange()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Now we can see the mice that lost the most weight are at the top.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Comments and pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Comments and pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Notice how in the previous example we have written a comment in the middle of the pipe chain. This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won’t cause any errors as long as you make sure that the comment is on its own line.</p>
<p>You can also write comments at the end of the line, just make sure it’s after the pipe operator <code>%&gt;%</code>.</p>
<p>For example, these comments are allowed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But this will cause an error, because the <code>#</code> is before the pipe, so R treats it as part of the comment (notice how the <code>%&gt;%</code> has changed colour?) and doesn’t know how the two lines relate to each other. It tries to run them separately, which for the first line is ok (it will just print <code>mousezempic_dosage_data</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># this comment will cause an error %&gt;%</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in eval(expr, envir, enclos): object 'weight_lost_g' not found</code></pre>
</div>
</div>
<p>But for the second line, there is an error that R doesn’t know what the <code>weight_lost_g</code> object is. That’s because it’s a column in the <code>mousezempic_dosage_data</code> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</p>
</div>
</div>
<p>You can also sort by multiple columns by passing multiple column names to <code>arrange()</code>. For example, to sort by the strain first and then by the amount of weight lost:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 BALB C       2B                   2.9  rep1      female     0.00187
 3 BALB C       2B                   3.2  rep2      female     0.00187
 4 BALB C       2B                   3.25 rep1      female     0.00178
 5 BALB C       2B                   3.25 rep3      male       0.00187
 6 BALB C       2B                   3.25 rep3      female     0.00191
 7 BALB C       2B                   3.3  rep1      male       0.00197
 8 BALB C       2B                   3.3  rep1      female     0.00195
 9 BALB C       2B                   3.32 rep3      female     0.00199
10 BALB C       2B                   3.35 rep2      female     0.00187
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Piping into View()">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Piping into View()
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn’t allow us to see the full effect of the sorting.</p>
<p>One handy trick you can use with pipes is to add <code>View()</code> at the end of your chain to open the data in a separate window. Try running this code, and you’ll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is a great way to check that your code has actually done what you intended!</p>
</div>
</div>
<section id="sec-sliceMinMax" class="level4" data-number="2.2.1.1">
<h4 data-number="2.2.1.1" class="anchored" data-anchor-id="sec-sliceMinMax"><span class="header-section-number">2.2.1.1</span> Extracting rows with the smallest or largest values</h4>
<p>Slice functions are used to select rows based on their position in the data frame. The <code>slice_min()</code> and <code>slice_max()</code> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</p>
<p>This is equivalent to using <code>arrange()</code> followed by <code>head()</code>, but is more concise:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.15 rep1      female     0.00172
 2 CD-1         3E                   3.4  rep1      female     0.00174
 3 CD-1         1A                   3.45 rep3      female     0.00176
 4 CD-1         2B                   3.25 rep1      female     0.00178
 5 CD-1         2B                   3.9  rep1      male       0.00178
 6 CD-1         2B                   2.9  rep2      female     0.00178
 7 BALB C       2B                   3.25 rep1      female     0.00178
 8 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
 9 CD-1         1A                   3.7  rep1      &lt;NA&gt;       0.0018 
10 CD-1         3E                   3.6  rep1      male       0.0018 
11 CD-1         3E                   3.8  rep1      male       0.0018 
12 CD-1         3E                   3.95 rep1      male       0.0018 
13 CD-1         2B                   3.55 rep1      female     0.0018 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
6 Black 6      3E                   5.95 rep3      male      0.00229
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But wait— neither of those pieces of code actually gave the number of rows we asked for! In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13. And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6. Why aren’t the <code>slice_</code> functions behaving as expected?</p>
<p>If we take a look at the help page (type <code>?slice_min</code> in the console), we learn that <code>slice_min()</code> and <code>slice_max()</code> have an argument called <code>with_ties</code> that is set to <code>TRUE</code> by default. If we want to make sure we only get the number of rows we asked for, we would have to set it to <code>FALSE</code>, like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to sort the <code>mousezempic_dosage_data</code> data frame from biggest to smallest initial weight?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% sort(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% sort(descending(initial_weight_g))</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which code would you use to extract the 3 mice with the highest initial weight from the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% slice_min(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. I’ve written the below code, but one of the comments is messing it up! Which one?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># comment B %&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Comment A</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Comment B</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Comment C</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Comment D</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to sort the <code>mousezempic_dosage_data</code> data frame from biggest to smallest initial weight is <code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code>. The <code>arrange()</code> function is used to sort the data frame (although there is a <code>sort()</code> function in R, that’s not part of dplyr and won’t work the same way), and the <code>desc()</code> function is used to sort in descending order.</p></li>
<li><p>The correct code to extract the 3 mice with the highest initial weight from the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)</code>. The <code>slice_max()</code> function is used to select the rows with the largest values in the <code>initial_weight_g</code> column, and the <code>n = 3</code> argument specifies that we want to keep 3 rows. The <code>arrange()</code> function is not needed in this case, because <code>slice_max()</code> will automatically sort the data frame by the specified column.</p></li>
<li><p>The comment that is messing up the code is Comment B. The <code>#</code> symbol is before the pipe operator <code>%&gt;%</code>, so R treats it as part of the comment and this breaks our chain of pipes. The other comments are fine, because they are either at the end of the line or on their own line. Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it’s in the wrong place!</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-filter" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F03_filter.png" class="img-fluid figure-img"></p>
<figcaption>Filter allows you to filter rows using a logical test</figcaption>
</figure>
</div>
<p>In <code>dplyr</code>, the <code>filter()</code> function is used to subset rows based on their values. You provide a logical test, and <code>filter()</code> will keep the rows where the test is <code>TRUE</code>. We can write these tests using the comparison operators we learned in the previous session (e.g.&nbsp;<code>==</code>, <code>&lt;</code> and <code>!=</code>, see <a href="session_1.html#sec-comparisons" class="quarto-xref">Section&nbsp;<span>1.3</span></a>).</p>
<p>For example, to filter the <code>mousezempic_dosage_data</code> data frame to only include mice that lost more than 6g:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or to only include mice from cage 3E:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 168 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.4  rep1      female     0.00174
 2 CD-1         3E                   3.6  rep1      male       0.0018 
 3 CD-1         3E                   3.8  rep1      female     0.00189
 4 CD-1         3E                   3.95 rep1      male       0.00185
 5 CD-1         3E                   3.8  rep1      male       0.0018 
 6 CD-1         3E                   3.8  rep1      female     0.00187
 7 CD-1         3E                   3.55 rep1      male       0.00183
 8 CD-1         3E                   3.2  rep1      female     0.00187
 9 CD-1         3E                   3.15 rep1      female     0.00172
10 CD-1         3E                   3.95 rep1      male       0.0018 
# ℹ 158 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<section id="combining-logical-tests" class="level4" data-number="2.2.2.1">
<h4 data-number="2.2.2.1" class="anchored" data-anchor-id="combining-logical-tests"><span class="header-section-number">2.2.2.1</span> Combining logical tests</h4>
<p>Sometimes we want to filter based on multiple conditions. Here we will show some more advanced operators that can be used to combine logical tests.</p>
<p>The <code>&amp;</code> operator is used to combine two logical tests with an ‘and’ condition. For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
1 CD-1         1A                   3.45 rep1      female     0.00193
2 CD-1         3E                   3.8  rep1      female     0.00189
3 CD-1         2B                   3.3  rep1      female     0.00181
4 CD-1         3E                   3.9  rep3      female     0.00191
5 BALB C       2B                   3.52 rep3      female     0.00194
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>|</code> operator is used to combine two logical tests with an ‘or’ condition. For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 2 CD-1         1A                   4.4  rep1      male       0.00198
 3 CD-1         1A                   3.32 rep1      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         1A                   3.6  rep2      female     0.0019 
 6 CD-1         1A                   3.2  rep2      female     0.00189
 7 CD-1         2B                   3.4  rep2      female     0.00185
 8 CD-1         2B                   2.9  rep2      female     0.00178
 9 CD-1         2B                   3.05 rep3      female     0.00188
10 Black 6      3E                   4.5  rep1      female     0.00211
# ℹ 20 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>%in%</code> operator can be used to filter based on a vector of multiple values (<code>c(x, y)</code>). It’s particularly useful when you have a few character values you want to filter on, as it is shorter to type than <code>|</code> (or).</p>
<p>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <code>|</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or we could use <code>%in%</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to filter the <code>mousezempic_dosage_data</code> data frame to only include mice from replicate 2?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate == 2)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate == rep2)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(replicate == "rep2")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate = "rep2")</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is NOT a valid way to filter the <code>mousezempic_dosage_data</code> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. Which option correctly describes what the following code is doing?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to filter the <code>mousezempic_dosage_data</code> data frame to only include mice from replicate 2 is <code>mousezempic_dosage_data %&gt;% filter(replicate == "rep2")</code>. Option A is incorrect because <code>2</code> is not a value of <code>replicate</code> (when filtering you need to know what values are actually in your columns! So make sure to <code>View()</code> your data first). Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on. Option D is incorrect because <code>=</code> is not the correct way to test for equality, you need to use <code>==</code>.</p></li>
<li><p>The invalid option is <code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code>. This is because the second filtering step is missing the name of the filter function, so R doesn’t know what to do with <code>(initial_weight_g &lt; 40)</code>. The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <code>filter()</code> functions in a row to apply multiple conditions, or the <code>&amp;</code> operator to combine them into a single <code>filter()</code> function. It’s just a matter of personal preference.</p></li>
<li><p>The correct description of the code is that it filters the data frame to only include mice from the “BALB C” and “Black 6” strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-missing" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</h3>
<p>Missing values are a common problem in real-world datasets. In R, missing values are represented by <code>NA</code>. In fact, if you look at the <code>mousezempic_dosage_data</code> data frame we’ve been using, you’ll see that some of the cells contain <code>NA</code>: try spotting them with the <code>View()</code> function.</p>
<p>You can also find missing values in a data frame using the <code>is.na()</code> function in combination with <code>filter()</code>. For example, to find all the rows in the <code>mousezempic_dosage_data</code> data frame that have a missing value for the <code>drug_dose_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 CD-1         1A                     NA rep1      &lt;NA&gt;           NA
2 Black 6      3E                     NA rep3      &lt;NA&gt;           NA
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The problem with missing values is that they can cause problems when you try to perform calculations on your data. For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <code>NA</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p><code>NA</code> values in R are therefore referred to as ‘contagious’: if you put an <code>NA</code> in you usually get an <code>NA</code> out. If you think about it, that makes sense— when we don’t know the value of a particular mouse’s drug dose, how can we calculate the average? That missing value could be anything.</p>
<p>For this reason, it’s important to deal with missing values before performing calculations. Many functions in R will have an argument called <code>na.rm</code> that you can set to <code>TRUE</code> to remove missing values before performing the calculation. For example, to calculate the mean of the <code>drug_dose_g</code> column with the missing values excluded:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.002009152</code></pre>
</div>
</div>
<p>This time, the result is a number, because the missing values have been removed before the calculation.</p>
<p>But not all functions have an <code>na.rm</code> argument. In these cases, you can remove rows with missing values. This can be done for a single column, using the <code>filter()</code> function together with <code>is.na()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 342 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 9 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
10 CD-1         1A                   3.3  rep1      &lt;NA&gt;       0.00186
# ℹ 332 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or, you can remove rows with missing values in any column using the <code>na.omit()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.2  rep1      female     0.00182
 9 CD-1         1A                   3.8  rep1      male       0.00191
10 CD-1         1A                   4.4  rep1      male       0.00198
# ℹ 323 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value. This can be done using the <code>replace_na()</code> function from the <code>tidyr</code> package. <code>replace_na()</code> takes a <code>list()</code> which contains each of the column names you want to edit, and the value that should be used.</p>
<p>For example, to replace missing values in the <code>weight_lost_g</code> columns with 0, replace missing values in the <code>sex</code> column with ‘unknown’ and leave the rest of the data frame unchanged:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need to load the tidyr package first</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex     drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;         &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male        0.00181
 2 CD-1         1A                   3.8  rep1      female      0.00186
 3 CD-1         1A                   3.25 rep1      female      0.00195
 4 CD-1         1A                   0    rep1      unknown    NA      
 5 CD-1         1A                   3.45 rep1      female      0.00193
 6 CD-1         1A                   3.65 rep1      male        0.0019 
 7 CD-1         1A                   3.62 rep1      female      0.00181
 8 CD-1         1A                   4.68 rep1      male        0.00195
 9 CD-1         1A                   3.48 rep1      unknown     0.00193
10 CD-1         1A                   4.25 rep1      unknown     0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>When deciding how to handle missing values, you might have prior knowledge that <code>NA</code> should be replaced with a specific value, or you might decide that removing rows with <code>NA</code> is the best approach for your analysis.</p>
<p>For example, maybe we knew that the mice were given a <code>weight_lost_g</code> of <code>NA</code> if they didn’t lose any weight, it would then make sense to replace those with 0 (as we did in the code above). However, if the <code>drug_dose_g</code> column was missing simply because the data was lost, we might choose to remove those rows entirely.</p>
<p>It’s important to think carefully about how missing values should be handled in your analysis.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What would be the result of running this R code: <code>mean(c(1, 2, 4, NA))</code></p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>2.333333</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>0</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>NA</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which line of code would you use to filter the <code>mousezempic_dosage_data</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(tail_length_mm != NA)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(is.na(tail_length_mm))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% na.omit()</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you replace missing values in the <code>initial_weight_g</code> column with the value 35?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(initial_weight_g = 35)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g == 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(35)</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The result of running the code <code>mean(c(1, 2, 4, NA))</code> is <code>NA</code>. This is because the <code>NA</code> value is ‘contagious’, so when you try to calculate the mean of a vector that contains an <code>NA</code>, the result will also be <code>NA</code>. If we wanted to calculate the mean of the vector without the <code>NA</code>, we would need to use the <code>na.rm = TRUE</code> argument.</li>
<li>The correct line of code to filter the <code>mousezempic_dosage_data</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column is <code>mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))</code>. The <code>!</code> symbol is used to negate the result of <code>is.na()</code>, so we are filtering to keep the rows where <code>tail_length_mm</code> is not <code>NA</code>. We can’t use the first option with the <code>!= NA</code> because <code>NA</code> is a special value in R that represents missing data, and it can’t be compared to anything, and the third option is incorrect because <code>na.omit()</code> removes entire rows with missing values, rather than just filtering based on a single column.</li>
<li>The correct line of code to replace missing values in the <code>initial_weight_g</code> column with the value 35 is <code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))</code>. The <code>replace_na()</code> function takes a <code>list()</code> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we’re not testing for equality, we’re assigning a value.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-select" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F04_select.png" class="img-fluid figure-img"></p>
<figcaption>Select allows you to select only certain columns</figcaption>
</figure>
</div>
<p>While <code>filter()</code> is used to subset rows, <code>select()</code> is used to subset columns. You can use <code>select()</code> to keep only the columns you’re interested in, or to drop columns you don’t need.</p>
<p>The <code>select()</code> function takes the names of the columns that you want to keep/remove (no vector notation <code>c()</code> or quotation marks <code>""</code> necessary). For example, to select only the <code>mouse_strain</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   mouse_strain initial_weight_g weight_lost_g
   &lt;chr&gt;                   &lt;dbl&gt;         &lt;dbl&gt;
 1 CD-1                     39.1          3.75
 2 CD-1                     39.5          3.8 
 3 CD-1                     40.3          3.25
 4 CD-1                     NA           NA   
 5 CD-1                     36.7          3.45
 6 CD-1                     39.3          3.65
 7 CD-1                     38.9          3.62
 8 CD-1                     39.2          4.68
 9 CD-1                     34.1          3.48
10 CD-1                     42            4.25
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>We can see that all the other columns have been removed from the data frame.</p>
<p>If you want to keep all columns except for a few, you can use <code>-</code> to drop columns. For example, to keep all columns except for <code>cage_number</code> and <code>sex</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 7
   mouse_strain weight_lost_g replicate drug_dose_g tail_length_mm
   &lt;chr&gt;                &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt;
 1 CD-1                  3.75 rep1          0.00181           18.7
 2 CD-1                  3.8  rep1          0.00186           17.4
 3 CD-1                  3.25 rep1          0.00195           18  
 4 CD-1                 NA    rep1         NA                 NA  
 5 CD-1                  3.45 rep1          0.00193           19.3
 6 CD-1                  3.65 rep1          0.0019            20.6
 7 CD-1                  3.62 rep1          0.00181           17.8
 8 CD-1                  4.68 rep1          0.00195           19.6
 9 CD-1                  3.48 rep1          0.00193           18.1
10 CD-1                  4.25 rep1          0.0019            20.2
# ℹ 334 more rows
# ℹ 2 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>There are also some helper functions that can be used to select columns based on their names :</p>
<table class="caption-top table">
<caption>There are several helper functions that can be used with the select function</caption>
<colgroup>
<col style="width: 14%">
<col style="width: 42%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>starts_with()</code></td>
<td>select column(s) that start with a certain string</td>
<td><p>select all columns starting with the letter i</p>
<p><code>select(starts_with("i"))</code></p></td>
</tr>
<tr class="even">
<td><code>ends_with()</code></td>
<td>select column(s) that end with a certain string</td>
<td><p>select all columns ending with _g</p>
<p><code>select(ends_with("_g"))</code></p></td>
</tr>
<tr class="odd">
<td><code>contains()</code></td>
<td>select column(s) that contain a certain string</td>
<td><p>select all columns containing the word ‘weight’</p>
<p><code>select(contains("weight"))</code></p></td>
</tr>
</tbody>
</table>
<p>You need to use quotation marks around the arguments in these helper functions, as they aren’t full column names, just strings of characters.</p>
<p>Try using these helper functions to select columns from the <code>mousezempic_dosage_data</code> data frame!</p>
<div class="callout callout-style-default callout-note callout-titled" title="Reordering columns">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Reordering columns
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/S02F05_reorder.png" class="img-fluid" alt="Relocate allows you to move columns around"> We can reorder columns using the <code>relocate()</code> function, which works similarly to <code>select()</code> (except it just moves columns around rather than dropping/keeping them). For example, to move the <code>sex</code> column to before the <code>cage_number</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain sex    cage_number weight_lost_g replicate drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 CD-1         male   1A                   3.75 rep1          0.00181
 2 CD-1         female 1A                   3.8  rep1          0.00186
 3 CD-1         female 1A                   3.25 rep1          0.00195
 4 CD-1         &lt;NA&gt;   1A                  NA    rep1         NA      
 5 CD-1         female 1A                   3.45 rep1          0.00193
 6 CD-1         male   1A                   3.65 rep1          0.0019 
 7 CD-1         female 1A                   3.62 rep1          0.00181
 8 CD-1         male   1A                   4.68 rep1          0.00195
 9 CD-1         &lt;NA&gt;   1A                   3.48 rep1          0.00193
10 CD-1         &lt;NA&gt;   1A                   4.25 rep1          0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Two useful helper functions here are the <code>everything()</code> and <code>last_col()</code> functions, which can be used to move columns to the start/end of the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   id_num mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
    &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1      1 CD-1         1A                   3.75 rep1      male       0.00181
 2      2 CD-1         1A                   3.8  rep1      female     0.00186
 3      3 CD-1         1A                   3.25 rep1      female     0.00195
 4      4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5      5 CD-1         1A                   3.45 rep1      female     0.00193
 6      6 CD-1         1A                   3.65 rep1      male       0.0019 
 7      7 CD-1         1A                   3.62 rep1      female     0.00181
 8      8 CD-1         1A                   4.68 rep1      male       0.00195
 9      9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10     10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 2 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   cage_number weight_lost_g replicate sex    drug_dose_g tail_length_mm
   &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;
 1 1A                   3.75 rep1      male       0.00181           18.7
 2 1A                   3.8  rep1      female     0.00186           17.4
 3 1A                   3.25 rep1      female     0.00195           18  
 4 1A                  NA    rep1      &lt;NA&gt;      NA                 NA  
 5 1A                   3.45 rep1      female     0.00193           19.3
 6 1A                   3.65 rep1      male       0.0019            20.6
 7 1A                   3.62 rep1      female     0.00181           17.8
 8 1A                   4.68 rep1      male       0.00195           19.6
 9 1A                   3.48 rep1      &lt;NA&gt;       0.00193           18.1
10 1A                   4.25 rep1      &lt;NA&gt;       0.0019            20.2
# ℹ 334 more rows
# ℹ 3 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;, mouse_strain &lt;chr&gt;</code></pre>
</div>
</div>
<p>Re-ordering columns isn’t necessary, but it makes it easier to see the data you’re most interested in within the console (since often not all of the columns will fit on the screen at once). For example, if we are doing a lot of computation on the <code>initial_weight_g</code> column, we’d probably like to have that near the start so we can easily check it.</p>
</div>
</div>
<p>Note that the output of the <code>select()</code> function is a new data frame, even if you only select a single column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from session 1 that class() tells us the type of an object</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
</div>
</div>
<p>Sometimes, we instead want to get the values of a column as a vector.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F06_pull.png" class="img-fluid figure-img"></p>
<figcaption>Pull allows you to pull acolumn out of a data frame as a vector</figcaption>
</figure>
</div>
<p>We can do this by using the <code>pull()</code> function, which extracts a single column from a data frame as a vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
</div>
<p>We can see that the class of the output is now a vector, rather than a data frame. This is important because some functions only accept vectors, not data frames, like <code>mean()</code> for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in mean.default(., na.rm = TRUE): argument is not numeric or logical:
returning NA</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 43.92193</code></pre>
</div>
</div>
<p>Note how both times we used <code>na.rm = TRUE</code> to remove missing values before calculating the mean.</p>
<p>You might remember that we used the <code>$</code> operator in the previous session to extract a single column from a data frame, so why use <code>pull()</code> instead? The main reason is that <code>pull()</code> works within a chain of pipes, whereas <code>$</code> doesn’t.</p>
<p>For example, let’s say we want to know the average initial weight of mice that lost at least 4g. We can do this by chaining <code>filter()</code> and <code>pull()</code> together:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 46.48023</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which line of code would NOT be a valid way to select the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% select(contains("g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(ends_with("_g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. How would I extract the <code>initial_weight_g</code> column from the <code>mousezempic_dosage_data</code> data frame as a vector?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% $initial_weight_g</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(initial_weight_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% pull(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you move the <code>sex</code> column to the end of the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% relocate(sex)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% reorder(sex, .after = last_col())</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The line of code that would NOT be a valid way to select the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% select(contains("g"))</code>. This line of code would select all columns that contain the letter ‘g’, which would include columns like <code>cage_number</code> and <code>tail_length_mm</code>. We need to specify either <code>ends_with("g")</code> or <code>contains("_g")</code> to only get those with <code>_g</code> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</li>
<li>The correct way to extract the <code>initial_weight_g</code> column from the <code>mousezempic_dosage_data</code> data frame as a vector is <code>mousezempic_dosage_data %&gt;% pull(initial_weight_g)</code>. The <code>pull()</code> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <code>filter()</code> is used to subset rows, <code>$</code> is not used in a pipe chain, and <code>select()</code> is outputs a data frame, not extract them as vectors.</li>
<li>The correct way to move the <code>sex</code> column to the end of the <code>mousezempic_dosage_data</code> data frame is using the <code>relocate()</code> function like this: <code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())</code>. The <code>last_col()</code> function is used to refer to the last column in the data frame. The other options are incorrect because <code>reorder()</code> is not a valid function, and you need to remember to include the brackets <code>()</code> when using <code>last_col()</code>.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-mutate" class="level3" data-number="2.2.5">
<h3 data-number="2.2.5" class="anchored" data-anchor-id="sec-mutate"><span class="header-section-number">2.2.5</span> Modifying data</h3>
<p>So far, we’ve learned how to filter rows and select columns from a data frame. But what if we want to change the data itself? This is where the <code>mutate()</code> function comes in.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F07_mutate.png" class="img-fluid figure-img"></p>
<figcaption>Mutate allows you to add new columns to a data frame</figcaption>
</figure>
</div>
<p>The <code>mutate()</code> function is used to add new columns to a data frame, or modify existing columns, often by performing some sort of calculation. For example, we can add a new column to <code>mousezempic_dosage_data</code> that contains the drug dose in mg (rather than g):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a new column called drug_dose_mg</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert drug_dose_g to mg by multiplying by 1000</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># just select the drug dose columns so we can compare them</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(drug_dose_g, drug_dose_mg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   drug_dose_g drug_dose_mg
         &lt;dbl&gt;        &lt;dbl&gt;
 1     0.00181         1.81
 2     0.00186         1.86
 3     0.00195         1.95
 4    NA              NA   
 5     0.00193         1.93
 6     0.0019          1.9 
 7     0.00181         1.81
 8     0.00195         1.95
 9     0.00193         1.93
10     0.0019          1.9 
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>You can see that the <code>drug_dose_mg</code> column has been added to the data frame, and it contains, for each row, the value of the <code>drug_dose_g</code> column multiplied by 1000 (<code>NA</code> values are preserved).</p>
<p>These calculations can be as complex as you like, and involve multiple different columns. For example, to add a new column to the <code>mousezempic_dosage_data</code> data frame that calculates the weight lost as a percentage of the initial weight:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate the % of initial weight that was lost</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_percent =</span> (weight_lost_g <span class="sc">/</span> initial_weight_g) <span class="sc">*</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 10
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 4 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, weight_lost_percent &lt;dbl&gt;</code></pre>
</div>
</div>
<p>A useful helper function for <code>mutate()</code> is the <code>case_when()</code> function, which allows you to create new columns based on multiple conditions. We do this with the notation <code>case_when(condition1 ~ value1, condition2 ~ value2, ...)</code>.</p>
<p>For example, to add a new column to the <code>mousezempic_dosage_data</code> data frame that categorises the mice based on how much weight they lost:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   weight_lost_g weight_loss_category
           &lt;dbl&gt; &lt;chr&gt;               
 1          3.75 Low                 
 2          3.8  Low                 
 3          3.25 Low                 
 4         NA    &lt;NA&gt;                
 5          3.45 Low                 
 6          3.65 Low                 
 7          3.62 Low                 
 8          4.68 Medium              
 9          3.48 Low                 
10          4.25 Medium              
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Note that the conditions are evaluated in order, and the first condition that is <code>TRUE</code> is the one that is used. So if a mouse lost 4.5g, it <code>case_when()</code> would first test if it fits the ‘Low’ category (by checking if 4.5 is less than 4, which it isn’t), and then if it fits the ‘Medium’ category (by checking if 4.5 is less than or equal to 5). Since it is, the mouse would be categorised as ‘Medium’.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Fallback with default value(s)">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Fallback with default value(s)
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above example, what would happen if a mouse lost -1g (gained weight)? It wouldn’t fit any of the conditions, so it would get an <code>NA</code> in the <code>weight_loss_category</code> column. Sometimes you might want this behaviour, but other times you would prefer to specify a ‘fallback’ category that will be assigned to everything that doesn’t fit in the other categories. You can do this by including a <code>.default =</code> argument at the end of the <code>case_when()</code> function. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span>,</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"Unknown"</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   weight_lost_g weight_loss_category
           &lt;dbl&gt; &lt;chr&gt;               
 1          3.75 Low                 
 2          3.8  Low                 
 3          3.25 Low                 
 4         NA    Unknown             
 5          3.45 Low                 
 6          3.65 Low                 
 7          3.62 Low                 
 8          4.68 Medium              
 9          3.48 Low                 
10          4.25 Medium              
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Notice how the <code>NA</code> value in the fourth row is now categorised as ‘Unknown’.</p>
</div>
</div>
<p>One final thing to note is that <code>mutate()</code> can be used to modify existing columns as well as add new ones. To do this, just use the name of the existing column as the ‘new’ one.</p>
<p>For example, let’s use <code>mutate()</code> together with <code>case_when()</code> to modify the <code>sex</code> column so that it uses <code>M</code> and <code>F</code> instead <code>male</code> and <code>female</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># modify sex column</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">case_when</span>(</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">~</span> <span class="st">"F"</span>,</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"male"</span> <span class="sc">~</span> <span class="st">"M"</span>,</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if neither, code it as 'X'</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"X"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      M         0.00181
 2 CD-1         1A                   3.8  rep1      F         0.00186
 3 CD-1         1A                   3.25 rep1      F         0.00195
 4 CD-1         1A                  NA    rep1      X        NA      
 5 CD-1         1A                   3.45 rep1      F         0.00193
 6 CD-1         1A                   3.65 rep1      M         0.0019 
 7 CD-1         1A                   3.62 rep1      F         0.00181
 8 CD-1         1A                   4.68 rep1      M         0.00195
 9 CD-1         1A                   3.48 rep1      X         0.00193
10 CD-1         1A                   4.25 rep1      X         0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What line of code would you use to add a new column to the <code>mousezempic_dosage_data</code> data frame that converts the <code>tail_length_mm</code> column to cm?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% create(tail_length_cm = tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm == tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% tail_length_cm = tail_length_mm / 10</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_rank =</span> <span class="fu">row_number</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hint: the row_number() function returns the number of each row in the data frame (1 being the first row and so on).</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the least weight.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Adds a new column to the data frame that ranks the mice</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Does nothing, because the <code>row_number()</code> function has no arguments</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. What is wrong with this R code?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span></span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: &lt;text&gt;:4:5: unexpected symbol
3:     weight_lost_g &lt; 4 ~ "Low"
4:     weight_lost_g
       ^</code></pre>
</div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>You didn’t include a <code>TRUE ~ ...</code> condition at the end of the <code>case_when()</code> function to act as a fallback</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>You can’t use the <code>case_when()</code> function with the <code>mutate()</code> function</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>weight_lost_g</code> is not a valid column name</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>You need to separate the conditions in the <code>case_when()</code> function with a comma</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>4. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_strain =</span> <span class="fu">case_when</span>(</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>    mouse_strain <span class="sc">==</span> <span class="st">"Black 6"</span> <span class="sc">~</span> <span class="st">"B6"</span>,</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> mouse_strain</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hint: if you’re not sure, try running the code, but pipe it into <code>View()</code> so that you can take a good look at what’s happening in the <code>mouse_strain</code> column.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Renames the strains of all the mice to “B6”, regardless of their original strain</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>This code will produce an error</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Adds a new column that categorises the mice based on their strain, so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Modifies the <code>mouse_strain</code> column so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged.</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The correct line of code to add a new column to the <code>mousezempic_dosage_data</code> data frame that converts the <code>tail_length_mm</code> column to cm is <code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)</code>.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% arrange(desc(weight_lost_g)) %&gt;% mutate(weight_lost_rank = row_number())</code> adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight. First, the <code>arrange(desc(weight_lost_g))</code> function sorts the data frame by the <code>weight_lost_g</code> column in descending order, and then the <code>mutate(weight_lost_rank = row_number())</code> function adds a new column that assigns a rank to each row based on its position (row number) in the sorted data frame.</li>
<li>The error is that the conditions in the <code>case_when()</code> function are not separated by commas. Each condition should be followed by a comma because these are like the arguments in a function. Remeber that it’s optional to include the <code>TRUE ~ ...</code> condition at the end of the <code>case_when()</code> function.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% mutate(mouse_strain = case_when(mouse_strain == "Black 6" ~ "B6", TRUE ~ mouse_strain))</code> modifies the <code>mouse_strain</code> column so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged. As we are calling our column <code>mouse_strain</code>, no new column is being created (we are modifying the existing one) and the <code>TRUE ~ mouse_strain</code> condition acts as a fallback to keep the original values (that already exist in the <code>mouse_strain</code> column) for any rows that don’t match our first condition (strain being “Black 6”).</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-summarise" class="level3" data-number="2.2.6">
<h3 data-number="2.2.6" class="anchored" data-anchor-id="sec-summarise"><span class="header-section-number">2.2.6</span> Summarising data</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F08_summarise.png" class="img-fluid figure-img"></p>
<figcaption>Summarise allows you to calculate summary statistics that collapse many rows into one</figcaption>
</figure>
</div>
<p>The <code>summarise()</code> (or <code>summarize()</code>, if you prefer US spelling) function is used to calculate summary statistics on your data. It takes similar arguments to <code>mutate()</code>, but instead of adding a new column to the data frame, it returns a new data frame with a single row and one column for each summary statistic you calculate.</p>
<p>For example, to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  mean_weight_lost
             &lt;dbl&gt;
1             4.20</code></pre>
</div>
</div>
<p>We can also calculate multiple summary statistics at once. For example, to calculate the mean, median, and standard deviation of the weight lost by the mice:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  mean_weight_lost median_weight_lost sd_weight_lost
             &lt;dbl&gt;              &lt;dbl&gt;          &lt;dbl&gt;
1             4.20               4.05          0.802</code></pre>
</div>
</div>
<p>The power of summarising data is really seen when combined with grouping, which we will cover in the next section.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_tail =</span> <span class="fu">mean</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_tail =</span> <span class="fu">min</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_tail =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Calculates the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Produces a data frame containing one column for each of the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Finds the average tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Produces a vector containing the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is NOT a valid way to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% pull(weight_lost_g) %&gt;% mean(na.rm = TRUE)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarize(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>1. The code <strong>produces a data frame</strong> containing one column for each of the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p>
<p>2. The line of code that is NOT a valid way to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)</code>. This line of code is incorrect because the <code>mean()</code> function is being used directly on the data frame, rather than within a <code>summarise()</code> function. The other options are valid ways to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame (although note that the second option uses <code>pull()</code> to extract the <code>weight_lost_g</code> column as a vector before calculating the mean, so the mean value is stored in a vector rather than in a data frame).</p>
</details>
</div>
</div>
</section>
</section>
<section id="sec-grouping" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec-grouping"><span class="header-section-number">2.3</span> Grouping</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F09_grouped_summarise.png" class="img-fluid figure-img"></p>
<figcaption>Using summarise with the .by option allows you to create separate summaries for different ‘groups’ of data</figcaption>
</figure>
</div>
<p>Grouping is a powerful concept in in <code>dplyr</code> that allows you to perform operations on subsets of your data. For example, you might want to calculate the mean weight lost by mice in each cage, or find the mouse with the longest tail in each strain.</p>
<p>We can group data using the <code>.by</code> argument that exists in many dplyr functions, like <code>summarise()</code> and <code>mutate()</code>, and passing it the name(s) of column(s) to group by. For example, to calculate the mean weight lost by mice in each cage:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># don't forget it's .by, not by!</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 3E                      4.72
3 2B                      3.71</code></pre>
</div>
</div>
<p>Like when we first learned the summarise function above, we give our new column a name (<code>mean_weight_lost</code>), and then we assign its value to be the mean of the <code>weight_lost_g</code> column (with <code>NA</code>s removed). But this time, we also added the <code>.by</code> argument to specify the column we want to group by (<code>cage_number</code>, in this case). This will return a data frame with the mean weight lost by mice in each cage.</p>
<p>Grouping is a powerful tool for exploring your data and can help you identify patterns that might not be obvious when looking at the data as a whole. For example, notice how this grouped summary reveals that mice in cage 3E lost more weight than those in the other two cages.</p>
<p>It’s also possible to group by multiple columns by passing a vector of column names to the <code>.by</code> argument. For example, to calculate the mean weight lost by mice in each cage and strain:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># group by both cage_number and mouse_strain</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain mean_weight_lost
  &lt;chr&gt;       &lt;chr&gt;                   &lt;dbl&gt;
1 1A          CD-1                     3.71
2 3E          CD-1                     3.71
3 2B          CD-1                     3.69
4 3E          Black 6                  5.08
5 2B          BALB C                   3.73</code></pre>
</div>
</div>
<p>Of course, <code>mean()</code> is not the only function that we can use within <code>summarise()</code>. We can use any function that takes a vector of values and returns a single value, like <code>median()</code>, <code>sd()</code>, or <code>max()</code>. We can also use multiple functions at once, by giving each column a name and specifying the function we want to use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n =</span> <span class="fu">n</span>(),</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_weight_lost =</span> <span class="fu">max</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_weight_lost =</span> <span class="fu">min</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 7
  cage_number     n mean_weight_lost median_weight_lost sd_weight_lost
  &lt;chr&gt;       &lt;int&gt;            &lt;dbl&gt;              &lt;dbl&gt;          &lt;dbl&gt;
1 1A             52             3.71               3.7           0.445
2 3E            168             4.72               4.78          0.783
3 2B            124             3.71               3.69          0.417
# ℹ 2 more variables: max_weight_lost &lt;dbl&gt;, min_weight_lost &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Here, we also used the <code>n()</code> function to calculate the number of mice in each cage. This is a special helper function that works within <code>summarise</code> to count the number of rows in each group.</p>
<div class="callout callout-style-default callout-note callout-titled" title="To `.by` or not to `.by`?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
To <code>.by</code> or not to <code>.by</code>?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the <code>dplyr</code> package, there are two ways to group data: using the <code>.by</code> argument within various functions (as we have covered so far), or using the <code>group_by()</code> function, then performing your operations and ungrouping with <code>ungroup()</code>.</p>
<p>For example, we’ve seen above how to calculate the mean weight lost by mice in each cage using the <code>.by</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 3E                      4.72
3 2B                      3.71</code></pre>
</div>
</div>
<p>But we can also do the same using <code>group_by()</code> and <code>ungroup()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cage_number) <span class="sc">%&gt;%</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 2B                      3.71
3 3E                      4.72</code></pre>
</div>
</div>
<p>The two methods are equivalent, but using the <code>.by</code> argument within functions can be more concise and easier to read. Still, it’s good to be aware of <code>group_by()</code> and <code>ungroup()</code> as they are widely used, particularly in older code.</p>
</div>
</div>
<p>Although grouping is most often used with <code>summarise()</code>, it can be used with <code>dplyr</code> functions too. For example mutate() function can also be used with grouping to add new columns to the data frame based on group-specific calculations. Let’s say we wanted to calculate the Z-score (also known as the <a href="https://en.wikipedia.org/wiki/Standard_score">standard score</a>) to standardise the weight lost by each mouse within each strain.</p>
<p>As a reminder, the formula for calculating the Z-score is <span class="math inline">\(\frac{x - \mu}{\sigma}\)</span>, where <span class="math inline">\(x\)</span> is the value (in our case the <code>weight_lost_g</code> column), <span class="math inline">\(\mu\)</span> is the mean, and <span class="math inline">\(\sigma\)</span> is the standard deviation.</p>
<p>We can calculate this for each mouse in each strain using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove NAs before calculating the mean and SD</span></span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_z =</span> (weight_lost_g <span class="sc">-</span> <span class="fu">mean</span>(weight_lost_g)) <span class="sc">/</span> <span class="fu">sd</span>(weight_lost_g), <span class="at">.by =</span> mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns</span></span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, weight_lost_g, weight_lost_z)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 342 × 3
   mouse_strain weight_lost_g weight_lost_z
   &lt;chr&gt;                &lt;dbl&gt;         &lt;dbl&gt;
 1 CD-1                  3.75         0.108
 2 CD-1                  3.8          0.217
 3 CD-1                  3.25        -0.983
 4 CD-1                  3.45        -0.547
 5 CD-1                  3.65        -0.110
 6 CD-1                  3.62        -0.165
 7 CD-1                  4.68         2.12 
 8 CD-1                  3.48        -0.492
 9 CD-1                  4.25         1.20 
10 CD-1                  3.3         -0.874
# ℹ 332 more rows</code></pre>
</div>
</div>
<p>Unlike when we used <code>.by</code> with <code>summarise()</code>, we still get the same number of rows as the original data frame, but now we have a new column <code>weight_lost_z</code> that contains the Z-score for each mouse within each strain. This could be useful for identifying outliers or comparing the weight lost by each mouse to the average for its strain.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which line of code would you use to calculate the median tail length of mice belonging to each strain in the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), by = mouse_strain)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">max_tail_len =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, replicate))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice for each strain in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice for each replicate in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Calculates the maximum tail length of mice in each unique combination of strain and replicate in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. I want to count how many male and how many female mice there are for each strain in the <code>mousezempic_dosage_data</code> data frame. Which line of code would I use?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = sex)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain, sex)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>4. I want to find the proportion of weight lost <strong>by each mouse</strong> in each cage in the <code>mousezempic_dosage_data</code> data frame. Which line of code would I use?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE, .by = cage_number))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The correct line of code to calculate the median tail length of mice belonging to each strain in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code>. Remember to use <code>na.rm = TRUE</code> to remove any missing values before calculating the median, and to use <code>.by</code> to specify the column to group by (not <code>by</code>). Seeing as we want to calculate the median (collapse down to a single value per group), we need to use <code>summarise()</code> rather than <code>mutate()</code>.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% summarise(max_tail_len = max(tail_length_mm, na.rm = TRUE), .by = c(mouse_strain, replicate))</code> calculates the maximum tail length of mice in each <strong>unique combination</strong> of strain and replicate in the <code>mousezempic_dosage_data</code> data frame.</li>
<li>The correct line of code to count how many male and how many female mice there are for each strain in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))</code>. We need to group by both <code>mouse_strain</code> and <code>sex</code> to get the count for each unique combination of strain and sex. Don’t forget that we specify the column names as a vector when grouping by multiple columns.</li>
<li>The correct line of code to find the proportion of weight lost <strong>by each mouse</strong> in each cage in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code>. We use <code>mutate()</code> because we want a value for each mouse (each row in our data), rather than to collapse down to a single value for each group (cage number in this case). Be careful that you use the <code>.by</code> argument within the <code>mutate()</code> function call, not within the <code>sum()</code> function by mistake (this is what is wrong with the third option).</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-reshaping" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="sec-reshaping"><span class="header-section-number">2.4</span> Reshaping and combining data</h2>
<p>In the previous section, we learned how to filter, select, mutate, and summarise data. These are the most common operations you’ll perform on your data, and all generally revolve around working with specific columns or rows of your data frame. But what if you need to change the structure of your entire data frame? Re-shaping (with <code>pivot_</code> functions) and combining (with <code>join</code> functions) data are two more advanced operations that can help you do this.</p>
<section id="sec-pivot" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="sec-pivot"><span class="header-section-number">2.4.1</span> Reshaping data with pivot functions</h3>
<p>Pivoting is a way to change the structure of your data frame by rotating it so that rows become columns and vice versa. This can be useful when you want to change the way your data is organised, for example, to make it easier to plot or analyse.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F10_pivot.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption>Pivot functions allow you to change the structure of your data frame</figcaption>
</figure>
</div>
<p>The <code>pivot_longer()</code> function is used to pivot data from wide to long format, and the <code>pivot_wider()</code> function is used to pivot data from long to wide format.</p>
<section id="pivot-wider" class="level4" data-number="2.4.1.1">
<h4 data-number="2.4.1.1" class="anchored" data-anchor-id="pivot-wider"><span class="header-section-number">2.4.1.1</span> Pivot wider</h4>
<p>A common use case for <code>pivot_wider()</code> is to make a <a href="https://en.wikipedia.org/wiki/Contingency_table">contingency table</a>, which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</p>
<p>For example, let’s say we want to create a table that shows how many mice there are of each strain, in each cage number. We can achieve this in a long format using <code>summarise()</code> as we learned in the previous section:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain n_mice
  &lt;chr&gt;       &lt;chr&gt;         &lt;int&gt;
1 1A          CD-1             52
2 3E          CD-1             44
3 2B          CD-1             56
4 3E          Black 6         124
5 2B          BALB C           68</code></pre>
</div>
</div>
<p>Notice how we have a row for each unique combination of <code>cage_number</code> and <code>mouse_strain</code>. Instead, we can use <code>pivot_wider()</code> to create a table where each row represents a unique <code>cage_number</code> and each column represents a unique <code>mouse_strain</code>.</p>
<p>To use a <code>pivot_wider()</code> you need to tell R two bits of information:</p>
<ol type="1">
<li>What the new columns should be (with <code>names_from</code>)</li>
<li>Where the values in these columns should come from (with <code>values_from</code>).</li>
</ol>
<p>In this case, we want the mouse strains to be our new columns (so <code>names_from = mouse_strain</code>), and the average weight lost to be the values in the table (<code>values_from = n_mice</code>). Putting it all together, here is the code to make our summary table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  cage_number `CD-1` `Black 6` `BALB C`
  &lt;chr&gt;        &lt;int&gt;     &lt;int&gt;    &lt;int&gt;
1 1A              52        NA       NA
2 3E              44       124       NA
3 2B              56        NA       68</code></pre>
</div>
</div>
<p>Notice how <code>NA</code> values are used where there is no data for a particular combination of <code>cage_number</code> and <code>mouse_strain</code>.</p>
</section>
<section id="pivot-longer" class="level4" data-number="2.4.1.2">
<h4 data-number="2.4.1.2" class="anchored" data-anchor-id="pivot-longer"><span class="header-section-number">2.4.1.2</span> Pivot longer</h4>
<p>While wide data can be easier to read, long data is often easier to work with. That’s because a lot of the tidyverse functions we’ve learned so far (like <code>filter()</code>, <code>mutate()</code>, and <code>summarise()</code>) work in a row-wise fashion, which is more natural with long data.</p>
<p>To demonstrate <code>pivot_longer()</code>, we will introduce a new data frame called <code>mousezempic_expression_data</code>. This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 453 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (1): group
dbl (10): id_num, TH_rep1, TH_rep2, TH_rep3, PRLH_rep1, PRLH_rep2, PRLH_rep3...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print it out</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 453 × 11
   id_num TH_rep1 TH_rep2 TH_rep3 PRLH_rep1 PRLH_rep2 PRLH_rep3 HPRT1_rep1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1      1   1431.   2737.   2036.      297.      428.      320.       741.
 2      2   2699.   4180.   3860.      388.      680.      500.       843.
 3      3   3596.   5194.   4128.      201.      346.      266.      1047.
 4      4   2723.   4442.   3728.      317.      576.      478.       541.
 5      5   2197.   3656.   2887.      114.      205.      173.       798.
 6      6   2316.   3949.   3355.      242.      383.      305.       763.
 7      7   1570.   2740.   2129.      294.      489.      427.       796.
 8      8   2683.   4749.   3563.      231.      400.      309.       922.
 9      9   3023.   5170.   3830.      289.      464.      390.       921.
10     10   3254.   5980.   4250.      337.      583.      436.       740.
# ℹ 443 more rows
# ℹ 3 more variables: HPRT1_rep2 &lt;dbl&gt;, HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<p>As you can see, the data is in wide format, with each row representing a different mouse (identified by its <code>id_num</code>) and each column representing a different measurement of a gene. To put this data into a long format (where each measurement is contained on a separate row), we can use <code>pivot_longer()</code>, specifying three arguments:</p>
<ol type="1">
<li><code>cols</code>: the columns to pivot from. You can use selection helpers like <code>contains()</code> or <code>starts_with()</code> to easily select multiple columns at once.</li>
<li><code>names_to</code>: the name of a new column that will contain the old column names.</li>
<li><code>values_to</code>: the name of a new column that will contain the values from the old columns.</li>
</ol>
<p>In this particular case here’s what the code would look like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="sc">%&gt;%</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>), <span class="at">names_to =</span> <span class="st">"measurement"</span>, <span class="at">values_to =</span> <span class="st">"expression_level"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,077 × 4
   id_num group     measurement expression_level
    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;                  &lt;dbl&gt;
 1      1 treatment TH_rep1                1431.
 2      1 treatment TH_rep2                2737.
 3      1 treatment TH_rep3                2036.
 4      1 treatment PRLH_rep1               297.
 5      1 treatment PRLH_rep2               428.
 6      1 treatment PRLH_rep3               320.
 7      1 treatment HPRT1_rep1              741.
 8      1 treatment HPRT1_rep2             1300.
 9      1 treatment HPRT1_rep3              988.
10      2 treatment TH_rep1                2699.
# ℹ 4,067 more rows</code></pre>
</div>
</div>
<p>Let’s go through those three arguments:</p>
<ol type="1">
<li><code>cols = contains("_rep")</code>: This tells R to pivot all columns that contain the string “_rep”. In this case, that’s the columns <code>TH_rep1</code>, <code>TH_rep2</code>, <code>TH_rep3</code>, <code>PRLH_rep1</code>, <code>PRLH_rep2</code>, <code>PRLH_rep3</code>, <code>HPRT1_rep1</code>, <code>HPRT1_rep2</code>, and <code>HPRT1_rep3</code>.</li>
<li><code>names_to = "measurement"</code>: This tells R to create a new column called ‘measurement’ that contains the names of the old columns. So, for example, the value of the ‘measurement’ column for the first row will be ‘TH_rep1’.</li>
<li><code>values_to = "expression_level"</code>: This tells R to create a new column called ‘expression_level’ that contains the values from the old columns. So, for example, the value of the ‘expression_level’ column for the first row will be the value of the <code>TH_rep1</code> column.</li>
</ol>
<p>After running this code, you’ll see that the data frame is now in long format, with each row representing a different measurement of a gene for a different mouse.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Perplexed by pivoting?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Perplexed by pivoting?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Pivoting can be a bit tricky to get your head around! Often when you’re doing analysis, you’ll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <a href="https://r4ds.hadley.nz/data-tidy.html#sec-pivoting">like those in the R for Data Science book</a>, or to just experiment with different arguments until you get the result you want.</p>
</div>
</div>
</section>
</section>
<section id="sec-join" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="sec-join"><span class="header-section-number">2.4.2</span> Combining data with join functions</h3>
<p>Another way you might need to change the structure of your data is by combining it with other data frames. This is where the <code>join</code> functions come in. There are many different types of joins in <code>dplyr</code> (which you can read about here if you’re interested), but the most useful one is the <code>left_join()</code> function.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F11_left_join.png" class="img-fluid figure-img"></p>
<figcaption>The left_join() function allows you to combine two data frames</figcaption>
</figure>
</div>
<p><code>left_join()</code> is used to combine two data frames based on a common column. It takes three arguments - the first data frame, the second data frame, and the common column to join on. The reason it’s called a ‘left join’ is because it keeps all the rows from the first data frame (the ‘left’ data frame), and only adds rows from the second data frame (the ‘right’ data frame) if they match the common column.</p>
<p>For example, we can join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column, which is common to both data frames:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># example of using it in a pipe</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here mousezempic_dosage_data is the left data frame (first) data frame passed on by the pipe</span></span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="st">"id_num"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 19
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 13 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, TH_rep1 &lt;dbl&gt;, TH_rep2 &lt;dbl&gt;, TH_rep3 &lt;dbl&gt;, PRLH_rep1 &lt;dbl&gt;,
#   PRLH_rep2 &lt;dbl&gt;, PRLH_rep3 &lt;dbl&gt;, HPRT1_rep1 &lt;dbl&gt;, HPRT1_rep2 &lt;dbl&gt;,
#   HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<p>This code will return a new data frame that contains all the columns from <code>mousezempic_dosage_data</code> and all the columns from <code>mousezempic_expression_data</code>, with the rows matched based on the <code>id_num</code> column. If there is no match for a particular <code>id_num</code> in the <code>mousezempic_expression_data</code> data frame, the columns from that data frame will be filled with <code>NA</code>. Rows that only appear in the <code>mousezempic_expression_data</code> data frame will not be included in the final data frame.</p>
<p>We can also use <code>left_join()</code> when our columns have different names (but contain the same value) using a named vector the format <code>c("name1" = "name2")</code>, where ‘name1’ is the name of the column in the left/first data frame (the one being piped in) and ‘name2’ is the name of the corresponding column in the right/second data frame.</p>
<p>For example, if the <code>id_num</code> column in the <code>mousezempic_dosage_data</code> data frame was called <code>mouse_id</code> instead, we could still join the two data frames using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># for the sake of this example, make a new column called mouse_id</span></span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_id =</span> id_num) <span class="sc">%&gt;%</span></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join the two data frames where mouse_id in the first data frame matches id_num in the second data frame</span></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"mouse_id"</span> <span class="ot">=</span> <span class="st">"id_num"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 20
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 14 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, mouse_id &lt;dbl&gt;, TH_rep1 &lt;dbl&gt;, TH_rep2 &lt;dbl&gt;, TH_rep3 &lt;dbl&gt;,
#   PRLH_rep1 &lt;dbl&gt;, PRLH_rep2 &lt;dbl&gt;, PRLH_rep3 &lt;dbl&gt;, HPRT1_rep1 &lt;dbl&gt;,
#   HPRT1_rep2 &lt;dbl&gt;, HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What does the following code do?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Pivots data into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots into a wide format where there is a column for each mouse strain.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>It just gives an error</p><p></p>
</div>
</div>

</div>
</div>
<p>2. I have run the following code to create a new column in the <code>mousezempic_dosage_data</code> data frame that gives the weight of the mice at the end of the experiment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   id_num initial_weight_g final_weight_g
    &lt;dbl&gt;            &lt;dbl&gt;          &lt;dbl&gt;
 1      1             39.1           35.4
 2      2             39.5           35.7
 3      3             40.3           37.0
 4      4             NA             NA  
 5      5             36.7           33.2
 6      6             39.3           35.6
 7      7             38.9           35.3
 8      8             39.2           34.5
 9      9             34.1           30.6
10     10             42             37.8
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</p>
<div class="choices">
<div class="choice">
<p><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")</code></p>
</div>
<div class="choice">
<p><code>pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")</code></p>
</div>
<div class="choice">
<p><code>pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)</code></p>
</div>
<div class="choice correct-choice">
<p><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code></p>
</div>
</div>
<div class="question">
<p>3. Which of the following is NOT a valid way to join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>left_join(mousezempic_dosage_data, mousezempic_expression_data, by = "id_num")</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = ("id_num" = "id_num"))</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>1. The code first calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <code>names_from = sex</code> )</p>
<p>2. The correct pivot function call to take the data from a wide format to a long format is <code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code>. This code tells R to pivot the <code>initial_weight_g</code> and <code>final_weight_g</code> columns into a long format, where there is a row for each mouse and each weight measurement. The <code>names_to</code> argument specifies to make a column called ‘timepoint’ that tells us whether the measurement is initial or final, and the <code>values_to</code> argument specifies the name of the new column that will contain these measurements.</p>
<p>3. The line of code that is NOT a valid way to join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column is <code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")</code>. This line of code is incorrect because the <code>.by</code> argument is not used in the <code>left_join()</code> function (this can be confusing! it’s <code>.by</code> when grouping by <code>by</code> when joining). The other options are valid ways to join the two data frames based on the <code>id_num</code> column: remember that we don’t have to use pipes to join data frames, we can use the <code>left_join()</code> function directly, and we can use a named vector to specify the columns to join on (although here it’s a bit redundant as the columns have the same name).</p>
</details>
</div>
</div>
</section>
</section>
<section id="sec-saving" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="sec-saving"><span class="header-section-number">2.5</span> Saving data to a file</h2>
<p>Once you’ve cleaned and transformed your data, you’ll often want to save it to a file so that you can use it in other programs or share it with others. The <code>write_csv()</code> and <code>write_tsv()</code> functions from the <code>readr</code> package are a great way to do this. They take two arguments - the data frame you want to save and the file path where you want to save it.</p>
<p>For example, let’s say I want to save my contingency table of the number of mice in each cage and strain to a CSV file called <code>cage_strain_counts.csv</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create the contingency table</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>cage_strain_counts <span class="ot">&lt;-</span> mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the data to a CSV file</span></span>
<span id="cb131-9"><a href="#cb131-9" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(cage_strain_counts, <span class="st">"cage_strain_counts.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>CSV files are particularly great because they can be easily read into other software, like Excel.</p>
<p>It’s also possible to use the <code>write_*()</code> functions along with a pipe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice) <span class="sc">%&gt;%</span></span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_csv</span>(<span class="st">"cage_strain_counts.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remember here that the first argument (the data frame to save) is passed on by the pipe, so the only argument in the brackets is the second one: the file path.</p>
</section>
<section id="summary" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="summary"><span class="header-section-number">2.6</span> Summary</h2>
<p>Here’s what we’ve covered in this session:</p>
<ul>
<li><p>The pipe operator <code>%&gt;%</code> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</p></li>
<li><p>The basic dplyr verbs <code>filter()</code>, <code>select()</code>, <code>mutate()</code>, and <code>arrange()</code> and how they can be used to tidy and analyse data.</p></li>
<li><p>Missing values (<code>NA</code>) and how to remove or replace them</p></li>
<li><p>The <code>summarise()</code> function and how it can be used to calculate summary statistics on your data, as well as the power of grouping data with the <code>.by</code> argument.</p></li>
<li><p>Reshaping data with <code>pivot_longer()</code> and <code>pivot_wider()</code> to change the structure of your data frame.</p></li>
<li><p>Combining data with <code>left_join()</code> to merge two data frames based on a common column.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="Why does data need to be tidy anyway?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why does data need to be tidy anyway?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this session, we’ve been focusing on making our data ‘tidy’: that is, structured in a consistent way that makes it easy to work with. A nice visual illustration of tidy data and its importance can be <a href="https://allisonhorst.com/other-r-fun">found here</a>.</p>
</div>
</div>
<section id="practice-questions" class="level3" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="practice-questions"><span class="header-section-number">2.6.1</span> Practice questions</h3>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session_1.html" class="pagination-link" aria-label="Session 1: Introduction to R">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb133" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this session:"</span></span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session 2: Working with data {#sec-session02}</span></span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a>In this session we will learn how to manipulate and summarise data using the <span class="in">`dplyr`</span> package.</span>
<span id="cb133-14"><a href="#cb133-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-15"><a href="#cb133-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb133-16"><a href="#cb133-16" aria-hidden="true" tabindex="-1"></a>At the end of this session, learners should be able to:</span>
<span id="cb133-17"><a href="#cb133-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-18"><a href="#cb133-18" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Use the pipe (<span class="in">`%&gt;%`</span>) to chain multiple functions together</span>
<span id="cb133-19"><a href="#cb133-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-20"><a href="#cb133-20" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Design chains of dplyr functions to manipulate data frames</span>
<span id="cb133-21"><a href="#cb133-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-22"><a href="#cb133-22" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Apply grouping for more complex analysis of data</span>
<span id="cb133-23"><a href="#cb133-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-24"><a href="#cb133-24" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Understand how pivot and join functions can be used to reshape and combine entire data frames</span>
<span id="cb133-25"><a href="#cb133-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-26"><a href="#cb133-26" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Recall how to save data frames to a file</span>
<span id="cb133-27"><a href="#cb133-27" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-28"><a href="#cb133-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-29"><a href="#cb133-29" aria-hidden="true" tabindex="-1"></a>Make sure you have <span class="in">`dplyr`</span> package loaded by running <span class="in">`library(dplyr)`</span>, otherwise you will not be able to run the code in this session.</span>
<span id="cb133-30"><a href="#cb133-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-33"><a href="#cb133-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-34"><a href="#cb133-34" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load dplyr</span></span>
<span id="cb133-35"><a href="#cb133-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb133-36"><a href="#cb133-36" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-37"><a href="#cb133-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-38"><a href="#cb133-38" aria-hidden="true" tabindex="-1"></a>You'll see some output from the <span class="in">`dplyr`</span> package when you load it, which is just telling you that some of <span class="in">`dplyr`</span>'s functions have the same name as other functions in R.</span>
<span id="cb133-39"><a href="#cb133-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-40"><a href="#cb133-40" aria-hidden="true" tabindex="-1"></a><span class="fu">## Chaining functions together with pipes {#sec-pipes}</span></span>
<span id="cb133-41"><a href="#cb133-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-42"><a href="#cb133-42" aria-hidden="true" tabindex="-1"></a>Pipes are a powerful feature of the <span class="in">`tidyverse`</span> that allow you to chain multiple functions together. Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</span>
<span id="cb133-43"><a href="#cb133-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-44"><a href="#cb133-44" aria-hidden="true" tabindex="-1"></a>For example, take the following code:</span>
<span id="cb133-45"><a href="#cb133-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-48"><a href="#cb133-48" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-49"><a href="#cb133-49" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb133-50"><a href="#cb133-50" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span>
<span id="cb133-51"><a href="#cb133-51" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-52"><a href="#cb133-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-53"><a href="#cb133-53" aria-hidden="true" tabindex="-1"></a>What do you think this code does? It calculates the mean of <span class="in">`my_vector`</span>, rounds the result to the nearest whole number, and then converts the result to a character. But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</span>
<span id="cb133-54"><a href="#cb133-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-55"><a href="#cb133-55" aria-hidden="true" tabindex="-1"></a>Instead, we can use the pipe operator (<span class="in">`%&gt;%`</span>) to chain these functions together in a more readable way:</span>
<span id="cb133-56"><a href="#cb133-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-59"><a href="#cb133-59" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-60"><a href="#cb133-60" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb133-61"><a href="#cb133-61" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb133-62"><a href="#cb133-62" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-63"><a href="#cb133-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-64"><a href="#cb133-64" aria-hidden="true" tabindex="-1"></a>See how the code reads naturally from left to right? You can think of the pipe as being like the phrase "and then". Here, we're telling R: "Take <span class="in">`my_vector`</span>, and then calculate the mean, and then round the result, and then convert it to a character."</span>
<span id="cb133-65"><a href="#cb133-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-66"><a href="#cb133-66" aria-hidden="true" tabindex="-1"></a>You'll notice that we didn't need to specify the input to each function. That's because the pipe automatically passes the output of the previous function as the first input to the next function. We can still specify additional arguments to each function if we need to. For example, if we wanted to round the mean to 2 decimal places, we could do this:</span>
<span id="cb133-67"><a href="#cb133-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-70"><a href="#cb133-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-71"><a href="#cb133-71" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb133-72"><a href="#cb133-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-73"><a href="#cb133-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-74"><a href="#cb133-74" aria-hidden="true" tabindex="-1"></a>R is clever enough to know that the first argument to <span class="in">`round()`</span> is still the output of <span class="in">`mean()`</span>, even though we've now specified the <span class="in">`digits`</span> argument.</span>
<span id="cb133-75"><a href="#cb133-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-76"><a href="#cb133-76" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Plenty of pipes"}</span>
<span id="cb133-77"><a href="#cb133-77" aria-hidden="true" tabindex="-1"></a>There is another style of pipe in R, called the 'base R pipe' <span class="in">`|&gt;`</span>, which is available in R version 4.1.0 and later. The base R pipe works in a similar way to the <span class="in">`magrittr`</span> pipe (<span class="in">`%&gt;%`</span>) that we use in this course, but it is not as flexible. We recommend using the <span class="in">`magrittr`</span> pipe for now.</span>
<span id="cb133-78"><a href="#cb133-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-79"><a href="#cb133-79" aria-hidden="true" tabindex="-1"></a>Fun fact: the <span class="in">`magrittr`</span> package is named after the <span class="co">[</span><span class="ot">artist René Magritte, who made a famous painting of a pipe</span><span class="co">](https://en.wikipedia.org/wiki/The_Treachery_of_Images)</span>.</span>
<span id="cb133-80"><a href="#cb133-80" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-81"><a href="#cb133-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-82"><a href="#cb133-82" aria-hidden="true" tabindex="-1"></a>To type the pipe operator more easily, you can use the keyboard shortcut {{&lt; kbd Cmd-shift-M &gt;}} (although once you get used to it, you might find it easier to type <span class="in">`%&gt;%`</span> manually).</span>
<span id="cb133-83"><a href="#cb133-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-84"><a href="#cb133-84" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-85"><a href="#cb133-85" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-86"><a href="#cb133-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-87"><a href="#cb133-87" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-88"><a href="#cb133-88" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What is NOT a valid way to re-write the following code using the pipe operator: <span class="in">`round(sqrt(sum(1:10)), 1)`</span>. If you're not sure, try running the different options in the console to see which one gives the same answer.</span>
<span id="cb133-89"><a href="#cb133-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-90"><a href="#cb133-90" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-91"><a href="#cb133-91" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-92"><a href="#cb133-92" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)`</span></span>
<span id="cb133-93"><a href="#cb133-93" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-94"><a href="#cb133-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-95"><a href="#cb133-95" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-96"><a href="#cb133-96" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span></span>
<span id="cb133-97"><a href="#cb133-97" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-98"><a href="#cb133-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-99"><a href="#cb133-99" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-100"><a href="#cb133-100" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb133-101"><a href="#cb133-101" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-102"><a href="#cb133-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-103"><a href="#cb133-103" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-104"><a href="#cb133-104" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb133-105"><a href="#cb133-105" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-106"><a href="#cb133-106" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-107"><a href="#cb133-107" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-108"><a href="#cb133-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-109"><a href="#cb133-109" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-110"><a href="#cb133-110" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is the output of the following code? <span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> Try to guess it before copy-pasting into R.</span>
<span id="cb133-111"><a href="#cb133-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-112"><a href="#cb133-112" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-113"><a href="#cb133-113" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-114"><a href="#cb133-114" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"`</span></span>
<span id="cb133-115"><a href="#cb133-115" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-116"><a href="#cb133-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-117"><a href="#cb133-117" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-118"><a href="#cb133-118" aria-hidden="true" tabindex="-1"></a><span class="in">`"a" "b" "c" "d" "e" "f"`</span></span>
<span id="cb133-119"><a href="#cb133-119" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-120"><a href="#cb133-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-121"><a href="#cb133-121" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-122"><a href="#cb133-122" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb133-123"><a href="#cb133-123" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-124"><a href="#cb133-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-125"><a href="#cb133-125" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-126"><a href="#cb133-126" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F"`</span></span>
<span id="cb133-127"><a href="#cb133-127" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-128"><a href="#cb133-128" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-129"><a href="#cb133-129" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-130"><a href="#cb133-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-131"><a href="#cb133-131" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-132"><a href="#cb133-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-133"><a href="#cb133-133" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-134"><a href="#cb133-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-135"><a href="#cb133-135" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb133-136"><a href="#cb133-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-137"><a href="#cb133-137" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The invalid option is <span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span>. This is because the <span class="in">`sqrt()`</span> function only takes one argument, so you can't specify <span class="in">`1`</span> as an argument in addition to what is being piped in from <span class="in">`sum(1:10)`</span>. Note that some options used the pipe to send <span class="in">`1:10`</span> to <span class="in">`sum()`</span> (like <span class="in">`1:10 %&gt;% sum()`</span>), and others just used <span class="in">`sum(1:10)`</span> directly. Both are valid ways to use the pipe, it's just a matter of personal preference.</span>
<span id="cb133-138"><a href="#cb133-138" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The output of the code <span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> is <span class="in">`"A" "B" "C" "D" "E" "F"`</span>. The <span class="in">`letters`</span> vector contains the lowercase alphabet, and the <span class="in">`head()`</span> function returns the first 6 elements of the vector. Finally, the <span class="in">`toupper()`</span> function then converts these elements to uppercase.</span>
<span id="cb133-139"><a href="#cb133-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-140"><a href="#cb133-140" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb133-141"><a href="#cb133-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-142"><a href="#cb133-142" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-143"><a href="#cb133-143" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-144"><a href="#cb133-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-145"><a href="#cb133-145" aria-hidden="true" tabindex="-1"></a><span class="fu">## Basic data manipulation {#sec-dataManip}</span></span>
<span id="cb133-146"><a href="#cb133-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-147"><a href="#cb133-147" aria-hidden="true" tabindex="-1"></a>To really see the power of the pipe, we will use it together with the <span class="in">`dplyr`</span> package that provides a set of functions to easily filter, sort, select, and summarise data frames. These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</span>
<span id="cb133-148"><a href="#cb133-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-149"><a href="#cb133-149" aria-hidden="true" tabindex="-1"></a>For example, even though we haven't covered the <span class="in">`dplyr`</span> functions yet, you can probably guess what the following code does:</span>
<span id="cb133-150"><a href="#cb133-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-153"><a href="#cb133-153" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-154"><a href="#cb133-154" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-155"><a href="#cb133-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-156"><a href="#cb133-156" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb133-157"><a href="#cb133-157" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-158"><a href="#cb133-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-159"><a href="#cb133-159" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb133-160"><a href="#cb133-160" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span>
<span id="cb133-161"><a href="#cb133-161" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-162"><a href="#cb133-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-163"><a href="#cb133-163" aria-hidden="true" tabindex="-1"></a>This code filters the <span class="in">`mousezempic_dosage_data`</span> data frame to only include data from cage 3E, then pulls out the <span class="in">`weight_lost_g`</span> column, and finally calculates the mean of the values in that column. The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <span class="in">`pull()`</span>) are specified in the brackets (like <span class="in">`round(digits = 2)`</span> from the previous example).</span>
<span id="cb133-164"><a href="#cb133-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-165"><a href="#cb133-165" aria-hidden="true" tabindex="-1"></a>We also used the enter key after each pipe <span class="in">`%&gt;%`</span> to break up the code into multiple lines to make it easier to read. This isn't required, but is a popular style in the R community, so all the code examples in this session will follow this format.</span>
<span id="cb133-166"><a href="#cb133-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-167"><a href="#cb133-167" aria-hidden="true" tabindex="-1"></a>We will now introduce some of the most commonly used <span class="in">`dplyr`</span> functions for manipulating data frames. To showcase these, we will use the <span class="in">`mousezempic_dosage_data`</span> that we practiced reading in last session. This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</span>
<span id="cb133-168"><a href="#cb133-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-171"><a href="#cb133-171" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-172"><a href="#cb133-172" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-173"><a href="#cb133-173" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in session 1</span></span>
<span id="cb133-174"><a href="#cb133-174" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb133-175"><a href="#cb133-175" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"~/Desktop/R_course/data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb133-176"><a href="#cb133-176" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-177"><a href="#cb133-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-180"><a href="#cb133-180" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-181"><a href="#cb133-181" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb133-182"><a href="#cb133-182" aria-hidden="true" tabindex="-1"></a><span class="co"># just for rendering the book not for students to see</span></span>
<span id="cb133-183"><a href="#cb133-183" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb133-184"><a href="#cb133-184" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb133-185"><a href="#cb133-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-186"><a href="#cb133-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-187"><a href="#cb133-187" aria-hidden="true" tabindex="-1"></a>Before we start, let's use what we learned in the previous session to take a look at <span class="in">`mousezempic_dosage_data`</span>:</span>
<span id="cb133-188"><a href="#cb133-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-191"><a href="#cb133-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-192"><a href="#cb133-192" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb133-193"><a href="#cb133-193" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data</span>
<span id="cb133-194"><a href="#cb133-194" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-195"><a href="#cb133-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-196"><a href="#cb133-196" aria-hidden="true" tabindex="-1"></a>You might also like to use <span class="in">`View()`</span> to open the data in a separate window and get a closer look.</span>
<span id="cb133-197"><a href="#cb133-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-198"><a href="#cb133-198" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Using RStudio autocomplete"}</span>
<span id="cb133-199"><a href="#cb133-199" aria-hidden="true" tabindex="-1"></a>Although it's great to give our data a descriptive name like <span class="in">`mousezempic_dosage_data`</span>, it can be a bit of a pain to type out every time. Luckily, RStudio has a handy autocomplete feature that can solve this problem. Just start typing the name of the object, and you'll see it will popup:</span>
<span id="cb133-200"><a href="#cb133-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-201"><a href="#cb133-201" aria-hidden="true" tabindex="-1"></a><span class="al">![RStudio autocomplete](images/S02F01_rstudio_autocomplete.png)</span></span>
<span id="cb133-202"><a href="#cb133-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-203"><a href="#cb133-203" aria-hidden="true" tabindex="-1"></a>You can then press {{&lt; kbd Tab &gt;}} to autocomplete it. If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</span>
<span id="cb133-204"><a href="#cb133-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-205"><a href="#cb133-205" aria-hidden="true" tabindex="-1"></a>Try using autocomplete this session to save yourself some typing!</span>
<span id="cb133-206"><a href="#cb133-206" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-207"><a href="#cb133-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-208"><a href="#cb133-208" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sorting data {#sec-sorting}</span></span>
<span id="cb133-209"><a href="#cb133-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-210"><a href="#cb133-210" aria-hidden="true" tabindex="-1"></a>Often, one of the first things you might want to do with a dataset is sort it. In <span class="in">`dplyr`</span>, this is called 'arranging' and is done with the <span class="in">`arrange()`</span> function.</span>
<span id="cb133-211"><a href="#cb133-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-212"><a href="#cb133-212" aria-hidden="true" tabindex="-1"></a><span class="al">![Arrange orders rows by their values in one or more columns](images/S02F02_arrange.png)</span></span>
<span id="cb133-213"><a href="#cb133-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-214"><a href="#cb133-214" aria-hidden="true" tabindex="-1"></a>By default, <span class="in">`arrange()`</span> sorts in ascending order (smallest values first). For example, let's sort the <span class="in">`mousezempic_dosage_data`</span> data frame by the <span class="in">`weight_lost_g`</span> column:</span>
<span id="cb133-215"><a href="#cb133-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-218"><a href="#cb133-218" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-219"><a href="#cb133-219" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-220"><a href="#cb133-220" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span>
<span id="cb133-221"><a href="#cb133-221" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-222"><a href="#cb133-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-223"><a href="#cb133-223" aria-hidden="true" tabindex="-1"></a>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</span>
<span id="cb133-224"><a href="#cb133-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-225"><a href="#cb133-225" aria-hidden="true" tabindex="-1"></a>Sometimes you might want to sort in descending order instead (largest values first). You can do this by putting the <span class="in">`desc()`</span> function around your column name, inside <span class="in">`arrange()`</span>:</span>
<span id="cb133-226"><a href="#cb133-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-229"><a href="#cb133-229" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-230"><a href="#cb133-230" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-231"><a href="#cb133-231" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb133-232"><a href="#cb133-232" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb133-233"><a href="#cb133-233" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-234"><a href="#cb133-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-235"><a href="#cb133-235" aria-hidden="true" tabindex="-1"></a>Now we can see the mice that lost the most weight are at the top.</span>
<span id="cb133-236"><a href="#cb133-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-237"><a href="#cb133-237" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Comments and pipes"}</span>
<span id="cb133-238"><a href="#cb133-238" aria-hidden="true" tabindex="-1"></a>Notice how in the previous example we have written a comment in the middle of the pipe chain. This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won't cause any errors as long as you make sure that the comment is on its own line.</span>
<span id="cb133-239"><a href="#cb133-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-240"><a href="#cb133-240" aria-hidden="true" tabindex="-1"></a>You can also write comments at the end of the line, just make sure it's after the pipe operator <span class="in">`%&gt;%`</span>.</span>
<span id="cb133-241"><a href="#cb133-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-242"><a href="#cb133-242" aria-hidden="true" tabindex="-1"></a>For example, these comments are allowed:</span>
<span id="cb133-243"><a href="#cb133-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-246"><a href="#cb133-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-247"><a href="#cb133-247" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb133-248"><a href="#cb133-248" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb133-249"><a href="#cb133-249" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb133-250"><a href="#cb133-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-251"><a href="#cb133-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-252"><a href="#cb133-252" aria-hidden="true" tabindex="-1"></a>But this will cause an error, because the <span class="in">`#`</span> is before the pipe, so R treats it as part of the comment (notice how the <span class="in">`%&gt;%`</span> has changed colour?) and doesn't know how the two lines relate to each other. It tries to run them separately, which for the first line is ok (it will just print <span class="in">`mousezempic_dosage_data`</span>):</span>
<span id="cb133-253"><a href="#cb133-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-256"><a href="#cb133-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-257"><a href="#cb133-257" aria-hidden="true" tabindex="-1"></a><span class="co">#| error: true</span></span>
<span id="cb133-258"><a href="#cb133-258" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># this comment will cause an error %&gt;%</span></span>
<span id="cb133-259"><a href="#cb133-259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb133-260"><a href="#cb133-260" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-261"><a href="#cb133-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-262"><a href="#cb133-262" aria-hidden="true" tabindex="-1"></a>But for the second line, there is an error that R doesn't know what the <span class="in">`weight_lost_g`</span> object is. That's because it's a column in the <span class="in">`mousezempic_dosage_data`</span> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</span>
<span id="cb133-263"><a href="#cb133-263" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-264"><a href="#cb133-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-265"><a href="#cb133-265" aria-hidden="true" tabindex="-1"></a>You can also sort by multiple columns by passing multiple column names to <span class="in">`arrange()`</span>. For example, to sort by the strain first and then by the amount of weight lost:</span>
<span id="cb133-266"><a href="#cb133-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-269"><a href="#cb133-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-270"><a href="#cb133-270" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb133-271"><a href="#cb133-271" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-272"><a href="#cb133-272" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span>
<span id="cb133-273"><a href="#cb133-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-274"><a href="#cb133-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-275"><a href="#cb133-275" aria-hidden="true" tabindex="-1"></a>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</span>
<span id="cb133-276"><a href="#cb133-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-277"><a href="#cb133-277" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Piping into View()"}</span>
<span id="cb133-278"><a href="#cb133-278" aria-hidden="true" tabindex="-1"></a>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn't allow us to see the full effect of the sorting.</span>
<span id="cb133-279"><a href="#cb133-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-280"><a href="#cb133-280" aria-hidden="true" tabindex="-1"></a>One handy trick you can use with pipes is to add <span class="in">`View()`</span> at the end of your chain to open the data in a separate window. Try running this code, and you'll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</span>
<span id="cb133-281"><a href="#cb133-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-284"><a href="#cb133-284" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-285"><a href="#cb133-285" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-286"><a href="#cb133-286" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb133-287"><a href="#cb133-287" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-288"><a href="#cb133-288" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb133-289"><a href="#cb133-289" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span>
<span id="cb133-290"><a href="#cb133-290" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-291"><a href="#cb133-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-292"><a href="#cb133-292" aria-hidden="true" tabindex="-1"></a>This is a great way to check that your code has actually done what you intended!</span>
<span id="cb133-293"><a href="#cb133-293" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-294"><a href="#cb133-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-295"><a href="#cb133-295" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Extracting rows with the smallest or largest values {#sec-sliceMinMax}</span></span>
<span id="cb133-296"><a href="#cb133-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-297"><a href="#cb133-297" aria-hidden="true" tabindex="-1"></a>Slice functions are used to select rows based on their position in the data frame. The <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</span>
<span id="cb133-298"><a href="#cb133-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-299"><a href="#cb133-299" aria-hidden="true" tabindex="-1"></a>This is equivalent to using <span class="in">`arrange()`</span> followed by <span class="in">`head()`</span>, but is more concise:</span>
<span id="cb133-300"><a href="#cb133-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-303"><a href="#cb133-303" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-304"><a href="#cb133-304" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb133-305"><a href="#cb133-305" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-306"><a href="#cb133-306" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb133-307"><a href="#cb133-307" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb133-308"><a href="#cb133-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-309"><a href="#cb133-309" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb133-310"><a href="#cb133-310" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-311"><a href="#cb133-311" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb133-312"><a href="#cb133-312" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb133-313"><a href="#cb133-313" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-314"><a href="#cb133-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-315"><a href="#cb133-315" aria-hidden="true" tabindex="-1"></a>But wait— neither of those pieces of code actually gave the number of rows we asked for! In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13. And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6. Why aren't the <span class="in">`slice_`</span> functions behaving as expected?</span>
<span id="cb133-316"><a href="#cb133-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-317"><a href="#cb133-317" aria-hidden="true" tabindex="-1"></a>If we take a look at the help page (type <span class="in">`?slice_min`</span> in the console), we learn that <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> have an argument called <span class="in">`with_ties`</span> that is set to <span class="in">`TRUE`</span> by default. If we want to make sure we only get the number of rows we asked for, we would have to set it to <span class="in">`FALSE`</span>, like so:</span>
<span id="cb133-318"><a href="#cb133-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-321"><a href="#cb133-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-322"><a href="#cb133-322" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb133-323"><a href="#cb133-323" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-324"><a href="#cb133-324" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb133-325"><a href="#cb133-325" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span>
<span id="cb133-326"><a href="#cb133-326" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-327"><a href="#cb133-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-328"><a href="#cb133-328" aria-hidden="true" tabindex="-1"></a>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</span>
<span id="cb133-329"><a href="#cb133-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-330"><a href="#cb133-330" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-331"><a href="#cb133-331" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-332"><a href="#cb133-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-333"><a href="#cb133-333" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-334"><a href="#cb133-334" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which code would you use to sort the <span class="in">`mousezempic_dosage_data`</span> data frame from biggest to smallest initial weight?</span>
<span id="cb133-335"><a href="#cb133-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-336"><a href="#cb133-336" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-337"><a href="#cb133-337" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-338"><a href="#cb133-338" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% sort(initial_weight_g)`</span></span>
<span id="cb133-339"><a href="#cb133-339" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-340"><a href="#cb133-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-341"><a href="#cb133-341" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-342"><a href="#cb133-342" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb133-343"><a href="#cb133-343" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-344"><a href="#cb133-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-345"><a href="#cb133-345" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-346"><a href="#cb133-346" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% sort(descending(initial_weight_g))`</span></span>
<span id="cb133-347"><a href="#cb133-347" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-348"><a href="#cb133-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-349"><a href="#cb133-349" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-350"><a href="#cb133-350" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb133-351"><a href="#cb133-351" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-352"><a href="#cb133-352" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-353"><a href="#cb133-353" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-354"><a href="#cb133-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-355"><a href="#cb133-355" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-356"><a href="#cb133-356" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Which code would you use to extract the 3 mice with the highest initial weight from the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb133-357"><a href="#cb133-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-358"><a href="#cb133-358" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-359"><a href="#cb133-359" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-360"><a href="#cb133-360" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)`</span></span>
<span id="cb133-361"><a href="#cb133-361" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-362"><a href="#cb133-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-363"><a href="#cb133-363" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-364"><a href="#cb133-364" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb133-365"><a href="#cb133-365" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-366"><a href="#cb133-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-367"><a href="#cb133-367" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-368"><a href="#cb133-368" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% slice_min(initial_weight_g, n = 3)`</span></span>
<span id="cb133-369"><a href="#cb133-369" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-370"><a href="#cb133-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-371"><a href="#cb133-371" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-372"><a href="#cb133-372" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb133-373"><a href="#cb133-373" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-374"><a href="#cb133-374" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-375"><a href="#cb133-375" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-376"><a href="#cb133-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-377"><a href="#cb133-377" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-378"><a href="#cb133-378" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> I've written the below code, but one of the comments is messing it up! Which one?</span>
<span id="cb133-379"><a href="#cb133-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-382"><a href="#cb133-382" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-383"><a href="#cb133-383" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-384"><a href="#cb133-384" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb133-385"><a href="#cb133-385" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># comment B %&gt;%</span></span>
<span id="cb133-386"><a href="#cb133-386" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb133-387"><a href="#cb133-387" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span>
<span id="cb133-388"><a href="#cb133-388" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-389"><a href="#cb133-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-390"><a href="#cb133-390" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-391"><a href="#cb133-391" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-392"><a href="#cb133-392" aria-hidden="true" tabindex="-1"></a>Comment A</span>
<span id="cb133-393"><a href="#cb133-393" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-394"><a href="#cb133-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-395"><a href="#cb133-395" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-396"><a href="#cb133-396" aria-hidden="true" tabindex="-1"></a>Comment B</span>
<span id="cb133-397"><a href="#cb133-397" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-398"><a href="#cb133-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-399"><a href="#cb133-399" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-400"><a href="#cb133-400" aria-hidden="true" tabindex="-1"></a>Comment C</span>
<span id="cb133-401"><a href="#cb133-401" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-402"><a href="#cb133-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-403"><a href="#cb133-403" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-404"><a href="#cb133-404" aria-hidden="true" tabindex="-1"></a>Comment D</span>
<span id="cb133-405"><a href="#cb133-405" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-406"><a href="#cb133-406" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-407"><a href="#cb133-407" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-408"><a href="#cb133-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-409"><a href="#cb133-409" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-410"><a href="#cb133-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-411"><a href="#cb133-411" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-412"><a href="#cb133-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-413"><a href="#cb133-413" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to sort the <span class="in">`mousezempic_dosage_data`</span> data frame from biggest to smallest initial weight is <span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span>. The <span class="in">`arrange()`</span> function is used to sort the data frame (although there is a <span class="in">`sort()`</span> function in R, that's not part of dplyr and won't work the same way), and the <span class="in">`desc()`</span> function is used to sort in descending order.</span>
<span id="cb133-414"><a href="#cb133-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-415"><a href="#cb133-415" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct code to extract the 3 mice with the highest initial weight from the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)`</span>. The <span class="in">`slice_max()`</span> function is used to select the rows with the largest values in the <span class="in">`initial_weight_g`</span> column, and the <span class="in">`n = 3`</span> argument specifies that we want to keep 3 rows. The <span class="in">`arrange()`</span> function is not needed in this case, because <span class="in">`slice_max()`</span> will automatically sort the data frame by the specified column.</span>
<span id="cb133-416"><a href="#cb133-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-417"><a href="#cb133-417" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The comment that is messing up the code is Comment B. The <span class="in">`#`</span> symbol is before the pipe operator <span class="in">`%&gt;%`</span>, so R treats it as part of the comment and this breaks our chain of pipes. The other comments are fine, because they are either at the end of the line or on their own line. Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it's in the wrong place!</span>
<span id="cb133-418"><a href="#cb133-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-419"><a href="#cb133-419" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-420"><a href="#cb133-420" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-421"><a href="#cb133-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-422"><a href="#cb133-422" aria-hidden="true" tabindex="-1"></a><span class="fu">### Filtering data (rows) {#sec-filter}</span></span>
<span id="cb133-423"><a href="#cb133-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-424"><a href="#cb133-424" aria-hidden="true" tabindex="-1"></a><span class="al">![Filter allows you to filter rows using a logical test](images/S02F03_filter.png)</span></span>
<span id="cb133-425"><a href="#cb133-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-426"><a href="#cb133-426" aria-hidden="true" tabindex="-1"></a>In <span class="in">`dplyr`</span>, the <span class="in">`filter()`</span> function is used to subset rows based on their values. You provide a logical test, and <span class="in">`filter()`</span> will keep the rows where the test is <span class="in">`TRUE`</span>. We can write these tests using the comparison operators we learned in the previous session (e.g. <span class="in">`==`</span>, <span class="in">`&lt;`</span> and <span class="in">`!=`</span>, see <span class="co">[</span><span class="ot">Section @sec-comparisons</span><span class="co">]</span>).</span>
<span id="cb133-427"><a href="#cb133-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-428"><a href="#cb133-428" aria-hidden="true" tabindex="-1"></a>For example, to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice that lost more than 6g:</span>
<span id="cb133-429"><a href="#cb133-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-432"><a href="#cb133-432" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-433"><a href="#cb133-433" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-434"><a href="#cb133-434" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span>
<span id="cb133-435"><a href="#cb133-435" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-436"><a href="#cb133-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-437"><a href="#cb133-437" aria-hidden="true" tabindex="-1"></a>Or to only include mice from cage 3E:</span>
<span id="cb133-438"><a href="#cb133-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-441"><a href="#cb133-441" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-442"><a href="#cb133-442" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-443"><a href="#cb133-443" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb133-444"><a href="#cb133-444" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span>
<span id="cb133-445"><a href="#cb133-445" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-446"><a href="#cb133-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-447"><a href="#cb133-447" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Combining logical tests</span></span>
<span id="cb133-448"><a href="#cb133-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-449"><a href="#cb133-449" aria-hidden="true" tabindex="-1"></a>Sometimes we want to filter based on multiple conditions. Here we will show some more advanced operators that can be used to combine logical tests.</span>
<span id="cb133-450"><a href="#cb133-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-451"><a href="#cb133-451" aria-hidden="true" tabindex="-1"></a>The <span class="in">`&amp;`</span> operator is used to combine two logical tests with an 'and' condition. For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</span>
<span id="cb133-452"><a href="#cb133-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-455"><a href="#cb133-455" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-456"><a href="#cb133-456" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-457"><a href="#cb133-457" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span>
<span id="cb133-458"><a href="#cb133-458" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-459"><a href="#cb133-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-460"><a href="#cb133-460" aria-hidden="true" tabindex="-1"></a>The <span class="in">`|`</span> operator is used to combine two logical tests with an 'or' condition. For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</span>
<span id="cb133-461"><a href="#cb133-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-464"><a href="#cb133-464" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-465"><a href="#cb133-465" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-466"><a href="#cb133-466" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span>
<span id="cb133-467"><a href="#cb133-467" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-468"><a href="#cb133-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-469"><a href="#cb133-469" aria-hidden="true" tabindex="-1"></a>The <span class="in">`%in%`</span> operator can be used to filter based on a vector of multiple values (<span class="in">`c(x, y)`</span>). It's particularly useful when you have a few character values you want to filter on, as it is shorter to type than <span class="in">`|`</span> (or).</span>
<span id="cb133-470"><a href="#cb133-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-471"><a href="#cb133-471" aria-hidden="true" tabindex="-1"></a>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <span class="in">`|`</span> like this:</span>
<span id="cb133-472"><a href="#cb133-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-475"><a href="#cb133-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-476"><a href="#cb133-476" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-477"><a href="#cb133-477" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span>
<span id="cb133-478"><a href="#cb133-478" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-479"><a href="#cb133-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-480"><a href="#cb133-480" aria-hidden="true" tabindex="-1"></a>Or we could use <span class="in">`%in%`</span> like this:</span>
<span id="cb133-481"><a href="#cb133-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-484"><a href="#cb133-484" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-485"><a href="#cb133-485" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-486"><a href="#cb133-486" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span>
<span id="cb133-487"><a href="#cb133-487" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-488"><a href="#cb133-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-489"><a href="#cb133-489" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-490"><a href="#cb133-490" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-491"><a href="#cb133-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-492"><a href="#cb133-492" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-493"><a href="#cb133-493" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which code would you use to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice from replicate 2?</span>
<span id="cb133-494"><a href="#cb133-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-495"><a href="#cb133-495" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-496"><a href="#cb133-496" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-497"><a href="#cb133-497" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == 2)`</span></span>
<span id="cb133-498"><a href="#cb133-498" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-499"><a href="#cb133-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-500"><a href="#cb133-500" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-501"><a href="#cb133-501" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == rep2)`</span></span>
<span id="cb133-502"><a href="#cb133-502" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-503"><a href="#cb133-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-504"><a href="#cb133-504" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-505"><a href="#cb133-505" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == "rep2")`</span></span>
<span id="cb133-506"><a href="#cb133-506" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-507"><a href="#cb133-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-508"><a href="#cb133-508" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-509"><a href="#cb133-509" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate = "rep2")`</span></span>
<span id="cb133-510"><a href="#cb133-510" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-511"><a href="#cb133-511" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-512"><a href="#cb133-512" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-513"><a href="#cb133-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-514"><a href="#cb133-514" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-515"><a href="#cb133-515" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is NOT a valid way to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</span>
<span id="cb133-516"><a href="#cb133-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-517"><a href="#cb133-517" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-518"><a href="#cb133-518" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-519"><a href="#cb133-519" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)`</span></span>
<span id="cb133-520"><a href="#cb133-520" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-521"><a href="#cb133-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-522"><a href="#cb133-522" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-523"><a href="#cb133-523" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span></span>
<span id="cb133-524"><a href="#cb133-524" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-525"><a href="#cb133-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-526"><a href="#cb133-526" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-527"><a href="#cb133-527" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)`</span></span>
<span id="cb133-528"><a href="#cb133-528" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-529"><a href="#cb133-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-530"><a href="#cb133-530" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-531"><a href="#cb133-531" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)`</span></span>
<span id="cb133-532"><a href="#cb133-532" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-533"><a href="#cb133-533" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-534"><a href="#cb133-534" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-535"><a href="#cb133-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-536"><a href="#cb133-536" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-537"><a href="#cb133-537" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Which option correctly describes what the following code is doing?</span>
<span id="cb133-538"><a href="#cb133-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-541"><a href="#cb133-541" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-542"><a href="#cb133-542" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-543"><a href="#cb133-543" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-544"><a href="#cb133-544" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb133-545"><a href="#cb133-545" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-546"><a href="#cb133-546" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span>
<span id="cb133-547"><a href="#cb133-547" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-548"><a href="#cb133-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-549"><a href="#cb133-549" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-550"><a href="#cb133-550" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-551"><a href="#cb133-551" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb133-552"><a href="#cb133-552" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-553"><a href="#cb133-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-554"><a href="#cb133-554" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-555"><a href="#cb133-555" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb133-556"><a href="#cb133-556" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-557"><a href="#cb133-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-558"><a href="#cb133-558" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-559"><a href="#cb133-559" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb133-560"><a href="#cb133-560" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-561"><a href="#cb133-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-562"><a href="#cb133-562" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-563"><a href="#cb133-563" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb133-564"><a href="#cb133-564" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-565"><a href="#cb133-565" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-566"><a href="#cb133-566" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-567"><a href="#cb133-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-568"><a href="#cb133-568" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-569"><a href="#cb133-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-570"><a href="#cb133-570" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-571"><a href="#cb133-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-572"><a href="#cb133-572" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice from replicate 2 is <span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == "rep2")`</span>. Option A is incorrect because <span class="in">`2`</span> is not a value of <span class="in">`replicate`</span> (when filtering you need to know what values are actually in your columns! So make sure to <span class="in">`View()`</span> your data first). Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on. Option D is incorrect because <span class="in">`=`</span> is not the correct way to test for equality, you need to use <span class="in">`==`</span>.</span>
<span id="cb133-573"><a href="#cb133-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-574"><a href="#cb133-574" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The invalid option is <span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span>. This is because the second filtering step is missing the name of the filter function, so R doesn't know what to do with <span class="in">`(initial_weight_g &lt; 40)`</span>. The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <span class="in">`filter()`</span> functions in a row to apply multiple conditions, or the <span class="in">`&amp;`</span> operator to combine them into a single <span class="in">`filter()`</span> function. It's just a matter of personal preference.</span>
<span id="cb133-575"><a href="#cb133-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-576"><a href="#cb133-576" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct description of the code is that it filters the data frame to only include mice from the "BALB C" and "Black 6" strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</span>
<span id="cb133-577"><a href="#cb133-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-578"><a href="#cb133-578" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-579"><a href="#cb133-579" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-580"><a href="#cb133-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-581"><a href="#cb133-581" aria-hidden="true" tabindex="-1"></a><span class="fu">### Dealing with missing values {#sec-missing}</span></span>
<span id="cb133-582"><a href="#cb133-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-583"><a href="#cb133-583" aria-hidden="true" tabindex="-1"></a>Missing values are a common problem in real-world datasets. In R, missing values are represented by <span class="in">`NA`</span>. In fact, if you look at the <span class="in">`mousezempic_dosage_data`</span> data frame we've been using, you'll see that some of the cells contain <span class="in">`NA`</span>: try spotting them with the <span class="in">`View()`</span> function.</span>
<span id="cb133-584"><a href="#cb133-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-585"><a href="#cb133-585" aria-hidden="true" tabindex="-1"></a>You can also find missing values in a data frame using the <span class="in">`is.na()`</span> function in combination with <span class="in">`filter()`</span>. For example, to find all the rows in the <span class="in">`mousezempic_dosage_data`</span> data frame that have a missing value for the <span class="in">`drug_dose_g`</span> column:</span>
<span id="cb133-586"><a href="#cb133-586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-589"><a href="#cb133-589" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-590"><a href="#cb133-590" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-591"><a href="#cb133-591" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb133-592"><a href="#cb133-592" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-593"><a href="#cb133-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-594"><a href="#cb133-594" aria-hidden="true" tabindex="-1"></a>The problem with missing values is that they can cause problems when you try to perform calculations on your data. For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <span class="in">`NA`</span>:</span>
<span id="cb133-595"><a href="#cb133-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-598"><a href="#cb133-598" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-599"><a href="#cb133-599" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb133-600"><a href="#cb133-600" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb133-601"><a href="#cb133-601" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span>
<span id="cb133-602"><a href="#cb133-602" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-603"><a href="#cb133-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-604"><a href="#cb133-604" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span> values in R are therefore referred to as 'contagious': if you put an <span class="in">`NA`</span> in you usually get an <span class="in">`NA`</span> out. If you think about it, that makes sense— when we don't know the value of a particular mouse's drug dose, how can we calculate the average? That missing value could be anything.</span>
<span id="cb133-605"><a href="#cb133-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-606"><a href="#cb133-606" aria-hidden="true" tabindex="-1"></a>For this reason, it's important to deal with missing values before performing calculations. Many functions in R will have an argument called <span class="in">`na.rm`</span> that you can set to <span class="in">`TRUE`</span> to remove missing values before performing the calculation. For example, to calculate the mean of the <span class="in">`drug_dose_g`</span> column with the missing values excluded:</span>
<span id="cb133-607"><a href="#cb133-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-610"><a href="#cb133-610" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-611"><a href="#cb133-611" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb133-612"><a href="#cb133-612" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb133-613"><a href="#cb133-613" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-614"><a href="#cb133-614" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-615"><a href="#cb133-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-616"><a href="#cb133-616" aria-hidden="true" tabindex="-1"></a>This time, the result is a number, because the missing values have been removed before the calculation.</span>
<span id="cb133-617"><a href="#cb133-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-618"><a href="#cb133-618" aria-hidden="true" tabindex="-1"></a>But not all functions have an <span class="in">`na.rm`</span> argument. In these cases, you can remove rows with missing values. This can be done for a single column, using the <span class="in">`filter()`</span> function together with <span class="in">`is.na()`</span>:</span>
<span id="cb133-619"><a href="#cb133-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-622"><a href="#cb133-622" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-623"><a href="#cb133-623" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb133-624"><a href="#cb133-624" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-625"><a href="#cb133-625" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb133-626"><a href="#cb133-626" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb133-627"><a href="#cb133-627" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-628"><a href="#cb133-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-629"><a href="#cb133-629" aria-hidden="true" tabindex="-1"></a>Or, you can remove rows with missing values in any column using the <span class="in">`na.omit()`</span> function:</span>
<span id="cb133-630"><a href="#cb133-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-633"><a href="#cb133-633" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-634"><a href="#cb133-634" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb133-635"><a href="#cb133-635" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-636"><a href="#cb133-636" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb133-637"><a href="#cb133-637" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-638"><a href="#cb133-638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-639"><a href="#cb133-639" aria-hidden="true" tabindex="-1"></a>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value. This can be done using the <span class="in">`replace_na()`</span> function from the <span class="in">`tidyr`</span> package. <span class="in">`replace_na()`</span> takes a <span class="in">`list()`</span> which contains each of the column names you want to edit, and the value that should be used.</span>
<span id="cb133-640"><a href="#cb133-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-641"><a href="#cb133-641" aria-hidden="true" tabindex="-1"></a>For example, to replace missing values in the <span class="in">`weight_lost_g`</span> columns with 0, replace missing values in the <span class="in">`sex`</span> column with 'unknown' and leave the rest of the data frame unchanged:</span>
<span id="cb133-642"><a href="#cb133-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-645"><a href="#cb133-645" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-646"><a href="#cb133-646" aria-hidden="true" tabindex="-1"></a><span class="co"># need to load the tidyr package first</span></span>
<span id="cb133-647"><a href="#cb133-647" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb133-648"><a href="#cb133-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-649"><a href="#cb133-649" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb133-650"><a href="#cb133-650" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-651"><a href="#cb133-651" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb133-652"><a href="#cb133-652" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb133-653"><a href="#cb133-653" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span>
<span id="cb133-654"><a href="#cb133-654" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-655"><a href="#cb133-655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-656"><a href="#cb133-656" aria-hidden="true" tabindex="-1"></a>When deciding how to handle missing values, you might have prior knowledge that <span class="in">`NA`</span> should be replaced with a specific value, or you might decide that removing rows with <span class="in">`NA`</span> is the best approach for your analysis.</span>
<span id="cb133-657"><a href="#cb133-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-658"><a href="#cb133-658" aria-hidden="true" tabindex="-1"></a>For example, maybe we knew that the mice were given a <span class="in">`weight_lost_g`</span> of <span class="in">`NA`</span> if they didn't lose any weight, it would then make sense to replace those with 0 (as we did in the code above). However, if the <span class="in">`drug_dose_g`</span> column was missing simply because the data was lost, we might choose to remove those rows entirely.</span>
<span id="cb133-659"><a href="#cb133-659" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-660"><a href="#cb133-660" aria-hidden="true" tabindex="-1"></a>It's important to think carefully about how missing values should be handled in your analysis.</span>
<span id="cb133-661"><a href="#cb133-661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-662"><a href="#cb133-662" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-663"><a href="#cb133-663" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-664"><a href="#cb133-664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-665"><a href="#cb133-665" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-666"><a href="#cb133-666" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What would be the result of running this R code: <span class="in">`mean(c(1, 2, 4, NA))`</span></span>
<span id="cb133-667"><a href="#cb133-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-668"><a href="#cb133-668" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-669"><a href="#cb133-669" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-670"><a href="#cb133-670" aria-hidden="true" tabindex="-1"></a>2.333333</span>
<span id="cb133-671"><a href="#cb133-671" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-672"><a href="#cb133-672" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-673"><a href="#cb133-673" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-674"><a href="#cb133-674" aria-hidden="true" tabindex="-1"></a>0</span>
<span id="cb133-675"><a href="#cb133-675" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-676"><a href="#cb133-676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-677"><a href="#cb133-677" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-678"><a href="#cb133-678" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span></span>
<span id="cb133-679"><a href="#cb133-679" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-680"><a href="#cb133-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-681"><a href="#cb133-681" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-682"><a href="#cb133-682" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb133-683"><a href="#cb133-683" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-684"><a href="#cb133-684" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-685"><a href="#cb133-685" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-686"><a href="#cb133-686" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-687"><a href="#cb133-687" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-688"><a href="#cb133-688" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Which line of code would you use to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column?</span>
<span id="cb133-689"><a href="#cb133-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-690"><a href="#cb133-690" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-691"><a href="#cb133-691" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-692"><a href="#cb133-692" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(tail_length_mm != NA)`</span></span>
<span id="cb133-693"><a href="#cb133-693" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-694"><a href="#cb133-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-695"><a href="#cb133-695" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-696"><a href="#cb133-696" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(is.na(tail_length_mm))`</span></span>
<span id="cb133-697"><a href="#cb133-697" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-698"><a href="#cb133-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-699"><a href="#cb133-699" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-700"><a href="#cb133-700" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% na.omit()`</span></span>
<span id="cb133-701"><a href="#cb133-701" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-702"><a href="#cb133-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-703"><a href="#cb133-703" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-704"><a href="#cb133-704" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))`</span></span>
<span id="cb133-705"><a href="#cb133-705" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-706"><a href="#cb133-706" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-707"><a href="#cb133-707" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-708"><a href="#cb133-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-709"><a href="#cb133-709" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-710"><a href="#cb133-710" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> How would you replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35?</span>
<span id="cb133-711"><a href="#cb133-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-712"><a href="#cb133-712" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-713"><a href="#cb133-713" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-714"><a href="#cb133-714" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))`</span></span>
<span id="cb133-715"><a href="#cb133-715" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-716"><a href="#cb133-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-717"><a href="#cb133-717" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-718"><a href="#cb133-718" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(initial_weight_g = 35)`</span></span>
<span id="cb133-719"><a href="#cb133-719" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-720"><a href="#cb133-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-721"><a href="#cb133-721" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-722"><a href="#cb133-722" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g == 35))`</span></span>
<span id="cb133-723"><a href="#cb133-723" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-724"><a href="#cb133-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-725"><a href="#cb133-725" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-726"><a href="#cb133-726" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(35)`</span></span>
<span id="cb133-727"><a href="#cb133-727" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-728"><a href="#cb133-728" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-729"><a href="#cb133-729" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-730"><a href="#cb133-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-731"><a href="#cb133-731" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-732"><a href="#cb133-732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-733"><a href="#cb133-733" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-734"><a href="#cb133-734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-735"><a href="#cb133-735" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb133-736"><a href="#cb133-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-737"><a href="#cb133-737" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The result of running the code <span class="in">`mean(c(1, 2, 4, NA))`</span> is <span class="in">`NA`</span>. This is because the <span class="in">`NA`</span> value is 'contagious', so when you try to calculate the mean of a vector that contains an <span class="in">`NA`</span>, the result will also be <span class="in">`NA`</span>. If we wanted to calculate the mean of the vector without the <span class="in">`NA`</span>, we would need to use the <span class="in">`na.rm = TRUE`</span> argument.</span>
<span id="cb133-738"><a href="#cb133-738" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct line of code to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column is <span class="in">`mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))`</span>. The <span class="in">`!`</span> symbol is used to negate the result of <span class="in">`is.na()`</span>, so we are filtering to keep the rows where <span class="in">`tail_length_mm`</span> is not <span class="in">`NA`</span>. We can't use the first option with the <span class="in">`!= NA`</span> because <span class="in">`NA`</span> is a special value in R that represents missing data, and it can't be compared to anything, and the third option is incorrect because <span class="in">`na.omit()`</span> removes entire rows with missing values, rather than just filtering based on a single column.</span>
<span id="cb133-739"><a href="#cb133-739" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct line of code to replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35 is <span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))`</span>. The <span class="in">`replace_na()`</span> function takes a <span class="in">`list()`</span> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we're not testing for equality, we're assigning a value.</span>
<span id="cb133-740"><a href="#cb133-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-741"><a href="#cb133-741" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb133-742"><a href="#cb133-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-743"><a href="#cb133-743" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-744"><a href="#cb133-744" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-745"><a href="#cb133-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-746"><a href="#cb133-746" aria-hidden="true" tabindex="-1"></a><span class="fu">### Selecting columns {#sec-select}</span></span>
<span id="cb133-747"><a href="#cb133-747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-748"><a href="#cb133-748" aria-hidden="true" tabindex="-1"></a><span class="al">![Select allows you to select only certain columns](images/S02F04_select.png)</span></span>
<span id="cb133-749"><a href="#cb133-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-750"><a href="#cb133-750" aria-hidden="true" tabindex="-1"></a>While <span class="in">`filter()`</span> is used to subset rows, <span class="in">`select()`</span> is used to subset columns. You can use <span class="in">`select()`</span> to keep only the columns you're interested in, or to drop columns you don't need.</span>
<span id="cb133-751"><a href="#cb133-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-752"><a href="#cb133-752" aria-hidden="true" tabindex="-1"></a>The <span class="in">`select()`</span> function takes the names of the columns that you want to keep/remove (no vector notation <span class="in">`c()`</span> or quotation marks <span class="in">`""`</span> necessary). For example, to select only the <span class="in">`mouse_strain`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame:</span>
<span id="cb133-753"><a href="#cb133-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-756"><a href="#cb133-756" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-757"><a href="#cb133-757" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-758"><a href="#cb133-758" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span>
<span id="cb133-759"><a href="#cb133-759" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-760"><a href="#cb133-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-761"><a href="#cb133-761" aria-hidden="true" tabindex="-1"></a>We can see that all the other columns have been removed from the data frame.</span>
<span id="cb133-762"><a href="#cb133-762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-763"><a href="#cb133-763" aria-hidden="true" tabindex="-1"></a>If you want to keep all columns except for a few, you can use <span class="in">`-`</span> to drop columns. For example, to keep all columns except for <span class="in">`cage_number`</span> and <span class="in">`sex`</span>:</span>
<span id="cb133-764"><a href="#cb133-764" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-767"><a href="#cb133-767" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-768"><a href="#cb133-768" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-769"><a href="#cb133-769" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span>
<span id="cb133-770"><a href="#cb133-770" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-771"><a href="#cb133-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-772"><a href="#cb133-772" aria-hidden="true" tabindex="-1"></a>There are also some helper functions that can be used to select columns based on their names :</span>
<span id="cb133-773"><a href="#cb133-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-774"><a href="#cb133-774" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb133-775"><a href="#cb133-775" aria-hidden="true" tabindex="-1"></a>| Function        | Description                                       | Example                                         |</span>
<span id="cb133-776"><a href="#cb133-776" aria-hidden="true" tabindex="-1"></a>+=================+===================================================+=================================================+</span>
<span id="cb133-777"><a href="#cb133-777" aria-hidden="true" tabindex="-1"></a>| <span class="in">`starts_with()`</span> | select column(s) that start with a certain string | select all columns starting with the letter i   |</span>
<span id="cb133-778"><a href="#cb133-778" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb133-779"><a href="#cb133-779" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(starts_with("i"))`</span>                      |</span>
<span id="cb133-780"><a href="#cb133-780" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb133-781"><a href="#cb133-781" aria-hidden="true" tabindex="-1"></a>| <span class="in">`ends_with()`</span>   | select column(s) that end with a certain string   | select all columns ending with <span class="sc">\_</span>g              |</span>
<span id="cb133-782"><a href="#cb133-782" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb133-783"><a href="#cb133-783" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(ends_with("_g"))`</span>                       |</span>
<span id="cb133-784"><a href="#cb133-784" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb133-785"><a href="#cb133-785" aria-hidden="true" tabindex="-1"></a>| <span class="in">`contains()`</span>    | select column(s) that contain a certain string    | select all columns containing the word 'weight' |</span>
<span id="cb133-786"><a href="#cb133-786" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb133-787"><a href="#cb133-787" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(contains("weight"))`</span>                    |</span>
<span id="cb133-788"><a href="#cb133-788" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb133-789"><a href="#cb133-789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-790"><a href="#cb133-790" aria-hidden="true" tabindex="-1"></a>: There are several helper functions that can be used with the select function</span>
<span id="cb133-791"><a href="#cb133-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-792"><a href="#cb133-792" aria-hidden="true" tabindex="-1"></a>You need to use quotation marks around the arguments in these helper functions, as they aren't full column names, just strings of characters.</span>
<span id="cb133-793"><a href="#cb133-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-794"><a href="#cb133-794" aria-hidden="true" tabindex="-1"></a>Try using these helper functions to select columns from the <span class="in">`mousezempic_dosage_data`</span> data frame!</span>
<span id="cb133-795"><a href="#cb133-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-796"><a href="#cb133-796" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Reordering columns"}</span>
<span id="cb133-797"><a href="#cb133-797" aria-hidden="true" tabindex="-1"></a><span class="al">![Relocate allows you to move columns around](images/S02F05_reorder.png)</span> We can reorder columns using the <span class="in">`relocate()`</span> function, which works similarly to <span class="in">`select()`</span> (except it just moves columns around rather than dropping/keeping them). For example, to move the <span class="in">`sex`</span> column to before the <span class="in">`cage_number`</span> column:</span>
<span id="cb133-798"><a href="#cb133-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-801"><a href="#cb133-801" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-802"><a href="#cb133-802" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-803"><a href="#cb133-803" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb133-804"><a href="#cb133-804" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span>
<span id="cb133-805"><a href="#cb133-805" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-806"><a href="#cb133-806" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-807"><a href="#cb133-807" aria-hidden="true" tabindex="-1"></a>Two useful helper functions here are the <span class="in">`everything()`</span> and <span class="in">`last_col()`</span> functions, which can be used to move columns to the start/end of the data frame.</span>
<span id="cb133-808"><a href="#cb133-808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-811"><a href="#cb133-811" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-812"><a href="#cb133-812" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb133-813"><a href="#cb133-813" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-814"><a href="#cb133-814" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span>
<span id="cb133-815"><a href="#cb133-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-816"><a href="#cb133-816" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb133-817"><a href="#cb133-817" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-818"><a href="#cb133-818" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span>
<span id="cb133-819"><a href="#cb133-819" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-820"><a href="#cb133-820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-821"><a href="#cb133-821" aria-hidden="true" tabindex="-1"></a>Re-ordering columns isn't necessary, but it makes it easier to see the data you're most interested in within the console (since often not all of the columns will fit on the screen at once). For example, if we are doing a lot of computation on the <span class="in">`initial_weight_g`</span> column, we'd probably like to have that near the start so we can easily check it.</span>
<span id="cb133-822"><a href="#cb133-822" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-823"><a href="#cb133-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-824"><a href="#cb133-824" aria-hidden="true" tabindex="-1"></a>Note that the output of the <span class="in">`select()`</span> function is a new data frame, even if you only select a single column:</span>
<span id="cb133-825"><a href="#cb133-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-828"><a href="#cb133-828" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-829"><a href="#cb133-829" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb133-830"><a href="#cb133-830" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-831"><a href="#cb133-831" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb133-832"><a href="#cb133-832" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from session 1 that class() tells us the type of an object</span></span>
<span id="cb133-833"><a href="#cb133-833" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb133-834"><a href="#cb133-834" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-835"><a href="#cb133-835" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-836"><a href="#cb133-836" aria-hidden="true" tabindex="-1"></a>Sometimes, we instead want to get the values of a column as a vector.</span>
<span id="cb133-837"><a href="#cb133-837" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-838"><a href="#cb133-838" aria-hidden="true" tabindex="-1"></a><span class="al">![Pull allows you to pull acolumn out of a data frame as a vector](images/S02F06_pull.png)</span></span>
<span id="cb133-839"><a href="#cb133-839" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-840"><a href="#cb133-840" aria-hidden="true" tabindex="-1"></a>We can do this by using the <span class="in">`pull()`</span> function, which extracts a single column from a data frame as a vector:</span>
<span id="cb133-841"><a href="#cb133-841" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-844"><a href="#cb133-844" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-845"><a href="#cb133-845" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb133-846"><a href="#cb133-846" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-847"><a href="#cb133-847" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb133-848"><a href="#cb133-848" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb133-849"><a href="#cb133-849" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-850"><a href="#cb133-850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-851"><a href="#cb133-851" aria-hidden="true" tabindex="-1"></a>We can see that the class of the output is now a vector, rather than a data frame. This is important because some functions only accept vectors, not data frames, like <span class="in">`mean()`</span> for example:</span>
<span id="cb133-852"><a href="#cb133-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-855"><a href="#cb133-855" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-856"><a href="#cb133-856" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb133-857"><a href="#cb133-857" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-858"><a href="#cb133-858" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-859"><a href="#cb133-859" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb133-860"><a href="#cb133-860" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-861"><a href="#cb133-861" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-862"><a href="#cb133-862" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-863"><a href="#cb133-863" aria-hidden="true" tabindex="-1"></a>Note how both times we used <span class="in">`na.rm = TRUE`</span> to remove missing values before calculating the mean.</span>
<span id="cb133-864"><a href="#cb133-864" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-865"><a href="#cb133-865" aria-hidden="true" tabindex="-1"></a>You might remember that we used the <span class="in">`$`</span> operator in the previous session to extract a single column from a data frame, so why use <span class="in">`pull()`</span> instead? The main reason is that <span class="in">`pull()`</span> works within a chain of pipes, whereas <span class="in">`$`</span> doesn't.</span>
<span id="cb133-866"><a href="#cb133-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-867"><a href="#cb133-867" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to know the average initial weight of mice that lost at least 4g. We can do this by chaining <span class="in">`filter()`</span> and <span class="in">`pull()`</span> together:</span>
<span id="cb133-868"><a href="#cb133-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-871"><a href="#cb133-871" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-872"><a href="#cb133-872" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-873"><a href="#cb133-873" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb133-874"><a href="#cb133-874" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-875"><a href="#cb133-875" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb133-876"><a href="#cb133-876" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb133-877"><a href="#cb133-877" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb133-878"><a href="#cb133-878" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-879"><a href="#cb133-879" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-880"><a href="#cb133-880" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-881"><a href="#cb133-881" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-882"><a href="#cb133-882" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-883"><a href="#cb133-883" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-884"><a href="#cb133-884" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-885"><a href="#cb133-885" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which line of code would NOT be a valid way to select the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb133-886"><a href="#cb133-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-887"><a href="#cb133-887" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-888"><a href="#cb133-888" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-889"><a href="#cb133-889" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)`</span></span>
<span id="cb133-890"><a href="#cb133-890" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-891"><a href="#cb133-891" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-892"><a href="#cb133-892" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-893"><a href="#cb133-893" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(contains("g"))`</span></span>
<span id="cb133-894"><a href="#cb133-894" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-895"><a href="#cb133-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-896"><a href="#cb133-896" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-897"><a href="#cb133-897" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(ends_with("_g"))`</span></span>
<span id="cb133-898"><a href="#cb133-898" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-899"><a href="#cb133-899" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-900"><a href="#cb133-900" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-901"><a href="#cb133-901" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)`</span></span>
<span id="cb133-902"><a href="#cb133-902" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-903"><a href="#cb133-903" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-904"><a href="#cb133-904" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-905"><a href="#cb133-905" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-906"><a href="#cb133-906" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-907"><a href="#cb133-907" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> How would I extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`mousezempic_dosage_data`</span> data frame as a vector?</span>
<span id="cb133-908"><a href="#cb133-908" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-909"><a href="#cb133-909" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-910"><a href="#cb133-910" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-911"><a href="#cb133-911" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(initial_weight_g)`</span></span>
<span id="cb133-912"><a href="#cb133-912" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-913"><a href="#cb133-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-914"><a href="#cb133-914" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-915"><a href="#cb133-915" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% $initial_weight_g`</span></span>
<span id="cb133-916"><a href="#cb133-916" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-917"><a href="#cb133-917" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-918"><a href="#cb133-918" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-919"><a href="#cb133-919" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(initial_weight_g)`</span></span>
<span id="cb133-920"><a href="#cb133-920" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-921"><a href="#cb133-921" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-922"><a href="#cb133-922" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-923"><a href="#cb133-923" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% pull(initial_weight_g)`</span></span>
<span id="cb133-924"><a href="#cb133-924" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-925"><a href="#cb133-925" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-926"><a href="#cb133-926" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-927"><a href="#cb133-927" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-928"><a href="#cb133-928" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-929"><a href="#cb133-929" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> How would you move the <span class="in">`sex`</span> column to the end of the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb133-930"><a href="#cb133-930" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-931"><a href="#cb133-931" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-932"><a href="#cb133-932" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-933"><a href="#cb133-933" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex)`</span></span>
<span id="cb133-934"><a href="#cb133-934" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-935"><a href="#cb133-935" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-936"><a href="#cb133-936" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-937"><a href="#cb133-937" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col)`</span></span>
<span id="cb133-938"><a href="#cb133-938" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-939"><a href="#cb133-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-940"><a href="#cb133-940" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-941"><a href="#cb133-941" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())`</span></span>
<span id="cb133-942"><a href="#cb133-942" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-943"><a href="#cb133-943" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-944"><a href="#cb133-944" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-945"><a href="#cb133-945" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% reorder(sex, .after = last_col())`</span></span>
<span id="cb133-946"><a href="#cb133-946" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-947"><a href="#cb133-947" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-948"><a href="#cb133-948" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-949"><a href="#cb133-949" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-950"><a href="#cb133-950" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-951"><a href="#cb133-951" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-952"><a href="#cb133-952" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-953"><a href="#cb133-953" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-954"><a href="#cb133-954" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb133-955"><a href="#cb133-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-956"><a href="#cb133-956" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The line of code that would NOT be a valid way to select the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% select(contains("g"))`</span>. This line of code would select all columns that contain the letter 'g', which would include columns like <span class="in">`cage_number`</span> and <span class="in">`tail_length_mm`</span>. We need to specify either <span class="in">`ends_with("g")`</span> or <span class="in">`contains("_g")`</span> to only get those with <span class="in">`_g`</span> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</span>
<span id="cb133-957"><a href="#cb133-957" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct way to extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`mousezempic_dosage_data`</span> data frame as a vector is <span class="in">`mousezempic_dosage_data %&gt;% pull(initial_weight_g)`</span>. The <span class="in">`pull()`</span> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <span class="in">`filter()`</span> is used to subset rows, <span class="in">`$`</span> is not used in a pipe chain, and <span class="in">`select()`</span> is outputs a data frame, not extract them as vectors.</span>
<span id="cb133-958"><a href="#cb133-958" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct way to move the <span class="in">`sex`</span> column to the end of the <span class="in">`mousezempic_dosage_data`</span> data frame is using the <span class="in">`relocate()`</span> function like this: <span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())`</span>. The <span class="in">`last_col()`</span> function is used to refer to the last column in the data frame. The other options are incorrect because <span class="in">`reorder()`</span> is not a valid function, and you need to remember to include the brackets <span class="in">`()`</span> when using <span class="in">`last_col()`</span>.</span>
<span id="cb133-959"><a href="#cb133-959" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-960"><a href="#cb133-960" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb133-961"><a href="#cb133-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-962"><a href="#cb133-962" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-963"><a href="#cb133-963" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-964"><a href="#cb133-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-965"><a href="#cb133-965" aria-hidden="true" tabindex="-1"></a><span class="fu">### Modifying data {#sec-mutate}</span></span>
<span id="cb133-966"><a href="#cb133-966" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-967"><a href="#cb133-967" aria-hidden="true" tabindex="-1"></a>So far, we've learned how to filter rows and select columns from a data frame. But what if we want to change the data itself? This is where the <span class="in">`mutate()`</span> function comes in.</span>
<span id="cb133-968"><a href="#cb133-968" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-969"><a href="#cb133-969" aria-hidden="true" tabindex="-1"></a><span class="al">![Mutate allows you to add new columns to a data frame](images/S02F07_mutate.png)</span></span>
<span id="cb133-970"><a href="#cb133-970" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-971"><a href="#cb133-971" aria-hidden="true" tabindex="-1"></a>The <span class="in">`mutate()`</span> function is used to add new columns to a data frame, or modify existing columns, often by performing some sort of calculation. For example, we can add a new column to <span class="in">`mousezempic_dosage_data`</span> that contains the drug dose in mg (rather than g):</span>
<span id="cb133-972"><a href="#cb133-972" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-975"><a href="#cb133-975" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-976"><a href="#cb133-976" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-977"><a href="#cb133-977" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a new column called drug_dose_mg</span></span>
<span id="cb133-978"><a href="#cb133-978" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert drug_dose_g to mg by multiplying by 1000</span></span>
<span id="cb133-979"><a href="#cb133-979" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-980"><a href="#cb133-980" aria-hidden="true" tabindex="-1"></a>  <span class="co"># just select the drug dose columns so we can compare them</span></span>
<span id="cb133-981"><a href="#cb133-981" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(drug_dose_g, drug_dose_mg)</span>
<span id="cb133-982"><a href="#cb133-982" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-983"><a href="#cb133-983" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-984"><a href="#cb133-984" aria-hidden="true" tabindex="-1"></a>You can see that the <span class="in">`drug_dose_mg`</span> column has been added to the data frame, and it contains, for each row, the value of the <span class="in">`drug_dose_g`</span> column multiplied by 1000 (<span class="in">`NA`</span> values are preserved).</span>
<span id="cb133-985"><a href="#cb133-985" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-986"><a href="#cb133-986" aria-hidden="true" tabindex="-1"></a>These calculations can be as complex as you like, and involve multiple different columns. For example, to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that calculates the weight lost as a percentage of the initial weight:</span>
<span id="cb133-987"><a href="#cb133-987" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-990"><a href="#cb133-990" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-991"><a href="#cb133-991" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-992"><a href="#cb133-992" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate the % of initial weight that was lost</span></span>
<span id="cb133-993"><a href="#cb133-993" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_percent =</span> (weight_lost_g <span class="sc">/</span> initial_weight_g) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb133-994"><a href="#cb133-994" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-995"><a href="#cb133-995" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-996"><a href="#cb133-996" aria-hidden="true" tabindex="-1"></a>A useful helper function for <span class="in">`mutate()`</span> is the <span class="in">`case_when()`</span> function, which allows you to create new columns based on multiple conditions. We do this with the notation <span class="in">`case_when(condition1 ~ value1, condition2 ~ value2, ...)`</span>.</span>
<span id="cb133-997"><a href="#cb133-997" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-998"><a href="#cb133-998" aria-hidden="true" tabindex="-1"></a>For example, to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that categorises the mice based on how much weight they lost:</span>
<span id="cb133-999"><a href="#cb133-999" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1002"><a href="#cb133-1002" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1003"><a href="#cb133-1003" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1004"><a href="#cb133-1004" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb133-1005"><a href="#cb133-1005" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb133-1006"><a href="#cb133-1006" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb133-1007"><a href="#cb133-1007" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb133-1008"><a href="#cb133-1008" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb133-1009"><a href="#cb133-1009" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb133-1010"><a href="#cb133-1010" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span>
<span id="cb133-1011"><a href="#cb133-1011" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1012"><a href="#cb133-1012" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1013"><a href="#cb133-1013" aria-hidden="true" tabindex="-1"></a>Note that the conditions are evaluated in order, and the first condition that is <span class="in">`TRUE`</span> is the one that is used. So if a mouse lost 4.5g, it <span class="in">`case_when()`</span> would first test if it fits the 'Low' category (by checking if 4.5 is less than 4, which it isn't), and then if it fits the 'Medium' category (by checking if 4.5 is less than or equal to 5). Since it is, the mouse would be categorised as 'Medium'.</span>
<span id="cb133-1014"><a href="#cb133-1014" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1015"><a href="#cb133-1015" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Fallback with default value(s)"}</span>
<span id="cb133-1016"><a href="#cb133-1016" aria-hidden="true" tabindex="-1"></a>In the above example, what would happen if a mouse lost -1g (gained weight)? It wouldn't fit any of the conditions, so it would get an <span class="in">`NA`</span> in the <span class="in">`weight_loss_category`</span> column. Sometimes you might want this behaviour, but other times you would prefer to specify a 'fallback' category that will be assigned to everything that doesn't fit in the other categories. You can do this by including a <span class="in">` .default = `</span> argument at the end of the <span class="in">`case_when()`</span> function. For example:</span>
<span id="cb133-1017"><a href="#cb133-1017" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1020"><a href="#cb133-1020" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1021"><a href="#cb133-1021" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1022"><a href="#cb133-1022" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb133-1023"><a href="#cb133-1023" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb133-1024"><a href="#cb133-1024" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb133-1025"><a href="#cb133-1025" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb133-1026"><a href="#cb133-1026" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span>,</span>
<span id="cb133-1027"><a href="#cb133-1027" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"Unknown"</span></span>
<span id="cb133-1028"><a href="#cb133-1028" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb133-1029"><a href="#cb133-1029" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span>
<span id="cb133-1030"><a href="#cb133-1030" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1031"><a href="#cb133-1031" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1032"><a href="#cb133-1032" aria-hidden="true" tabindex="-1"></a>Notice how the <span class="in">`NA`</span> value in the fourth row is now categorised as 'Unknown'.</span>
<span id="cb133-1033"><a href="#cb133-1033" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1034"><a href="#cb133-1034" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1035"><a href="#cb133-1035" aria-hidden="true" tabindex="-1"></a>One final thing to note is that <span class="in">`mutate()`</span> can be used to modify existing columns as well as add new ones. To do this, just use the name of the existing column as the 'new' one.</span>
<span id="cb133-1036"><a href="#cb133-1036" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1037"><a href="#cb133-1037" aria-hidden="true" tabindex="-1"></a>For example, let's use <span class="in">`mutate()`</span> together with <span class="in">`case_when()`</span> to modify the <span class="in">`sex`</span> column so that it uses <span class="in">`M`</span> and <span class="in">`F`</span> instead <span class="in">`male`</span> and <span class="in">`female`</span>:</span>
<span id="cb133-1038"><a href="#cb133-1038" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1041"><a href="#cb133-1041" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1042"><a href="#cb133-1042" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1043"><a href="#cb133-1043" aria-hidden="true" tabindex="-1"></a>  <span class="co"># modify sex column</span></span>
<span id="cb133-1044"><a href="#cb133-1044" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">case_when</span>(</span>
<span id="cb133-1045"><a href="#cb133-1045" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">~</span> <span class="st">"F"</span>,</span>
<span id="cb133-1046"><a href="#cb133-1046" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"male"</span> <span class="sc">~</span> <span class="st">"M"</span>,</span>
<span id="cb133-1047"><a href="#cb133-1047" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if neither, code it as 'X'</span></span>
<span id="cb133-1048"><a href="#cb133-1048" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"X"</span>))</span>
<span id="cb133-1049"><a href="#cb133-1049" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1050"><a href="#cb133-1050" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1051"><a href="#cb133-1051" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-1052"><a href="#cb133-1052" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-1053"><a href="#cb133-1053" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1054"><a href="#cb133-1054" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1055"><a href="#cb133-1055" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What line of code would you use to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that converts the <span class="in">`tail_length_mm`</span> column to cm?</span>
<span id="cb133-1056"><a href="#cb133-1056" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1057"><a href="#cb133-1057" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1058"><a href="#cb133-1058" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1059"><a href="#cb133-1059" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% create(tail_length_cm = tail_length_mm / 10)`</span></span>
<span id="cb133-1060"><a href="#cb133-1060" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1061"><a href="#cb133-1061" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1062"><a href="#cb133-1062" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1063"><a href="#cb133-1063" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm == tail_length_mm / 10)`</span></span>
<span id="cb133-1064"><a href="#cb133-1064" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1065"><a href="#cb133-1065" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1066"><a href="#cb133-1066" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1067"><a href="#cb133-1067" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)`</span></span>
<span id="cb133-1068"><a href="#cb133-1068" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1069"><a href="#cb133-1069" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1070"><a href="#cb133-1070" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1071"><a href="#cb133-1071" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% tail_length_cm = tail_length_mm / 10`</span></span>
<span id="cb133-1072"><a href="#cb133-1072" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1073"><a href="#cb133-1073" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1074"><a href="#cb133-1074" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1075"><a href="#cb133-1075" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1076"><a href="#cb133-1076" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1077"><a href="#cb133-1077" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb133-1078"><a href="#cb133-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1081"><a href="#cb133-1081" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1082"><a href="#cb133-1082" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1083"><a href="#cb133-1083" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1084"><a href="#cb133-1084" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1085"><a href="#cb133-1085" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_rank =</span> <span class="fu">row_number</span>())</span>
<span id="cb133-1086"><a href="#cb133-1086" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1087"><a href="#cb133-1087" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1088"><a href="#cb133-1088" aria-hidden="true" tabindex="-1"></a>Hint: the row_number() function returns the number of each row in the data frame (1 being the first row and so on).</span>
<span id="cb133-1089"><a href="#cb133-1089" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1090"><a href="#cb133-1090" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1091"><a href="#cb133-1091" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1092"><a href="#cb133-1092" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the least weight.</span>
<span id="cb133-1093"><a href="#cb133-1093" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1094"><a href="#cb133-1094" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1095"><a href="#cb133-1095" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1096"><a href="#cb133-1096" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight.</span>
<span id="cb133-1097"><a href="#cb133-1097" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1098"><a href="#cb133-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1099"><a href="#cb133-1099" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1100"><a href="#cb133-1100" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice</span>
<span id="cb133-1101"><a href="#cb133-1101" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1102"><a href="#cb133-1102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1103"><a href="#cb133-1103" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1104"><a href="#cb133-1104" aria-hidden="true" tabindex="-1"></a>Does nothing, because the <span class="in">`row_number()`</span> function has no arguments</span>
<span id="cb133-1105"><a href="#cb133-1105" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1106"><a href="#cb133-1106" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1107"><a href="#cb133-1107" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1108"><a href="#cb133-1108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1109"><a href="#cb133-1109" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1110"><a href="#cb133-1110" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> What is wrong with this R code?</span>
<span id="cb133-1111"><a href="#cb133-1111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1114"><a href="#cb133-1114" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1115"><a href="#cb133-1115" aria-hidden="true" tabindex="-1"></a><span class="co">#| error: true</span></span>
<span id="cb133-1116"><a href="#cb133-1116" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1117"><a href="#cb133-1117" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb133-1118"><a href="#cb133-1118" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span></span>
<span id="cb133-1119"><a href="#cb133-1119" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span></span>
<span id="cb133-1120"><a href="#cb133-1120" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb133-1121"><a href="#cb133-1121" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb133-1122"><a href="#cb133-1122" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1123"><a href="#cb133-1123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1124"><a href="#cb133-1124" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1125"><a href="#cb133-1125" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1126"><a href="#cb133-1126" aria-hidden="true" tabindex="-1"></a>You didn't include a <span class="in">`TRUE ~ ...`</span> condition at the end of the <span class="in">`case_when()`</span> function to act as a fallback</span>
<span id="cb133-1127"><a href="#cb133-1127" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1128"><a href="#cb133-1128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1129"><a href="#cb133-1129" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1130"><a href="#cb133-1130" aria-hidden="true" tabindex="-1"></a>You can't use the <span class="in">`case_when()`</span> function with the <span class="in">`mutate()`</span> function</span>
<span id="cb133-1131"><a href="#cb133-1131" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1132"><a href="#cb133-1132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1133"><a href="#cb133-1133" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1134"><a href="#cb133-1134" aria-hidden="true" tabindex="-1"></a><span class="in">`weight_lost_g`</span> is not a valid column name</span>
<span id="cb133-1135"><a href="#cb133-1135" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1136"><a href="#cb133-1136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1137"><a href="#cb133-1137" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1138"><a href="#cb133-1138" aria-hidden="true" tabindex="-1"></a>You need to separate the conditions in the <span class="in">`case_when()`</span> function with a comma</span>
<span id="cb133-1139"><a href="#cb133-1139" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1140"><a href="#cb133-1140" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1141"><a href="#cb133-1141" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1142"><a href="#cb133-1142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1143"><a href="#cb133-1143" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1144"><a href="#cb133-1144" aria-hidden="true" tabindex="-1"></a>4<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb133-1145"><a href="#cb133-1145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1148"><a href="#cb133-1148" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1149"><a href="#cb133-1149" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1150"><a href="#cb133-1150" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1151"><a href="#cb133-1151" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_strain =</span> <span class="fu">case_when</span>(</span>
<span id="cb133-1152"><a href="#cb133-1152" aria-hidden="true" tabindex="-1"></a>    mouse_strain <span class="sc">==</span> <span class="st">"Black 6"</span> <span class="sc">~</span> <span class="st">"B6"</span>,</span>
<span id="cb133-1153"><a href="#cb133-1153" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> mouse_strain</span>
<span id="cb133-1154"><a href="#cb133-1154" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb133-1155"><a href="#cb133-1155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1156"><a href="#cb133-1156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1157"><a href="#cb133-1157" aria-hidden="true" tabindex="-1"></a>Hint: if you're not sure, try running the code, but pipe it into <span class="in">`View()`</span> so that you can take a good look at what's happening in the <span class="in">`mouse_strain`</span> column.</span>
<span id="cb133-1158"><a href="#cb133-1158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1159"><a href="#cb133-1159" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1160"><a href="#cb133-1160" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1161"><a href="#cb133-1161" aria-hidden="true" tabindex="-1"></a>Renames the strains of all the mice to "B6", regardless of their original strain</span>
<span id="cb133-1162"><a href="#cb133-1162" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1163"><a href="#cb133-1163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1164"><a href="#cb133-1164" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1165"><a href="#cb133-1165" aria-hidden="true" tabindex="-1"></a>This code will produce an error</span>
<span id="cb133-1166"><a href="#cb133-1166" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1167"><a href="#cb133-1167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1168"><a href="#cb133-1168" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1169"><a href="#cb133-1169" aria-hidden="true" tabindex="-1"></a>Adds a new column that categorises the mice based on their strain, so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged.</span>
<span id="cb133-1170"><a href="#cb133-1170" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1171"><a href="#cb133-1171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1172"><a href="#cb133-1172" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1173"><a href="#cb133-1173" aria-hidden="true" tabindex="-1"></a>Modifies the <span class="in">`mouse_strain`</span> column so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged.</span>
<span id="cb133-1174"><a href="#cb133-1174" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1175"><a href="#cb133-1175" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1176"><a href="#cb133-1176" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1177"><a href="#cb133-1177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1178"><a href="#cb133-1178" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-1179"><a href="#cb133-1179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1180"><a href="#cb133-1180" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-1181"><a href="#cb133-1181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1182"><a href="#cb133-1182" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb133-1183"><a href="#cb133-1183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1184"><a href="#cb133-1184" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct line of code to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that converts the <span class="in">`tail_length_mm`</span> column to cm is <span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)`</span>.</span>
<span id="cb133-1185"><a href="#cb133-1185" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(weight_lost_g)) %&gt;% mutate(weight_lost_rank = row_number())`</span> adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight. First, the <span class="in">`arrange(desc(weight_lost_g))`</span> function sorts the data frame by the <span class="in">`weight_lost_g`</span> column in descending order, and then the <span class="in">`mutate(weight_lost_rank = row_number())`</span> function adds a new column that assigns a rank to each row based on its position (row number) in the sorted data frame.</span>
<span id="cb133-1186"><a href="#cb133-1186" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The error is that the conditions in the <span class="in">`case_when()`</span> function are not separated by commas. Each condition should be followed by a comma because these are like the arguments in a function. Remeber that it's optional to include the <span class="in">`TRUE ~ ...`</span> condition at the end of the <span class="in">`case_when()`</span> function.</span>
<span id="cb133-1187"><a href="#cb133-1187" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% mutate(mouse_strain = case_when(mouse_strain == "Black 6" ~ "B6", TRUE ~ mouse_strain))`</span> modifies the <span class="in">`mouse_strain`</span> column so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged. As we are calling our column <span class="in">`mouse_strain`</span>, no new column is being created (we are modifying the existing one) and the <span class="in">`TRUE ~ mouse_strain`</span> condition acts as a fallback to keep the original values (that already exist in the <span class="in">`mouse_strain`</span> column) for any rows that don't match our first condition (strain being "Black 6").</span>
<span id="cb133-1188"><a href="#cb133-1188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1189"><a href="#cb133-1189" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb133-1190"><a href="#cb133-1190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1191"><a href="#cb133-1191" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-1192"><a href="#cb133-1192" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1193"><a href="#cb133-1193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1194"><a href="#cb133-1194" aria-hidden="true" tabindex="-1"></a><span class="fu">### Summarising data {#sec-summarise}</span></span>
<span id="cb133-1195"><a href="#cb133-1195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1196"><a href="#cb133-1196" aria-hidden="true" tabindex="-1"></a><span class="al">![Summarise allows you to calculate summary statistics that collapse many rows into one](images/S02F08_summarise.png)</span></span>
<span id="cb133-1197"><a href="#cb133-1197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1198"><a href="#cb133-1198" aria-hidden="true" tabindex="-1"></a>The <span class="in">`summarise()`</span> (or <span class="in">`summarize()`</span>, if you prefer US spelling) function is used to calculate summary statistics on your data. It takes similar arguments to <span class="in">`mutate()`</span>, but instead of adding a new column to the data frame, it returns a new data frame with a single row and one column for each summary statistic you calculate.</span>
<span id="cb133-1199"><a href="#cb133-1199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1200"><a href="#cb133-1200" aria-hidden="true" tabindex="-1"></a>For example, to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame:</span>
<span id="cb133-1201"><a href="#cb133-1201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1204"><a href="#cb133-1204" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1205"><a href="#cb133-1205" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1206"><a href="#cb133-1206" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb133-1207"><a href="#cb133-1207" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1208"><a href="#cb133-1208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1209"><a href="#cb133-1209" aria-hidden="true" tabindex="-1"></a>We can also calculate multiple summary statistics at once. For example, to calculate the mean, median, and standard deviation of the weight lost by the mice:</span>
<span id="cb133-1210"><a href="#cb133-1210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1213"><a href="#cb133-1213" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1214"><a href="#cb133-1214" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1215"><a href="#cb133-1215" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1216"><a href="#cb133-1216" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1217"><a href="#cb133-1217" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1218"><a href="#cb133-1218" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-1219"><a href="#cb133-1219" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb133-1220"><a href="#cb133-1220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1221"><a href="#cb133-1221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1222"><a href="#cb133-1222" aria-hidden="true" tabindex="-1"></a>The power of summarising data is really seen when combined with grouping, which we will cover in the next section.</span>
<span id="cb133-1223"><a href="#cb133-1223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1224"><a href="#cb133-1224" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-1225"><a href="#cb133-1225" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-1226"><a href="#cb133-1226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1227"><a href="#cb133-1227" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1228"><a href="#cb133-1228" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb133-1229"><a href="#cb133-1229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1232"><a href="#cb133-1232" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1233"><a href="#cb133-1233" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1234"><a href="#cb133-1234" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1235"><a href="#cb133-1235" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_tail =</span> <span class="fu">mean</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1236"><a href="#cb133-1236" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_tail =</span> <span class="fu">min</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1237"><a href="#cb133-1237" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_tail =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb133-1238"><a href="#cb133-1238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1239"><a href="#cb133-1239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1240"><a href="#cb133-1240" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1241"><a href="#cb133-1241" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1242"><a href="#cb133-1242" aria-hidden="true" tabindex="-1"></a>Calculates the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1243"><a href="#cb133-1243" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1244"><a href="#cb133-1244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1245"><a href="#cb133-1245" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1246"><a href="#cb133-1246" aria-hidden="true" tabindex="-1"></a>Produces a data frame containing one column for each of the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1247"><a href="#cb133-1247" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1248"><a href="#cb133-1248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1249"><a href="#cb133-1249" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1250"><a href="#cb133-1250" aria-hidden="true" tabindex="-1"></a>Finds the average tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1251"><a href="#cb133-1251" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1252"><a href="#cb133-1252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1253"><a href="#cb133-1253" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1254"><a href="#cb133-1254" aria-hidden="true" tabindex="-1"></a>Produces a vector containing the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1255"><a href="#cb133-1255" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1256"><a href="#cb133-1256" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1257"><a href="#cb133-1257" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1258"><a href="#cb133-1258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1259"><a href="#cb133-1259" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1260"><a href="#cb133-1260" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is NOT a valid way to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb133-1261"><a href="#cb133-1261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1262"><a href="#cb133-1262" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1263"><a href="#cb133-1263" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1264"><a href="#cb133-1264" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb133-1265"><a href="#cb133-1265" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1266"><a href="#cb133-1266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1267"><a href="#cb133-1267" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1268"><a href="#cb133-1268" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% pull(weight_lost_g) %&gt;% mean(na.rm = TRUE)`</span></span>
<span id="cb133-1269"><a href="#cb133-1269" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1270"><a href="#cb133-1270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1271"><a href="#cb133-1271" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1272"><a href="#cb133-1272" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarize(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb133-1273"><a href="#cb133-1273" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1274"><a href="#cb133-1274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1275"><a href="#cb133-1275" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1276"><a href="#cb133-1276" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)`</span></span>
<span id="cb133-1277"><a href="#cb133-1277" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1278"><a href="#cb133-1278" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1279"><a href="#cb133-1279" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1280"><a href="#cb133-1280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1281"><a href="#cb133-1281" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-1282"><a href="#cb133-1282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1283"><a href="#cb133-1283" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-1284"><a href="#cb133-1284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1285"><a href="#cb133-1285" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The code **produces a data frame** containing one column for each of the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1286"><a href="#cb133-1286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1287"><a href="#cb133-1287" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> The line of code that is NOT a valid way to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)`</span>. This line of code is incorrect because the <span class="in">`mean()`</span> function is being used directly on the data frame, rather than within a <span class="in">`summarise()`</span> function. The other options are valid ways to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame (although note that the second option uses <span class="in">`pull()`</span> to extract the <span class="in">`weight_lost_g`</span> column as a vector before calculating the mean, so the mean value is stored in a vector rather than in a data frame).</span>
<span id="cb133-1288"><a href="#cb133-1288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1289"><a href="#cb133-1289" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-1290"><a href="#cb133-1290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1291"><a href="#cb133-1291" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1292"><a href="#cb133-1292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1293"><a href="#cb133-1293" aria-hidden="true" tabindex="-1"></a><span class="fu">## Grouping {#sec-grouping}</span></span>
<span id="cb133-1294"><a href="#cb133-1294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1295"><a href="#cb133-1295" aria-hidden="true" tabindex="-1"></a><span class="al">![Using summarise with the .by option allows you to create separate summaries for different 'groups' of data](images/S02F09_grouped_summarise.png)</span></span>
<span id="cb133-1296"><a href="#cb133-1296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1297"><a href="#cb133-1297" aria-hidden="true" tabindex="-1"></a>Grouping is a powerful concept in in <span class="in">`dplyr`</span> that allows you to perform operations on subsets of your data. For example, you might want to calculate the mean weight lost by mice in each cage, or find the mouse with the longest tail in each strain.</span>
<span id="cb133-1298"><a href="#cb133-1298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1299"><a href="#cb133-1299" aria-hidden="true" tabindex="-1"></a>We can group data using the <span class="in">`.by`</span> argument that exists in many dplyr functions, like <span class="in">`summarise()`</span> and <span class="in">`mutate()`</span>, and passing it the name(s) of column(s) to group by. For example, to calculate the mean weight lost by mice in each cage:</span>
<span id="cb133-1300"><a href="#cb133-1300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1303"><a href="#cb133-1303" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1304"><a href="#cb133-1304" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1305"><a href="#cb133-1305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1306"><a href="#cb133-1306" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1307"><a href="#cb133-1307" aria-hidden="true" tabindex="-1"></a>    <span class="co"># don't forget it's .by, not by!</span></span>
<span id="cb133-1308"><a href="#cb133-1308" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span>
<span id="cb133-1309"><a href="#cb133-1309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1310"><a href="#cb133-1310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1311"><a href="#cb133-1311" aria-hidden="true" tabindex="-1"></a>Like when we first learned the summarise function above, we give our new column a name (<span class="in">`mean_weight_lost`</span>), and then we assign its value to be the mean of the <span class="in">`weight_lost_g`</span> column (with <span class="in">`NA`</span>s removed). But this time, we also added the <span class="in">`.by`</span> argument to specify the column we want to group by (<span class="in">`cage_number`</span>, in this case). This will return a data frame with the mean weight lost by mice in each cage.</span>
<span id="cb133-1312"><a href="#cb133-1312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1313"><a href="#cb133-1313" aria-hidden="true" tabindex="-1"></a>Grouping is a powerful tool for exploring your data and can help you identify patterns that might not be obvious when looking at the data as a whole. For example, notice how this grouped summary reveals that mice in cage 3E lost more weight than those in the other two cages.</span>
<span id="cb133-1314"><a href="#cb133-1314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1315"><a href="#cb133-1315" aria-hidden="true" tabindex="-1"></a>It's also possible to group by multiple columns by passing a vector of column names to the <span class="in">`.by`</span> argument. For example, to calculate the mean weight lost by mice in each cage and strain:</span>
<span id="cb133-1316"><a href="#cb133-1316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1319"><a href="#cb133-1319" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1320"><a href="#cb133-1320" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1321"><a href="#cb133-1321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1322"><a href="#cb133-1322" aria-hidden="true" tabindex="-1"></a>  <span class="co"># group by both cage_number and mouse_strain</span></span>
<span id="cb133-1323"><a href="#cb133-1323" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span>
<span id="cb133-1324"><a href="#cb133-1324" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1325"><a href="#cb133-1325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1326"><a href="#cb133-1326" aria-hidden="true" tabindex="-1"></a>Of course, <span class="in">`mean()`</span> is not the only function that we can use within <span class="in">`summarise()`</span>. We can use any function that takes a vector of values and returns a single value, like <span class="in">`median()`</span>, <span class="in">`sd()`</span>, or <span class="in">`max()`</span>. We can also use multiple functions at once, by giving each column a name and specifying the function we want to use:</span>
<span id="cb133-1327"><a href="#cb133-1327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1330"><a href="#cb133-1330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1331"><a href="#cb133-1331" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1332"><a href="#cb133-1332" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1333"><a href="#cb133-1333" aria-hidden="true" tabindex="-1"></a>    <span class="at">n =</span> <span class="fu">n</span>(),</span>
<span id="cb133-1334"><a href="#cb133-1334" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1335"><a href="#cb133-1335" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1336"><a href="#cb133-1336" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1337"><a href="#cb133-1337" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_weight_lost =</span> <span class="fu">max</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1338"><a href="#cb133-1338" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_weight_lost =</span> <span class="fu">min</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1339"><a href="#cb133-1339" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span>
<span id="cb133-1340"><a href="#cb133-1340" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1341"><a href="#cb133-1341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1342"><a href="#cb133-1342" aria-hidden="true" tabindex="-1"></a>Here, we also used the <span class="in">`n()`</span> function to calculate the number of mice in each cage. This is a special helper function that works within <span class="in">`summarise`</span> to count the number of rows in each group.</span>
<span id="cb133-1343"><a href="#cb133-1343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1344"><a href="#cb133-1344" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="To <span class="in">`.by`</span> or not to <span class="in">`.by`</span>?"}</span>
<span id="cb133-1345"><a href="#cb133-1345" aria-hidden="true" tabindex="-1"></a>In the <span class="in">`dplyr`</span> package, there are two ways to group data: using the <span class="in">`.by`</span> argument within various functions (as we have covered so far), or using the <span class="in">`group_by()`</span> function, then performing your operations and ungrouping with <span class="in">`ungroup()`</span>.</span>
<span id="cb133-1346"><a href="#cb133-1346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1347"><a href="#cb133-1347" aria-hidden="true" tabindex="-1"></a>For example, we've seen above how to calculate the mean weight lost by mice in each cage using the <span class="in">`.by`</span> argument:</span>
<span id="cb133-1348"><a href="#cb133-1348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1351"><a href="#cb133-1351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1352"><a href="#cb133-1352" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1353"><a href="#cb133-1353" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> cage_number)</span>
<span id="cb133-1354"><a href="#cb133-1354" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1355"><a href="#cb133-1355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1356"><a href="#cb133-1356" aria-hidden="true" tabindex="-1"></a>But we can also do the same using <span class="in">`group_by()`</span> and <span class="in">`ungroup()`</span>:</span>
<span id="cb133-1359"><a href="#cb133-1359" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1360"><a href="#cb133-1360" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1361"><a href="#cb133-1361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cage_number) <span class="sc">%&gt;%</span></span>
<span id="cb133-1362"><a href="#cb133-1362" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1363"><a href="#cb133-1363" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb133-1364"><a href="#cb133-1364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1365"><a href="#cb133-1365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1366"><a href="#cb133-1366" aria-hidden="true" tabindex="-1"></a>The two methods are equivalent, but using the <span class="in">`.by`</span> argument within functions can be more concise and easier to read. Still, it's good to be aware of <span class="in">`group_by()`</span> and <span class="in">`ungroup()`</span> as they are widely used, particularly in older code.</span>
<span id="cb133-1367"><a href="#cb133-1367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1368"><a href="#cb133-1368" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1369"><a href="#cb133-1369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1370"><a href="#cb133-1370" aria-hidden="true" tabindex="-1"></a>Although grouping is most often used with <span class="in">`summarise()`</span>, it can be used with <span class="in">`dplyr`</span> functions too. For example mutate() function can also be used with grouping to add new columns to the data frame based on group-specific calculations. Let's say we wanted to calculate the Z-score (also known as the <span class="co">[</span><span class="ot">standard score</span><span class="co">](https://en.wikipedia.org/wiki/Standard_score)</span>) to standardise the weight lost by each mouse within each strain.</span>
<span id="cb133-1371"><a href="#cb133-1371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1372"><a href="#cb133-1372" aria-hidden="true" tabindex="-1"></a>As a reminder, the formula for calculating the Z-score is $\frac{x - \mu}{\sigma}$, where $x$ is the value (in our case the <span class="in">`weight_lost_g`</span> column), $\mu$ is the mean, and $\sigma$ is the standard deviation.</span>
<span id="cb133-1373"><a href="#cb133-1373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1374"><a href="#cb133-1374" aria-hidden="true" tabindex="-1"></a>We can calculate this for each mouse in each strain using the following code:</span>
<span id="cb133-1375"><a href="#cb133-1375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1378"><a href="#cb133-1378" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1379"><a href="#cb133-1379" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1380"><a href="#cb133-1380" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove NAs before calculating the mean and SD</span></span>
<span id="cb133-1381"><a href="#cb133-1381" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1382"><a href="#cb133-1382" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_z =</span> (weight_lost_g <span class="sc">-</span> <span class="fu">mean</span>(weight_lost_g)) <span class="sc">/</span> <span class="fu">sd</span>(weight_lost_g), <span class="at">.by =</span> mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb133-1383"><a href="#cb133-1383" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns</span></span>
<span id="cb133-1384"><a href="#cb133-1384" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, weight_lost_g, weight_lost_z)</span>
<span id="cb133-1385"><a href="#cb133-1385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1386"><a href="#cb133-1386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1387"><a href="#cb133-1387" aria-hidden="true" tabindex="-1"></a>Unlike when we used <span class="in">`.by`</span> with <span class="in">`summarise()`</span>, we still get the same number of rows as the original data frame, but now we have a new column <span class="in">`weight_lost_z`</span> that contains the Z-score for each mouse within each strain. This could be useful for identifying outliers or comparing the weight lost by each mouse to the average for its strain.</span>
<span id="cb133-1388"><a href="#cb133-1388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1389"><a href="#cb133-1389" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-1390"><a href="#cb133-1390" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-1391"><a href="#cb133-1391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1392"><a href="#cb133-1392" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1393"><a href="#cb133-1393" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which line of code would you use to calculate the median tail length of mice belonging to each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb133-1394"><a href="#cb133-1394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1395"><a href="#cb133-1395" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1396"><a href="#cb133-1396" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1397"><a href="#cb133-1397" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm), .by = mouse_strain)`</span></span>
<span id="cb133-1398"><a href="#cb133-1398" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1399"><a href="#cb133-1399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1400"><a href="#cb133-1400" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1401"><a href="#cb133-1401" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span></span>
<span id="cb133-1402"><a href="#cb133-1402" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1403"><a href="#cb133-1403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1404"><a href="#cb133-1404" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1405"><a href="#cb133-1405" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), by = mouse_strain)`</span></span>
<span id="cb133-1406"><a href="#cb133-1406" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1407"><a href="#cb133-1407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1408"><a href="#cb133-1408" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1409"><a href="#cb133-1409" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span></span>
<span id="cb133-1410"><a href="#cb133-1410" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1411"><a href="#cb133-1411" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1412"><a href="#cb133-1412" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1413"><a href="#cb133-1413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1414"><a href="#cb133-1414" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1415"><a href="#cb133-1415" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb133-1416"><a href="#cb133-1416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1419"><a href="#cb133-1419" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1420"><a href="#cb133-1420" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1421"><a href="#cb133-1421" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1422"><a href="#cb133-1422" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">max_tail_len =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, replicate))</span>
<span id="cb133-1423"><a href="#cb133-1423" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1424"><a href="#cb133-1424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1425"><a href="#cb133-1425" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1426"><a href="#cb133-1426" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1427"><a href="#cb133-1427" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb133-1428"><a href="#cb133-1428" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1429"><a href="#cb133-1429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1430"><a href="#cb133-1430" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1431"><a href="#cb133-1431" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice for each replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb133-1432"><a href="#cb133-1432" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1433"><a href="#cb133-1433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1434"><a href="#cb133-1434" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1435"><a href="#cb133-1435" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb133-1436"><a href="#cb133-1436" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1437"><a href="#cb133-1437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1438"><a href="#cb133-1438" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1439"><a href="#cb133-1439" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of mice in each unique combination of strain and replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1440"><a href="#cb133-1440" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1441"><a href="#cb133-1441" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1442"><a href="#cb133-1442" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1443"><a href="#cb133-1443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1444"><a href="#cb133-1444" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1445"><a href="#cb133-1445" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> I want to count how many male and how many female mice there are for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame. Which line of code would I use?</span>
<span id="cb133-1446"><a href="#cb133-1446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1447"><a href="#cb133-1447" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1448"><a href="#cb133-1448" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1449"><a href="#cb133-1449" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = sex)`</span></span>
<span id="cb133-1450"><a href="#cb133-1450" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1451"><a href="#cb133-1451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1452"><a href="#cb133-1452" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1453"><a href="#cb133-1453" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain)`</span></span>
<span id="cb133-1454"><a href="#cb133-1454" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1455"><a href="#cb133-1455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1456"><a href="#cb133-1456" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1457"><a href="#cb133-1457" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))`</span></span>
<span id="cb133-1458"><a href="#cb133-1458" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1459"><a href="#cb133-1459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1460"><a href="#cb133-1460" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1461"><a href="#cb133-1461" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain, sex)`</span></span>
<span id="cb133-1462"><a href="#cb133-1462" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1463"><a href="#cb133-1463" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1464"><a href="#cb133-1464" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1465"><a href="#cb133-1465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1466"><a href="#cb133-1466" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1467"><a href="#cb133-1467" aria-hidden="true" tabindex="-1"></a>4<span class="sc">\.</span> I want to find the proportion of weight lost **by each mouse** in each cage in the <span class="in">`mousezempic_dosage_data`</span> data frame. Which line of code would I use?</span>
<span id="cb133-1468"><a href="#cb133-1468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1469"><a href="#cb133-1469" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1470"><a href="#cb133-1470" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1471"><a href="#cb133-1471" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span></span>
<span id="cb133-1472"><a href="#cb133-1472" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1473"><a href="#cb133-1473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1474"><a href="#cb133-1474" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1475"><a href="#cb133-1475" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span></span>
<span id="cb133-1476"><a href="#cb133-1476" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1477"><a href="#cb133-1477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1478"><a href="#cb133-1478" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1479"><a href="#cb133-1479" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE, .by = cage_number))`</span></span>
<span id="cb133-1480"><a href="#cb133-1480" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1481"><a href="#cb133-1481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1482"><a href="#cb133-1482" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1483"><a href="#cb133-1483" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb133-1484"><a href="#cb133-1484" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1485"><a href="#cb133-1485" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1486"><a href="#cb133-1486" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1487"><a href="#cb133-1487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1488"><a href="#cb133-1488" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb133-1489"><a href="#cb133-1489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1490"><a href="#cb133-1490" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-1491"><a href="#cb133-1491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1492"><a href="#cb133-1492" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb133-1493"><a href="#cb133-1493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1494"><a href="#cb133-1494" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct line of code to calculate the median tail length of mice belonging to each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span>. Remember to use <span class="in">`na.rm = TRUE`</span> to remove any missing values before calculating the median, and to use <span class="in">`.by`</span> to specify the column to group by (not <span class="in">`by`</span>). Seeing as we want to calculate the median (collapse down to a single value per group), we need to use <span class="in">`summarise()`</span> rather than <span class="in">`mutate()`</span>.</span>
<span id="cb133-1495"><a href="#cb133-1495" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% summarise(max_tail_len = max(tail_length_mm, na.rm = TRUE), .by = c(mouse_strain, replicate))`</span> calculates the maximum tail length of mice in each **unique combination** of strain and replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb133-1496"><a href="#cb133-1496" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct line of code to count how many male and how many female mice there are for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))`</span>. We need to group by both <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> to get the count for each unique combination of strain and sex. Don't forget that we specify the column names as a vector when grouping by multiple columns.</span>
<span id="cb133-1497"><a href="#cb133-1497" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The correct line of code to find the proportion of weight lost **by each mouse** in each cage in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span>. We use <span class="in">`mutate()`</span> because we want a value for each mouse (each row in our data), rather than to collapse down to a single value for each group (cage number in this case). Be careful that you use the <span class="in">`.by`</span> argument within the <span class="in">`mutate()`</span> function call, not within the <span class="in">`sum()`</span> function by mistake (this is what is wrong with the third option).</span>
<span id="cb133-1498"><a href="#cb133-1498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1499"><a href="#cb133-1499" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb133-1500"><a href="#cb133-1500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1501"><a href="#cb133-1501" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-1502"><a href="#cb133-1502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1503"><a href="#cb133-1503" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1504"><a href="#cb133-1504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1505"><a href="#cb133-1505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1506"><a href="#cb133-1506" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reshaping and combining data {#sec-reshaping}</span></span>
<span id="cb133-1507"><a href="#cb133-1507" aria-hidden="true" tabindex="-1"></a>In the previous section, we learned how to filter, select, mutate, and summarise data. These are the most common operations you'll perform on your data, and all generally revolve around working with specific columns or rows of your data frame. But what if you need to change the structure of your entire data frame? Re-shaping (with <span class="in">`pivot_`</span> functions) and combining (with <span class="in">`join`</span> functions) data are two more advanced operations that can help you do this.</span>
<span id="cb133-1508"><a href="#cb133-1508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1509"><a href="#cb133-1509" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reshaping data with pivot functions {#sec-pivot}</span></span>
<span id="cb133-1510"><a href="#cb133-1510" aria-hidden="true" tabindex="-1"></a>Pivoting is a way to change the structure of your data frame by rotating it so that rows become columns and vice versa. This can be useful when you want to change the way your data is organised, for example, to make it easier to plot or analyse.</span>
<span id="cb133-1511"><a href="#cb133-1511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1512"><a href="#cb133-1512" aria-hidden="true" tabindex="-1"></a><span class="al">![Pivot functions allow you to change the structure of your data frame](images/S02F10_pivot.png)</span>{width=50%}</span>
<span id="cb133-1513"><a href="#cb133-1513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1514"><a href="#cb133-1514" aria-hidden="true" tabindex="-1"></a>The <span class="in">`pivot_longer()`</span> function is used to pivot data from wide to long format, and the <span class="in">`pivot_wider()`</span> function is used to pivot data from long to wide format.</span>
<span id="cb133-1515"><a href="#cb133-1515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1516"><a href="#cb133-1516" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot wider</span></span>
<span id="cb133-1517"><a href="#cb133-1517" aria-hidden="true" tabindex="-1"></a>A common use case for <span class="in">`pivot_wider()`</span> is to make a <span class="co">[</span><span class="ot">contingency table</span><span class="co">](https://en.wikipedia.org/wiki/Contingency_table)</span>,</span>
<span id="cb133-1518"><a href="#cb133-1518" aria-hidden="true" tabindex="-1"></a>which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</span>
<span id="cb133-1519"><a href="#cb133-1519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1520"><a href="#cb133-1520" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to create a table that shows how many mice there are of each strain, in each cage number. We can achieve this in a long format using <span class="in">`summarise()`</span> as we learned in the previous section:</span>
<span id="cb133-1521"><a href="#cb133-1521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1524"><a href="#cb133-1524" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1525"><a href="#cb133-1525" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1526"><a href="#cb133-1526" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1527"><a href="#cb133-1527" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb133-1528"><a href="#cb133-1528" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span>
<span id="cb133-1529"><a href="#cb133-1529" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1530"><a href="#cb133-1530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1531"><a href="#cb133-1531" aria-hidden="true" tabindex="-1"></a>Notice how we have a row for each unique combination of <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span>. Instead, we can use <span class="in">`pivot_wider()`</span> to create a table where each row represents a unique <span class="in">`cage_number`</span> and each column represents a unique <span class="in">`mouse_strain`</span>.</span>
<span id="cb133-1532"><a href="#cb133-1532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1533"><a href="#cb133-1533" aria-hidden="true" tabindex="-1"></a>To use a <span class="in">`pivot_wider()`</span> you need to tell R two bits of information:</span>
<span id="cb133-1534"><a href="#cb133-1534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1535"><a href="#cb133-1535" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>What the new columns should be (with <span class="in">`names_from`</span>)</span>
<span id="cb133-1536"><a href="#cb133-1536" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Where the values in these columns should come from (with <span class="in">`values_from`</span>).</span>
<span id="cb133-1537"><a href="#cb133-1537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1538"><a href="#cb133-1538" aria-hidden="true" tabindex="-1"></a>In this case, we want the mouse strains to be our new columns (so <span class="in">`names_from = mouse_strain`</span>), and the average weight lost to be the values in the table (<span class="in">`values_from = n_mice`</span>). Putting it all together, here is the code to make our summary table:</span>
<span id="cb133-1539"><a href="#cb133-1539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1542"><a href="#cb133-1542" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1543"><a href="#cb133-1543" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1544"><a href="#cb133-1544" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1545"><a href="#cb133-1545" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb133-1546"><a href="#cb133-1546" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1547"><a href="#cb133-1547" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb133-1548"><a href="#cb133-1548" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1549"><a href="#cb133-1549" aria-hidden="true" tabindex="-1"></a>Notice how <span class="in">`NA`</span> values are used where there is no data for a particular combination of <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span>.</span>
<span id="cb133-1550"><a href="#cb133-1550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1551"><a href="#cb133-1551" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot longer</span></span>
<span id="cb133-1552"><a href="#cb133-1552" aria-hidden="true" tabindex="-1"></a>While wide data can be easier to read, long data is often easier to work with. That's because a lot of the tidyverse functions we've learned so far (like <span class="in">`filter()`</span>, <span class="in">`mutate()`</span>, and <span class="in">`summarise()`</span>) work in a row-wise fashion, which is more natural with long data.</span>
<span id="cb133-1553"><a href="#cb133-1553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1554"><a href="#cb133-1554" aria-hidden="true" tabindex="-1"></a>To demonstrate <span class="in">`pivot_longer()`</span>, we will introduce a new data frame called <span class="in">`mousezempic_expression_data`</span>. This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate:</span>
<span id="cb133-1555"><a href="#cb133-1555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1558"><a href="#cb133-1558" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1559"><a href="#cb133-1559" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data</span></span>
<span id="cb133-1560"><a href="#cb133-1560" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb133-1561"><a href="#cb133-1561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1562"><a href="#cb133-1562" aria-hidden="true" tabindex="-1"></a><span class="co"># print it out</span></span>
<span id="cb133-1563"><a href="#cb133-1563" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data</span>
<span id="cb133-1564"><a href="#cb133-1564" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1565"><a href="#cb133-1565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1566"><a href="#cb133-1566" aria-hidden="true" tabindex="-1"></a>As you can see, the data is in wide format, with each row representing a different mouse (identified by its <span class="in">`id_num`</span>) and each column representing a different measurement of a gene. To put this data into a long format (where each measurement is contained on a separate row), we can use <span class="in">`pivot_longer()`</span>, specifying three arguments:</span>
<span id="cb133-1567"><a href="#cb133-1567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1568"><a href="#cb133-1568" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="in">`cols`</span>: the columns to pivot from. You can use selection helpers like <span class="in">`contains()`</span> or <span class="in">`starts_with()`</span> to easily select multiple columns at once.</span>
<span id="cb133-1569"><a href="#cb133-1569" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="in">`names_to`</span>: the name of a new column that will contain the old column names.</span>
<span id="cb133-1570"><a href="#cb133-1570" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="in">`values_to`</span>: the name of a new column that will contain the values from the old columns.</span>
<span id="cb133-1571"><a href="#cb133-1571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1572"><a href="#cb133-1572" aria-hidden="true" tabindex="-1"></a>In this particular case here's what the code would look like:</span>
<span id="cb133-1573"><a href="#cb133-1573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1576"><a href="#cb133-1576" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1577"><a href="#cb133-1577" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1578"><a href="#cb133-1578" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>), <span class="at">names_to =</span> <span class="st">"measurement"</span>, <span class="at">values_to =</span> <span class="st">"expression_level"</span>)</span>
<span id="cb133-1579"><a href="#cb133-1579" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1580"><a href="#cb133-1580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1581"><a href="#cb133-1581" aria-hidden="true" tabindex="-1"></a>Let's go through those three arguments:</span>
<span id="cb133-1582"><a href="#cb133-1582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1583"><a href="#cb133-1583" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="in">`cols = contains("_rep")`</span>: This tells R to pivot all columns that contain the string "_rep". In this case, that's the columns <span class="in">`TH_rep1`</span>, <span class="in">`TH_rep2`</span>, <span class="in">`TH_rep3`</span>, <span class="in">`PRLH_rep1`</span>, <span class="in">`PRLH_rep2`</span>, <span class="in">`PRLH_rep3`</span>, <span class="in">`HPRT1_rep1`</span>, <span class="in">`HPRT1_rep2`</span>, and <span class="in">`HPRT1_rep3`</span>.</span>
<span id="cb133-1584"><a href="#cb133-1584" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="in">`names_to = "measurement"`</span>: This tells R to create a new column called 'measurement' that contains the names of the old columns. So, for example, the value of the 'measurement' column for the first row will be 'TH_rep1'.</span>
<span id="cb133-1585"><a href="#cb133-1585" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="in">`values_to = "expression_level"`</span>: This tells R to create a new column called 'expression_level' that contains the values from the old columns. So, for example, the value of the 'expression_level' column for the first row will be the value of the <span class="in">`TH_rep1`</span> column.</span>
<span id="cb133-1586"><a href="#cb133-1586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1587"><a href="#cb133-1587" aria-hidden="true" tabindex="-1"></a>After running this code, you'll see that the data frame is now in long format, with each row representing a different measurement of a gene for a different mouse.</span>
<span id="cb133-1588"><a href="#cb133-1588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1589"><a href="#cb133-1589" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Perplexed by pivoting?"}</span>
<span id="cb133-1590"><a href="#cb133-1590" aria-hidden="true" tabindex="-1"></a>Pivoting can be a bit tricky to get your head around! Often when you're doing analysis, you'll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <span class="co">[</span><span class="ot">like those in the R for Data Science book</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html#sec-pivoting)</span>, or to just experiment with different arguments until you get the result you want.</span>
<span id="cb133-1591"><a href="#cb133-1591" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1592"><a href="#cb133-1592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1593"><a href="#cb133-1593" aria-hidden="true" tabindex="-1"></a><span class="fu">### Combining data with join functions {#sec-join}</span></span>
<span id="cb133-1594"><a href="#cb133-1594" aria-hidden="true" tabindex="-1"></a>Another way you might need to change the structure of your data is by combining it with other data frames. This is where the <span class="in">`join`</span> functions come in. There are many different types of joins in <span class="in">`dplyr`</span> (which you can read about here if you're interested), but the most useful one is the <span class="in">`left_join()`</span> function.</span>
<span id="cb133-1595"><a href="#cb133-1595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1596"><a href="#cb133-1596" aria-hidden="true" tabindex="-1"></a><span class="al">![The left_join() function allows you to combine two data frames](images/S02F11_left_join.png)</span></span>
<span id="cb133-1597"><a href="#cb133-1597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1598"><a href="#cb133-1598" aria-hidden="true" tabindex="-1"></a><span class="in">`left_join()`</span> is used to combine two data frames based on a common column. It takes three arguments - the first data frame, the second data frame, and the common column to join on. The reason it's called a 'left join' is because it keeps all the rows from the first data frame (the 'left' data frame), and only adds rows from the second data frame (the 'right' data frame) if they match the common column.</span>
<span id="cb133-1599"><a href="#cb133-1599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1600"><a href="#cb133-1600" aria-hidden="true" tabindex="-1"></a>For example, we can join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column, which is common to both data frames:</span>
<span id="cb133-1601"><a href="#cb133-1601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1604"><a href="#cb133-1604" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1605"><a href="#cb133-1605" aria-hidden="true" tabindex="-1"></a><span class="co"># example of using it in a pipe</span></span>
<span id="cb133-1606"><a href="#cb133-1606" aria-hidden="true" tabindex="-1"></a><span class="co"># here mousezempic_dosage_data is the left data frame (first) data frame passed on by the pipe</span></span>
<span id="cb133-1607"><a href="#cb133-1607" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1608"><a href="#cb133-1608" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="st">"id_num"</span>)</span>
<span id="cb133-1609"><a href="#cb133-1609" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1610"><a href="#cb133-1610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1611"><a href="#cb133-1611" aria-hidden="true" tabindex="-1"></a>This code will return a new data frame that contains all the columns from <span class="in">`mousezempic_dosage_data`</span> and all the columns from <span class="in">`mousezempic_expression_data`</span>, with the rows matched based on the <span class="in">`id_num`</span> column. If there is no match for a particular <span class="in">`id_num`</span> in the <span class="in">`mousezempic_expression_data`</span> data frame, the columns from that data frame will be filled with <span class="in">`NA`</span>. Rows that only appear in the <span class="in">`mousezempic_expression_data`</span> data frame will not be included in the final data frame.</span>
<span id="cb133-1612"><a href="#cb133-1612" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1613"><a href="#cb133-1613" aria-hidden="true" tabindex="-1"></a>We can also use <span class="in">`left_join()`</span> when our columns have different names (but contain the same value) using a named vector the format <span class="in">`c("name1" = "name2")`</span>, where 'name1' is the name of the column in the left/first data frame (the one being piped in) and 'name2' is the name of the corresponding column in the right/second data frame.</span>
<span id="cb133-1614"><a href="#cb133-1614" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1615"><a href="#cb133-1615" aria-hidden="true" tabindex="-1"></a>For example, if the <span class="in">`id_num`</span> column in the <span class="in">`mousezempic_dosage_data`</span> data frame was called <span class="in">`mouse_id`</span> instead, we could still join the two data frames using the following code:</span>
<span id="cb133-1616"><a href="#cb133-1616" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1619"><a href="#cb133-1619" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1620"><a href="#cb133-1620" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1621"><a href="#cb133-1621" aria-hidden="true" tabindex="-1"></a>  <span class="co"># for the sake of this example, make a new column called mouse_id</span></span>
<span id="cb133-1622"><a href="#cb133-1622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_id =</span> id_num) <span class="sc">%&gt;%</span></span>
<span id="cb133-1623"><a href="#cb133-1623" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join the two data frames where mouse_id in the first data frame matches id_num in the second data frame</span></span>
<span id="cb133-1624"><a href="#cb133-1624" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"mouse_id"</span> <span class="ot">=</span> <span class="st">"id_num"</span>))</span>
<span id="cb133-1625"><a href="#cb133-1625" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1626"><a href="#cb133-1626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1627"><a href="#cb133-1627" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb133-1628"><a href="#cb133-1628" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb133-1629"><a href="#cb133-1629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1630"><a href="#cb133-1630" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1631"><a href="#cb133-1631" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What does the following code do?</span>
<span id="cb133-1632"><a href="#cb133-1632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1635"><a href="#cb133-1635" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1636"><a href="#cb133-1636" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1637"><a href="#cb133-1637" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1638"><a href="#cb133-1638" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1639"><a href="#cb133-1639" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-1640"><a href="#cb133-1640" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1641"><a href="#cb133-1641" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span>
<span id="cb133-1642"><a href="#cb133-1642" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1643"><a href="#cb133-1643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1644"><a href="#cb133-1644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1645"><a href="#cb133-1645" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1646"><a href="#cb133-1646" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1647"><a href="#cb133-1647" aria-hidden="true" tabindex="-1"></a>Pivots data into a wide format where there is a column for each sex.</span>
<span id="cb133-1648"><a href="#cb133-1648" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1649"><a href="#cb133-1649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1650"><a href="#cb133-1650" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1651"><a href="#cb133-1651" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots into a wide format where there is a column for each sex.</span>
<span id="cb133-1652"><a href="#cb133-1652" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1653"><a href="#cb133-1653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1654"><a href="#cb133-1654" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1655"><a href="#cb133-1655" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots into a wide format where there is a column for each mouse strain.</span>
<span id="cb133-1656"><a href="#cb133-1656" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1657"><a href="#cb133-1657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1658"><a href="#cb133-1658" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1659"><a href="#cb133-1659" aria-hidden="true" tabindex="-1"></a>It just gives an error</span>
<span id="cb133-1660"><a href="#cb133-1660" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1661"><a href="#cb133-1661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1662"><a href="#cb133-1662" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1663"><a href="#cb133-1663" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1664"><a href="#cb133-1664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1665"><a href="#cb133-1665" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> I have run the following code to create a new column in the <span class="in">`mousezempic_dosage_data`</span> data frame that gives the weight of the mice at the end of the experiment.</span>
<span id="cb133-1666"><a href="#cb133-1666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1669"><a href="#cb133-1669" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1670"><a href="#cb133-1670" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1671"><a href="#cb133-1671" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb133-1672"><a href="#cb133-1672" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb133-1673"><a href="#cb133-1673" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb133-1674"><a href="#cb133-1674" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span>
<span id="cb133-1675"><a href="#cb133-1675" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1676"><a href="#cb133-1676" aria-hidden="true" tabindex="-1"></a>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</span>
<span id="cb133-1677"><a href="#cb133-1677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1678"><a href="#cb133-1678" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1679"><a href="#cb133-1679" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1680"><a href="#cb133-1680" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")`</span></span>
<span id="cb133-1681"><a href="#cb133-1681" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1682"><a href="#cb133-1682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1683"><a href="#cb133-1683" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1684"><a href="#cb133-1684" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")`</span></span>
<span id="cb133-1685"><a href="#cb133-1685" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1686"><a href="#cb133-1686" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1687"><a href="#cb133-1687" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1688"><a href="#cb133-1688" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)`</span></span>
<span id="cb133-1689"><a href="#cb133-1689" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1690"><a href="#cb133-1690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1691"><a href="#cb133-1691" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1692"><a href="#cb133-1692" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span></span>
<span id="cb133-1693"><a href="#cb133-1693" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1694"><a href="#cb133-1694" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1695"><a href="#cb133-1695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1696"><a href="#cb133-1696" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb133-1697"><a href="#cb133-1697" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Which of the following is NOT a valid way to join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column?</span>
<span id="cb133-1698"><a href="#cb133-1698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1699"><a href="#cb133-1699" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb133-1700"><a href="#cb133-1700" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1701"><a href="#cb133-1701" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = "id_num")`</span></span>
<span id="cb133-1702"><a href="#cb133-1702" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1703"><a href="#cb133-1703" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1704"><a href="#cb133-1704" aria-hidden="true" tabindex="-1"></a><span class="in">`left_join(mousezempic_dosage_data, mousezempic_expression_data, by = "id_num")`</span></span>
<span id="cb133-1705"><a href="#cb133-1705" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1706"><a href="#cb133-1706" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb133-1707"><a href="#cb133-1707" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")`</span></span>
<span id="cb133-1708"><a href="#cb133-1708" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1709"><a href="#cb133-1709" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb133-1710"><a href="#cb133-1710" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = ("id_num" = "id_num"))`</span></span>
<span id="cb133-1711"><a href="#cb133-1711" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1712"><a href="#cb133-1712" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1713"><a href="#cb133-1713" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1714"><a href="#cb133-1714" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1715"><a href="#cb133-1715" aria-hidden="true" tabindex="-1"></a>&lt;details&gt; &lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb133-1716"><a href="#cb133-1716" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The code first calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <span class="in">`names_from = sex`</span> )</span>
<span id="cb133-1717"><a href="#cb133-1717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1718"><a href="#cb133-1718" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> The correct pivot function call to take the data from a wide format to a long format is <span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span>. This code tells R to pivot the <span class="in">`initial_weight_g`</span> and <span class="in">`final_weight_g`</span> columns into a long format, where there is a row for each mouse and each weight measurement. The <span class="in">`names_to`</span> argument specifies to make a column called 'timepoint' that tells us whether the measurement is initial or final, and the <span class="in">`values_to`</span> argument specifies the name of the new column that will contain these measurements.</span>
<span id="cb133-1719"><a href="#cb133-1719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1720"><a href="#cb133-1720" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> The line of code that is NOT a valid way to join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column is <span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")`</span>. This line of code is incorrect because the <span class="in">`.by`</span> argument is not used in the <span class="in">`left_join()`</span> function (this can be confusing! it's <span class="in">`.by`</span> when grouping by <span class="in">`by`</span> when joining). The other options are valid ways to join the two data frames based on the <span class="in">`id_num`</span> column: remember that we don't have to use pipes to join data frames, we can use the <span class="in">`left_join()`</span> function directly, and we can use a named vector to specify the columns to join on (although here it's a bit redundant as the columns have the same name).</span>
<span id="cb133-1721"><a href="#cb133-1721" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1722"><a href="#cb133-1722" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb133-1723"><a href="#cb133-1723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1724"><a href="#cb133-1724" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1725"><a href="#cb133-1725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1726"><a href="#cb133-1726" aria-hidden="true" tabindex="-1"></a><span class="fu">## Saving data to a file {#sec-saving}</span></span>
<span id="cb133-1727"><a href="#cb133-1727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1728"><a href="#cb133-1728" aria-hidden="true" tabindex="-1"></a>Once you've cleaned and transformed your data, you'll often want to save it to a file so that you can use it in other programs or share it with others. The <span class="in">`write_csv()`</span> and <span class="in">`write_tsv()`</span> functions from the <span class="in">`readr`</span> package are a great way to do this. They take two arguments - the data frame you want to save and the file path where you want to save it.</span>
<span id="cb133-1729"><a href="#cb133-1729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1730"><a href="#cb133-1730" aria-hidden="true" tabindex="-1"></a>For example, let's say I want to save my contingency table of the number of mice in each cage and strain to a CSV file called <span class="in">`cage_strain_counts.csv`</span>:</span>
<span id="cb133-1731"><a href="#cb133-1731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1734"><a href="#cb133-1734" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1735"><a href="#cb133-1735" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1736"><a href="#cb133-1736" aria-hidden="true" tabindex="-1"></a><span class="co"># create the contingency table</span></span>
<span id="cb133-1737"><a href="#cb133-1737" aria-hidden="true" tabindex="-1"></a>cage_strain_counts <span class="ot">&lt;-</span> mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1738"><a href="#cb133-1738" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1739"><a href="#cb133-1739" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb133-1740"><a href="#cb133-1740" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1741"><a href="#cb133-1741" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb133-1742"><a href="#cb133-1742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1743"><a href="#cb133-1743" aria-hidden="true" tabindex="-1"></a><span class="co"># save the data to a CSV file</span></span>
<span id="cb133-1744"><a href="#cb133-1744" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(cage_strain_counts, <span class="st">"cage_strain_counts.csv"</span>)</span>
<span id="cb133-1745"><a href="#cb133-1745" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1746"><a href="#cb133-1746" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1747"><a href="#cb133-1747" aria-hidden="true" tabindex="-1"></a>CSV files are particularly great because they can be easily read into other software, like Excel.</span>
<span id="cb133-1748"><a href="#cb133-1748" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1749"><a href="#cb133-1749" aria-hidden="true" tabindex="-1"></a>It's also possible to use the <span class="in">`write_*()`</span> functions along with a pipe:</span>
<span id="cb133-1750"><a href="#cb133-1750" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1753"><a href="#cb133-1753" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb133-1754"><a href="#cb133-1754" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb133-1755"><a href="#cb133-1755" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb133-1756"><a href="#cb133-1756" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb133-1757"><a href="#cb133-1757" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb133-1758"><a href="#cb133-1758" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb133-1759"><a href="#cb133-1759" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice) <span class="sc">%&gt;%</span></span>
<span id="cb133-1760"><a href="#cb133-1760" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_csv</span>(<span class="st">"cage_strain_counts.csv"</span>)</span>
<span id="cb133-1761"><a href="#cb133-1761" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb133-1762"><a href="#cb133-1762" aria-hidden="true" tabindex="-1"></a>Remember here that the first argument (the data frame to save) is passed on by the pipe, so the only argument in the brackets is the second one: the file path.</span>
<span id="cb133-1763"><a href="#cb133-1763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1764"><a href="#cb133-1764" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb133-1765"><a href="#cb133-1765" aria-hidden="true" tabindex="-1"></a>Here's what we've covered in this session:</span>
<span id="cb133-1766"><a href="#cb133-1766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1767"><a href="#cb133-1767" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The pipe operator <span class="in">`%&gt;%`</span> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</span>
<span id="cb133-1768"><a href="#cb133-1768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1769"><a href="#cb133-1769" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The basic dplyr verbs <span class="in">`filter()`</span>, <span class="in">`select()`</span>, <span class="in">`mutate()`</span>, and <span class="in">`arrange()`</span> and how they can be used to tidy and analyse data.</span>
<span id="cb133-1770"><a href="#cb133-1770" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1771"><a href="#cb133-1771" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Missing values (<span class="in">`NA`</span>) and how to remove or replace them</span>
<span id="cb133-1772"><a href="#cb133-1772" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1773"><a href="#cb133-1773" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The <span class="in">`summarise()`</span> function and how it can be used to calculate summary statistics on your data, as well as the power of grouping data with the <span class="in">`.by`</span> argument.</span>
<span id="cb133-1774"><a href="#cb133-1774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1775"><a href="#cb133-1775" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Reshaping data with <span class="in">`pivot_longer()`</span> and <span class="in">`pivot_wider()`</span> to change the structure of your data frame.</span>
<span id="cb133-1776"><a href="#cb133-1776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1777"><a href="#cb133-1777" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Combining data with <span class="in">`left_join()`</span> to merge two data frames based on a common column.</span>
<span id="cb133-1778"><a href="#cb133-1778" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1779"><a href="#cb133-1779" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Why does data need to be tidy anyway?"}</span>
<span id="cb133-1780"><a href="#cb133-1780" aria-hidden="true" tabindex="-1"></a>In this session, we've been focusing on making our data 'tidy': that is, structured in a consistent way that makes it easy to work with. A nice visual illustration of tidy data and its importance can be <span class="co">[</span><span class="ot">found here</span><span class="co">](https://allisonhorst.com/other-r-fun)</span>.</span>
<span id="cb133-1781"><a href="#cb133-1781" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb133-1782"><a href="#cb133-1782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-1783"><a href="#cb133-1783" aria-hidden="true" tabindex="-1"></a><span class="fu">### Practice questions</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>