<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.29">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>6&nbsp; Session 6: RNA-seq part 2 – WEHI R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./further_reading.html" rel="next">
<link href="./session_5.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-cc48a02baa7d92beca338a7447a78b59.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-8b16d95f9cb9693e68e2bdbbe3456629.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session_6.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Session 6: RNA-seq part 2</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">WEHI R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Session 3: Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Session 5: RNA-seq part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_6.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Session 6: RNA-seq part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this session:</h2>
   
  <ul>
  <li><a href="#linear-modelling" id="toc-linear-modelling" class="nav-link active" data-scroll-target="#linear-modelling"><span class="header-section-number">6.1</span> Linear modelling</a>
  <ul class="collapse">
  <li><a href="#design-matrix" id="toc-design-matrix" class="nav-link" data-scroll-target="#design-matrix"><span class="header-section-number">6.1.1</span> Design matrix</a></li>
  <li><a href="#contrasts" id="toc-contrasts" class="nav-link" data-scroll-target="#contrasts"><span class="header-section-number">6.1.2</span> Contrasts</a></li>
  <li><a href="#variance-modelling-with-voom" id="toc-variance-modelling-with-voom" class="nav-link" data-scroll-target="#variance-modelling-with-voom"><span class="header-section-number">6.1.3</span> Variance modelling with voom</a></li>
  <li><a href="#fitting-the-linear-model" id="toc-fitting-the-linear-model" class="nav-link" data-scroll-target="#fitting-the-linear-model"><span class="header-section-number">6.1.4</span> Fitting the linear model</a></li>
  </ul></li>
  <li><a href="#statistical-testing" id="toc-statistical-testing" class="nav-link" data-scroll-target="#statistical-testing"><span class="header-section-number">6.2</span> Statistical testing</a></li>
  <li><a href="#ma-plot" id="toc-ma-plot" class="nav-link" data-scroll-target="#ma-plot"><span class="header-section-number">6.3</span> MA Plot</a></li>
  <li><a href="#volcano-plot" id="toc-volcano-plot" class="nav-link" data-scroll-target="#volcano-plot"><span class="header-section-number">6.4</span> Volcano Plot</a></li>
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap"><span class="header-section-number">6.5</span> Heatmap</a></li>
  <li><a href="#saving-all-the-results" id="toc-saving-all-the-results" class="nav-link" data-scroll-target="#saving-all-the-results"><span class="header-section-number">6.6</span> Saving all the results</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">6.7</span> References</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Session 6: RNA-seq part 2</span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>The aim of this session is to complete the analysis of the RNA-seq data using the <code>limma</code> and <code>edgeR</code> packages. With the results of the analysis, we want to be able to produce publication quality figures and write out the results to a file for sharing or publication.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>fit a linear model to the data</li>
<li>construct design and contrast matrices for the linear model</li>
<li>model the mean-variance relationship of the data using the <code>voom()</code> function</li>
<li>perform statistical testing to identify differentially expressed genes</li>
<li>visualise the results using an MA plot</li>
<li>write out the results to a file</li>
</ul>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: limma</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">parse_factor</span>(</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Basal"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>samplenames <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"10_6_5_11"</span>, <span class="st">"9_6_5_11"</span>, <span class="st">"purep53"</span>, <span class="st">"JMS8-2"</span>, <span class="st">"JMS8-3"</span>,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"JMS8-4"</span>, <span class="st">"JMS8-5"</span>, <span class="st">"JMS9-P7c"</span>, <span class="st">"JMS9-P8c"</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co"># create DGEList object</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">readDGE</span>(</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir</span>(<span class="at">path =</span> <span class="st">"data/counts"</span>, <span class="at">pattern =</span> <span class="st">"GSM*"</span>),</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"data/counts"</span>,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">columns =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> group,</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> samplenames</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="co"># add gene annotation information</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>genes <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/Ses3_geneAnnot.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 27179 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (2): SYMBOL, TXCHROM
dbl (1): ENTREZID

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter lowly expressed genes</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep, ]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># normalise the data for library size and RNA composition</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="linear-modelling" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="linear-modelling"><span class="header-section-number">6.1</span> Linear modelling</h2>
<p>To identify differentially expressed genes from our normalised gene expression data, we will use the <code>limma</code> package to fit linear models to the genes. A linear model is a broad class of statistical models that fit the value of an response (or dependent) variable as a linear function of one or more explanatory (or independent) variables (also called covariates).</p>
<p>The general form of a linear model looks like this:</p>
<p><span class="math inline">\(Y = \beta_{0} + \beta_{1}X_{1} + \beta_{2}X_{2}... + \epsilon\)</span></p>
<p>This equation is saying that a response variable of interest <span class="math inline">\(Y\)</span> is equal to a constant (<span class="math inline">\(\beta_{0}\)</span>) plus the sum of the covariates (<span class="math inline">\(X_{i}\)</span>) each multiplied by a constant coefficient (<span class="math inline">\(\beta_{i}\)</span>) and some error term (<span class="math inline">\(\epsilon\)</span>). The error term is the difference between the observed value and the predicted value of the model and is assumed to have a normal distribution with mean 0 and some variance.</p>
<p>Our experiment is quite simple, since there is only a single covariate, the cell type. The true benefit of using linear models is in its ability to accommodate more complex designs including multiple covariates.</p>
<p>To fit the linear models in the limma-voom framework we need two objects in addition to our data:</p>
<ul>
<li>A design matrix, representing the covariates.</li>
<li>A contrast matrix, representing the specific comparison we wish to make.</li>
</ul>
<section id="design-matrix" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="design-matrix"><span class="header-section-number">6.1.1</span> Design matrix</h3>
<p>The first step to fitting a linear model is to specify a design matrix. The design matrix specifies the values of the covariates for each sample, and is represented as a matrix due to the mathematical convenience.</p>
<p>To generate a design matrix. We use the function <code>model.matrix()</code>, with the expression <code>~0 + group</code>. This returns a matrix representing the design where there is no intercept term and group is the only covariate. If we omit the <code>0</code> then there would be an intercept in the model, and if we included more covariates then more columns would be generated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span><span class="dv">0</span> <span class="sc">+</span> group, <span class="at">data =</span> dge<span class="sc">$</span>samples)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          groupBasal groupLP groupML
10_6_5_11          0       1       0
9_6_5_11           0       0       1
purep53            1       0       0
JMS8-2             1       0       0
JMS8-3             0       0       1
JMS8-4             0       1       0
JMS8-5             1       0       0
JMS9-P7c           0       0       1
JMS9-P8c           0       1       0
attr(,"assign")
[1] 1 1 1
attr(,"contrasts")
attr(,"contrasts")$group
[1] "contr.treatment"</code></pre>
</div>
</div>
<p>There are 9 rows, one for each sample. Along the columns are the names of the groups. The values in the cells denote membership of the particular sample for a particular group, as our groups in this case are mutually exclusive, each row contains only a single 1 to denote membership in a single group.</p>
</section>
<section id="contrasts" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="contrasts"><span class="header-section-number">6.1.2</span> Contrasts</h3>
<p>Contrast matrices are companions to design matrices. They are used to specify the comparison of interest. In our case, we have three experimental groups: <code>Basal</code>, <code>LP</code> and <code>ML</code>. So if we are to perform differential expression analysis, we are most likely interested in the differences between only two of the groups at a time. Contrast matrices let use specify which comparison we’d like to make, and are also represented as a matrix just like the design.</p>
<p>A contrast matrix can be made using the <code>makeContrasts()</code> function. Within this function, we specify the name of each specific contrast and the formula for that contrast. For example, the <code>BasalvsLP</code> contrasts compares the difference between the <code>Basal</code> and <code>LP</code> groups. Note that the name of the phenotype groups must be written exactly as they are in the column names of our design matrix (see above).</p>
<p>In addition to the individual contrasts, the function must know about the design of the model. This is passed through the <code>levels</code> argument, which either accepts a matrix with the column names corresponding to levels of your experimental groups, or the levels themselves as a character vector. Here we first simplify the column names of our design matrix to make it easier to read. Then we create the contrast matrix using the <code>makeContrasts()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>contr_matrix <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">BasalvsLP =</span> <span class="st">"Basal - LP"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">BasalvsML =</span> <span class="st">"Basal - ML"</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">LPvsML =</span> <span class="st">"LP - ML"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> design <span class="co"># alternatively 'levels = colnames(design)'</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>contr_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Contrasts
Levels  BasalvsLP BasalvsML LPvsML
  Basal         1         1      0
  LP           -1         0      1
  ML            0        -1     -1</code></pre>
</div>
</div>
<p>There are two things to note about a design matrix. First, the sum of the numbers in each row is equal to 0. Second, the way that you set up the equation in the matrix will determine the interpretation of the log-fold-change calculated later, as well as the meaning of up-regulated and down-regulated genes. For example, the contrast <code>Basal - LP</code> will give a positive log-fold-change for genes that are up-regulated in the <code>Basal</code> group compared to the <code>LP</code> group. If we had set up the contrast as <code>LP - Basal</code>, then the opposite would be true, and we would get a negative log-fold-change for genes that are up-regulated in the <code>Basal</code> group compared to the <code>LP</code> group.</p>
</section>
<section id="variance-modelling-with-voom" class="level3" data-number="6.1.3">
<h3 data-number="6.1.3" class="anchored" data-anchor-id="variance-modelling-with-voom"><span class="header-section-number">6.1.3</span> Variance modelling with voom</h3>
<p>We are now ready to fit our linear models. Limma fits linear models to the data with the assumption that the underlying data is normally distributed. Count data is generally not normally distributed, but log transforming count data gives it a roughly normal distribution sufficient for linear models to work well. To do this, limma transforms the raw count data to log-cpm using library sizes and the normalisation factors we calculated previously.</p>
<p>In addition to the normalisation steps, the limma-voom pipeline uses the <code>voom()</code> function to generate weights for the individual genes based on a modelled mean-variance relationship. This modelling allows use to get more information out of small sample sizes as the weights prevent our model from being more heavily influenced by more variable data points. The weights will then allow the linear model to be less influenced by genes with high variability, which is important for small sample sizes. The voom function also generates a mean-variance plot, which is useful for visualising the data.</p>
<p>The <code>voom()</code> function takes as arguments, our <code>DGEList</code> object and our design matrix. It also optionally outputs a plot of the mean-variance relationship of our data, called the ‘voom-plot’.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">voom</span>(dge, design, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_6_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>The output of <code>voom()</code> (our variable <code>v</code>) is an <code>EList</code> object which contains the following elements:</p>
<ul>
<li><code>genes</code> - a data frame of gene annotation data.</li>
<li><code>targets</code> - data frame of sample data.</li>
<li><code>E</code> - numeric matrix of normalised log-cpm values.</li>
<li><code>weights</code> - numeric matrix of precision weights.</li>
<li><code>design</code> - the design matrix.</li>
</ul>
<p>The MA plot shows the mean-variance relationship of the data. The x-axis is the average log2 expression of each gene, and the y-axis is the log2 standard deviation of each gene. The idea is that the variance of a gene can be estimated as a function of its main expression, and that this estimate is more accurate than the direct calculation of the variance.</p>
</section>
<section id="fitting-the-linear-model" class="level3" data-number="6.1.4">
<h3 data-number="6.1.4" class="anchored" data-anchor-id="fitting-the-linear-model"><span class="header-section-number">6.1.4</span> Fitting the linear model</h3>
<p>We are now ready to fit our linear model with <code>lmFit()</code>, which calculates coefficients we defined in our design matrix <code>design</code>. The resulting object, <code>vfit</code> is a <code>MArrayLM</code> object. It contains a information about our genes (the same data frame as <code>genes</code> from our <code>EList</code> object <code>v</code> above), the design matrix and a number of statistical outputs. Of most interest to us is the coefficients, stored in an element called <code>coefficients</code>. The first rows of this matrix is shown below. Each gene is row and is labelled using the EntrezID. Each column gives coefficients for each of our phenotype groups. These coefficients are weighted averages of the log-cpm of each gene in each group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(v)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(vfit<span class="sc">$</span>coefficients)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Basal        LP        ML
497097  3.0238407 -4.490714 -3.944753
20671   0.2678024 -2.489068 -2.025190
27395   4.3267906  3.900754  4.365115
18777   5.2066347  4.975762  5.653810
21399   5.2105491  4.901520  4.876117
58175  -1.9300214  3.581004  3.133716</code></pre>
</div>
</div>
<p>We can then use <code>contrasts.fit()</code> to calculate coefficients for each contrast (or ‘comparison’) we specified in our <code>contr_matrix</code>. The output is also an object of the class <code>MArrayLM</code> (also known as an <code>MArrayLM</code> object). When we inspect the <code>coefficients</code> element now, we can see that each column is a contrast that we specified in our contrast matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(vfit, <span class="at">contrasts =</span> contr_matrix)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(vfit<span class="sc">$</span>coefficients)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Contrasts
          BasalvsLP   BasalvsML      LPvsML
  497097  7.5145551  6.96859344 -0.54596165
  20671   2.7568708  2.29299235 -0.46387848
  27395   0.4260362 -0.03832424 -0.46436046
  18777   0.2308732 -0.44717502 -0.67804818
  21399   0.3090295  0.33443212  0.02540261
  58175  -5.5110259 -5.06373764  0.44728824</code></pre>
</div>
</div>
<p>With these values we essentially want to know whether or not the difference in values is significantly different from 0. If the difference is not significantly different from 0, then we have failed to establish a difference in the expression of a particularly gene between two groups. If the difference is significantly greater than 0, then we have up-regulation in the first group of the contrast, and if the difference is significantly less than 0, then we have down-regulation in the first group of the contrast.</p>
</section>
</section>
<section id="statistical-testing" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="statistical-testing"><span class="header-section-number">6.2</span> Statistical testing</h2>
<p>To actually test if the values are significantly different from 0, we use the <code>eBayes()</code> function to compute moderated t-statistics, moderated F-statistics and log-odds of differential expression for each gene, given a fitted linear model. ‘Moderated’ refers to empirical Bayes moderation, which borrows information across genes to obtain more accurate measures of variability for each gene. This also increases our power to detect differentially expressed genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>efit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(vfit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now look at the number of differentially expressed genes using the <code>decideTests()</code> function. The output of this function is a matrix where each column is a contrast (comparison of interest) and each row is a gene. The numbers 1, -1 and 0 mean up-regulated, down-regulated or not significantly differentially expressed, respectively.</p>
<p>Note that <code>decideTests()</code> also accounts for multiple testing. The default method is Benjamini and Hochberg<span class="citation" data-cites="benjamini1995"><sup><a href="#ref-benjamini1995" role="doc-biblioref">1</a></sup></span> but several others are also available.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">decideTests</span>(efit)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>dt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>TestResults matrix
        Contrasts
         BasalvsLP BasalvsML LPvsML
  497097         1         1      0
  20671          1         1      0
  27395          0         0      0
  18777          0        -1     -1
  21399          0         1      0
16619 more rows ...</code></pre>
</div>
</div>
<p>To obtain the total number of differentially expressed genes for each comparison, we can add the function <code>summary()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       BasalvsLP BasalvsML LPvsML
Down        4501      4850   2701
NotSig      7306      6996  11821
Up          4817      4778   2102</code></pre>
</div>
</div>
<p>The function <code>topTable()</code> can be used to obtain more information on the differentially expressed genes for each contrast. <code>topTable()</code> takes as arguments the <code>MArrayLM</code> object output by <code>eBayes()</code> (<code>efit</code>), the contrast name of interest and the number of top differentially expressed genes to output. Note that the contrast name must be given in quotes and must be exactly as written in the contrast matrix <code>contr_matrix</code>.</p>
<p>It outputs a data frame with the following information:</p>
<ul>
<li><strong>Gene details</strong> - gene information, from the <code>gene</code> element of the <code>MArrayLM</code> object (<code>efit</code>).</li>
<li><code>logFC</code> - the log2 fold change of the contrast.</li>
<li><code>AveExpr</code> - the average log2 expression of that gene.</li>
<li><code>t</code> - moderated t-statistic.</li>
<li><code>P.Value</code> - p value.</li>
<li><code>adj.P.Val</code> - adjusted p value.</li>
<li><code>B</code> - log-odds that the gene is differentially expressed.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>top <span class="ot">&lt;-</span> <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsLP"</span>, <span class="at">n =</span> <span class="cn">Inf</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(top)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       ENTREZID   SYMBOL TXCHROM     logFC  AveExpr         t      P.Value
12521     12521     Cd82    chr2 -4.095480 7.069340 -35.47147 3.381469e-12
22249     22249   Unc13b    chr4 -4.350553 5.662874 -32.38385 8.656449e-12
16324     16324    Inhbb    chr1 -4.721420 6.460624 -30.81124 1.446147e-11
14245     14245    Lpin1   chr12 -3.768976 6.293719 -29.95387 1.934041e-11
218518   218518 Marveld2   chr13 -5.215230 4.929711 -30.87474 1.415794e-11
12759     12759      Clu   chr14 -5.306850 8.856284 -29.55444 2.220734e-11
          adj.P.Val        B
12521  4.655035e-08 18.41399
22249  4.655035e-08 17.42593
16324  4.655035e-08 17.06368
14245  4.655035e-08 16.86037
218518 4.655035e-08 16.76898
12759  4.655035e-08 16.70270</code></pre>
</div>
</div>
</section>
<section id="ma-plot" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="ma-plot"><span class="header-section-number">6.3</span> MA Plot</h2>
<p>The MA plot is a plot of log-fold-change (M-values) against log-expression averages (A-values), this is a common plot in RNA sequencing analysis to visualise the result of differential expression tests. It can be created using the <code>plotMA()</code> from the <code>limma</code> package. Creating this plot requires 3 pieces of information:</p>
<ul>
<li><code>object = efit</code>: The the fitted object containing the log-fold-change and log-expression averages</li>
<li><code>coef = 1</code>: The column number of the contrast to plot since there are 3 different contrasts fitted within the object.</li>
<li><code>status = dt[, 1]</code>: A vector of numerics denoting whether a gene is up-regulated or down-regulated.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMA</span>(efit, <span class="at">coef =</span> <span class="dv">1</span>, <span class="at">status =</span> dt[, <span class="st">"BasalvsLP"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_6_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>As before, we can try to re-create this plot using <code>ggplot2</code> to give us more flexibility in plotting. The information we’ll need is the log-fold-change, the average log-expression and the differential expression status of each gene. We can extract this information from the <code>efit</code> and <code>top</code> data frame we created above, as well as gene names for annotation purposes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>ma_plot_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">logFC =</span> efit<span class="sc">$</span>coefficients[, <span class="st">"BasalvsLP"</span>],</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">AvgExpr =</span> efit<span class="sc">$</span>Amean,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">status =</span> <span class="fu">factor</span>(<span class="fu">as.numeric</span>(dt[, <span class="st">"BasalvsLP"</span>])),</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> efit<span class="sc">$</span>genes<span class="sc">$</span>SYMBOL,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">p_value =</span> efit<span class="sc">$</span>p.value[, <span class="st">"BasalvsLP"</span>]</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With our tibble prepared, we can use <code>ggplot2</code> to create the plot as we have done before</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ma_plot_data <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> AvgExpr, <span class="at">y =</span> logFC, <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_6_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>While the coordinates of the points are the same, there is still quite a bit of work to do to make this plot look like the one produced by <code>plotMA()</code>. First we have the fix up the colours of the points. We also have to make it such that the non-significant points are plotted below the other points. To do this we would need to make it so that the non-significant points are plotted first. We can do this by reordering the levels of the <code>status</code> factor so that the non-significant points are plotted first. Labels can also be added using <code>ggrepel</code> for the top 15 most DE genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>ma_plot_data_fct <span class="ot">&lt;-</span> ma_plot_data <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">status =</span> <span class="fu">fct_recode</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>      status,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Not DE"</span> <span class="ot">=</span> <span class="st">"0"</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Down"</span> <span class="ot">=</span> <span class="st">"-1"</span>,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Up"</span> <span class="ot">=</span> <span class="st">"1"</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">status =</span> <span class="fu">fct_relevel</span>(status, <span class="st">"Not DE"</span>, <span class="st">"Down"</span>, <span class="st">"Up"</span>))</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>top_de_genes <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p_value) <span class="sc">%&gt;%</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">15</span>)</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>scale_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Not DE"</span> <span class="ot">=</span> <span class="st">"black"</span>,</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Down"</span> <span class="ot">=</span> <span class="st">"blue"</span>,</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Up"</span> <span class="ot">=</span> <span class="st">"red"</span></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>ma_plot <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> AvgExpr, <span class="at">y =</span> logFC, <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">==</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">!=</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> top_de_genes,</span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene, <span class="at">col =</span> <span class="cn">NULL</span>),</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.overlaps =</span> <span class="dv">20</span>,</span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> scale_cols)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="volcano-plot" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="volcano-plot"><span class="header-section-number">6.4</span> Volcano Plot</h2>
<p>Since we have the tibble of the data prepared, we can also choose to create an alternative plot called a volcano plot. Which is also very popular in RNA-seq analysis. The volcano plot is a scatter plot of the negative log10 p-value against the log2 fold-change. This plot is useful for visualising the significance of the differentially expressed genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>volcano_plot <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logFC, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p_value), <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">==</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">!=</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> top_de_genes,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene, <span class="at">col =</span> <span class="cn">NULL</span>),</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.overlaps =</span> <span class="dv">20</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"red"</span>))</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>volcano_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_6_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="heatmap" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="heatmap"><span class="header-section-number">6.5</span> Heatmap</h2>
<p>Finally we can create a heatmap of the differentially expressed genes. The heatmap is a useful way to visualise the expression of the differentially expressed genes across the samples. We can use the <code>pheatmap</code> package to create the heatmap.</p>
<p>First we need to install the package if you haven’t already</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"pheatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data we will use for the heatmap is going to be the log2 cpm expression values of the top differentially expressed genes. We can use the <code>topTable()</code> function to get the top differentially expressed genes for each contrast. We can then use the <code>cpm()</code> function to get the expression values of these genes. We will take the top 50 genes from each contrast and combine them into a single data frame. We will then use extract the expression values for these genes from the full expression matrix. For more interpretable labels, we will use the gene symbols as the row names of the expression matrix.</p>
<p>Note that by default the ENTREZID column is encoded as a numeric value, so to use it to subset our expression matrix we need to convert it to a character vector in order to select the matching IDs rather than trying to select the numerical rows of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsLP"</span>, <span class="at">n =</span> <span class="dv">50</span>),</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsML"</span>, <span class="at">n =</span> <span class="dv">50</span>),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"LPvsML"</span>, <span class="at">n =</span> <span class="dv">50</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ENTREZID, SYMBOL, TXCHROM) <span class="sc">%&gt;%</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>top_gene_expr <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)[<span class="fu">as.character</span>(top_genes<span class="sc">$</span>ENTREZID), ]</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(top_gene_expr) <span class="ot">&lt;-</span> top_genes<span class="sc">$</span>SYMBOL</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To create the heatmap we will use the <code>pheatmap()</code> function with the expression matrix. We will add in arguments to provide our own colour scale, to scale the data along the rows with a z-score, adding in sample annotation and clustering the data by rows and columns. We will also add in the option to show the row names and column names, and set the font size of the row names. You can experiment with removing each of these options to see how they affect the heatmap, as well as investigate additional options in the <code>pheatmap()</code> documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples <span class="sc">%&gt;%</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(group)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>heatmap <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  top_gene_expr,</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>),</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotation_col =</span> sample_anno,</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">border_color =</span> <span class="cn">NA</span>,</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_colnames =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontsize_row =</span> <span class="dv">6</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_6_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="saving-all-the-results" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="saving-all-the-results"><span class="header-section-number">6.6</span> Saving all the results</h2>
<p>Now that we have produced a series of plots and the top genes table, we can save these files to a directory. These figures may form the basis for publication figures, and the results table can partially presented as a result, or the full table can be included as a supplementary file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"output"</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(top, <span class="st">"output/top_genes_BasalvsLP.csv"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/ma_plot_BasalvsLP.png"</span>, ma_plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 10 x 5 in image</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/volcano_plot_BasalvsLP.png"</span>, volcano_plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 10 x 5 in image</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/heatmap_BasalvsLP.png"</span>, heatmap<span class="sc">$</span>gtable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 10 x 5 in image</code></pre>
</div>
</div>
<p>With that we have come to the conclusion of the RNA-seq analysis and this course. You have now learned how to use various <code>tidyverse</code> packages to operate on data and create publication quality figures. You have also learned how to use the <code>limma</code> and <code>edgeR</code> packages to perform differential expression analysis on RNA-seq data. You should now be able to apply these skills to perform analysis on your own data.</p>
</section>
<section id="references" class="level2" data-number="6.7">
<h2 data-number="6.7" class="anchored" data-anchor-id="references"><span class="header-section-number">6.7</span> References</h2>


<!-- -->

<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-benjamini1995" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Benjamini Y, Hochberg Y. Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing. Journal of the Royal Statistical Society Series B: Statistical Methodology [Internet]. 1995 Jan 1;57(1):289–300. Available from: <a href="http://dx.doi.org/10.1111/j.2517-6161.1995.tb02031.x">http://dx.doi.org/10.1111/j.2517-6161.1995.tb02031.x</a></div>
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session_5.html" class="pagination-link" aria-label="Session 5: RNA-seq part 1">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Session 5: RNA-seq part 1</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./further_reading.html" class="pagination-link" aria-label="Further reading">
        <span class="nav-page-text">Further reading</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb38" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this session:"</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session 6: RNA-seq part 2</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>The aim of this session is to complete the analysis of the RNA-seq data using the <span class="in">`limma`</span> and <span class="in">`edgeR`</span> packages. With the results of the analysis, we want to be able to produce publication quality figures and write out the results to a file for sharing or publication.</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>fit a linear model to the data</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>construct design and contrast matrices for the linear model</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>model the mean-variance relationship of the data using the <span class="in">`voom()`</span> function</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>perform statistical testing to identify differentially expressed genes</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>visualise the results using an MA plot</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>write out the results to a file</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb38-28"><a href="#cb38-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb38-29"><a href="#cb38-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb38-30"><a href="#cb38-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb38-31"><a href="#cb38-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-32"><a href="#cb38-32" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">parse_factor</span>(</span>
<span id="cb38-33"><a href="#cb38-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Basal"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>, <span class="st">"Basal"</span>, <span class="st">"ML"</span>, <span class="st">"LP"</span>),</span>
<span id="cb38-34"><a href="#cb38-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>)</span>
<span id="cb38-35"><a href="#cb38-35" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-36"><a href="#cb38-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-37"><a href="#cb38-37" aria-hidden="true" tabindex="-1"></a>samplenames <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb38-38"><a href="#cb38-38" aria-hidden="true" tabindex="-1"></a>  <span class="st">"10_6_5_11"</span>, <span class="st">"9_6_5_11"</span>, <span class="st">"purep53"</span>, <span class="st">"JMS8-2"</span>, <span class="st">"JMS8-3"</span>,</span>
<span id="cb38-39"><a href="#cb38-39" aria-hidden="true" tabindex="-1"></a>  <span class="st">"JMS8-4"</span>, <span class="st">"JMS8-5"</span>, <span class="st">"JMS9-P7c"</span>, <span class="st">"JMS9-P8c"</span></span>
<span id="cb38-40"><a href="#cb38-40" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-41"><a href="#cb38-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-42"><a href="#cb38-42" aria-hidden="true" tabindex="-1"></a><span class="co"># create DGEList object</span></span>
<span id="cb38-43"><a href="#cb38-43" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">readDGE</span>(</span>
<span id="cb38-44"><a href="#cb38-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir</span>(<span class="at">path =</span> <span class="st">"data/counts"</span>, <span class="at">pattern =</span> <span class="st">"GSM*"</span>),</span>
<span id="cb38-45"><a href="#cb38-45" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"data/counts"</span>,</span>
<span id="cb38-46"><a href="#cb38-46" aria-hidden="true" tabindex="-1"></a>  <span class="at">columns =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb38-47"><a href="#cb38-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> group,</span>
<span id="cb38-48"><a href="#cb38-48" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> samplenames</span>
<span id="cb38-49"><a href="#cb38-49" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-50"><a href="#cb38-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-51"><a href="#cb38-51" aria-hidden="true" tabindex="-1"></a><span class="co"># add gene annotation information</span></span>
<span id="cb38-52"><a href="#cb38-52" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>genes <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/Ses3_geneAnnot.tsv"</span>)</span>
<span id="cb38-53"><a href="#cb38-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-54"><a href="#cb38-54" aria-hidden="true" tabindex="-1"></a><span class="co"># filter lowly expressed genes</span></span>
<span id="cb38-55"><a href="#cb38-55" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge)</span>
<span id="cb38-56"><a href="#cb38-56" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep, ]</span>
<span id="cb38-57"><a href="#cb38-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-58"><a href="#cb38-58" aria-hidden="true" tabindex="-1"></a><span class="co"># normalise the data for library size and RNA composition</span></span>
<span id="cb38-59"><a href="#cb38-59" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span>
<span id="cb38-60"><a href="#cb38-60" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-61"><a href="#cb38-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-62"><a href="#cb38-62" aria-hidden="true" tabindex="-1"></a><span class="fu">## Linear modelling</span></span>
<span id="cb38-63"><a href="#cb38-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-64"><a href="#cb38-64" aria-hidden="true" tabindex="-1"></a>To identify differentially expressed genes from our normalised gene expression data, we will use the <span class="in">`limma`</span> package to fit linear models to the genes. A linear model is a broad class of statistical models that fit the value of an response (or dependent) variable as a linear function of one or more explanatory (or independent) variables (also called covariates).</span>
<span id="cb38-65"><a href="#cb38-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-66"><a href="#cb38-66" aria-hidden="true" tabindex="-1"></a>The general form of a linear model looks like this:</span>
<span id="cb38-67"><a href="#cb38-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-68"><a href="#cb38-68" aria-hidden="true" tabindex="-1"></a>$Y = \beta_{0} + \beta_{1}X_{1} + \beta_{2}X_{2}... + \epsilon$</span>
<span id="cb38-69"><a href="#cb38-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-70"><a href="#cb38-70" aria-hidden="true" tabindex="-1"></a>This equation is saying that a response variable of interest $Y$ is equal to a constant ($\beta_{0}$) plus the sum of the covariates ($X_{i}$) each multiplied by a constant coefficient ($\beta_{i}$) and some error term ($\epsilon$). The error term is the difference between the observed value and the predicted value of the model and is assumed to have a normal distribution with mean 0 and some variance.</span>
<span id="cb38-71"><a href="#cb38-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-72"><a href="#cb38-72" aria-hidden="true" tabindex="-1"></a>Our experiment is quite simple, since there is only a single covariate, the cell type. The true benefit of using linear models is in its ability to accommodate more complex designs including multiple covariates.</span>
<span id="cb38-73"><a href="#cb38-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-74"><a href="#cb38-74" aria-hidden="true" tabindex="-1"></a>To fit the linear models in the limma-voom framework we need two objects in addition to our data:</span>
<span id="cb38-75"><a href="#cb38-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-76"><a href="#cb38-76" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A design matrix, representing the covariates.</span>
<span id="cb38-77"><a href="#cb38-77" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A contrast matrix, representing the specific comparison we wish to make.</span>
<span id="cb38-78"><a href="#cb38-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-79"><a href="#cb38-79" aria-hidden="true" tabindex="-1"></a><span class="fu">### Design matrix</span></span>
<span id="cb38-80"><a href="#cb38-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-81"><a href="#cb38-81" aria-hidden="true" tabindex="-1"></a>The first step to fitting a linear model is to specify a design matrix. The design matrix specifies the values of the covariates for each sample, and is represented as a matrix due to the mathematical convenience.</span>
<span id="cb38-82"><a href="#cb38-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-83"><a href="#cb38-83" aria-hidden="true" tabindex="-1"></a>To generate a design matrix. We use the function <span class="in">`model.matrix()`</span>, with the expression <span class="in">`~0 + group`</span>. This returns a matrix representing the design where there is no intercept term and group is the only covariate. If we omit the <span class="in">`0`</span> then there would be an intercept in the model, and if we included more covariates then more columns would be generated.</span>
<span id="cb38-84"><a href="#cb38-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-87"><a href="#cb38-87" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-88"><a href="#cb38-88" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span><span class="dv">0</span> <span class="sc">+</span> group, <span class="at">data =</span> dge<span class="sc">$</span>samples)</span>
<span id="cb38-89"><a href="#cb38-89" aria-hidden="true" tabindex="-1"></a>design</span>
<span id="cb38-90"><a href="#cb38-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-91"><a href="#cb38-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-92"><a href="#cb38-92" aria-hidden="true" tabindex="-1"></a>There are 9 rows, one for each sample. Along the columns are the names of the groups. The values in the cells denote membership of the particular sample for a particular group, as our groups in this case are mutually exclusive, each row contains only a single 1 to denote membership in a single group.</span>
<span id="cb38-93"><a href="#cb38-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-94"><a href="#cb38-94" aria-hidden="true" tabindex="-1"></a><span class="fu">### Contrasts</span></span>
<span id="cb38-95"><a href="#cb38-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-96"><a href="#cb38-96" aria-hidden="true" tabindex="-1"></a>Contrast matrices are companions to design matrices. They are used to specify the comparison of interest. In our case, we have three experimental groups: <span class="in">`Basal`</span>, <span class="in">`LP`</span> and <span class="in">`ML`</span>. So if we are to perform differential expression analysis, we are most likely interested in the differences between only two of the groups at a time. Contrast matrices let use specify which comparison we'd like to make, and are also represented as a matrix just like the design.</span>
<span id="cb38-97"><a href="#cb38-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-98"><a href="#cb38-98" aria-hidden="true" tabindex="-1"></a>A contrast matrix can be made using the <span class="in">`makeContrasts()`</span> function. Within this function, we specify the name of each specific contrast and the formula for that contrast. For example, the <span class="in">`BasalvsLP`</span> contrasts compares the difference between the <span class="in">`Basal`</span> and <span class="in">`LP`</span> groups. Note that the name of the phenotype groups must be written exactly as they are in the column names of our design matrix (see above).</span>
<span id="cb38-99"><a href="#cb38-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-100"><a href="#cb38-100" aria-hidden="true" tabindex="-1"></a>In addition to the individual contrasts, the function must know about the design of the model. This is passed through the <span class="in">`levels`</span> argument, which either accepts a matrix with the column names corresponding to levels of your experimental groups, or the levels themselves as a character vector. Here we first simplify the column names of our design matrix to make it easier to read. Then we create the contrast matrix using the <span class="in">`makeContrasts()`</span> function.</span>
<span id="cb38-101"><a href="#cb38-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-104"><a href="#cb38-104" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-105"><a href="#cb38-105" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>)</span>
<span id="cb38-106"><a href="#cb38-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-107"><a href="#cb38-107" aria-hidden="true" tabindex="-1"></a>contr_matrix <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(</span>
<span id="cb38-108"><a href="#cb38-108" aria-hidden="true" tabindex="-1"></a>  <span class="at">BasalvsLP =</span> <span class="st">"Basal - LP"</span>,</span>
<span id="cb38-109"><a href="#cb38-109" aria-hidden="true" tabindex="-1"></a>  <span class="at">BasalvsML =</span> <span class="st">"Basal - ML"</span>,</span>
<span id="cb38-110"><a href="#cb38-110" aria-hidden="true" tabindex="-1"></a>  <span class="at">LPvsML =</span> <span class="st">"LP - ML"</span>,</span>
<span id="cb38-111"><a href="#cb38-111" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> design <span class="co"># alternatively 'levels = colnames(design)'</span></span>
<span id="cb38-112"><a href="#cb38-112" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-113"><a href="#cb38-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-114"><a href="#cb38-114" aria-hidden="true" tabindex="-1"></a>contr_matrix</span>
<span id="cb38-115"><a href="#cb38-115" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-116"><a href="#cb38-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-117"><a href="#cb38-117" aria-hidden="true" tabindex="-1"></a>There are two things to note about a design matrix. First, the sum of the numbers in each row is equal to 0. Second, the way that you set up the equation in the matrix will determine the interpretation of the log-fold-change calculated later, as well as the meaning of up-regulated and down-regulated genes. For example, the contrast <span class="in">`Basal - LP`</span> will give a positive log-fold-change for genes that are up-regulated in the <span class="in">`Basal`</span> group compared to the <span class="in">`LP`</span> group. If we had set up the contrast as <span class="in">`LP - Basal`</span>, then the opposite would be true, and we would get a negative log-fold-change for genes that are up-regulated in the <span class="in">`Basal`</span> group compared to the <span class="in">`LP`</span> group.</span>
<span id="cb38-118"><a href="#cb38-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-119"><a href="#cb38-119" aria-hidden="true" tabindex="-1"></a><span class="fu">### Variance modelling with voom</span></span>
<span id="cb38-120"><a href="#cb38-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-121"><a href="#cb38-121" aria-hidden="true" tabindex="-1"></a>We are now ready to fit our linear models. Limma fits linear models to the data with the assumption that the underlying data is normally distributed. Count data is generally not normally distributed, but log transforming count data gives it a roughly normal distribution sufficient for linear models to work well. To do this, limma transforms the raw count data to log-cpm using library sizes and the normalisation factors we calculated previously.</span>
<span id="cb38-122"><a href="#cb38-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-123"><a href="#cb38-123" aria-hidden="true" tabindex="-1"></a>In addition to the normalisation steps, the limma-voom pipeline uses the <span class="in">`voom()`</span> function to generate weights for the individual genes based on a modelled mean-variance relationship. This modelling allows use to get more information out of small sample sizes as the weights prevent our model from being more heavily influenced by more variable data points. The weights will then allow the linear model to be less influenced by genes with high variability, which is important for small sample sizes. The voom function also generates a mean-variance plot, which is useful for visualising the data.</span>
<span id="cb38-124"><a href="#cb38-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-125"><a href="#cb38-125" aria-hidden="true" tabindex="-1"></a>The <span class="in">`voom()`</span> function takes as arguments, our <span class="in">`DGEList`</span> object and our design matrix. It also optionally outputs a plot of the mean-variance relationship of our data, called the 'voom-plot'.</span>
<span id="cb38-126"><a href="#cb38-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-129"><a href="#cb38-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-130"><a href="#cb38-130" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">voom</span>(dge, design, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-131"><a href="#cb38-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-132"><a href="#cb38-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-133"><a href="#cb38-133" aria-hidden="true" tabindex="-1"></a>The output of <span class="in">`voom()`</span> (our variable <span class="in">`v`</span>) is an <span class="in">`EList`</span> object which contains the following elements:</span>
<span id="cb38-134"><a href="#cb38-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-135"><a href="#cb38-135" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`genes`</span> - a data frame of gene annotation data.</span>
<span id="cb38-136"><a href="#cb38-136" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`targets`</span> - data frame of sample data.</span>
<span id="cb38-137"><a href="#cb38-137" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`E`</span> - numeric matrix of normalised log-cpm values.</span>
<span id="cb38-138"><a href="#cb38-138" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`weights`</span> - numeric matrix of precision weights.</span>
<span id="cb38-139"><a href="#cb38-139" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`design`</span> - the design matrix.</span>
<span id="cb38-140"><a href="#cb38-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-141"><a href="#cb38-141" aria-hidden="true" tabindex="-1"></a>The MA plot shows the mean-variance relationship of the data. The x-axis is the average log2 expression of each gene, and the y-axis is the log2 standard deviation of each gene. The idea is that the variance of a gene can be estimated as a function of its main expression, and that this estimate is more accurate than the direct calculation of the variance.</span>
<span id="cb38-142"><a href="#cb38-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-143"><a href="#cb38-143" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fitting the linear model</span></span>
<span id="cb38-144"><a href="#cb38-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-145"><a href="#cb38-145" aria-hidden="true" tabindex="-1"></a>We are now ready to fit our linear model with <span class="in">`lmFit()`</span>, which calculates coefficients we defined in our design matrix <span class="in">`design`</span>. The resulting object, <span class="in">`vfit`</span> is a <span class="in">`MArrayLM`</span> object. It contains a information about our genes (the same data frame as <span class="in">`genes`</span> from our <span class="in">`EList`</span> object <span class="in">`v`</span> above), the design matrix and a number of statistical outputs. Of most interest to us is the coefficients, stored in an element called <span class="in">`coefficients`</span>. The first rows of this matrix is shown below. Each gene is row and is labelled using the EntrezID. Each column gives coefficients for each of our phenotype groups. These coefficients are weighted averages of the log-cpm of each gene in each group.</span>
<span id="cb38-146"><a href="#cb38-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-149"><a href="#cb38-149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-150"><a href="#cb38-150" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(v)</span>
<span id="cb38-151"><a href="#cb38-151" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(vfit<span class="sc">$</span>coefficients)</span>
<span id="cb38-152"><a href="#cb38-152" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-153"><a href="#cb38-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-154"><a href="#cb38-154" aria-hidden="true" tabindex="-1"></a>We can then use <span class="in">`contrasts.fit()`</span> to calculate coefficients for each contrast (or 'comparison') we specified in our <span class="in">`contr_matrix`</span>. The output is also an object of the class <span class="in">`MArrayLM`</span> (also known as an <span class="in">`MArrayLM`</span> object). When we inspect the <span class="in">`coefficients`</span> element now, we can see that each column is a contrast that we specified in our contrast matrix.</span>
<span id="cb38-155"><a href="#cb38-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-158"><a href="#cb38-158" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-159"><a href="#cb38-159" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(vfit, <span class="at">contrasts =</span> contr_matrix)</span>
<span id="cb38-160"><a href="#cb38-160" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(vfit<span class="sc">$</span>coefficients)</span>
<span id="cb38-161"><a href="#cb38-161" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-162"><a href="#cb38-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-163"><a href="#cb38-163" aria-hidden="true" tabindex="-1"></a>With these values we essentially want to know whether or not the difference in values is significantly different from 0. If the difference is not significantly different from 0, then we have failed to establish a difference in the expression of a particularly gene between two groups. If the difference is significantly greater than 0, then we have up-regulation in the first group of the contrast, and if the difference is significantly less than 0, then we have down-regulation in the first group of the contrast.</span>
<span id="cb38-164"><a href="#cb38-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-165"><a href="#cb38-165" aria-hidden="true" tabindex="-1"></a><span class="fu">## Statistical testing</span></span>
<span id="cb38-166"><a href="#cb38-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-167"><a href="#cb38-167" aria-hidden="true" tabindex="-1"></a>To actually test if the values are significantly different from 0, we use the <span class="in">`eBayes()`</span> function to compute moderated t-statistics, moderated F-statistics and log-odds of differential expression for each gene, given a fitted linear model. 'Moderated' refers to empirical Bayes moderation, which borrows information across genes to obtain more accurate measures of variability for each gene. This also increases our power to detect differentially expressed genes.</span>
<span id="cb38-168"><a href="#cb38-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-171"><a href="#cb38-171" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-172"><a href="#cb38-172" aria-hidden="true" tabindex="-1"></a>efit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(vfit)</span>
<span id="cb38-173"><a href="#cb38-173" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-174"><a href="#cb38-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-175"><a href="#cb38-175" aria-hidden="true" tabindex="-1"></a>We can now look at the number of differentially expressed genes using the <span class="in">`decideTests()`</span> function. The output of this function is a matrix where each column is a contrast (comparison of interest) and each row is a gene. The numbers 1, -1 and 0 mean up-regulated, down-regulated or not significantly differentially expressed, respectively.</span>
<span id="cb38-176"><a href="#cb38-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-177"><a href="#cb38-177" aria-hidden="true" tabindex="-1"></a>Note that <span class="in">`decideTests()`</span> also accounts for multiple testing. The default method is Benjamini and Hochberg <span class="co">[</span><span class="ot">@benjamini1995</span><span class="co">]</span> but several others are also available.</span>
<span id="cb38-178"><a href="#cb38-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-181"><a href="#cb38-181" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-182"><a href="#cb38-182" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">decideTests</span>(efit)</span>
<span id="cb38-183"><a href="#cb38-183" aria-hidden="true" tabindex="-1"></a>dt</span>
<span id="cb38-184"><a href="#cb38-184" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-185"><a href="#cb38-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-186"><a href="#cb38-186" aria-hidden="true" tabindex="-1"></a>To obtain the total number of differentially expressed genes for each comparison, we can add the function <span class="in">`summary()`</span>:</span>
<span id="cb38-187"><a href="#cb38-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-190"><a href="#cb38-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-191"><a href="#cb38-191" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(dt)</span>
<span id="cb38-192"><a href="#cb38-192" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-193"><a href="#cb38-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-194"><a href="#cb38-194" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`topTable()`</span> can be used to obtain more information on the differentially expressed genes for each contrast. <span class="in">`topTable()`</span> takes as arguments the <span class="in">`MArrayLM`</span> object output by <span class="in">`eBayes()`</span> (<span class="in">`efit`</span>), the contrast name of interest and the number of top differentially expressed genes to output. Note that the contrast name must be given in quotes and must be exactly as written in the contrast matrix <span class="in">`contr_matrix`</span>.</span>
<span id="cb38-195"><a href="#cb38-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-196"><a href="#cb38-196" aria-hidden="true" tabindex="-1"></a>It outputs a data frame with the following information:</span>
<span id="cb38-197"><a href="#cb38-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-198"><a href="#cb38-198" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Gene details** - gene information, from the <span class="in">`gene`</span> element of the <span class="in">`MArrayLM`</span> object (<span class="in">`efit`</span>).</span>
<span id="cb38-199"><a href="#cb38-199" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`logFC`</span> - the log2 fold change of the contrast.</span>
<span id="cb38-200"><a href="#cb38-200" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`AveExpr`</span> - the average log2 expression of that gene.</span>
<span id="cb38-201"><a href="#cb38-201" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`t`</span> - moderated t-statistic.</span>
<span id="cb38-202"><a href="#cb38-202" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`P.Value`</span> - p value.</span>
<span id="cb38-203"><a href="#cb38-203" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`adj.P.Val`</span> - adjusted p value.</span>
<span id="cb38-204"><a href="#cb38-204" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`B`</span> - log-odds that the gene is differentially expressed.</span>
<span id="cb38-205"><a href="#cb38-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-208"><a href="#cb38-208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-209"><a href="#cb38-209" aria-hidden="true" tabindex="-1"></a>top <span class="ot">&lt;-</span> <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsLP"</span>, <span class="at">n =</span> <span class="cn">Inf</span>)</span>
<span id="cb38-210"><a href="#cb38-210" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(top)</span>
<span id="cb38-211"><a href="#cb38-211" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-212"><a href="#cb38-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-213"><a href="#cb38-213" aria-hidden="true" tabindex="-1"></a><span class="fu">## MA Plot</span></span>
<span id="cb38-214"><a href="#cb38-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-215"><a href="#cb38-215" aria-hidden="true" tabindex="-1"></a>The MA plot is a plot of log-fold-change (M-values) against log-expression averages (A-values), this is a common plot in RNA sequencing analysis to visualise the result of differential expression tests. It can be created using the <span class="in">`plotMA()`</span> from the <span class="in">`limma`</span> package. Creating this plot requires 3 pieces of information:</span>
<span id="cb38-216"><a href="#cb38-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-217"><a href="#cb38-217" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`object = efit`</span>: The the fitted object containing the log-fold-change and log-expression averages</span>
<span id="cb38-218"><a href="#cb38-218" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`coef = 1`</span>: The column number of the contrast to plot since there are 3 different contrasts fitted within the object.</span>
<span id="cb38-219"><a href="#cb38-219" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`status = dt[, 1]`</span>: A vector of numerics denoting whether a gene is up-regulated or down-regulated.</span>
<span id="cb38-220"><a href="#cb38-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-223"><a href="#cb38-223" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-224"><a href="#cb38-224" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMA</span>(efit, <span class="at">coef =</span> <span class="dv">1</span>, <span class="at">status =</span> dt[, <span class="st">"BasalvsLP"</span>])</span>
<span id="cb38-225"><a href="#cb38-225" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-226"><a href="#cb38-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-227"><a href="#cb38-227" aria-hidden="true" tabindex="-1"></a>As before, we can try to re-create this plot using <span class="in">`ggplot2`</span> to give us more flexibility in plotting. The information we'll need is the log-fold-change, the average log-expression and the differential expression status of each gene. We can extract this information from the <span class="in">`efit`</span> and <span class="in">`top`</span> data frame we created above, as well as gene names for annotation purposes.</span>
<span id="cb38-228"><a href="#cb38-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-231"><a href="#cb38-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-232"><a href="#cb38-232" aria-hidden="true" tabindex="-1"></a>ma_plot_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb38-233"><a href="#cb38-233" aria-hidden="true" tabindex="-1"></a>  <span class="at">logFC =</span> efit<span class="sc">$</span>coefficients[, <span class="st">"BasalvsLP"</span>],</span>
<span id="cb38-234"><a href="#cb38-234" aria-hidden="true" tabindex="-1"></a>  <span class="at">AvgExpr =</span> efit<span class="sc">$</span>Amean,</span>
<span id="cb38-235"><a href="#cb38-235" aria-hidden="true" tabindex="-1"></a>  <span class="at">status =</span> <span class="fu">factor</span>(<span class="fu">as.numeric</span>(dt[, <span class="st">"BasalvsLP"</span>])),</span>
<span id="cb38-236"><a href="#cb38-236" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> efit<span class="sc">$</span>genes<span class="sc">$</span>SYMBOL,</span>
<span id="cb38-237"><a href="#cb38-237" aria-hidden="true" tabindex="-1"></a>  <span class="at">p_value =</span> efit<span class="sc">$</span>p.value[, <span class="st">"BasalvsLP"</span>]</span>
<span id="cb38-238"><a href="#cb38-238" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-239"><a href="#cb38-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-240"><a href="#cb38-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-241"><a href="#cb38-241" aria-hidden="true" tabindex="-1"></a>With our tibble prepared, we can use <span class="in">`ggplot2`</span> to create the plot as we have done before</span>
<span id="cb38-242"><a href="#cb38-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-245"><a href="#cb38-245" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-246"><a href="#cb38-246" aria-hidden="true" tabindex="-1"></a>ma_plot_data <span class="sc">%&gt;%</span></span>
<span id="cb38-247"><a href="#cb38-247" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> AvgExpr, <span class="at">y =</span> logFC, <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb38-248"><a href="#cb38-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb38-249"><a href="#cb38-249" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-250"><a href="#cb38-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-251"><a href="#cb38-251" aria-hidden="true" tabindex="-1"></a>While the coordinates of the points are the same, there is still quite a bit of work to do to make this plot look like the one produced by <span class="in">`plotMA()`</span>. First we have the fix up the colours of the points. We also have to make it such that the non-significant points are plotted below the other points. To do this we would need to make it so that the non-significant points are plotted first. We can do this by reordering the levels of the <span class="in">`status`</span> factor so that the non-significant points are plotted first. Labels can also be added using <span class="in">`ggrepel`</span> for the top 15 most DE genes.</span>
<span id="cb38-252"><a href="#cb38-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-255"><a href="#cb38-255" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-256"><a href="#cb38-256" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb38-257"><a href="#cb38-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-258"><a href="#cb38-258" aria-hidden="true" tabindex="-1"></a>ma_plot_data_fct <span class="ot">&lt;-</span> ma_plot_data <span class="sc">%&gt;%</span></span>
<span id="cb38-259"><a href="#cb38-259" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb38-260"><a href="#cb38-260" aria-hidden="true" tabindex="-1"></a>    <span class="at">status =</span> <span class="fu">fct_recode</span>(</span>
<span id="cb38-261"><a href="#cb38-261" aria-hidden="true" tabindex="-1"></a>      status,</span>
<span id="cb38-262"><a href="#cb38-262" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Not DE"</span> <span class="ot">=</span> <span class="st">"0"</span>,</span>
<span id="cb38-263"><a href="#cb38-263" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Down"</span> <span class="ot">=</span> <span class="st">"-1"</span>,</span>
<span id="cb38-264"><a href="#cb38-264" aria-hidden="true" tabindex="-1"></a>      <span class="st">"Up"</span> <span class="ot">=</span> <span class="st">"1"</span></span>
<span id="cb38-265"><a href="#cb38-265" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb38-266"><a href="#cb38-266" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb38-267"><a href="#cb38-267" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">status =</span> <span class="fu">fct_relevel</span>(status, <span class="st">"Not DE"</span>, <span class="st">"Down"</span>, <span class="st">"Up"</span>))</span>
<span id="cb38-268"><a href="#cb38-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-269"><a href="#cb38-269" aria-hidden="true" tabindex="-1"></a>top_de_genes <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb38-270"><a href="#cb38-270" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p_value) <span class="sc">%&gt;%</span></span>
<span id="cb38-271"><a href="#cb38-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">15</span>)</span>
<span id="cb38-272"><a href="#cb38-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-273"><a href="#cb38-273" aria-hidden="true" tabindex="-1"></a>scale_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb38-274"><a href="#cb38-274" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Not DE"</span> <span class="ot">=</span> <span class="st">"black"</span>,</span>
<span id="cb38-275"><a href="#cb38-275" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Down"</span> <span class="ot">=</span> <span class="st">"blue"</span>,</span>
<span id="cb38-276"><a href="#cb38-276" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Up"</span> <span class="ot">=</span> <span class="st">"red"</span></span>
<span id="cb38-277"><a href="#cb38-277" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-278"><a href="#cb38-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-279"><a href="#cb38-279" aria-hidden="true" tabindex="-1"></a>ma_plot <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb38-280"><a href="#cb38-280" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> AvgExpr, <span class="at">y =</span> logFC, <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb38-281"><a href="#cb38-281" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">==</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb38-282"><a href="#cb38-282" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">!=</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb38-283"><a href="#cb38-283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb38-284"><a href="#cb38-284" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> top_de_genes,</span>
<span id="cb38-285"><a href="#cb38-285" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene, <span class="at">col =</span> <span class="cn">NULL</span>),</span>
<span id="cb38-286"><a href="#cb38-286" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.overlaps =</span> <span class="dv">20</span>,</span>
<span id="cb38-287"><a href="#cb38-287" aria-hidden="true" tabindex="-1"></a>    <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb38-288"><a href="#cb38-288" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb38-289"><a href="#cb38-289" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> scale_cols)</span>
<span id="cb38-290"><a href="#cb38-290" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-291"><a href="#cb38-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-292"><a href="#cb38-292" aria-hidden="true" tabindex="-1"></a><span class="fu">## Volcano Plot</span></span>
<span id="cb38-293"><a href="#cb38-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-294"><a href="#cb38-294" aria-hidden="true" tabindex="-1"></a>Since we have the tibble of the data prepared, we can also choose to create an alternative plot called a volcano plot. Which is also very popular in RNA-seq analysis. The volcano plot is a scatter plot of the negative log10 p-value against the log2 fold-change. This plot is useful for visualising the significance of the differentially expressed genes.</span>
<span id="cb38-295"><a href="#cb38-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-298"><a href="#cb38-298" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-299"><a href="#cb38-299" aria-hidden="true" tabindex="-1"></a>volcano_plot <span class="ot">&lt;-</span> ma_plot_data_fct <span class="sc">%&gt;%</span></span>
<span id="cb38-300"><a href="#cb38-300" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logFC, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p_value), <span class="at">col =</span> status)) <span class="sc">+</span></span>
<span id="cb38-301"><a href="#cb38-301" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">==</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb38-302"><a href="#cb38-302" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> ma_plot_data_fct <span class="sc">%&gt;%</span> <span class="fu">filter</span>(status <span class="sc">!=</span> <span class="st">"Not DE"</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb38-303"><a href="#cb38-303" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb38-304"><a href="#cb38-304" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> top_de_genes,</span>
<span id="cb38-305"><a href="#cb38-305" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene, <span class="at">col =</span> <span class="cn">NULL</span>),</span>
<span id="cb38-306"><a href="#cb38-306" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.overlaps =</span> <span class="dv">20</span>,</span>
<span id="cb38-307"><a href="#cb38-307" aria-hidden="true" tabindex="-1"></a>    <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb38-308"><a href="#cb38-308" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb38-309"><a href="#cb38-309" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"red"</span>))</span>
<span id="cb38-310"><a href="#cb38-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-311"><a href="#cb38-311" aria-hidden="true" tabindex="-1"></a>volcano_plot</span>
<span id="cb38-312"><a href="#cb38-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-313"><a href="#cb38-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-314"><a href="#cb38-314" aria-hidden="true" tabindex="-1"></a><span class="fu">## Heatmap</span></span>
<span id="cb38-315"><a href="#cb38-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-316"><a href="#cb38-316" aria-hidden="true" tabindex="-1"></a>Finally we can create a heatmap of the differentially expressed genes. The heatmap is a useful way to visualise the expression of the differentially expressed genes across the samples. We can use the <span class="in">`pheatmap`</span> package to create the heatmap.</span>
<span id="cb38-317"><a href="#cb38-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-318"><a href="#cb38-318" aria-hidden="true" tabindex="-1"></a>First we need to install the package if you haven't already</span>
<span id="cb38-319"><a href="#cb38-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-320"><a href="#cb38-320" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval = FALSE}</span></span>
<span id="cb38-321"><a href="#cb38-321" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb38-322"><a href="#cb38-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-323"><a href="#cb38-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-324"><a href="#cb38-324" aria-hidden="true" tabindex="-1"></a>The data we will use for the heatmap is going to be the log2 cpm expression values of the top differentially expressed genes. We can use the <span class="in">`topTable()`</span> function to get the top differentially expressed genes for each contrast. We can then use the <span class="in">`cpm()`</span> function to get the expression values of these genes. We will take the top 50 genes from each contrast and combine them into a single data frame. We will then use extract the expression values for these genes from the full expression matrix. For more interpretable labels, we will use the gene symbols as the row names of the expression matrix.</span>
<span id="cb38-325"><a href="#cb38-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-326"><a href="#cb38-326" aria-hidden="true" tabindex="-1"></a>Note that by default the ENTREZID column is encoded as a numeric value, so to use it to subset our expression matrix we need to convert it to a character vector in order to select the matching IDs rather than trying to select the numerical rows of the data.</span>
<span id="cb38-327"><a href="#cb38-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-330"><a href="#cb38-330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-331"><a href="#cb38-331" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb38-332"><a href="#cb38-332" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsLP"</span>, <span class="at">n =</span> <span class="dv">50</span>),</span>
<span id="cb38-333"><a href="#cb38-333" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"BasalvsML"</span>, <span class="at">n =</span> <span class="dv">50</span>),</span>
<span id="cb38-334"><a href="#cb38-334" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTable</span>(efit, <span class="at">coef =</span> <span class="st">"LPvsML"</span>, <span class="at">n =</span> <span class="dv">50</span>)</span>
<span id="cb38-335"><a href="#cb38-335" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb38-336"><a href="#cb38-336" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(ENTREZID, SYMBOL, TXCHROM) <span class="sc">%&gt;%</span></span>
<span id="cb38-337"><a href="#cb38-337" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb38-338"><a href="#cb38-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-339"><a href="#cb38-339" aria-hidden="true" tabindex="-1"></a>top_gene_expr <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)[<span class="fu">as.character</span>(top_genes<span class="sc">$</span>ENTREZID), ]</span>
<span id="cb38-340"><a href="#cb38-340" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(top_gene_expr) <span class="ot">&lt;-</span> top_genes<span class="sc">$</span>SYMBOL</span>
<span id="cb38-341"><a href="#cb38-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-342"><a href="#cb38-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-343"><a href="#cb38-343" aria-hidden="true" tabindex="-1"></a>To create the heatmap we will use the <span class="in">`pheatmap()`</span> function with the expression matrix. We will add in arguments to provide our own colour scale, to scale the data along the rows with a z-score, adding in sample annotation and clustering the data by rows and columns. We will also add in the option to show the row names and column names, and set the font size of the row names. You can experiment with removing each of these options to see how they affect the heatmap, as well as investigate additional options in the <span class="in">`pheatmap()`</span> documentation.</span>
<span id="cb38-344"><a href="#cb38-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-347"><a href="#cb38-347" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb38-348"><a href="#cb38-348" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb38-349"><a href="#cb38-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-350"><a href="#cb38-350" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples <span class="sc">%&gt;%</span></span>
<span id="cb38-351"><a href="#cb38-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(group)</span>
<span id="cb38-352"><a href="#cb38-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-353"><a href="#cb38-353" aria-hidden="true" tabindex="-1"></a>heatmap <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(</span>
<span id="cb38-354"><a href="#cb38-354" aria-hidden="true" tabindex="-1"></a>  top_gene_expr,</span>
<span id="cb38-355"><a href="#cb38-355" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>),</span>
<span id="cb38-356"><a href="#cb38-356" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb38-357"><a href="#cb38-357" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotation_col =</span> sample_anno,</span>
<span id="cb38-358"><a href="#cb38-358" aria-hidden="true" tabindex="-1"></a>  <span class="at">border_color =</span> <span class="cn">NA</span>,</span>
<span id="cb38-359"><a href="#cb38-359" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-360"><a href="#cb38-360" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-361"><a href="#cb38-361" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-362"><a href="#cb38-362" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_colnames =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-363"><a href="#cb38-363" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontsize_row =</span> <span class="dv">6</span></span>
<span id="cb38-364"><a href="#cb38-364" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-365"><a href="#cb38-365" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-366"><a href="#cb38-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-367"><a href="#cb38-367" aria-hidden="true" tabindex="-1"></a><span class="fu">## Saving all the results</span></span>
<span id="cb38-368"><a href="#cb38-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-369"><a href="#cb38-369" aria-hidden="true" tabindex="-1"></a>Now that we have produced a series of plots and the top genes table, we can save these files to a directory. These figures may form the basis for publication figures, and the results table can partially presented as a result, or the full table can be included as a supplementary file.</span>
<span id="cb38-370"><a href="#cb38-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-371"><a href="#cb38-371" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, warning=FALSE}</span></span>
<span id="cb38-372"><a href="#cb38-372" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"output"</span>)</span>
<span id="cb38-373"><a href="#cb38-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-374"><a href="#cb38-374" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(top, <span class="st">"output/top_genes_BasalvsLP.csv"</span>)</span>
<span id="cb38-375"><a href="#cb38-375" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/ma_plot_BasalvsLP.png"</span>, ma_plot)</span>
<span id="cb38-376"><a href="#cb38-376" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/volcano_plot_BasalvsLP.png"</span>, volcano_plot)</span>
<span id="cb38-377"><a href="#cb38-377" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"output/heatmap_BasalvsLP.png"</span>, heatmap<span class="sc">$</span>gtable)</span>
<span id="cb38-378"><a href="#cb38-378" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb38-379"><a href="#cb38-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-380"><a href="#cb38-380" aria-hidden="true" tabindex="-1"></a>With that we have come to the conclusion of the RNA-seq analysis and this course. You have now learned how to use various <span class="in">`tidyverse`</span> packages to operate on data and create publication quality figures. You have also learned how to use the <span class="in">`limma`</span> and <span class="in">`edgeR`</span> packages to perform differential expression analysis on RNA-seq data. You should now be able to apply these skills to perform analysis on your own data.</span>
<span id="cb38-381"><a href="#cb38-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-382"><a href="#cb38-382" aria-hidden="true" tabindex="-1"></a><span class="fu">## References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>