<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.24">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>4&nbsp; Session 4: Putting it all together – WEHI R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./summary.html" rel="next">
<link href="./session_3.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2c5f73ba8ffe94769966358c3b6948a5.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-c081ddf1e2ccd80e94c6a2eac056c367.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-FN78SCFVJJ"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-FN78SCFVJJ', { 'anonymize_ip': true});
</script>


</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session_4.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">WEHI R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Session 3: Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_4.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Course summary &amp; concluding remarks</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this session:</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">4.1</span> Introduction</a>
  <ul class="collapse">
  <li><a href="#sec-research-questions" id="toc-sec-research-questions" class="nav-link" data-scroll-target="#sec-research-questions"><span class="header-section-number">4.1.1</span> Today’s research questions</a></li>
  </ul></li>
  <li><a href="#sec-tidy-data" id="toc-sec-tidy-data" class="nav-link" data-scroll-target="#sec-tidy-data"><span class="header-section-number">4.2</span> Tidy data</a></li>
  <li><a href="#sec-reshaping" id="toc-sec-reshaping" class="nav-link" data-scroll-target="#sec-reshaping"><span class="header-section-number">4.3</span> Reshaping and combining data (to make it tidy)</a>
  <ul class="collapse">
  <li><a href="#sec-pivot" id="toc-sec-pivot" class="nav-link" data-scroll-target="#sec-pivot"><span class="header-section-number">4.3.1</span> Reshaping data with pivot functions</a></li>
  <li><a href="#sec-separate" id="toc-sec-separate" class="nav-link" data-scroll-target="#sec-separate"><span class="header-section-number">4.3.2</span> Separating data in a column</a></li>
  <li><a href="#sec-join" id="toc-sec-join" class="nav-link" data-scroll-target="#sec-join"><span class="header-section-number">4.3.3</span> Combining data with join functions</a></li>
  </ul></li>
  <li><a href="#the-process-of-data-analysis" id="toc-the-process-of-data-analysis" class="nav-link" data-scroll-target="#the-process-of-data-analysis"><span class="header-section-number">4.4</span> The process of data analysis</a></li>
  <li><a href="#question-1-the-relationship-between-dosage-and-weight-loss" id="toc-question-1-the-relationship-between-dosage-and-weight-loss" class="nav-link" data-scroll-target="#question-1-the-relationship-between-dosage-and-weight-loss"><span class="header-section-number">4.5</span> Question 1: The relationship between dosage and weight loss</a>
  <ul class="collapse">
  <li><a href="#the-effect-of-strain-and-sex" id="toc-the-effect-of-strain-and-sex" class="nav-link" data-scroll-target="#the-effect-of-strain-and-sex"><span class="header-section-number">4.5.1</span> The effect of strain and sex</a></li>
  </ul></li>
  <li><a href="#question-2-the-effect-of-mousezempic-on-gene-expression" id="toc-question-2-the-effect-of-mousezempic-on-gene-expression" class="nav-link" data-scroll-target="#question-2-the-effect-of-mousezempic-on-gene-expression"><span class="header-section-number">4.6</span> Question 2: The effect of mousezempic on gene expression</a>
  <ul class="collapse">
  <li><a href="#do-the-treatment-and-control-groups-differ" id="toc-do-the-treatment-and-control-groups-differ" class="nav-link" data-scroll-target="#do-the-treatment-and-control-groups-differ"><span class="header-section-number">4.6.1</span> Do the treatment and control groups differ?</a></li>
  <li><a href="#an-aside-string-manipulation" id="toc-an-aside-string-manipulation" class="nav-link" data-scroll-target="#an-aside-string-manipulation"><span class="header-section-number">4.6.2</span> An aside: string manipulation</a></li>
  <li><a href="#the-relationship-between-gene-expression-and-weight-loss" id="toc-the-relationship-between-gene-expression-and-weight-loss" class="nav-link" data-scroll-target="#the-relationship-between-gene-expression-and-weight-loss"><span class="header-section-number">4.6.3</span> The relationship between gene expression and weight loss</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">4.7</span> Summary</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-session04" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Session 4: Putting it all together</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this session we will combine all the skills we have learned so far to perform a complete analysis of a small dataset.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this session, learners should be able to:</p>
<ol type="1">
<li><p>Explain the concept and purpose of ‘tidy’ data</p></li>
<li><p>Understand how pivot, separate and join functions can be used to reshape and combine entire data frames</p></li>
<li><p>Describe the key steps in data analysis</p></li>
<li><p>Perform basic data analyses in R</p></li>
</ol>
</div>
</div>
<section id="introduction" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">4.1</span> Introduction</h2>
<p>For our analysis, we will use the <code>dosage</code> and <code>genes</code> data frames, which contain information about the mice treated with mousezempic and their gene expression levels, respectively.</p>
<p>The first thing we need to do is load the packages we’ll need for today:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse package contains most of what we need</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># let's also load patchwork to put plots together</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>And read in the two datasets:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in dosage data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>dosage <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 344 Columns: 9
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): mouse_strain, cage_number, replicate, sex
dbl (5): weight_lost_g, drug_dose_g, tail_length_mm, initial_weight_g, id_num

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in expression data</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 343 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (1): group
dbl (10): Th_rep1, Th_rep2, Th_rep3, Prlh_rep1, Prlh_rep2, Prlh_rep3, Hprt1_...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="How do we read in data again?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>How do we read in data again?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that we used <code>read_csv()</code> for the mouse dosage data and <code>read_tsv()</code> for the expression data. We chose these functions to match the file extensions of the two datasets, as described when we first learned about reading in data in session 1 (see <a href="session_1.html#sec-readinFunctions" class="quarto-xref"><span>Section 1.8.2</span></a>).</p>
</div>
</div>
<p>Before we go further, let’s take a moment to explore the <code>genes</code> dataset, as it’s new in this session.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>genes</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 343 × 11
   group        Th_rep1 Th_rep2 Th_rep3 Prlh_rep1 Prlh_rep2 Prlh_rep3 Hprt1_rep1
   &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1 grp_treatme…   3239.   5964.   4123.      225.      359.      314.       858.
 2 grp_treatme…   2665.   4867.   4240.      253.      460.      358.       709.
 3 grp_treatme…   1536.   2501.   2054.      367.      552.      448.       975.
 4 grp_treatme…   2732.   4800.   3981.      258.      411.      339.       725.
 5 grp_treatme…   2384.   3460.   3115.      347.      651.      491.       896.
 6 grp_treatme…   2972.   4880.   4028.      303.      456.      377.       865.
 7 grp_treatme…   1936.   3666.   2964.      302.      441.      355.       625.
 8 grp_treatme…   2801.   4977.   3795.      246.      424.      321.       856.
 9 grp_treatme…   3239.   5198.   3930.      385.      654.      508.       623.
10 grp_treatme…   2665.   3820.   2958.      230.      391.      301.       740.
# ℹ 333 more rows
# ℹ 3 more variables: Hprt1_rep2 &lt;dbl&gt;, Hprt1_rep3 &lt;dbl&gt;, id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This data frame contains expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking mousezempic, as well as one housekeeping gene (HPRT1), all measured in triplicate. Each mouse has a unique ID number (<code>id_num</code> column), which matches with those in the <code>dosage</code> dataset, but there are also additional control mice, who have not received mousezempic (as described in the <code>group</code> column).</p>
<section id="sec-research-questions" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="sec-research-questions"><span class="header-section-number">4.1.1</span> Today’s research questions</h3>
<p>In today’s session, we will use the <code>dosage</code> and <code>genes</code> datasets to explore the following questions:</p>
<ol type="1">
<li>Does the amount of weight lost by mice increase as dosage of mousezempic increases? Is this relationship affected by the mouse strain or sex?</li>
<li>How does mousezempic affect the expression of the TH and PRLH genes? Is the expression of either of these genes correlated with the amount of weight mice lose?</li>
</ol>
<p>Of course, there are countless questions we could ask of our data! In the interest of time, we will stick to these two key analyses but various others are explored in the practice questions at the end of this session’s notes.</p>
</section>
</section>
<section id="sec-tidy-data" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="sec-tidy-data"><span class="header-section-number">4.2</span> Tidy data</h2>
<p>In order to answer the research questions outlined above (or any others for that matter), we need to ensure that our data is <strong>“tidy”</strong>.</p>
<p>This is because the tidyverse packages we are using revolve around the concept of “<a href="https://r4ds.hadley.nz/data-tidy.html">tidy data</a>”, which is a specific way of organising data into a table so that it is easy to work with. Tidy data is roughly defined as tabular data that contains:</p>
<ol type="1">
<li>Individual variables in columns</li>
<li>Individual observations in rows</li>
<li>One data value in each cell</li>
</ol>
<p>In other words, each column in our datasets should represent a particular measurement (like expression of the TH gene or the sex of the mice), and each row should contain all of the measurements available for a particular mouse.</p>
<div class="callout callout-style-default callout-note callout-titled" title="*Why* is tidy data easy to work with?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span><em>Why</em> is tidy data easy to work with?
</div>
</div>
<div class="callout-body-container callout-body">
<p><a href="https://allisonhorst.com/other-r-fun">This cartoon from Allison Horst</a> sums it up well: <img src="images/S04F01_tidy_data.jpg" class="img-fluid" style="width:80.0%" alt="Tidy data can be analysed in a more routine/systematic manner"></p>
</div>
</div>
<p>Although tidy data is the easiest to work with, it’s often necessary to alter the format of your data for plotting or table displays. It’s a good idea to keep your core data in a tidy format and treating plot or table outputs as representations of that tidy data.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Data wrangling?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Data wrangling?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most data you encounter will not be tidy, so the first part of data analysis is usually called “data wrangling” or “data cleaning” and involves tidying up your data (by “wrangling” it into shape) so it is easier to use for downstream analysis. In this session, we will practice data wrangling by tidying up the <code>genes</code> data frame.</p>
</div>
</div>
</section>
<section id="sec-reshaping" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="sec-reshaping"><span class="header-section-number">4.3</span> Reshaping and combining data (to make it tidy)</h2>
<p>In session 2, we learned a range of functions that work on either the rows or columns of our data: <code>filter()</code>, <code>select()</code>, <code>mutate()</code> and <code>summarise()</code>. By cleverly combining these functions you can answer a lot of different questions about your data, but sometimes you may need to reshape or even combine different datasets to get the <em>right</em> rows and columns. In this section, we will cover how the <code>pivot</code>, <code>separate</code> and <code>join</code> families of functions can be used to help us get our data in the right (tidy!) format to answer the research questions outlined above.</p>
<section id="sec-pivot" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="sec-pivot"><span class="header-section-number">4.3.1</span> Reshaping data with pivot functions</h3>
<p>Pivoting is a way to change the shape of your data:</p>
<ul>
<li><p>Pivoting <strong>longer</strong> reshapes the data to transfer data stored in columns into rows, resulting in more rows and fewer columns.</p></li>
<li><p>Pivoting <strong>wider</strong> is the reverse, moving data from rows into columns.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S04F02_pivot.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption>Pivot functions allow you to change the structure of your data frame</figcaption>
</figure>
</div>
<p>As shown in the figure above, we use the <code>pivot_longer()</code> function to pivot data from wide to long format, and the <code>pivot_wider()</code> function to pivot data from long to wide format.</p>
<section id="pivot-wider" class="level4" data-number="4.3.1.1">
<h4 data-number="4.3.1.1" class="anchored" data-anchor-id="pivot-wider"><span class="header-section-number">4.3.1.1</span> Pivot wider</h4>
<p>A common use case for <code>pivot_wider()</code> is to make a <a href="https://en.wikipedia.org/wiki/Contingency_table">contingency table</a>, which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</p>
<p>For example, before we can address our first research question of whether the relationship between weight lost and drug dosage is affected by mouse strain or sex, we probably want to check that we have sufficient <em>n</em> for each mouse strain/sex combination. We can achieve this in a long format using <code>summarise()</code> as we learned in session 2:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 3
  mouse_strain sex    n_mice
  &lt;chr&gt;        &lt;chr&gt;   &lt;int&gt;
1 CD-1         male       73
2 CD-1         female     73
3 CD-1         &lt;NA&gt;        6
4 Black 6      female     58
5 Black 6      male       61
6 Black 6      &lt;NA&gt;        5
7 BALB C       female     34
8 BALB C       male       34</code></pre>
</div>
</div>
<p>This summarises by each combination of <code>mouse_strain</code> and <code>sex</code>, with the <code>n()</code> function giving the count of data belonging to that combination.</p>
<p>To get a contingency table, we want the rows to represent one categorical variable (in this case <code>sex</code>) and the columns to represent another categorical variable (in this case <code>mouse_strain</code>), with the values in the cells being the counts (coming from the <code>n_mice</code> column). In doing so, our table will become wider, because we are adding additional columns. So, we can use the <code>pivot_wider()</code> function to create our contingency table.</p>
<p><code>pivot_wider()</code> requires two pieces of information:</p>
<ol type="1">
<li><code>names_from</code>, which is the column that the new columns will come from</li>
<li><code>values_from</code>, which is the column that the values of the cells will come from</li>
</ol>
<p>In this case, our <code>names_from</code> will be <code>mouse_strain</code>, and <code>values_from</code> will be <code>n_mice</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  sex    `CD-1` `Black 6` `BALB C`
  &lt;chr&gt;   &lt;int&gt;     &lt;int&gt;    &lt;int&gt;
1 male       73        61       34
2 female     73        58       34
3 &lt;NA&gt;        6         5       NA</code></pre>
</div>
</div>
<p>This has transformed our data into a contingency table, with <code>NA</code> used to indicate where no data corresponding data exists for the specific <code>sex</code> and <code>mouse_strain</code> combination. It allows us to see that we have a decent number of mice for each strain/sex, giving us confidence that we can address whether or not these variables affect the relationship between drug dosage and weight loss.</p>
</section>
<section id="pivot-longer" class="level4" data-number="4.3.1.2">
<h4 data-number="4.3.1.2" class="anchored" data-anchor-id="pivot-longer"><span class="header-section-number">4.3.1.2</span> Pivot longer</h4>
<p>Although wide format data is easy to read, it is difficult to use tidyverse functions on because of how variables are stored in column names rather than in columns of their own. For example, in the contingency table we created above, the different strains of the mice are now in the names of the columns. Another example of this is in our new <code>genes</code> dataset, where the gene names and replicates are both encoded as column names:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>genes</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 343 × 11
   group        Th_rep1 Th_rep2 Th_rep3 Prlh_rep1 Prlh_rep2 Prlh_rep3 Hprt1_rep1
   &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1 grp_treatme…   3239.   5964.   4123.      225.      359.      314.       858.
 2 grp_treatme…   2665.   4867.   4240.      253.      460.      358.       709.
 3 grp_treatme…   1536.   2501.   2054.      367.      552.      448.       975.
 4 grp_treatme…   2732.   4800.   3981.      258.      411.      339.       725.
 5 grp_treatme…   2384.   3460.   3115.      347.      651.      491.       896.
 6 grp_treatme…   2972.   4880.   4028.      303.      456.      377.       865.
 7 grp_treatme…   1936.   3666.   2964.      302.      441.      355.       625.
 8 grp_treatme…   2801.   4977.   3795.      246.      424.      321.       856.
 9 grp_treatme…   3239.   5198.   3930.      385.      654.      508.       623.
10 grp_treatme…   2665.   3820.   2958.      230.      391.      301.       740.
# ℹ 333 more rows
# ℹ 3 more variables: Hprt1_rep2 &lt;dbl&gt;, Hprt1_rep3 &lt;dbl&gt;, id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This can be pretty tricky to work with— we’d have to type out all of the <code>Th_rep1 Th_rep2 Th_rep3</code> columns every time we want to do something with the TH gene, and if we ever wanted to re-use the code for a different gene or a different number of replicates we’d need to change the column names accordingly. Instead, it would be better to have a column for gene name, and another column for the replicate so that we can write code that will cope with any number of different genes and replicates.</p>
<p>To achieve this, we can use the <code>pivot_longer()</code> function, which will create a pair of new columns from the original column name and the value of the corresponding cell. <code>pivot_longer()</code> requires three different arguments:</p>
<ol type="1">
<li><code>cols</code>: the columns to pivot from. You can use selection helpers like <code>contains()</code> or <code>starts_with()</code> to easily select multiple columns at once.</li>
<li><code>names_to</code>: the name of a new column that will contain the original column names.</li>
<li><code>values_to</code>: the name of a new column that will contain the values from the original columns.</li>
</ol>
<p>In our case, here’s what the code would look like:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save the long format data as a variable so we can use it later</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>genes_long <span class="ot">&lt;-</span> genes <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>),</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"measurement"</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>genes_long</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 4
   group         id_num measurement expression
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;            &lt;dbl&gt;
 1 grp_treatment      1 Th_rep1          3239.
 2 grp_treatment      1 Th_rep2          5964.
 3 grp_treatment      1 Th_rep3          4123.
 4 grp_treatment      1 Prlh_rep1         225.
 5 grp_treatment      1 Prlh_rep2         359.
 6 grp_treatment      1 Prlh_rep3         314.
 7 grp_treatment      1 Hprt1_rep1        858.
 8 grp_treatment      1 Hprt1_rep2       1292.
 9 grp_treatment      1 Hprt1_rep3        988.
10 grp_treatment      2 Th_rep1          2665.
# ℹ 3,077 more rows</code></pre>
</div>
</div>
<p>Take a moment to try and understand the values provided to each argument, then check your answer below ⬇️</p>
<details>
<summary>
Solution
</summary>
<ol type="1">
<li>For the <code>cols</code> argument, we used the selection helper <code>contains("_rep")</code> to select all columns containing the string <code>"rep"</code> (this saves us typing out all nine of the column names)</li>
<li>For the <code>names_to</code> argument, we decided to call this new column ‘measurement’, as it contains both the gene name and replicate (so is a unique measurement of a particular gene in a particular replicate)</li>
<li>For the <code>values_to</code> argument, we have called this new column ‘expression level’ as the numbers represent the gene expression in each of the different measurements</li>
</ol>
</details>
<div class="callout callout-style-default callout-note callout-titled" title="Perplexed by pivoting?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Perplexed by pivoting?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Pivoting can be a bit tricky to get your head around! Often when you’re doing analysis, you’ll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <a href="https://r4ds.hadley.nz/data-tidy.html#sec-pivoting">like those in the R for Data Science book</a>, ask an AI assistant or just try things until it works!</p>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What does the following code do?</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Pivots data into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>dosage</code> data frame, then pivots into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>dosage</code> data frame, then pivots into a wide format where there is a column for each mouse strain.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>It just gives an error</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. I have run the following code to create a new column in the <code>dosage</code> data frame that gives the weight of the mice at the end of the experiment.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   id_num initial_weight_g final_weight_g
    &lt;dbl&gt;            &lt;dbl&gt;          &lt;dbl&gt;
 1      1             39.1           35.4
 2      2             39.5           35.7
 3      3             40.3           37.0
 4      4             NA             NA  
 5      5             36.7           33.2
 6      6             39.3           35.6
 7      7             38.9           35.3
 8      8             39.2           34.5
 9      9             34.1           30.6
10     10             42             37.8
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code></p><p></p>
</div>
</div>

</div>
</div>
<ol start="3" type="1">
<li><p>Using the <code>dosage</code> data frame, write R code to:</p>
<ol type="a">
<li>Make a data frame that shows the number of mice of each strain, in each replicate.</li>
<li>Pivot this data frame into a wide format to create a contingency table.</li>
<li>Pivot the wide data frame from (b) back into a long format.</li>
</ol></li>
</ol>
<details>
<summary>
Solutions
</summary>
<p>1. The code first calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>dosage</code> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <code>names_from = sex</code> )</p>
<p>2. The correct pivot function call to take the data from a wide format to a long format is <code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code>. This code tells R to pivot the <code>initial_weight_g</code> and <code>final_weight_g</code> columns into a long format, where there is a row for each mouse and each weight measurement. The <code>names_to</code> argument specifies to make a column called ‘timepoint’ that tells us whether the measurement is initial or final, and the <code>values_to</code> argument specifies the name of the new column that will contain these measurements.</p>
<p>3. Here’s how you could write R code to achieve the tasks:</p>
<pre><code>a.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate))`
b.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice)`
c.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice) %&gt;% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")`</code></pre>
</details>
</div>
</div>
</section>
</section>
<section id="sec-separate" class="level3" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="sec-separate"><span class="header-section-number">4.3.2</span> Separating data in a column</h3>
<p>Now that we’ve pivoted it into a long format, our <code>genes</code> data certainly looks a lot better! But, it’s not quite tidy yet; recall that our for data to be tidy, <a href="#sec-tidy-data">each column should contain only one variable</a>. but our <code>measurement</code> column actually contains <strong>two</strong> pieces of information: the gene being measured and the replicate number, which are separated by an underscore. In order to make our data tidy, we can use the <code>separate()</code> function to split these into columns of their own.</p>
<p>To use the separate function, we need to give it three arguments:</p>
<ol type="1">
<li>The name of the column that needs to be separated (in this case, <code>measurement</code>)</li>
<li><code>into</code>, which is a vector of the new column names that the data will be placed into</li>
<li><code>sep</code>, which is the character that separates the two pieces of data</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="ot">&lt;-</span> genes_long <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>      measurement, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"replicate"</span>), </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">sep =</span> <span class="st">"_"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We can now see that the gene name and replicate are contained within their own columns, officially making the data tidy!</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>genes_tidy</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 5
   group         id_num gene  replicate expression
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
 1 grp_treatment      1 Th    rep1           3239.
 2 grp_treatment      1 Th    rep2           5964.
 3 grp_treatment      1 Th    rep3           4123.
 4 grp_treatment      1 Prlh  rep1            225.
 5 grp_treatment      1 Prlh  rep2            359.
 6 grp_treatment      1 Prlh  rep3            314.
 7 grp_treatment      1 Hprt1 rep1            858.
 8 grp_treatment      1 Hprt1 rep2           1292.
 9 grp_treatment      1 Hprt1 rep3            988.
10 grp_treatment      2 Th    rep1           2665.
# ℹ 3,077 more rows</code></pre>
</div>
</div>
</section>
<section id="sec-join" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="sec-join"><span class="header-section-number">4.3.3</span> Combining data with join functions</h3>
<p>To address our <a href="#sec-research-questions">second research question</a>, we need to look at the relationship between gene expression data (in the <code>genes</code> data frame) and mouse strain (in the <code>dosage</code> data frame). But how can we achieve this when the information is contained across two separate data frames? So long as we have some way of connecting the observations in each data frame (i.e.&nbsp;a shared column), we can use the <code>join</code> family of functions to combine them.</p>
<p>There are many different types of <code>join</code> functions that you can use, but broadly they will either:</p>
<ul>
<li>Add new variables from one data frame to another</li>
<li>Filter observations in one data frame based on whether or not they match observations in another data frame.</li>
<li>Do both of these things at the same time</li>
</ul>
<p>Probably the most commonly used is the <code>left_join()</code> function. You need to provide it with two pieces of information:</p>
<ol type="1">
<li><p>The two data frames you want to join</p></li>
<li><p>The name of a column (or columns) to join by</p></li>
</ol>
<p>The reason it’s called a left-join is because it retains all rows from the left data frame while adding on columns from the right data frame only when the column you’re joining by matches.</p>
<p>For example, in the following cartoon, where we are joining by column A:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S04F03_left_join.png" class="img-fluid figure-img"></p>
<figcaption>The left_join() function allows you to combine two data frames</figcaption>
</figure>
</div>
<p>We can see that:</p>
<ul>
<li>A new column, D has been added from the right data frame</li>
<li>All of the observations from the left data frame are present in the result</li>
<li>The first row of the result contains an <code>NA</code> in the D column, because in the right data frame that the D column comes from, there is no matching observation ‘a’</li>
<li>The last row from the right data frame is missing from the result, because it has no matching value of ‘d’ in the left data frame</li>
</ul>
<div class="callout callout-style-default callout-warning callout-titled" title="Beware multiple matches!">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Warning</span>Beware multiple matches!
</div>
</div>
<div class="callout-body-container callout-body">
<p>If the values from the left data frame match to multiple rows of the column in the right data frame, the <code>left_join()</code> will duplicate the data from the left data frame for each match to the right. This can cause issues with downstream summarisation if not carefully considered.</p>
</div>
</div>
<p>Using this knowledge, let’s try joining our <code>dosage</code> and <code>genes_tidy</code> data frames based on their shared column, <code>id_num</code> (an ID number uniquely assigned to each mouse):</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's save the result of this as a variable</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>combined_data <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    genes_tidy, <span class="co"># the left data frame</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    dosage, <span class="co"># the right data frame</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id_num"</span>) <span class="co"># name of the column to join by</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>combined_data</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 13
   group         id_num gene  replicate.x expression mouse_strain cage_number
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;      
 1 grp_treatment      1 Th    rep1             3239. CD-1         1A         
 2 grp_treatment      1 Th    rep2             5964. CD-1         1A         
 3 grp_treatment      1 Th    rep3             4123. CD-1         1A         
 4 grp_treatment      1 Prlh  rep1              225. CD-1         1A         
 5 grp_treatment      1 Prlh  rep2              359. CD-1         1A         
 6 grp_treatment      1 Prlh  rep3              314. CD-1         1A         
 7 grp_treatment      1 Hprt1 rep1              858. CD-1         1A         
 8 grp_treatment      1 Hprt1 rep2             1292. CD-1         1A         
 9 grp_treatment      1 Hprt1 rep3              988. CD-1         1A         
10 grp_treatment      2 Th    rep1             2665. CD-1         1A         
# ℹ 3,077 more rows
# ℹ 6 more variables: weight_lost_g &lt;dbl&gt;, replicate.y &lt;chr&gt;, sex &lt;chr&gt;,
#   drug_dose_g &lt;dbl&gt;, tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
</div>
<p>We now have the gene expression information and the dosage/metadata for each mouse all together in a single data frame. Notice how all of the 3,087 rows in the <code>genes_tidy</code> data frame are present in the <code>combined_data</code>, and the relevant information from the <code>dosage</code> data frame has been joined where available.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Two replicate columns?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Two replicate columns?
</div>
</div>
<div class="callout-body-container callout-body">
<p>One other thing you might have noticed is that there are now two replicate columns: <code>replicate.x</code> and <code>replicate.y</code>. This is because both <code>genes_tidy</code> and <code>dosage</code> each contained their own replicate column, but we only told <code>left_join()</code> to match up our data by <code>id_num</code>. Since you can’t have two columns with the same name, R has added the <code>.x</code> and <code>.y</code> to differentiate them. If you want R to use more informative names than just <code>.x</code> and <code>.y</code> , you can use the <code>suffix</code> argument:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    genes_tidy, <span class="co"># the left data frame</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    dosage, <span class="co"># the right data frame</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id_num"</span>, <span class="co"># name of the column to join by</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">"_genes"</span>, <span class="st">"_dosage"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 13
   group        id_num gene  replicate_genes expression mouse_strain cage_number
   &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;                &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;      
 1 grp_treatme…      1 Th    rep1                 3239. CD-1         1A         
 2 grp_treatme…      1 Th    rep2                 5964. CD-1         1A         
 3 grp_treatme…      1 Th    rep3                 4123. CD-1         1A         
 4 grp_treatme…      1 Prlh  rep1                  225. CD-1         1A         
 5 grp_treatme…      1 Prlh  rep2                  359. CD-1         1A         
 6 grp_treatme…      1 Prlh  rep3                  314. CD-1         1A         
 7 grp_treatme…      1 Hprt1 rep1                  858. CD-1         1A         
 8 grp_treatme…      1 Hprt1 rep2                 1292. CD-1         1A         
 9 grp_treatme…      1 Hprt1 rep3                  988. CD-1         1A         
10 grp_treatme…      2 Th    rep1                 2665. CD-1         1A         
# ℹ 3,077 more rows
# ℹ 6 more variables: weight_lost_g &lt;dbl&gt;, replicate_dosage &lt;chr&gt;, sex &lt;chr&gt;,
#   drug_dose_g &lt;dbl&gt;, tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
</div>
<p>It is now clearer which original dataset each replicate column comes from.</p>
</div>
</div>
<section id="joining-with-mismatched-column-names" class="level4" data-number="4.3.3.1">
<h4 data-number="4.3.3.1" class="anchored" data-anchor-id="joining-with-mismatched-column-names"><span class="header-section-number">4.3.3.1</span> Joining with mismatched column names</h4>
<p>Sometimes, the column containing the matching information you want to join over will have different names in different data frames. For example what might be called “id_num” in one data frame could be called “mouse_id” in another data frame. In this case, the <code>by</code> argument of <code>left_join()</code> can be formatted using the <code>join_by(name1 == name2)</code> helper function to specify which column in the left data frame matches which column on the right.</p>
<p>As an example, we’ll rename the <code>id_num</code> column in the <code>dosage</code> data frame to <code>mouse_id</code> and repeat the join to <code>genes_tidy</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename id_num to mouse_id</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>renamed_dosage <span class="ot">&lt;-</span> dosage <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we can use the rename() function to rename a column</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">mouse_id =</span> id_num)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># now perform the left join</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  genes_tidy, <span class="co"># left data frame </span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  renamed_dosage, <span class="co"># right data frame</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join by the id_num column in genes_tidy, which is equivalent to the mouse_id column in renamed_dosage</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">join_by</span>(id_num <span class="sc">==</span> mouse_id)) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 13
   group         id_num gene  replicate.x expression mouse_strain cage_number
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;      
 1 grp_treatment      1 Th    rep1             3239. CD-1         1A         
 2 grp_treatment      1 Th    rep2             5964. CD-1         1A         
 3 grp_treatment      1 Th    rep3             4123. CD-1         1A         
 4 grp_treatment      1 Prlh  rep1              225. CD-1         1A         
 5 grp_treatment      1 Prlh  rep2              359. CD-1         1A         
 6 grp_treatment      1 Prlh  rep3              314. CD-1         1A         
 7 grp_treatment      1 Hprt1 rep1              858. CD-1         1A         
 8 grp_treatment      1 Hprt1 rep2             1292. CD-1         1A         
 9 grp_treatment      1 Hprt1 rep3              988. CD-1         1A         
10 grp_treatment      2 Th    rep1             2665. CD-1         1A         
# ℹ 3,077 more rows
# ℹ 6 more variables: weight_lost_g &lt;dbl&gt;, replicate.y &lt;chr&gt;, sex &lt;chr&gt;,
#   drug_dose_g &lt;dbl&gt;, tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Notice how the column name from the left data frame (in this case, <code>id_num</code>) is retained in the joined result.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that an older syntax exists for the <code>by</code> argument that would look like <code>by = c("mouse_id" = "id_num")</code>. You might encounter this in older books/posts/examples online, or when using AI assistants like ChatGPT, but we recommend you stick to the <code>join_by()</code> syntax instead, as it’s easier to understand.</p>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which of the following is NOT a valid way to join the <code>dosage</code> data frame with the <code>genes</code> data frame based on the <code>id_num</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>dosage %&gt;% left_join(genes, by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>left_join(dosage, genes, by = "id_num")</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>dosage %&gt;% left_join(genes, .by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>dosage %&gt;% left_join(genes, by = ("id_num" = "id_num"))</code></p><p></p>
</div>
</div>

</div>
</div>
<ol start="2" type="1">
<li>Let’s say I have two data frames, <code>df1</code> and <code>df2</code>, that I want to join based a shared ‘key’ column, that is called ‘key’ in <code>df1</code> and ‘item_key’ in <code>df2</code>. Write R code to join these two data frames using the <code>left_join()</code> function.</li>
</ol>
<details>
<summary>
Solutions
</summary>
<p>1. The line of code that is NOT a valid way to join the <code>dosage</code> data frame with the <code>genes</code> data frame based on the <code>id_num</code> column is <code>dosage %&gt;% left_join(genes, .by = "id_num")</code>. This line of code is incorrect because the <code>.by</code> argument is not used in the <code>left_join()</code> function (this can be confusing! it’s <code>.by</code> when grouping by <code>by</code> when joining). The other options are valid ways to join the two data frames based on the <code>id_num</code> column: remember that we don’t have to use pipes to join data frames, we can use the <code>left_join()</code> function directly, and we can use a named vector to specify the columns to join on (although here it’s a bit redundant as the columns have the same name).</p>
<p>2. To join the two data frames you could use <code>df1 %&gt;% left_join(df2, by = c("key" = "item_key"))</code> (with pipe) or <code>left_join(df1, df2, by = c("key" = "item_key"))</code> (without pipe).</p>
</details>
</div>
</div>
</section>
</section>
</section>
<section id="the-process-of-data-analysis" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="the-process-of-data-analysis"><span class="header-section-number">4.4</span> The process of data analysis</h2>
<p>With all that data wrangling out of the way, we can finally move on to doing some analysis to answer our <a href="#sec-research-questions">research questions</a>! But wait— how do you actually <em>do</em> data analysis? Where should you start?</p>
<p>Well, the creators of the tidyverse packages we use in this course have proposed the following ‘process of data science’:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S04F04_analysis_steps.png" class="img-fluid figure-img"></p>
<figcaption>The ‘process of data science’, according to the tidyverse (source: <a href="https://r4ds.hadley.nz/intro.html#fig-ds-diagram" class="uri">https://r4ds.hadley.nz/intro.html#fig-ds-diagram</a>)</figcaption>
</figure>
</div>
<p>In fact, we’ve already done the first two steps: reading in our data and tidying it up! The main portion of the analysis however is depicted as a cycle: transforming your data with the dplyr functions we learned in <a href="session_2.html" class="quarto-xref"><span>Chapter 2</span></a>, visualising it with the ggplot2 functions we learned in <a href="session_3.html" class="quarto-xref"><span>Chapter 3</span></a>, modelling it (with statistics, which we don’t cover in this course as they’re quite field-specific) and repeating. This process of iterating over many different research/analysis questions to develop a deeper understanding of your data allows you to really explore every aspect. Finally, you need to be able to clearly communicate the results of your analysis and interpret them in the biological context!</p>
<p>As we go through the research questions in the rest of this session, keep these steps in mind and try to think about how you could communicate the results.</p>
</section>
<section id="question-1-the-relationship-between-dosage-and-weight-loss" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="question-1-the-relationship-between-dosage-and-weight-loss"><span class="header-section-number">4.5</span> Question 1: The relationship between dosage and weight loss</h2>
<p>Let’s start by tackling the first part of our first research question: Does the amount of weight lost by mice increase as dosage of mousezempic increases?</p>
<p>Luckily, our <code>dosage</code> data was pretty tidy already so we didn’t have to do much cleaning. To begin our analysis process, let’s try investigating the relationship between these variables with visualisation. As both the dosage and weight loss variables are continuous, we can represent them on a scatterplot using <code>geom_point()</code></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dosage, <span class="fu">aes</span>(<span class="at">x =</span> drug_dose_g, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We can clearly see a positive linear relationship between drug dosage and weight lost.</p>
<p>However, this plot certainly isn’t publication ready! Practice the R skills you have learned throughout the course so far and try to improve the look of the plot by working through the exercises below ⬇️</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercise: changing the x axis">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercise: changing the x axis
</div>
</div>
<div class="callout-body-container callout-body">
<p>For starters, it would probably make more sense to display the drug doses in milligrams rather than in grams, given that the values are so small. Take a moment to think about how you might achieve this using the dplyr functions we learned in session 2, and then check your answer below.</p>
<details>
<summary>
Solution
</summary>
<p>To convert g to mg we need to multiply by 1000. So, we need some kind of function that allows us to take the existing <code>drug_dose_g</code> column and multiply it by 1000 to create a new column that we can use for our plotting.</p>
<p>1. Use the <code>mutate()</code> function to create a new column in the <code>dosage</code> data frame called (for example) <code>drug_dose_mg</code></p>
<p>2. Either save the modified <code>dosage</code> data frame as a variable in R, or pipe it directly into the ggplot function call</p>
<p>3. Make the ggplot as before</p>
</details>
<p>Learning how to break a problem down into simple steps is a crucial skill to learn for data analysis! Even if there are some steps that you don’t immediately know how to solve with code, it gives you a good starting point for your googling or to ask ChatGPT.</p>
<p>Now, try to convert these steps into code and create a plot where the x axis displays the drug dose in mg, like this one:</p>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 2 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<details>
<summary>
Solution
</summary>
<p>Following the steps outlined above, here are two example code chunks that will make the correct plot. There are of course many more ways to approach the problem, so it’s ok if your solution doesn’t match perfectly!</p>
<p>This first solution saves the modified dosage data as a new data frame, which could be useful if you want to use the <code>drug_dose_mg</code> column in other plots or calculations:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># step two: save the modified version of dosage as its own </span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>dosage_mg <span class="ot">&lt;-</span> dosage <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step one: use mutate to create drug_dose_mg column</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step three: plot same as before</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(dosage_mg, <span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The second one instead pipes this data frame directly from the <code>mutate()</code> function call straight into the <code>ggplot()</code> function. This approach is convenient when the data manipulation you’re doing (in this case adding the <code>drug_dose_mg</code> column) is only for a specific plot, so you don’t clutter up your environment panel:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step one: use mutate to create drug_dose_mg column</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step two: pipe directly into the ggplot() function</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># note that because the pipe passes our modified dosage data in as the first argument,</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we don't need to specify the data frame we are plotting, just the aes() </span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step three: plot same as before </span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</details>
<p>Now our x axis is much easier to read!</p>
</div>
</div>
<p>Still, there are a few things we could do to make our plot look a bit better. In the next set of practice problems, we will use our ggplot2 knowledge to improve the plot visually:</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercise: making visual improvements">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercise: making visual improvements
</div>
</div>
<div class="callout-body-container callout-body">
<p>Take a moment to look at the plot and think of what you could change to make it more informative and visually appealing. Of course this is subjective, but some suggestions are provided below.</p>
<details>
<summary>
Solution
</summary>
<p>1. The plot could do with some nicely formatted labels on the x and y axes, as well as a title.</p>
<p>2. Some of the points on the plot are overlapping, so it could be helpful to change the points from filled circles to unfilled circles so they don’t all just merge into a blob.</p>
<p>3. This is a matter of personal taste, but I think the default ggplot2 grey background is kind of basic so I always change it 😝</p>
</details>
<p>Here’s my plot. Click below to see the R code that made it:</p>
<div class="cell">
<details class="code-fold">
<summary>Show the R code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see the previous practice q for the explanation of this</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># dplyr code to create the mg column</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create the plot</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the points to unfilled circles</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># note that we don't need to put it in aes() because</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># it isn't tied to a particular variable</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># also increasing the stroke (width of the circle line) to 0.75</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># so the points stand out more</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">1</span>, <span class="at">stroke =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add axis labels and title</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Association between drug dosage and weight lost"</span>,</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Drug dosage (mg)"</span>,</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Weight lost (g)"</span>) <span class="sc">+</span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the theme</span></span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<section id="the-effect-of-strain-and-sex" class="level3" data-number="4.5.1">
<h3 data-number="4.5.1" class="anchored" data-anchor-id="the-effect-of-strain-and-sex"><span class="header-section-number">4.5.1</span> The effect of strain and sex</h3>
<p>The final part of the research question was to investigate whether mouse strain or sex has any effect on this relationship. Above, we learned how to use the <code>summarise()</code> and <code>pivot_wider()</code> functions to create a contingency table that counts up how many mice fall into each of the mouse strain/sex combinations. This is an important first step for the analysis, as it allows us to double check that there isn’t a category with so few mice that would prevent us from drawing reliable conclusions.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  sex    `CD-1` `Black 6` `BALB C`
  &lt;chr&gt;   &lt;int&gt;     &lt;int&gt;    &lt;int&gt;
1 male       73        61       34
2 female     73        58       34
3 &lt;NA&gt;        6         5       NA</code></pre>
</div>
</div>
<p>Given that there are &gt;30 mice in each category, we can proceed with the analysis.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercise: adding in extra information">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercise: adding in extra information
</div>
</div>
<div class="callout-body-container callout-body">
<p>Now, let’s try adding the mouse strain and sex information to the scatterplot we created above. Try to think about what aesthetics would be good to use to show those two variables.</p>
<details>
<summary>
Solution
</summary>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see the previous practice q for the explanation of this</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># dplyr code to create the mg column</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create the plot</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g, <span class="at">colour =</span> mouse_strain, <span class="at">shape =</span> sex)) <span class="sc">+</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add axis labels and title</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Association between drug dosage and weight lost"</span>,</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Drug dosage (mg)"</span>,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Weight lost (g)"</span>,</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>       <span class="co"># adding nicer labels to the legend</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>       <span class="co"># you just use the aesthetic name </span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">colour =</span> <span class="st">"Mouse strain"</span>,</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">shape =</span> <span class="st">"Sex"</span>) <span class="sc">+</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the theme</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Note that:</p>
<ul>
<li><p>It’s up to you which aesthetics should represent each variable, but in this example the <code>colour</code> aesthetic was chosen for the mouse strains because there are more strains than sexes, and it’s easier to distinguish between different colours than different shapes</p></li>
<li><p>We used the <code>labs()</code> function to change the legend labels for our mouse strain and sex variables. It just gives a nice professional touch to the plot</p></li>
</ul>
</details>
</div>
</div>
<p>Our plot reveals that Black 6 mice received higher drug dosages than the other mice, and also lost more weight. This shows us why it’s so important to explore the relationships between lots of different variables in large datasets. Imagine if we had just plotted the weight loss for mice of each strain:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dosage, <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span> </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We might think that the drug is more effective on Black 6 mice, because they lost more weight, but our scatterplot above shows it’s more likely to be due to their higher dosage.</p>
<p>Always try to look at all the variables you have, because you never know what associations you might find!</p>
</section>
</section>
<section id="question-2-the-effect-of-mousezempic-on-gene-expression" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="question-2-the-effect-of-mousezempic-on-gene-expression"><span class="header-section-number">4.6</span> Question 2: The effect of mousezempic on gene expression</h2>
<p>Our second research question was about the effect of mousezempic on the expression of the TH and PRLH genes. To answer this question, we can use the <code>genes_tidy</code> dataset that we prepared in <a href="#sec-reshaping" class="quarto-xref"><span>Section 4.3</span></a>, by pivoting the <code>genes</code> dataset to be longer and separating the measurement column so each variable is contained within its own column.</p>
<p>Let’s take a look at the <code>genes_tidy</code> data to remind ourselves of the information it contains. You might like to use the <code>View()</code> function, or just type the name of the data like so:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>genes_tidy</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 5
   group         id_num gene  replicate expression
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
 1 grp_treatment      1 Th    rep1           3239.
 2 grp_treatment      1 Th    rep2           5964.
 3 grp_treatment      1 Th    rep3           4123.
 4 grp_treatment      1 Prlh  rep1            225.
 5 grp_treatment      1 Prlh  rep2            359.
 6 grp_treatment      1 Prlh  rep3            314.
 7 grp_treatment      1 Hprt1 rep1            858.
 8 grp_treatment      1 Hprt1 rep2           1292.
 9 grp_treatment      1 Hprt1 rep3            988.
10 grp_treatment      2 Th    rep1           2665.
# ℹ 3,077 more rows</code></pre>
</div>
</div>
<p>Each mouse has a unique <code>id_num</code>, and belongs to a <code>group</code> (either treatment or control). There are also gene expression measurements in triplicate for each of three different genes: the TH and PRLH genes we are interested in and HPRT1, a housekeeping gene.</p>
<section id="do-the-treatment-and-control-groups-differ" class="level3" data-number="4.6.1">
<h3 data-number="4.6.1" class="anchored" data-anchor-id="do-the-treatment-and-control-groups-differ"><span class="header-section-number">4.6.1</span> Do the treatment and control groups differ?</h3>
<p>To investigate differences in gene expression between the treatment an control groups, a good first step would be to make a plot. Boxplots, violin plots or beeswarm plots would be a good choice to visualise this type of data.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercise: plotting the relationship between mousezempic and gene expression">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Practice exercise: plotting the relationship between mousezempic and gene expression
</div>
</div>
<div class="callout-body-container callout-body">
<p>Take a moment to think about how you could achieve this, and check your answer below. As an additional exercise, how could you prevent the HPRT1 gene from being shown on your plot?</p>
<details>
<summary>
Solution
</summary>
<p>Since we want to plot the expression of two different genes, we have a couple of options.</p>
<p>Firstly, we could use a single <code>ggplot()</code> function call, and facet the plot by the gene. However, if we want to exlude the HPRT1 gene, we’ll have to first use the <code>filter()</code> function from dplyr to remove it from our data frame.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that != is the test for inequality in R</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># also note we use Hprt1 not HPRT1: needs to match </span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the data frame exactly for filter() to work</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we will pipe this filtered data into the ggplot</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># because we are unlikely to want to re-use it</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># to facet write ~ then the name of the column</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># need scales = "free" because otherwise it will</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># use the same scales for both genes, and the </span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># expression levels are quite different</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Another option would be to plot each of TH and PRLH separately, with two different <code>ggplot()</code> function calls and then combine them with patchwork (make sure this package is loaded!). To do this, we would also have to use the <code>filter()</code> function, this time to filter to just the rows that contain the gene we want to plot</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first plot: TH</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>th_plot <span class="ot">&lt;-</span> genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Th"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># should add a title to remind us which one this is</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"TH gene expression in control vs treatment groups"</span>) <span class="sc">+</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co"># second plot: PRLH</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>prlh_plot <span class="ot">&lt;-</span> genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Prlh"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># should add a title to remind us which one this is</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"PRLH gene expression in control vs treatment groups"</span>) <span class="sc">+</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a><span class="co"># combine them using patchwork + syntax</span></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>th_plot <span class="sc">+</span> prlh_plot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>The first method with faceting requires less code, but the second method of making two separate plots that are combined with patchwork allows you to be more flexible with the layout and include extra plots in the figure if desired.</p>
<p>But either solution is perfectly valid, and allows us to see that the expression of the Th and Prlh genes is higher in the treatment group than in the control group.</p>
</details>
</div>
</div>
<p>R also provides a range of functions that we could use to formally test the relationship between these variables, although they are out of the scope of this course. A helpful guide on performing statistical tests in R <a href="https://epirhandbook.com/en/new_pages/stat_tests.html#base-r">can be found here</a>.</p>
</section>
<section id="an-aside-string-manipulation" class="level3" data-number="4.6.2">
<h3 data-number="4.6.2" class="anchored" data-anchor-id="an-aside-string-manipulation"><span class="header-section-number">4.6.2</span> An aside: string manipulation</h3>
<p>Let’s now take a moment to improve the appearance of our plot from the exercise above:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>In session three we learned how to change the x and y axis labels on our graph to be more informative, but how can we change the values on the axes themselves? Currently, we have ‘grp_control’ and ‘grp_treatment’, but to make the plot look more publication-ready, it would be nice if it could just say ‘Control’ and ‘Treatment’.</p>
<p>We can do this with the help of the <a href="https://stringr.tidyverse.org/">stringr</a> tidyverse package, which contains a set of functions to manipulate string (character) data in R. Being able to programatically manipulate strings is an essential tool for dealing with real-world data, particularly when you need to extract sample information from names or (as in our case) remove irrelevant characters that only take up space on plots.</p>
<p>Some useful functions for string manipulation include:</p>
<ul>
<li><code>str_replace(string, pattern, replacement)</code>: replaces the <code>pattern</code> in the <code>string</code> with the <code>replacement</code></li>
<li><code>str_extract(string, pattern)</code>: extracts the part of the <code>string</code> matching the <code>pattern</code></li>
<li><code>str_remove(string, pattern)</code>: removes the part of the <code>string</code> matching the <code>pattern</code></li>
</ul>
<p>The full list of functions can be found in the <a href="https://stringr.tidyverse.org/reference/index.html">stringr package documentation</a>.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Using stringr functions on a data frame">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Using stringr functions on a data frame
</div>
</div>
<div class="callout-body-container callout-body">
<p>Remember from session 2 that when we use the pipe operator <code>%&gt;%</code>, it passes whatever precedes it into the function after it as the first argument. This means that the functions we use within a chain of pipes on a data frame should (generally) have a data frame as their input, and return a data frame as their output.</p>
<p>But stringr functions operate on character vectors (their first argument is <code>string</code>, as shown above), so how can we use them on our data frames? The answer is to wrap it in the <code>mutate()</code> function from the dplyr package, which allows us to either edit an exisiting column or create a new one based on our string manipulation.</p>
<p>For example, let’s say we want to change the replicate column in our <code>genes_tidy</code> data frame to read ‘Replicate 1’, ‘Replicate 2’, etc instead of ‘rep1’, ‘rep2’. We would do this by using the <code>str_replace()</code> function inside the <code>mutate()</code> function like so:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">replicate =</span> <span class="fu">str_replace</span>(replicate, <span class="st">"rep"</span>, <span class="st">"Replicate "</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,087 × 5
   group         id_num gene  replicate   expression
   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt;
 1 grp_treatment      1 Th    Replicate 1      3239.
 2 grp_treatment      1 Th    Replicate 2      5964.
 3 grp_treatment      1 Th    Replicate 3      4123.
 4 grp_treatment      1 Prlh  Replicate 1       225.
 5 grp_treatment      1 Prlh  Replicate 2       359.
 6 grp_treatment      1 Prlh  Replicate 3       314.
 7 grp_treatment      1 Hprt1 Replicate 1       858.
 8 grp_treatment      1 Hprt1 Replicate 2      1292.
 9 grp_treatment      1 Hprt1 Replicate 3       988.
10 grp_treatment      2 Th    Replicate 1      2665.
# ℹ 3,077 more rows</code></pre>
</div>
</div>
<p>This applies to lots of other functions, not just the ones from the stringr package— if you want to use a function that takes a vector as input on a column of a data frame, wrap it in the <code>mutate()</code> function first!</p>
</div>
</div>
<p>‘Patterns’ are what the stringr package will look for in your <code>string</code>. For example, in this case, our pattern could be <code>grp_</code> because this is what we want to remove from the strings in our ‘group’ column. This is an example of a <strong>fixed pattern</strong>, where you specify exactly what string you want to find. There is also another way to write patterns, called <a href="https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions"><strong>regular expressions</strong></a> which allow you to, for example, remove everything before the first underscore or all numbers. These are extremely powerful, but can get a little confusing so you might like to get help from AI to write them.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Cleaning up the text on our plot">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Cleaning up the text on our plot
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using the stringr package, remove the ‘grp_’ from the x axis labels, and then use ggplot2 functions to add other labels to the plot. Here’s an example of what this might look like:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<details>
<summary>
Solution
</summary>
<p>The code that generates the plot above is:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># use str_remove function to remove 'grp_' from the </span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># beginning of the group column. need to wrap it in</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the mutate function because we are working on a </span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># column in a data frame not just on a vector </span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">str_remove</span>(group, <span class="st">"grp_"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add the other labels </span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Effect of mousezempic on gene expression"</span>,</span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Group"</span>, </span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Expression (arbitrary units)"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</details>
</div>
</div>
</section>
<section id="the-relationship-between-gene-expression-and-weight-loss" class="level3" data-number="4.6.3">
<h3 data-number="4.6.3" class="anchored" data-anchor-id="the-relationship-between-gene-expression-and-weight-loss"><span class="header-section-number">4.6.3</span> The relationship between gene expression and weight loss</h3>
<p>The second part of research question 2 asked us to look at whether TH and/or PRLH expression is associated with the amount of weight lost by mice taking mousezempic. To do this, we will need to use the <code>combined_data</code> that we generated in <a href="#sec-join" class="quarto-xref"><span>Section 4.3.3</span></a>, which is the result of joining the <code>dosage</code> and <code>genes_tidy</code> data frames.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Plotting the relationship between gene expression and weight loss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Plotting the relationship between gene expression and weight loss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <code>combined_data</code> data frame and your choice of ggplot2 functions to see whether there is a relationship between the expression of the TH &amp; PRLH genes and weight lost by mice.</p>
<details>
<summary>
Solution
</summary>
<p>Since both variables are continuous, a scatterplot is a good choice to visualise their relationship. Here is some sample code:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first plot: TH</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>th_plot <span class="ot">&lt;-</span> combined_data <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Th"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"TH gene expression (arbitrary units)"</span>,</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Weight loss (g)"</span>) <span class="sc">+</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co"># second plot: PRLH</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>prlh_plot <span class="ot">&lt;-</span> combined_data <span class="sc">%&gt;%</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Prlh"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"PRLH gene expression (arbitrary units)"</span>,</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Weight loss (g)"</span>) <span class="sc">+</span></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="co"># combine them using patchwork + syntax</span></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>th_plot <span class="sc">+</span> prlh_plot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 6 rows containing missing values or values outside the scale range
(`geom_point()`).
Removed 6 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session_4_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We can se there is no association between the expression of TH and weight lost, and the same is true for PRLH expression and weight lost. This probably makes sense, as there are many complex regulatory pathways involved in hunger.</p>
</details>
</div>
</div>
</section>
</section>
<section id="summary" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="summary"><span class="header-section-number">4.7</span> Summary</h2>
<p>Today, we’ve tried to put together all of the skills we’ve learned in this R course and use them to address some sample data analysis questions. We’ve covered:</p>
<ul>
<li><p>The concept of tidy data, that is stored in a consistent format so that we can easily analyse it with functions from the tidyverse</p></li>
<li><p>How to reshape data frames using the <code>pivot_wider()</code>, <code>pivot_longer()</code> and <code>separate()</code> functions</p></li>
<li><p>How to merge data frames with common columns using the <code>left_join()</code> function</p></li>
<li><p>The ‘stringr’ package for string manipulation, and some of the basic ways we can use it to clean up character columns in our data frames</p></li>
<li><p>The overall process of data science, and how to apply it in practice to break down our research questions into a series of steps/function calls in R</p></li>
</ul>
<p>Of course, a real data analysis will likely be much longer and more complex than the toy examples we have tackled today. However, we hope that it has provided you with a basic, yet versatile toolkit for transforming the messy data you will find in the wild into a neat, tidy format that you can use to explore your research questions and produce gorgeous data visualisations.</p>
<!--
### Practice questions
If you have time, try having a go at addressing these additional research questions. There are many different ways to approach them, so don't worry if yours doesn't match exactly with the solutions.

::: {.callout-important title="Practice exercise: looking at individual replicates"}

ask students to facet the plot by replicate

:::

Normalisation is a common step in many data analyses, as it allows us to more easily compare data across samples or experimental groups. For example, if we were to plot the expression of all genes across the three replicates:


::: {.cell}

```{.r .cell-code}
genes_tidy %>%
  ggplot(aes(x = replicate, y = expression)) + 
    geom_boxplot()
```

::: {.cell-output-display}
![](session_4_files/figure-html/unnamed-chunk-31-1.png){width=2000}
:::
:::


We can see that replicate 2 has higher expression on average compared to replicates 1 and 3. 

There are many different ways to normalise data, depending on the type, but for the sake of this question we'll:

1. Calculate the average HPRT1 expression in each replicate
2. Divide all TH and PRLH measurements by the average HPRT1 expression calculated above

Take a moment to think about how you could achieve this using either the 
-->


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session_3.html" class="pagination-link" aria-label="Session 3: Plotting with ggplot2">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Session 3: Plotting with ggplot2</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./summary.html" class="pagination-link" aria-label="Course summary &amp; concluding remarks">
        <span class="nav-page-text">Course summary &amp; concluding remarks</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb47" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this session:"</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session 4: Putting it all together {#sec-session04}</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>In this session we will combine all the skills we have learned so far to perform a complete analysis of a small dataset.</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>At the end of this session, learners should be able to:</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Explain the concept and purpose of 'tidy' data</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Understand how pivot, separate and join functions can be used to reshape and combine entire data frames</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Describe the key steps in data analysis</span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Perform basic data analyses in R</span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a><span class="fu">## Introduction</span></span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a>For our analysis, we will use the <span class="in">`dosage`</span> and <span class="in">`genes`</span> data frames, which contain information about the mice treated with mousezempic and their gene expression levels, respectively.</span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-31"><a href="#cb47-31" aria-hidden="true" tabindex="-1"></a>The first thing we need to do is load the packages we'll need for today:</span>
<span id="cb47-32"><a href="#cb47-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-35"><a href="#cb47-35" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-36"><a href="#cb47-36" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb47-37"><a href="#cb47-37" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse package contains most of what we need</span></span>
<span id="cb47-38"><a href="#cb47-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb47-39"><a href="#cb47-39" aria-hidden="true" tabindex="-1"></a><span class="co"># let's also load patchwork to put plots together</span></span>
<span id="cb47-40"><a href="#cb47-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb47-41"><a href="#cb47-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-42"><a href="#cb47-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-43"><a href="#cb47-43" aria-hidden="true" tabindex="-1"></a>And read in the two datasets:</span>
<span id="cb47-44"><a href="#cb47-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-47"><a href="#cb47-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-48"><a href="#cb47-48" aria-hidden="true" tabindex="-1"></a><span class="co"># read in dosage data</span></span>
<span id="cb47-49"><a href="#cb47-49" aria-hidden="true" tabindex="-1"></a>dosage <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb47-50"><a href="#cb47-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-51"><a href="#cb47-51" aria-hidden="true" tabindex="-1"></a><span class="co"># read in expression data</span></span>
<span id="cb47-52"><a href="#cb47-52" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb47-53"><a href="#cb47-53" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-54"><a href="#cb47-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-55"><a href="#cb47-55" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="How do we read in data again?"}</span>
<span id="cb47-56"><a href="#cb47-56" aria-hidden="true" tabindex="-1"></a>Note that we used <span class="in">`read_csv()`</span> for the mouse dosage data and <span class="in">`read_tsv()`</span> for the expression data. We chose these functions to match the file extensions of the two datasets, as described when we first learned about reading in data in session 1 (see @sec-readinFunctions).</span>
<span id="cb47-57"><a href="#cb47-57" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-58"><a href="#cb47-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-59"><a href="#cb47-59" aria-hidden="true" tabindex="-1"></a>Before we go further, let's take a moment to explore the <span class="in">`genes`</span> dataset, as it's new in this session.</span>
<span id="cb47-60"><a href="#cb47-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-63"><a href="#cb47-63" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-64"><a href="#cb47-64" aria-hidden="true" tabindex="-1"></a>genes</span>
<span id="cb47-65"><a href="#cb47-65" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-66"><a href="#cb47-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-67"><a href="#cb47-67" aria-hidden="true" tabindex="-1"></a>This data frame contains expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking mousezempic, as well as one housekeeping gene (HPRT1), all measured in triplicate. Each mouse has a unique ID number (<span class="in">`id_num`</span> column), which matches with those in the <span class="in">`dosage`</span> dataset, but there are also additional control mice, who have not received mousezempic (as described in the <span class="in">`group`</span> column).</span>
<span id="cb47-68"><a href="#cb47-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-69"><a href="#cb47-69" aria-hidden="true" tabindex="-1"></a><span class="fu">### Today's research questions {#sec-research-questions}</span></span>
<span id="cb47-70"><a href="#cb47-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-71"><a href="#cb47-71" aria-hidden="true" tabindex="-1"></a>In today's session, we will use the <span class="in">`dosage`</span> and <span class="in">`genes`</span> datasets to explore the following questions:</span>
<span id="cb47-72"><a href="#cb47-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-73"><a href="#cb47-73" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Does the amount of weight lost by mice increase as dosage of mousezempic increases? Is this relationship affected by the mouse strain or sex?</span>
<span id="cb47-74"><a href="#cb47-74" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>How does mousezempic affect the expression of the TH and PRLH genes? Is the expression of either of these genes correlated with the amount of weight mice lose?</span>
<span id="cb47-75"><a href="#cb47-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-76"><a href="#cb47-76" aria-hidden="true" tabindex="-1"></a>Of course, there are countless questions we could ask of our data! In the interest of time, we will stick to these two key analyses but various others are explored in the practice questions at the end of this session's notes.</span>
<span id="cb47-77"><a href="#cb47-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-78"><a href="#cb47-78" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidy data {#sec-tidy-data}</span></span>
<span id="cb47-79"><a href="#cb47-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-80"><a href="#cb47-80" aria-hidden="true" tabindex="-1"></a>In order to answer the research questions outlined above (or any others for that matter), we need to ensure that our data is **"tidy"**.</span>
<span id="cb47-81"><a href="#cb47-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-82"><a href="#cb47-82" aria-hidden="true" tabindex="-1"></a>This is because the tidyverse packages we are using revolve around the concept of "<span class="co">[</span><span class="ot">tidy data</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html)</span>", which is a specific way of organising data into a table so that it is easy to work with. Tidy data is roughly defined as tabular data that contains:</span>
<span id="cb47-83"><a href="#cb47-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-84"><a href="#cb47-84" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Individual variables in columns</span>
<span id="cb47-85"><a href="#cb47-85" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Individual observations in rows</span>
<span id="cb47-86"><a href="#cb47-86" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>One data value in each cell</span>
<span id="cb47-87"><a href="#cb47-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-88"><a href="#cb47-88" aria-hidden="true" tabindex="-1"></a>In other words, each column in our datasets should represent a particular measurement (like expression of the TH gene or the sex of the mice), and each row should contain all of the measurements available for a particular mouse.</span>
<span id="cb47-89"><a href="#cb47-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-90"><a href="#cb47-90" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="*Why* is tidy data easy to work with?"}</span>
<span id="cb47-91"><a href="#cb47-91" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">This cartoon from Allison Horst</span><span class="co">](https://allisonhorst.com/other-r-fun)</span> sums it up well: <span class="al">![Tidy data can be analysed in a more routine/systematic manner](images/S04F01_tidy_data.jpg)</span>{width="80%"}</span>
<span id="cb47-92"><a href="#cb47-92" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-93"><a href="#cb47-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-94"><a href="#cb47-94" aria-hidden="true" tabindex="-1"></a>Although tidy data is the easiest to work with, it's often necessary to alter the format of your data for plotting or table displays. It's a good idea to keep your core data in a tidy format and treating plot or table outputs as representations of that tidy data.</span>
<span id="cb47-95"><a href="#cb47-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-96"><a href="#cb47-96" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Data wrangling?"}</span>
<span id="cb47-97"><a href="#cb47-97" aria-hidden="true" tabindex="-1"></a>Most data you encounter will not be tidy, so the first part of data analysis is usually called "data wrangling" or "data cleaning" and involves tidying up your data (by "wrangling" it into shape) so it is easier to use for downstream analysis. In this session, we will practice data wrangling by tidying up the <span class="in">`genes`</span> data frame.</span>
<span id="cb47-98"><a href="#cb47-98" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-99"><a href="#cb47-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-100"><a href="#cb47-100" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reshaping and combining data (to make it tidy) {#sec-reshaping}</span></span>
<span id="cb47-101"><a href="#cb47-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-102"><a href="#cb47-102" aria-hidden="true" tabindex="-1"></a>In session 2, we learned a range of functions that work on either the rows or columns of our data: <span class="in">`filter()`</span>, <span class="in">`select()`</span>, <span class="in">`mutate()`</span> and <span class="in">`summarise()`</span>. By cleverly combining these functions you can answer a lot of different questions about your data, but sometimes you may need to reshape or even combine different datasets to get the *right* rows and columns. In this section, we will cover how the <span class="in">`pivot`</span>, <span class="in">`separate`</span> and <span class="in">`join`</span> families of functions can be used to help us get our data in the right (tidy!) format to answer the research questions outlined above.</span>
<span id="cb47-103"><a href="#cb47-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-104"><a href="#cb47-104" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reshaping data with pivot functions {#sec-pivot}</span></span>
<span id="cb47-105"><a href="#cb47-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-106"><a href="#cb47-106" aria-hidden="true" tabindex="-1"></a>Pivoting is a way to change the shape of your data:</span>
<span id="cb47-107"><a href="#cb47-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-108"><a href="#cb47-108" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Pivoting **longer** reshapes the data to transfer data stored in columns into rows, resulting in more rows and fewer columns.</span>
<span id="cb47-109"><a href="#cb47-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-110"><a href="#cb47-110" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Pivoting **wider** is the reverse, moving data from rows into columns.</span>
<span id="cb47-111"><a href="#cb47-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-112"><a href="#cb47-112" aria-hidden="true" tabindex="-1"></a><span class="al">![Pivot functions allow you to change the structure of your data frame](images/S04F02_pivot.png)</span>{width="50%"}</span>
<span id="cb47-113"><a href="#cb47-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-114"><a href="#cb47-114" aria-hidden="true" tabindex="-1"></a>As shown in the figure above, we use the <span class="in">`pivot_longer()`</span> function to pivot data from wide to long format, and the <span class="in">`pivot_wider()`</span> function to pivot data from long to wide format.</span>
<span id="cb47-115"><a href="#cb47-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-116"><a href="#cb47-116" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot wider</span></span>
<span id="cb47-117"><a href="#cb47-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-118"><a href="#cb47-118" aria-hidden="true" tabindex="-1"></a>A common use case for <span class="in">`pivot_wider()`</span> is to make a <span class="co">[</span><span class="ot">contingency table</span><span class="co">](https://en.wikipedia.org/wiki/Contingency_table)</span>, which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</span>
<span id="cb47-119"><a href="#cb47-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-120"><a href="#cb47-120" aria-hidden="true" tabindex="-1"></a>For example, before we can address our first research question of whether the relationship between weight lost and drug dosage is affected by mouse strain or sex, we probably want to check that we have sufficient *n* for each mouse strain/sex combination. We can achieve this in a long format using <span class="in">`summarise()`</span> as we learned in session 2:</span>
<span id="cb47-121"><a href="#cb47-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-124"><a href="#cb47-124" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-125"><a href="#cb47-125" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-126"><a href="#cb47-126" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex))</span>
<span id="cb47-127"><a href="#cb47-127" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-128"><a href="#cb47-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-129"><a href="#cb47-129" aria-hidden="true" tabindex="-1"></a>This summarises by each combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span>, with the <span class="in">`n()`</span> function giving the count of data belonging to that combination.</span>
<span id="cb47-130"><a href="#cb47-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-131"><a href="#cb47-131" aria-hidden="true" tabindex="-1"></a>To get a contingency table, we want the rows to represent one categorical variable (in this case <span class="in">`sex`</span>) and the columns to represent another categorical variable (in this case <span class="in">`mouse_strain`</span>), with the values in the cells being the counts (coming from the <span class="in">`n_mice`</span> column). In doing so, our table will become wider, because we are adding additional columns. So, we can use the <span class="in">`pivot_wider()`</span> function to create our contingency table.</span>
<span id="cb47-132"><a href="#cb47-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-133"><a href="#cb47-133" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_wider()`</span> requires two pieces of information:</span>
<span id="cb47-134"><a href="#cb47-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-135"><a href="#cb47-135" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span><span class="in">`names_from`</span>, which is the column that the new columns will come from</span>
<span id="cb47-136"><a href="#cb47-136" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`values_from`</span>, which is the column that the values of the cells will come from</span>
<span id="cb47-137"><a href="#cb47-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-138"><a href="#cb47-138" aria-hidden="true" tabindex="-1"></a>In this case, our <span class="in">`names_from`</span> will be <span class="in">`mouse_strain`</span>, and <span class="in">`values_from`</span> will be <span class="in">`n_mice`</span>.</span>
<span id="cb47-139"><a href="#cb47-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-142"><a href="#cb47-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-143"><a href="#cb47-143" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-144"><a href="#cb47-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb47-145"><a href="#cb47-145" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb47-146"><a href="#cb47-146" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-147"><a href="#cb47-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-148"><a href="#cb47-148" aria-hidden="true" tabindex="-1"></a>This has transformed our data into a contingency table, with <span class="in">`NA`</span> used to indicate where no data corresponding data exists for the specific <span class="in">`sex`</span> and <span class="in">`mouse_strain`</span> combination. It allows us to see that we have a decent number of mice for each strain/sex, giving us confidence that we can address whether or not these variables affect the relationship between drug dosage and weight loss.</span>
<span id="cb47-149"><a href="#cb47-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-150"><a href="#cb47-150" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot longer</span></span>
<span id="cb47-151"><a href="#cb47-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-152"><a href="#cb47-152" aria-hidden="true" tabindex="-1"></a>Although wide format data is easy to read, it is difficult to use tidyverse functions on because of how variables are stored in column names rather than in columns of their own. For example, in the contingency table we created above, the different strains of the mice are now in the names of the columns. Another example of this is in our new <span class="in">`genes`</span> dataset, where the gene names and replicates are both encoded as column names:</span>
<span id="cb47-153"><a href="#cb47-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-156"><a href="#cb47-156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-157"><a href="#cb47-157" aria-hidden="true" tabindex="-1"></a>genes</span>
<span id="cb47-158"><a href="#cb47-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-159"><a href="#cb47-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-160"><a href="#cb47-160" aria-hidden="true" tabindex="-1"></a>This can be pretty tricky to work with--- we'd have to type out all of the <span class="in">`Th_rep1 Th_rep2 Th_rep3`</span> columns every time we want to do something with the TH gene, and if we ever wanted to re-use the code for a different gene or a different number of replicates we'd need to change the column names accordingly. Instead, it would be better to have a column for gene name, and another column for the replicate so that we can write code that will cope with any number of different genes and replicates.</span>
<span id="cb47-161"><a href="#cb47-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-162"><a href="#cb47-162" aria-hidden="true" tabindex="-1"></a>To achieve this, we can use the <span class="in">`pivot_longer()`</span> function, which will create a pair of new columns from the original column name and the value of the corresponding cell. <span class="in">`pivot_longer()`</span> requires three different arguments:</span>
<span id="cb47-163"><a href="#cb47-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-164"><a href="#cb47-164" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span><span class="in">`cols`</span>: the columns to pivot from. You can use selection helpers like <span class="in">`contains()`</span> or <span class="in">`starts_with()`</span> to easily select multiple columns at once.</span>
<span id="cb47-165"><a href="#cb47-165" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`names_to`</span>: the name of a new column that will contain the original column names.</span>
<span id="cb47-166"><a href="#cb47-166" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span><span class="in">`values_to`</span>: the name of a new column that will contain the values from the original columns.</span>
<span id="cb47-167"><a href="#cb47-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-168"><a href="#cb47-168" aria-hidden="true" tabindex="-1"></a>In our case, here's what the code would look like:</span>
<span id="cb47-169"><a href="#cb47-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-172"><a href="#cb47-172" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-173"><a href="#cb47-173" aria-hidden="true" tabindex="-1"></a><span class="co"># save the long format data as a variable so we can use it later</span></span>
<span id="cb47-174"><a href="#cb47-174" aria-hidden="true" tabindex="-1"></a>genes_long <span class="ot">&lt;-</span> genes <span class="sc">%&gt;%</span></span>
<span id="cb47-175"><a href="#cb47-175" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb47-176"><a href="#cb47-176" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>),</span>
<span id="cb47-177"><a href="#cb47-177" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"measurement"</span>,</span>
<span id="cb47-178"><a href="#cb47-178" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb47-179"><a href="#cb47-179" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb47-180"><a href="#cb47-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-181"><a href="#cb47-181" aria-hidden="true" tabindex="-1"></a>genes_long</span>
<span id="cb47-182"><a href="#cb47-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-183"><a href="#cb47-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-184"><a href="#cb47-184" aria-hidden="true" tabindex="-1"></a>Take a moment to try and understand the values provided to each argument, then check your answer below ⬇️</span>
<span id="cb47-185"><a href="#cb47-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-186"><a href="#cb47-186" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-187"><a href="#cb47-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-188"><a href="#cb47-188" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-189"><a href="#cb47-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-190"><a href="#cb47-190" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>For the <span class="in">`cols`</span> argument, we used the selection helper <span class="in">`contains("_rep")`</span> to select all columns containing the string <span class="in">`"rep"`</span> (this saves us typing out all nine of the column names)</span>
<span id="cb47-191"><a href="#cb47-191" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>For the <span class="in">`names_to`</span> argument, we decided to call this new column 'measurement', as it contains both the gene name and replicate (so is a unique measurement of a particular gene in a particular replicate)</span>
<span id="cb47-192"><a href="#cb47-192" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>For the <span class="in">`values_to`</span> argument, we have called this new column 'expression level' as the numbers represent the gene expression in each of the different measurements</span>
<span id="cb47-193"><a href="#cb47-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-194"><a href="#cb47-194" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-195"><a href="#cb47-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-196"><a href="#cb47-196" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Perplexed by pivoting?"}</span>
<span id="cb47-197"><a href="#cb47-197" aria-hidden="true" tabindex="-1"></a>Pivoting can be a bit tricky to get your head around! Often when you're doing analysis, you'll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <span class="co">[</span><span class="ot">like those in the R for Data Science book</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html#sec-pivoting)</span>, ask an AI assistant or just try things until it works!</span>
<span id="cb47-198"><a href="#cb47-198" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-199"><a href="#cb47-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-200"><a href="#cb47-200" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb47-201"><a href="#cb47-201" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb47-202"><a href="#cb47-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-203"><a href="#cb47-203" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb47-204"><a href="#cb47-204" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What does the following code do?</span>
<span id="cb47-205"><a href="#cb47-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-208"><a href="#cb47-208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-209"><a href="#cb47-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb47-210"><a href="#cb47-210" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-211"><a href="#cb47-211" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb47-212"><a href="#cb47-212" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb47-213"><a href="#cb47-213" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb47-214"><a href="#cb47-214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span>
<span id="cb47-215"><a href="#cb47-215" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-216"><a href="#cb47-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-217"><a href="#cb47-217" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb47-218"><a href="#cb47-218" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-219"><a href="#cb47-219" aria-hidden="true" tabindex="-1"></a>Pivots data into a wide format where there is a column for each sex.</span>
<span id="cb47-220"><a href="#cb47-220" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-221"><a href="#cb47-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-222"><a href="#cb47-222" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb47-223"><a href="#cb47-223" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`dosage`</span> data frame, then pivots into a wide format where there is a column for each sex.</span>
<span id="cb47-224"><a href="#cb47-224" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-225"><a href="#cb47-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-226"><a href="#cb47-226" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-227"><a href="#cb47-227" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`dosage`</span> data frame, then pivots into a wide format where there is a column for each mouse strain.</span>
<span id="cb47-228"><a href="#cb47-228" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-229"><a href="#cb47-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-230"><a href="#cb47-230" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-231"><a href="#cb47-231" aria-hidden="true" tabindex="-1"></a>It just gives an error</span>
<span id="cb47-232"><a href="#cb47-232" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-233"><a href="#cb47-233" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-234"><a href="#cb47-234" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-235"><a href="#cb47-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-236"><a href="#cb47-236" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb47-237"><a href="#cb47-237" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> I have run the following code to create a new column in the <span class="in">`dosage`</span> data frame that gives the weight of the mice at the end of the experiment.</span>
<span id="cb47-238"><a href="#cb47-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-241"><a href="#cb47-241" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-242"><a href="#cb47-242" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-243"><a href="#cb47-243" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb47-244"><a href="#cb47-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb47-245"><a href="#cb47-245" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb47-246"><a href="#cb47-246" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span>
<span id="cb47-247"><a href="#cb47-247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-248"><a href="#cb47-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-249"><a href="#cb47-249" aria-hidden="true" tabindex="-1"></a>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</span>
<span id="cb47-250"><a href="#cb47-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-251"><a href="#cb47-251" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb47-252"><a href="#cb47-252" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-253"><a href="#cb47-253" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")`</span></span>
<span id="cb47-254"><a href="#cb47-254" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-255"><a href="#cb47-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-256"><a href="#cb47-256" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-257"><a href="#cb47-257" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")`</span></span>
<span id="cb47-258"><a href="#cb47-258" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-259"><a href="#cb47-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-260"><a href="#cb47-260" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-261"><a href="#cb47-261" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)`</span></span>
<span id="cb47-262"><a href="#cb47-262" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-263"><a href="#cb47-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-264"><a href="#cb47-264" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb47-265"><a href="#cb47-265" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span></span>
<span id="cb47-266"><a href="#cb47-266" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-267"><a href="#cb47-267" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-268"><a href="#cb47-268" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-269"><a href="#cb47-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-270"><a href="#cb47-270" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Using the <span class="in">`dosage`</span> data frame, write R code to:</span>
<span id="cb47-271"><a href="#cb47-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-272"><a href="#cb47-272" aria-hidden="true" tabindex="-1"></a>    a.  Make a data frame that shows the number of mice of each strain, in each replicate.</span>
<span id="cb47-273"><a href="#cb47-273" aria-hidden="true" tabindex="-1"></a>    b.  Pivot this data frame into a wide format to create a contingency table.</span>
<span id="cb47-274"><a href="#cb47-274" aria-hidden="true" tabindex="-1"></a>    c.  Pivot the wide data frame from (b) back into a long format.</span>
<span id="cb47-275"><a href="#cb47-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-276"><a href="#cb47-276" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-277"><a href="#cb47-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-278"><a href="#cb47-278" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solutions<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-279"><a href="#cb47-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-280"><a href="#cb47-280" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The code first calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`dosage`</span> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <span class="in">`names_from = sex`</span> )</span>
<span id="cb47-281"><a href="#cb47-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-282"><a href="#cb47-282" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> The correct pivot function call to take the data from a wide format to a long format is <span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span>. This code tells R to pivot the <span class="in">`initial_weight_g`</span> and <span class="in">`final_weight_g`</span> columns into a long format, where there is a row for each mouse and each weight measurement. The <span class="in">`names_to`</span> argument specifies to make a column called 'timepoint' that tells us whether the measurement is initial or final, and the <span class="in">`values_to`</span> argument specifies the name of the new column that will contain these measurements.</span>
<span id="cb47-283"><a href="#cb47-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-284"><a href="#cb47-284" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Here's how you could write R code to achieve the tasks:</span>
<span id="cb47-285"><a href="#cb47-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-286"><a href="#cb47-286" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb47-287"><a href="#cb47-287" aria-hidden="true" tabindex="-1"></a><span class="in">a.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate))`</span></span>
<span id="cb47-288"><a href="#cb47-288" aria-hidden="true" tabindex="-1"></a><span class="in">b.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice)`</span></span>
<span id="cb47-289"><a href="#cb47-289" aria-hidden="true" tabindex="-1"></a><span class="in">c.  `dosage %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice) %&gt;% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")`</span></span>
<span id="cb47-290"><a href="#cb47-290" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-291"><a href="#cb47-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-292"><a href="#cb47-292" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-293"><a href="#cb47-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-294"><a href="#cb47-294" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-295"><a href="#cb47-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-296"><a href="#cb47-296" aria-hidden="true" tabindex="-1"></a><span class="fu">### Separating data in a column {#sec-separate}</span></span>
<span id="cb47-297"><a href="#cb47-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-298"><a href="#cb47-298" aria-hidden="true" tabindex="-1"></a>Now that we've pivoted it into a long format, our <span class="in">`genes`</span> data certainly looks a lot better! But, it's not quite tidy yet; recall that our for data to be tidy, <span class="co">[</span><span class="ot">each column should contain only one variable</span><span class="co">](#sec-tidy-data)</span>. but our <span class="in">`measurement`</span> column actually contains **two** pieces of information: the gene being measured and the replicate number, which are separated by an underscore. In order to make our data tidy, we can use the <span class="in">`separate()`</span> function to split these into columns of their own.</span>
<span id="cb47-299"><a href="#cb47-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-300"><a href="#cb47-300" aria-hidden="true" tabindex="-1"></a>To use the separate function, we need to give it three arguments:</span>
<span id="cb47-301"><a href="#cb47-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-302"><a href="#cb47-302" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The name of the column that needs to be separated (in this case, <span class="in">`measurement`</span>)</span>
<span id="cb47-303"><a href="#cb47-303" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`into`</span>, which is a vector of the new column names that the data will be placed into</span>
<span id="cb47-304"><a href="#cb47-304" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span><span class="in">`sep`</span>, which is the character that separates the two pieces of data</span>
<span id="cb47-305"><a href="#cb47-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-308"><a href="#cb47-308" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-309"><a href="#cb47-309" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="ot">&lt;-</span> genes_long <span class="sc">%&gt;%</span></span>
<span id="cb47-310"><a href="#cb47-310" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(</span>
<span id="cb47-311"><a href="#cb47-311" aria-hidden="true" tabindex="-1"></a>      measurement, </span>
<span id="cb47-312"><a href="#cb47-312" aria-hidden="true" tabindex="-1"></a>      <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"replicate"</span>), </span>
<span id="cb47-313"><a href="#cb47-313" aria-hidden="true" tabindex="-1"></a>      <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb47-314"><a href="#cb47-314" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-315"><a href="#cb47-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-316"><a href="#cb47-316" aria-hidden="true" tabindex="-1"></a>We can now see that the gene name and replicate are contained within their own columns, officially making the data tidy!</span>
<span id="cb47-317"><a href="#cb47-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-320"><a href="#cb47-320" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-321"><a href="#cb47-321" aria-hidden="true" tabindex="-1"></a>genes_tidy</span>
<span id="cb47-322"><a href="#cb47-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-323"><a href="#cb47-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-324"><a href="#cb47-324" aria-hidden="true" tabindex="-1"></a><span class="fu">### Combining data with join functions {#sec-join}</span></span>
<span id="cb47-325"><a href="#cb47-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-326"><a href="#cb47-326" aria-hidden="true" tabindex="-1"></a>To address our <span class="co">[</span><span class="ot">second research question</span><span class="co">](#sec-research-questions)</span>, we need to look at the relationship between gene expression data (in the <span class="in">`genes`</span> data frame) and mouse strain (in the <span class="in">`dosage`</span> data frame). But how can we achieve this when the information is contained across two separate data frames? So long as we have some way of connecting the observations in each data frame (i.e. a shared column), we can use the <span class="in">`join`</span> family of functions to combine them.</span>
<span id="cb47-327"><a href="#cb47-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-328"><a href="#cb47-328" aria-hidden="true" tabindex="-1"></a>There are many different types of <span class="in">`join`</span> functions that you can use, but broadly they will either:</span>
<span id="cb47-329"><a href="#cb47-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-330"><a href="#cb47-330" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add new variables from one data frame to another</span>
<span id="cb47-331"><a href="#cb47-331" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Filter observations in one data frame based on whether or not they match observations in another data frame.</span>
<span id="cb47-332"><a href="#cb47-332" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Do both of these things at the same time</span>
<span id="cb47-333"><a href="#cb47-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-334"><a href="#cb47-334" aria-hidden="true" tabindex="-1"></a>Probably the most commonly used is the <span class="in">`left_join()`</span> function. You need to provide it with two pieces of information:</span>
<span id="cb47-335"><a href="#cb47-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-336"><a href="#cb47-336" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The two data frames you want to join</span>
<span id="cb47-337"><a href="#cb47-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-338"><a href="#cb47-338" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The name of a column (or columns) to join by</span>
<span id="cb47-339"><a href="#cb47-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-340"><a href="#cb47-340" aria-hidden="true" tabindex="-1"></a>The reason it's called a left-join is because it retains all rows from the left data frame while adding on columns from the right data frame only when the column you're joining by matches.</span>
<span id="cb47-341"><a href="#cb47-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-342"><a href="#cb47-342" aria-hidden="true" tabindex="-1"></a>For example, in the following cartoon, where we are joining by column A:</span>
<span id="cb47-343"><a href="#cb47-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-344"><a href="#cb47-344" aria-hidden="true" tabindex="-1"></a><span class="al">![The left_join() function allows you to combine two data frames](images/S04F03_left_join.png)</span></span>
<span id="cb47-345"><a href="#cb47-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-346"><a href="#cb47-346" aria-hidden="true" tabindex="-1"></a>We can see that:</span>
<span id="cb47-347"><a href="#cb47-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-348"><a href="#cb47-348" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A new column, D has been added from the right data frame</span>
<span id="cb47-349"><a href="#cb47-349" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All of the observations from the left data frame are present in the result</span>
<span id="cb47-350"><a href="#cb47-350" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The first row of the result contains an <span class="in">`NA`</span> in the D column, because in the right data frame that the D column comes from, there is no matching observation 'a'</span>
<span id="cb47-351"><a href="#cb47-351" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The last row from the right data frame is missing from the result, because it has no matching value of 'd' in the left data frame</span>
<span id="cb47-352"><a href="#cb47-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-353"><a href="#cb47-353" aria-hidden="true" tabindex="-1"></a>::: {.callout-warning title="Beware multiple matches!"}</span>
<span id="cb47-354"><a href="#cb47-354" aria-hidden="true" tabindex="-1"></a>If the values from the left data frame match to multiple rows of the column in the right data frame, the <span class="in">`left_join()`</span> will duplicate the data from the left data frame for each match to the right. This can cause issues with downstream summarisation if not carefully considered.</span>
<span id="cb47-355"><a href="#cb47-355" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-356"><a href="#cb47-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-357"><a href="#cb47-357" aria-hidden="true" tabindex="-1"></a>Using this knowledge, let's try joining our <span class="in">`dosage`</span> and <span class="in">`genes_tidy`</span> data frames based on their shared column, <span class="in">`id_num`</span> (an ID number uniquely assigned to each mouse):</span>
<span id="cb47-358"><a href="#cb47-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-361"><a href="#cb47-361" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-362"><a href="#cb47-362" aria-hidden="true" tabindex="-1"></a><span class="co"># let's save the result of this as a variable</span></span>
<span id="cb47-363"><a href="#cb47-363" aria-hidden="true" tabindex="-1"></a>combined_data <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb47-364"><a href="#cb47-364" aria-hidden="true" tabindex="-1"></a>    genes_tidy, <span class="co"># the left data frame</span></span>
<span id="cb47-365"><a href="#cb47-365" aria-hidden="true" tabindex="-1"></a>    dosage, <span class="co"># the right data frame</span></span>
<span id="cb47-366"><a href="#cb47-366" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id_num"</span>) <span class="co"># name of the column to join by</span></span>
<span id="cb47-367"><a href="#cb47-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-368"><a href="#cb47-368" aria-hidden="true" tabindex="-1"></a>combined_data</span>
<span id="cb47-369"><a href="#cb47-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-370"><a href="#cb47-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-371"><a href="#cb47-371" aria-hidden="true" tabindex="-1"></a>We now have the gene expression information and the dosage/metadata for each mouse all together in a single data frame. Notice how all of the 3,087 rows in the <span class="in">`genes_tidy`</span> data frame are present in the <span class="in">`combined_data`</span>, and the relevant information from the <span class="in">`dosage`</span> data frame has been joined where available.</span>
<span id="cb47-372"><a href="#cb47-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-373"><a href="#cb47-373" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Two replicate columns?"}</span>
<span id="cb47-374"><a href="#cb47-374" aria-hidden="true" tabindex="-1"></a>One other thing you might have noticed is that there are now two replicate columns: <span class="in">`replicate.x`</span> and <span class="in">`replicate.y`</span>. This is because both <span class="in">`genes_tidy`</span> and <span class="in">`dosage`</span> each contained their own replicate column, but we only told <span class="in">`left_join()`</span> to match up our data by <span class="in">`id_num`</span>. Since you can't have two columns with the same name, R has added the <span class="in">`.x`</span> and <span class="in">`.y`</span> to differentiate them. If you want R to use more informative names than just <span class="in">`.x`</span> and <span class="in">`.y`</span> , you can use the <span class="in">`suffix`</span> argument:</span>
<span id="cb47-375"><a href="#cb47-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-378"><a href="#cb47-378" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-379"><a href="#cb47-379" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(</span>
<span id="cb47-380"><a href="#cb47-380" aria-hidden="true" tabindex="-1"></a>    genes_tidy, <span class="co"># the left data frame</span></span>
<span id="cb47-381"><a href="#cb47-381" aria-hidden="true" tabindex="-1"></a>    dosage, <span class="co"># the right data frame</span></span>
<span id="cb47-382"><a href="#cb47-382" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id_num"</span>, <span class="co"># name of the column to join by</span></span>
<span id="cb47-383"><a href="#cb47-383" aria-hidden="true" tabindex="-1"></a>    <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">"_genes"</span>, <span class="st">"_dosage"</span>))</span>
<span id="cb47-384"><a href="#cb47-384" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-385"><a href="#cb47-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-386"><a href="#cb47-386" aria-hidden="true" tabindex="-1"></a>It is now clearer which original dataset each replicate column comes from.</span>
<span id="cb47-387"><a href="#cb47-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-388"><a href="#cb47-388" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-389"><a href="#cb47-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-390"><a href="#cb47-390" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Joining with mismatched column names</span></span>
<span id="cb47-391"><a href="#cb47-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-392"><a href="#cb47-392" aria-hidden="true" tabindex="-1"></a>Sometimes, the column containing the matching information you want to join over will have different names in different data frames. For example what might be called "id_num" in one data frame could be called "mouse_id" in another data frame. In this case, the <span class="in">`by`</span> argument of <span class="in">`left_join()`</span> can be formatted using the <span class="in">`join_by(name1 == name2)`</span> helper function to specify which column in the left data frame matches which column on the right.</span>
<span id="cb47-393"><a href="#cb47-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-394"><a href="#cb47-394" aria-hidden="true" tabindex="-1"></a>As an example, we'll rename the <span class="in">`id_num`</span> column in the <span class="in">`dosage`</span> data frame to <span class="in">`mouse_id`</span> and repeat the join to <span class="in">`genes_tidy`</span>:</span>
<span id="cb47-395"><a href="#cb47-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-398"><a href="#cb47-398" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-399"><a href="#cb47-399" aria-hidden="true" tabindex="-1"></a><span class="co"># rename id_num to mouse_id</span></span>
<span id="cb47-400"><a href="#cb47-400" aria-hidden="true" tabindex="-1"></a>renamed_dosage <span class="ot">&lt;-</span> dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-401"><a href="#cb47-401" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we can use the rename() function to rename a column</span></span>
<span id="cb47-402"><a href="#cb47-402" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">mouse_id =</span> id_num)</span>
<span id="cb47-403"><a href="#cb47-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-404"><a href="#cb47-404" aria-hidden="true" tabindex="-1"></a><span class="co"># now perform the left join</span></span>
<span id="cb47-405"><a href="#cb47-405" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(</span>
<span id="cb47-406"><a href="#cb47-406" aria-hidden="true" tabindex="-1"></a>  genes_tidy, <span class="co"># left data frame </span></span>
<span id="cb47-407"><a href="#cb47-407" aria-hidden="true" tabindex="-1"></a>  renamed_dosage, <span class="co"># right data frame</span></span>
<span id="cb47-408"><a href="#cb47-408" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join by the id_num column in genes_tidy, which is equivalent to the mouse_id column in renamed_dosage</span></span>
<span id="cb47-409"><a href="#cb47-409" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">join_by</span>(id_num <span class="sc">==</span> mouse_id)) </span>
<span id="cb47-410"><a href="#cb47-410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-411"><a href="#cb47-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-412"><a href="#cb47-412" aria-hidden="true" tabindex="-1"></a>Notice how the column name from the left data frame (in this case, <span class="in">`id_num`</span>) is retained in the joined result.</span>
<span id="cb47-413"><a href="#cb47-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-414"><a href="#cb47-414" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb47-415"><a href="#cb47-415" aria-hidden="true" tabindex="-1"></a>Note that an older syntax exists for the <span class="in">`by`</span> argument that would look like <span class="in">`by = c("mouse_id" = "id_num")`</span>. You might encounter this in older books/posts/examples online, or when using AI assistants like ChatGPT, but we recommend you stick to the <span class="in">`join_by()`</span> syntax instead, as it's easier to understand.</span>
<span id="cb47-416"><a href="#cb47-416" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-417"><a href="#cb47-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-418"><a href="#cb47-418" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb47-419"><a href="#cb47-419" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb47-420"><a href="#cb47-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-421"><a href="#cb47-421" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb47-422"><a href="#cb47-422" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which of the following is NOT a valid way to join the <span class="in">`dosage`</span> data frame with the <span class="in">`genes`</span> data frame based on the <span class="in">`id_num`</span> column?</span>
<span id="cb47-423"><a href="#cb47-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-424"><a href="#cb47-424" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb47-425"><a href="#cb47-425" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-426"><a href="#cb47-426" aria-hidden="true" tabindex="-1"></a><span class="in">`dosage %&gt;% left_join(genes, by = "id_num")`</span></span>
<span id="cb47-427"><a href="#cb47-427" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-428"><a href="#cb47-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-429"><a href="#cb47-429" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-430"><a href="#cb47-430" aria-hidden="true" tabindex="-1"></a><span class="in">`left_join(dosage, genes, by = "id_num")`</span></span>
<span id="cb47-431"><a href="#cb47-431" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-432"><a href="#cb47-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-433"><a href="#cb47-433" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb47-434"><a href="#cb47-434" aria-hidden="true" tabindex="-1"></a><span class="in">`dosage %&gt;% left_join(genes, .by = "id_num")`</span></span>
<span id="cb47-435"><a href="#cb47-435" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-436"><a href="#cb47-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-437"><a href="#cb47-437" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb47-438"><a href="#cb47-438" aria-hidden="true" tabindex="-1"></a><span class="in">`dosage %&gt;% left_join(genes, by = ("id_num" = "id_num"))`</span></span>
<span id="cb47-439"><a href="#cb47-439" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-440"><a href="#cb47-440" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-441"><a href="#cb47-441" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-442"><a href="#cb47-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-443"><a href="#cb47-443" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Let's say I have two data frames, <span class="in">`df1`</span> and <span class="in">`df2`</span>, that I want to join based a shared 'key' column, that is called 'key' in <span class="in">`df1`</span> and 'item_key' in <span class="in">`df2`</span>. Write R code to join these two data frames using the <span class="in">`left_join()`</span> function.</span>
<span id="cb47-444"><a href="#cb47-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-445"><a href="#cb47-445" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-446"><a href="#cb47-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-447"><a href="#cb47-447" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solutions<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-448"><a href="#cb47-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-449"><a href="#cb47-449" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The line of code that is NOT a valid way to join the <span class="in">`dosage`</span> data frame with the <span class="in">`genes`</span> data frame based on the <span class="in">`id_num`</span> column is <span class="in">`dosage %&gt;% left_join(genes, .by = "id_num")`</span>. This line of code is incorrect because the <span class="in">`.by`</span> argument is not used in the <span class="in">`left_join()`</span> function (this can be confusing! it's <span class="in">`.by`</span> when grouping by <span class="in">`by`</span> when joining). The other options are valid ways to join the two data frames based on the <span class="in">`id_num`</span> column: remember that we don't have to use pipes to join data frames, we can use the <span class="in">`left_join()`</span> function directly, and we can use a named vector to specify the columns to join on (although here it's a bit redundant as the columns have the same name).</span>
<span id="cb47-450"><a href="#cb47-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-451"><a href="#cb47-451" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> To join the two data frames you could use <span class="in">`df1 %&gt;% left_join(df2, by = c("key" = "item_key"))`</span> (with pipe) or <span class="in">`left_join(df1, df2, by = c("key" = "item_key"))`</span> (without pipe).</span>
<span id="cb47-452"><a href="#cb47-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-453"><a href="#cb47-453" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-454"><a href="#cb47-454" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-455"><a href="#cb47-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-456"><a href="#cb47-456" aria-hidden="true" tabindex="-1"></a><span class="fu">## The process of data analysis</span></span>
<span id="cb47-457"><a href="#cb47-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-458"><a href="#cb47-458" aria-hidden="true" tabindex="-1"></a>With all that data wrangling out of the way, we can finally move on to doing some analysis to answer our <span class="co">[</span><span class="ot">research questions</span><span class="co">](#sec-research-questions)</span>! But wait— how do you actually *do* data analysis? Where should you start?</span>
<span id="cb47-459"><a href="#cb47-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-460"><a href="#cb47-460" aria-hidden="true" tabindex="-1"></a>Well, the creators of the tidyverse packages we use in this course have proposed the following 'process of data science':</span>
<span id="cb47-461"><a href="#cb47-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-462"><a href="#cb47-462" aria-hidden="true" tabindex="-1"></a><span class="al">![The 'process of data science', according to the tidyverse (source: &lt;https://r4ds.hadley.nz/intro.html#fig-ds-diagram&gt;)](images/S04F04_analysis_steps.png)</span></span>
<span id="cb47-463"><a href="#cb47-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-464"><a href="#cb47-464" aria-hidden="true" tabindex="-1"></a>In fact, we've already done the first two steps: reading in our data and tidying it up! The main portion of the analysis however is depicted as a cycle: transforming your data with the dplyr functions we learned in @sec-session02, visualising it with the ggplot2 functions we learned in @sec-session03, modelling it (with statistics, which we don't cover in this course as they're quite field-specific) and repeating. This process of iterating over many different research/analysis questions to develop a deeper understanding of your data allows you to really explore every aspect. Finally, you need to be able to clearly communicate the results of your analysis and interpret them in the biological context!</span>
<span id="cb47-465"><a href="#cb47-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-466"><a href="#cb47-466" aria-hidden="true" tabindex="-1"></a>As we go through the research questions in the rest of this session, keep these steps in mind and try to think about how you could communicate the results.</span>
<span id="cb47-467"><a href="#cb47-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-468"><a href="#cb47-468" aria-hidden="true" tabindex="-1"></a><span class="fu">## Question 1: The relationship between dosage and weight loss</span></span>
<span id="cb47-469"><a href="#cb47-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-470"><a href="#cb47-470" aria-hidden="true" tabindex="-1"></a>Let's start by tackling the first part of our first research question: Does the amount of weight lost by mice increase as dosage of mousezempic increases?</span>
<span id="cb47-471"><a href="#cb47-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-472"><a href="#cb47-472" aria-hidden="true" tabindex="-1"></a>Luckily, our <span class="in">`dosage`</span> data was pretty tidy already so we didn't have to do much cleaning. To begin our analysis process, let's try investigating the relationship between these variables with visualisation. As both the dosage and weight loss variables are continuous, we can represent them on a scatterplot using <span class="in">`geom_point()`</span></span>
<span id="cb47-473"><a href="#cb47-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-476"><a href="#cb47-476" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-477"><a href="#cb47-477" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb47-478"><a href="#cb47-478" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dosage, <span class="fu">aes</span>(<span class="at">x =</span> drug_dose_g, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-479"><a href="#cb47-479" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb47-480"><a href="#cb47-480" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-481"><a href="#cb47-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-482"><a href="#cb47-482" aria-hidden="true" tabindex="-1"></a>We can clearly see a positive linear relationship between drug dosage and weight lost.</span>
<span id="cb47-483"><a href="#cb47-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-484"><a href="#cb47-484" aria-hidden="true" tabindex="-1"></a>However, this plot certainly isn't publication ready! Practice the R skills you have learned throughout the course so far and try to improve the look of the plot by working through the exercises below ⬇️</span>
<span id="cb47-485"><a href="#cb47-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-486"><a href="#cb47-486" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercise: changing the x axis"}</span>
<span id="cb47-487"><a href="#cb47-487" aria-hidden="true" tabindex="-1"></a>For starters, it would probably make more sense to display the drug doses in milligrams rather than in grams, given that the values are so small. Take a moment to think about how you might achieve this using the dplyr functions we learned in session 2, and then check your answer below.</span>
<span id="cb47-488"><a href="#cb47-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-489"><a href="#cb47-489" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-490"><a href="#cb47-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-491"><a href="#cb47-491" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-492"><a href="#cb47-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-493"><a href="#cb47-493" aria-hidden="true" tabindex="-1"></a>To convert g to mg we need to multiply by 1000. So, we need some kind of function that allows us to take the existing <span class="in">`drug_dose_g`</span> column and multiply it by 1000 to create a new column that we can use for our plotting.</span>
<span id="cb47-494"><a href="#cb47-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-495"><a href="#cb47-495" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Use the <span class="in">`mutate()`</span> function to create a new column in the <span class="in">`dosage`</span> data frame called (for example) <span class="in">`drug_dose_mg`</span></span>
<span id="cb47-496"><a href="#cb47-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-497"><a href="#cb47-497" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Either save the modified <span class="in">`dosage`</span> data frame as a variable in R, or pipe it directly into the ggplot function call</span>
<span id="cb47-498"><a href="#cb47-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-499"><a href="#cb47-499" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Make the ggplot as before</span>
<span id="cb47-500"><a href="#cb47-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-501"><a href="#cb47-501" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-502"><a href="#cb47-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-503"><a href="#cb47-503" aria-hidden="true" tabindex="-1"></a>Learning how to break a problem down into simple steps is a crucial skill to learn for data analysis! Even if there are some steps that you don't immediately know how to solve with code, it gives you a good starting point for your googling or to ask ChatGPT.</span>
<span id="cb47-504"><a href="#cb47-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-505"><a href="#cb47-505" aria-hidden="true" tabindex="-1"></a>Now, try to convert these steps into code and create a plot where the x axis displays the drug dose in mg, like this one:</span>
<span id="cb47-506"><a href="#cb47-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-509"><a href="#cb47-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-510"><a href="#cb47-510" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb47-511"><a href="#cb47-511" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-512"><a href="#cb47-512" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-513"><a href="#cb47-513" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-514"><a href="#cb47-514" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span>
<span id="cb47-515"><a href="#cb47-515" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-516"><a href="#cb47-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-517"><a href="#cb47-517" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-518"><a href="#cb47-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-519"><a href="#cb47-519" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-520"><a href="#cb47-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-521"><a href="#cb47-521" aria-hidden="true" tabindex="-1"></a>Following the steps outlined above, here are two example code chunks that will make the correct plot. There are of course many more ways to approach the problem, so it's ok if your solution doesn't match perfectly!</span>
<span id="cb47-522"><a href="#cb47-522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-523"><a href="#cb47-523" aria-hidden="true" tabindex="-1"></a>This first solution saves the modified dosage data as a new data frame, which could be useful if you want to use the <span class="in">`drug_dose_mg`</span> column in other plots or calculations:</span>
<span id="cb47-524"><a href="#cb47-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-527"><a href="#cb47-527" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-528"><a href="#cb47-528" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb47-529"><a href="#cb47-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-530"><a href="#cb47-530" aria-hidden="true" tabindex="-1"></a><span class="co"># step two: save the modified version of dosage as its own </span></span>
<span id="cb47-531"><a href="#cb47-531" aria-hidden="true" tabindex="-1"></a>dosage_mg <span class="ot">&lt;-</span> dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-532"><a href="#cb47-532" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step one: use mutate to create drug_dose_mg column</span></span>
<span id="cb47-533"><a href="#cb47-533" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb47-534"><a href="#cb47-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-535"><a href="#cb47-535" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step three: plot same as before</span></span>
<span id="cb47-536"><a href="#cb47-536" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(dosage_mg, <span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-537"><a href="#cb47-537" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span>
<span id="cb47-538"><a href="#cb47-538" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-539"><a href="#cb47-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-540"><a href="#cb47-540" aria-hidden="true" tabindex="-1"></a>The second one instead pipes this data frame directly from the <span class="in">`mutate()`</span> function call straight into the <span class="in">`ggplot()`</span> function. This approach is convenient when the data manipulation you're doing (in this case adding the <span class="in">`drug_dose_mg`</span> column) is only for a specific plot, so you don't clutter up your environment panel:</span>
<span id="cb47-541"><a href="#cb47-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-544"><a href="#cb47-544" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-545"><a href="#cb47-545" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb47-546"><a href="#cb47-546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-547"><a href="#cb47-547" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-548"><a href="#cb47-548" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step one: use mutate to create drug_dose_mg column</span></span>
<span id="cb47-549"><a href="#cb47-549" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-550"><a href="#cb47-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-551"><a href="#cb47-551" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step two: pipe directly into the ggplot() function</span></span>
<span id="cb47-552"><a href="#cb47-552" aria-hidden="true" tabindex="-1"></a>  <span class="co"># note that because the pipe passes our modified dosage data in as the first argument,</span></span>
<span id="cb47-553"><a href="#cb47-553" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we don't need to specify the data frame we are plotting, just the aes() </span></span>
<span id="cb47-554"><a href="#cb47-554" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-555"><a href="#cb47-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-556"><a href="#cb47-556" aria-hidden="true" tabindex="-1"></a>  <span class="co"># step three: plot same as before </span></span>
<span id="cb47-557"><a href="#cb47-557" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span>
<span id="cb47-558"><a href="#cb47-558" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-559"><a href="#cb47-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-560"><a href="#cb47-560" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-561"><a href="#cb47-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-562"><a href="#cb47-562" aria-hidden="true" tabindex="-1"></a>Now our x axis is much easier to read!</span>
<span id="cb47-563"><a href="#cb47-563" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-564"><a href="#cb47-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-565"><a href="#cb47-565" aria-hidden="true" tabindex="-1"></a>Still, there are a few things we could do to make our plot look a bit better. In the next set of practice problems, we will use our ggplot2 knowledge to improve the plot visually:</span>
<span id="cb47-566"><a href="#cb47-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-567"><a href="#cb47-567" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercise: making visual improvements"}</span>
<span id="cb47-568"><a href="#cb47-568" aria-hidden="true" tabindex="-1"></a>Take a moment to look at the plot and think of what you could change to make it more informative and visually appealing. Of course this is subjective, but some suggestions are provided below.</span>
<span id="cb47-569"><a href="#cb47-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-570"><a href="#cb47-570" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-571"><a href="#cb47-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-572"><a href="#cb47-572" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-573"><a href="#cb47-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-574"><a href="#cb47-574" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The plot could do with some nicely formatted labels on the x and y axes, as well as a title.</span>
<span id="cb47-575"><a href="#cb47-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-576"><a href="#cb47-576" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Some of the points on the plot are overlapping, so it could be helpful to change the points from filled circles to unfilled circles so they don't all just merge into a blob.</span>
<span id="cb47-577"><a href="#cb47-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-578"><a href="#cb47-578" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> This is a matter of personal taste, but I think the default ggplot2 grey background is kind of basic so I always change it 😝</span>
<span id="cb47-579"><a href="#cb47-579" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-580"><a href="#cb47-580" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-581"><a href="#cb47-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-582"><a href="#cb47-582" aria-hidden="true" tabindex="-1"></a>Here's my plot. Click below to see the R code that made it:</span>
<span id="cb47-583"><a href="#cb47-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-586"><a href="#cb47-586" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-587"><a href="#cb47-587" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb47-588"><a href="#cb47-588" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Show the R code"</span></span>
<span id="cb47-589"><a href="#cb47-589" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb47-590"><a href="#cb47-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-591"><a href="#cb47-591" aria-hidden="true" tabindex="-1"></a><span class="co"># see the previous practice q for the explanation of this</span></span>
<span id="cb47-592"><a href="#cb47-592" aria-hidden="true" tabindex="-1"></a><span class="co"># dplyr code to create the mg column</span></span>
<span id="cb47-593"><a href="#cb47-593" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-594"><a href="#cb47-594" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-595"><a href="#cb47-595" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create the plot</span></span>
<span id="cb47-596"><a href="#cb47-596" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-597"><a href="#cb47-597" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the points to unfilled circles</span></span>
<span id="cb47-598"><a href="#cb47-598" aria-hidden="true" tabindex="-1"></a>    <span class="co"># note that we don't need to put it in aes() because</span></span>
<span id="cb47-599"><a href="#cb47-599" aria-hidden="true" tabindex="-1"></a>    <span class="co"># it isn't tied to a particular variable</span></span>
<span id="cb47-600"><a href="#cb47-600" aria-hidden="true" tabindex="-1"></a>    <span class="co"># also increasing the stroke (width of the circle line) to 0.75</span></span>
<span id="cb47-601"><a href="#cb47-601" aria-hidden="true" tabindex="-1"></a>    <span class="co"># so the points stand out more</span></span>
<span id="cb47-602"><a href="#cb47-602" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">1</span>, <span class="at">stroke =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb47-603"><a href="#cb47-603" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add axis labels and title</span></span>
<span id="cb47-604"><a href="#cb47-604" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Association between drug dosage and weight lost"</span>,</span>
<span id="cb47-605"><a href="#cb47-605" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Drug dosage (mg)"</span>,</span>
<span id="cb47-606"><a href="#cb47-606" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Weight lost (g)"</span>) <span class="sc">+</span></span>
<span id="cb47-607"><a href="#cb47-607" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the theme</span></span>
<span id="cb47-608"><a href="#cb47-608" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb47-609"><a href="#cb47-609" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-610"><a href="#cb47-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-611"><a href="#cb47-611" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-612"><a href="#cb47-612" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-613"><a href="#cb47-613" aria-hidden="true" tabindex="-1"></a><span class="fu">### The effect of strain and sex</span></span>
<span id="cb47-614"><a href="#cb47-614" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-615"><a href="#cb47-615" aria-hidden="true" tabindex="-1"></a>The final part of the research question was to investigate whether mouse strain or sex has any effect on this relationship. Above, we learned how to use the <span class="in">`summarise()`</span> and <span class="in">`pivot_wider()`</span> functions to create a contingency table that counts up how many mice fall into each of the mouse strain/sex combinations. This is an important first step for the analysis, as it allows us to double check that there isn't a category with so few mice that would prevent us from drawing reliable conclusions.</span>
<span id="cb47-616"><a href="#cb47-616" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-619"><a href="#cb47-619" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-620"><a href="#cb47-620" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-621"><a href="#cb47-621" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n_mice =</span> <span class="fu">n</span>(), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb47-622"><a href="#cb47-622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb47-623"><a href="#cb47-623" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-624"><a href="#cb47-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-625"><a href="#cb47-625" aria-hidden="true" tabindex="-1"></a>Given that there are &gt;30 mice in each category, we can proceed with the analysis.</span>
<span id="cb47-626"><a href="#cb47-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-627"><a href="#cb47-627" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercise: adding in extra information"}</span>
<span id="cb47-628"><a href="#cb47-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-629"><a href="#cb47-629" aria-hidden="true" tabindex="-1"></a>Now, let's try adding the mouse strain and sex information to the scatterplot we created above. Try to think about what aesthetics would be good to use to show those two variables.</span>
<span id="cb47-630"><a href="#cb47-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-631"><a href="#cb47-631" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-632"><a href="#cb47-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-633"><a href="#cb47-633" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-634"><a href="#cb47-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-637"><a href="#cb47-637" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-638"><a href="#cb47-638" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb47-639"><a href="#cb47-639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-640"><a href="#cb47-640" aria-hidden="true" tabindex="-1"></a><span class="co"># see the previous practice q for the explanation of this</span></span>
<span id="cb47-641"><a href="#cb47-641" aria-hidden="true" tabindex="-1"></a><span class="co"># dplyr code to create the mg column</span></span>
<span id="cb47-642"><a href="#cb47-642" aria-hidden="true" tabindex="-1"></a>dosage <span class="sc">%&gt;%</span></span>
<span id="cb47-643"><a href="#cb47-643" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-644"><a href="#cb47-644" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create the plot</span></span>
<span id="cb47-645"><a href="#cb47-645" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_mg, <span class="at">y =</span> weight_lost_g, <span class="at">colour =</span> mouse_strain, <span class="at">shape =</span> sex)) <span class="sc">+</span></span>
<span id="cb47-646"><a href="#cb47-646" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb47-647"><a href="#cb47-647" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add axis labels and title</span></span>
<span id="cb47-648"><a href="#cb47-648" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Association between drug dosage and weight lost"</span>,</span>
<span id="cb47-649"><a href="#cb47-649" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Drug dosage (mg)"</span>,</span>
<span id="cb47-650"><a href="#cb47-650" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Weight lost (g)"</span>,</span>
<span id="cb47-651"><a href="#cb47-651" aria-hidden="true" tabindex="-1"></a>       <span class="co"># adding nicer labels to the legend</span></span>
<span id="cb47-652"><a href="#cb47-652" aria-hidden="true" tabindex="-1"></a>       <span class="co"># you just use the aesthetic name </span></span>
<span id="cb47-653"><a href="#cb47-653" aria-hidden="true" tabindex="-1"></a>       <span class="at">colour =</span> <span class="st">"Mouse strain"</span>,</span>
<span id="cb47-654"><a href="#cb47-654" aria-hidden="true" tabindex="-1"></a>       <span class="at">shape =</span> <span class="st">"Sex"</span>) <span class="sc">+</span></span>
<span id="cb47-655"><a href="#cb47-655" aria-hidden="true" tabindex="-1"></a>    <span class="co"># change the theme</span></span>
<span id="cb47-656"><a href="#cb47-656" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb47-657"><a href="#cb47-657" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-658"><a href="#cb47-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-659"><a href="#cb47-659" aria-hidden="true" tabindex="-1"></a>Note that:</span>
<span id="cb47-660"><a href="#cb47-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-661"><a href="#cb47-661" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>It's up to you which aesthetics should represent each variable, but in this example the <span class="in">`colour`</span> aesthetic was chosen for the mouse strains because there are more strains than sexes, and it's easier to distinguish between different colours than different shapes</span>
<span id="cb47-662"><a href="#cb47-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-663"><a href="#cb47-663" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>We used the <span class="in">`labs()`</span> function to change the legend labels for our mouse strain and sex variables. It just gives a nice professional touch to the plot</span>
<span id="cb47-664"><a href="#cb47-664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-665"><a href="#cb47-665" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-666"><a href="#cb47-666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-667"><a href="#cb47-667" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-668"><a href="#cb47-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-669"><a href="#cb47-669" aria-hidden="true" tabindex="-1"></a>Our plot reveals that Black 6 mice received higher drug dosages than the other mice, and also lost more weight. This shows us why it's so important to explore the relationships between lots of different variables in large datasets. Imagine if we had just plotted the weight loss for mice of each strain:</span>
<span id="cb47-670"><a href="#cb47-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-673"><a href="#cb47-673" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-674"><a href="#cb47-674" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb47-675"><a href="#cb47-675" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dosage, <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span> </span>
<span id="cb47-676"><a href="#cb47-676" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb47-677"><a href="#cb47-677" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb47-678"><a href="#cb47-678" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-679"><a href="#cb47-679" aria-hidden="true" tabindex="-1"></a>We might think that the drug is more effective on Black 6 mice, because they lost more weight, but our scatterplot above shows it's more likely to be due to their higher dosage.</span>
<span id="cb47-680"><a href="#cb47-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-681"><a href="#cb47-681" aria-hidden="true" tabindex="-1"></a>Always try to look at all the variables you have, because you never know what associations you might find!</span>
<span id="cb47-682"><a href="#cb47-682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-683"><a href="#cb47-683" aria-hidden="true" tabindex="-1"></a><span class="fu">## Question 2: The effect of mousezempic on gene expression</span></span>
<span id="cb47-684"><a href="#cb47-684" aria-hidden="true" tabindex="-1"></a>Our second research question was about the effect of mousezempic on the expression of the TH and PRLH genes. To answer this question, we can use the <span class="in">`genes_tidy`</span> dataset that we prepared in @sec-reshaping, by pivoting the <span class="in">`genes`</span> dataset to be longer and separating the measurement column so each variable is contained within its own column.</span>
<span id="cb47-685"><a href="#cb47-685" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-686"><a href="#cb47-686" aria-hidden="true" tabindex="-1"></a>Let's take a look at the <span class="in">`genes_tidy`</span> data to remind ourselves of the information it contains. You might like to use the <span class="in">`View()`</span> function, or just type the name of the data like so:</span>
<span id="cb47-687"><a href="#cb47-687" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-690"><a href="#cb47-690" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-691"><a href="#cb47-691" aria-hidden="true" tabindex="-1"></a>genes_tidy</span>
<span id="cb47-692"><a href="#cb47-692" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-693"><a href="#cb47-693" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-694"><a href="#cb47-694" aria-hidden="true" tabindex="-1"></a>Each mouse has a unique <span class="in">`id_num`</span>, and belongs to a <span class="in">`group`</span> (either treatment or control). There are also gene expression measurements in triplicate for each of three different genes: the TH and PRLH genes we are interested in and HPRT1, a housekeeping gene.</span>
<span id="cb47-695"><a href="#cb47-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-696"><a href="#cb47-696" aria-hidden="true" tabindex="-1"></a><span class="fu">### Do the treatment and control groups differ?</span></span>
<span id="cb47-697"><a href="#cb47-697" aria-hidden="true" tabindex="-1"></a>To investigate differences in gene expression between the treatment an control groups, a good first step would be to make a plot. Boxplots, violin plots or beeswarm plots would be a good choice to visualise this type of data.</span>
<span id="cb47-698"><a href="#cb47-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-699"><a href="#cb47-699" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercise: plotting the relationship between mousezempic and gene expression"}</span>
<span id="cb47-700"><a href="#cb47-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-701"><a href="#cb47-701" aria-hidden="true" tabindex="-1"></a>Take a moment to think about how you could achieve this, and check your answer below. As an additional exercise, how could you prevent the HPRT1 gene from being shown on your plot?</span>
<span id="cb47-702"><a href="#cb47-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-703"><a href="#cb47-703" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-704"><a href="#cb47-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-705"><a href="#cb47-705" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-706"><a href="#cb47-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-707"><a href="#cb47-707" aria-hidden="true" tabindex="-1"></a>Since we want to plot the expression of two different genes, we have a couple of options.</span>
<span id="cb47-708"><a href="#cb47-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-709"><a href="#cb47-709" aria-hidden="true" tabindex="-1"></a>Firstly, we could use a single <span class="in">`ggplot()`</span> function call, and facet the plot by the gene. However, if we want to exlude the HPRT1 gene, we'll have to first use the <span class="in">`filter()`</span> function from dplyr to remove it from our data frame.</span>
<span id="cb47-710"><a href="#cb47-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-713"><a href="#cb47-713" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-714"><a href="#cb47-714" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-715"><a href="#cb47-715" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that != is the test for inequality in R</span></span>
<span id="cb47-716"><a href="#cb47-716" aria-hidden="true" tabindex="-1"></a>  <span class="co"># also note we use Hprt1 not HPRT1: needs to match </span></span>
<span id="cb47-717"><a href="#cb47-717" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the data frame exactly for filter() to work</span></span>
<span id="cb47-718"><a href="#cb47-718" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-719"><a href="#cb47-719" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we will pipe this filtered data into the ggplot</span></span>
<span id="cb47-720"><a href="#cb47-720" aria-hidden="true" tabindex="-1"></a>  <span class="co"># because we are unlikely to want to re-use it</span></span>
<span id="cb47-721"><a href="#cb47-721" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-722"><a href="#cb47-722" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb47-723"><a href="#cb47-723" aria-hidden="true" tabindex="-1"></a>    <span class="co"># to facet write ~ then the name of the column</span></span>
<span id="cb47-724"><a href="#cb47-724" aria-hidden="true" tabindex="-1"></a>    <span class="co"># need scales = "free" because otherwise it will</span></span>
<span id="cb47-725"><a href="#cb47-725" aria-hidden="true" tabindex="-1"></a>    <span class="co"># use the same scales for both genes, and the </span></span>
<span id="cb47-726"><a href="#cb47-726" aria-hidden="true" tabindex="-1"></a>    <span class="co"># expression levels are quite different</span></span>
<span id="cb47-727"><a href="#cb47-727" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb47-728"><a href="#cb47-728" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb47-729"><a href="#cb47-729" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-730"><a href="#cb47-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-731"><a href="#cb47-731" aria-hidden="true" tabindex="-1"></a>Another option would be to plot each of TH and PRLH separately, with two different <span class="in">`ggplot()`</span> function calls and then combine them with patchwork (make sure this package is loaded!). To do this, we would also have to use the <span class="in">`filter()`</span> function, this time to filter to just the rows that contain the gene we want to plot </span>
<span id="cb47-732"><a href="#cb47-732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-735"><a href="#cb47-735" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-736"><a href="#cb47-736" aria-hidden="true" tabindex="-1"></a><span class="co"># first plot: TH</span></span>
<span id="cb47-737"><a href="#cb47-737" aria-hidden="true" tabindex="-1"></a>th_plot <span class="ot">&lt;-</span> genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-738"><a href="#cb47-738" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb47-739"><a href="#cb47-739" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Th"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-740"><a href="#cb47-740" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-741"><a href="#cb47-741" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb47-742"><a href="#cb47-742" aria-hidden="true" tabindex="-1"></a>    <span class="co"># should add a title to remind us which one this is</span></span>
<span id="cb47-743"><a href="#cb47-743" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"TH gene expression in control vs treatment groups"</span>) <span class="sc">+</span></span>
<span id="cb47-744"><a href="#cb47-744" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb47-745"><a href="#cb47-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-746"><a href="#cb47-746" aria-hidden="true" tabindex="-1"></a><span class="co"># second plot: PRLH</span></span>
<span id="cb47-747"><a href="#cb47-747" aria-hidden="true" tabindex="-1"></a>prlh_plot <span class="ot">&lt;-</span> genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-748"><a href="#cb47-748" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb47-749"><a href="#cb47-749" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Prlh"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-750"><a href="#cb47-750" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-751"><a href="#cb47-751" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb47-752"><a href="#cb47-752" aria-hidden="true" tabindex="-1"></a>    <span class="co"># should add a title to remind us which one this is</span></span>
<span id="cb47-753"><a href="#cb47-753" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb47-754"><a href="#cb47-754" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"PRLH gene expression in control vs treatment groups"</span>) <span class="sc">+</span></span>
<span id="cb47-755"><a href="#cb47-755" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb47-756"><a href="#cb47-756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-757"><a href="#cb47-757" aria-hidden="true" tabindex="-1"></a><span class="co"># combine them using patchwork + syntax</span></span>
<span id="cb47-758"><a href="#cb47-758" aria-hidden="true" tabindex="-1"></a>th_plot <span class="sc">+</span> prlh_plot</span>
<span id="cb47-759"><a href="#cb47-759" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-760"><a href="#cb47-760" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-761"><a href="#cb47-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-762"><a href="#cb47-762" aria-hidden="true" tabindex="-1"></a>The first method with faceting requires less code, but the second method of making two separate plots that are combined with patchwork allows you to be more flexible with the layout and  include extra plots in the figure if desired.</span>
<span id="cb47-763"><a href="#cb47-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-764"><a href="#cb47-764" aria-hidden="true" tabindex="-1"></a>But either solution is perfectly valid, and allows us to see that the expression of the Th and Prlh genes is higher in the treatment group than in the control group.</span>
<span id="cb47-765"><a href="#cb47-765" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-766"><a href="#cb47-766" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-767"><a href="#cb47-767" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-768"><a href="#cb47-768" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-769"><a href="#cb47-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-770"><a href="#cb47-770" aria-hidden="true" tabindex="-1"></a>R also provides a range of functions that we could use to formally test the relationship between these variables, although they are out of the scope of this course. A helpful guide on performing statistical tests in R <span class="co">[</span><span class="ot">can be found here</span><span class="co">](https://epirhandbook.com/en/new_pages/stat_tests.html#base-r)</span>.</span>
<span id="cb47-771"><a href="#cb47-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-772"><a href="#cb47-772" aria-hidden="true" tabindex="-1"></a><span class="fu">### An aside: string manipulation</span></span>
<span id="cb47-773"><a href="#cb47-773" aria-hidden="true" tabindex="-1"></a>Let's now take a moment to improve the appearance of our plot from the exercise above:</span>
<span id="cb47-774"><a href="#cb47-774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-777"><a href="#cb47-777" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-778"><a href="#cb47-778" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb47-779"><a href="#cb47-779" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-780"><a href="#cb47-780" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-781"><a href="#cb47-781" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-782"><a href="#cb47-782" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb47-783"><a href="#cb47-783" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb47-784"><a href="#cb47-784" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb47-785"><a href="#cb47-785" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-786"><a href="#cb47-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-787"><a href="#cb47-787" aria-hidden="true" tabindex="-1"></a>In session three we learned how to change the x and y axis labels on our graph to be more informative, but how can we change the values on the axes themselves? Currently, we have 'grp_control' and 'grp_treatment', but to make the plot look more publication-ready, it would be nice if it could just say 'Control' and 'Treatment'.</span>
<span id="cb47-788"><a href="#cb47-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-789"><a href="#cb47-789" aria-hidden="true" tabindex="-1"></a>We can do this with the help of the <span class="co">[</span><span class="ot">stringr</span><span class="co">](https://stringr.tidyverse.org/)</span> tidyverse package, which contains a set of functions to manipulate string (character) data in R. Being able to programatically manipulate strings is an essential tool for dealing with real-world data, particularly when you need to extract sample information from names or (as in our case) remove irrelevant characters that only take up space on plots.</span>
<span id="cb47-790"><a href="#cb47-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-791"><a href="#cb47-791" aria-hidden="true" tabindex="-1"></a>Some useful functions for string manipulation include:</span>
<span id="cb47-792"><a href="#cb47-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-793"><a href="#cb47-793" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_replace(string, pattern, replacement)`</span>: replaces the <span class="in">`pattern`</span> in the <span class="in">`string`</span> with the <span class="in">`replacement`</span></span>
<span id="cb47-794"><a href="#cb47-794" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_extract(string, pattern)`</span>: extracts the part of the <span class="in">`string`</span> matching the <span class="in">`pattern`</span></span>
<span id="cb47-795"><a href="#cb47-795" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_remove(string, pattern)`</span>: removes the part of the <span class="in">`string`</span> matching the <span class="in">`pattern`</span></span>
<span id="cb47-796"><a href="#cb47-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-797"><a href="#cb47-797" aria-hidden="true" tabindex="-1"></a>The full list of functions can be found in the <span class="co">[</span><span class="ot">stringr package documentation</span><span class="co">](https://stringr.tidyverse.org/reference/index.html)</span>.</span>
<span id="cb47-798"><a href="#cb47-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-799"><a href="#cb47-799" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Using stringr functions on a data frame"}</span>
<span id="cb47-800"><a href="#cb47-800" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-801"><a href="#cb47-801" aria-hidden="true" tabindex="-1"></a>Remember from session 2 that when we use the pipe operator <span class="in">`%&gt;%`</span>, it passes whatever precedes it into the function after it as the first argument. This means that the functions we use within a chain of pipes on a data frame should (generally) have a data frame as their input, and return a data frame as their output.</span>
<span id="cb47-802"><a href="#cb47-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-803"><a href="#cb47-803" aria-hidden="true" tabindex="-1"></a>But stringr functions operate on character vectors (their first argument is <span class="in">`string`</span>, as shown above), so how can we use them on our data frames? The answer is to wrap it in the <span class="in">`mutate()`</span> function from the dplyr package, which allows us to either edit an exisiting column or create a new one based on our string manipulation.</span>
<span id="cb47-804"><a href="#cb47-804" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-805"><a href="#cb47-805" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to change the replicate column in our <span class="in">`genes_tidy`</span> data frame to read 'Replicate 1', 'Replicate 2', etc instead of 'rep1', 'rep2'. We would do this by using the <span class="in">`str_replace()`</span> function inside the <span class="in">`mutate()`</span> function like so:</span>
<span id="cb47-806"><a href="#cb47-806" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-809"><a href="#cb47-809" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-810"><a href="#cb47-810" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-811"><a href="#cb47-811" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">replicate =</span> <span class="fu">str_replace</span>(replicate, <span class="st">"rep"</span>, <span class="st">"Replicate "</span>))</span>
<span id="cb47-812"><a href="#cb47-812" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-813"><a href="#cb47-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-814"><a href="#cb47-814" aria-hidden="true" tabindex="-1"></a>This applies to lots of other functions, not just the ones from the stringr package--- if you want to use a function that takes a vector as input on a column of a data frame, wrap it in the <span class="in">`mutate()`</span> function first!</span>
<span id="cb47-815"><a href="#cb47-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-816"><a href="#cb47-816" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-817"><a href="#cb47-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-818"><a href="#cb47-818" aria-hidden="true" tabindex="-1"></a>'Patterns' are what the stringr package will look for in your <span class="in">`string`</span>. For example, in this case, our pattern could be <span class="in">`grp_`</span> because this is what we want to remove from the strings in our 'group' column. This is an example of a **fixed pattern**, where you specify exactly what string you want to find. There is also another way to write patterns, called [**regular expressions**](https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions) which allow you to, for example, remove everything before the first underscore or all numbers. These are extremely powerful, but can get a little confusing so you might like to get help from AI to write them.</span>
<span id="cb47-819"><a href="#cb47-819" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-820"><a href="#cb47-820" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Cleaning up the text on our plot"}</span>
<span id="cb47-821"><a href="#cb47-821" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-822"><a href="#cb47-822" aria-hidden="true" tabindex="-1"></a>Using the stringr package, remove the 'grp_' from the x axis labels, and then use ggplot2 functions to add other labels to the plot. Here's an example of what this might look like:</span>
<span id="cb47-823"><a href="#cb47-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-826"><a href="#cb47-826" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-827"><a href="#cb47-827" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb47-828"><a href="#cb47-828" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-829"><a href="#cb47-829" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">str_remove</span>(group, <span class="st">"grp_"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb47-830"><a href="#cb47-830" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-831"><a href="#cb47-831" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-832"><a href="#cb47-832" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb47-833"><a href="#cb47-833" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb47-834"><a href="#cb47-834" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb47-835"><a href="#cb47-835" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb47-836"><a href="#cb47-836" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Effect of mousezempic on gene expression"</span>,</span>
<span id="cb47-837"><a href="#cb47-837" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Group"</span>, <span class="at">y =</span> <span class="st">"Expression (arbitrary units)"</span>)</span>
<span id="cb47-838"><a href="#cb47-838" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-839"><a href="#cb47-839" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-840"><a href="#cb47-840" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-841"><a href="#cb47-841" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-842"><a href="#cb47-842" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-843"><a href="#cb47-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-844"><a href="#cb47-844" aria-hidden="true" tabindex="-1"></a>The code that generates the plot above is:</span>
<span id="cb47-845"><a href="#cb47-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-848"><a href="#cb47-848" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-849"><a href="#cb47-849" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb47-850"><a href="#cb47-850" aria-hidden="true" tabindex="-1"></a>genes_tidy <span class="sc">%&gt;%</span></span>
<span id="cb47-851"><a href="#cb47-851" aria-hidden="true" tabindex="-1"></a>  <span class="co"># use str_remove function to remove 'grp_' from the </span></span>
<span id="cb47-852"><a href="#cb47-852" aria-hidden="true" tabindex="-1"></a>  <span class="co"># beginning of the group column. need to wrap it in</span></span>
<span id="cb47-853"><a href="#cb47-853" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the mutate function because we are working on a </span></span>
<span id="cb47-854"><a href="#cb47-854" aria-hidden="true" tabindex="-1"></a>  <span class="co"># column in a data frame not just on a vector </span></span>
<span id="cb47-855"><a href="#cb47-855" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">str_remove</span>(group, <span class="st">"grp_"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb47-856"><a href="#cb47-856" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">!=</span> <span class="st">"Hprt1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-857"><a href="#cb47-857" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb47-858"><a href="#cb47-858" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb47-859"><a href="#cb47-859" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb47-860"><a href="#cb47-860" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb47-861"><a href="#cb47-861" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add the other labels </span></span>
<span id="cb47-862"><a href="#cb47-862" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb47-863"><a href="#cb47-863" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Effect of mousezempic on gene expression"</span>,</span>
<span id="cb47-864"><a href="#cb47-864" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Group"</span>, </span>
<span id="cb47-865"><a href="#cb47-865" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Expression (arbitrary units)"</span>)</span>
<span id="cb47-866"><a href="#cb47-866" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-867"><a href="#cb47-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-868"><a href="#cb47-868" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-869"><a href="#cb47-869" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-870"><a href="#cb47-870" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-871"><a href="#cb47-871" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-872"><a href="#cb47-872" aria-hidden="true" tabindex="-1"></a><span class="fu">### The relationship between gene expression and weight loss</span></span>
<span id="cb47-873"><a href="#cb47-873" aria-hidden="true" tabindex="-1"></a>The second part of research question 2 asked us to look at whether TH and/or PRLH expression is associated with the amount of weight lost by mice taking mousezempic. To do this, we will need to use the <span class="in">`combined_data`</span> that we generated in @sec-join, which is the result of joining the <span class="in">`dosage`</span> and <span class="in">`genes_tidy`</span> data frames. </span>
<span id="cb47-874"><a href="#cb47-874" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-875"><a href="#cb47-875" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Plotting the relationship between gene expression and weight loss"}</span>
<span id="cb47-876"><a href="#cb47-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-877"><a href="#cb47-877" aria-hidden="true" tabindex="-1"></a>Use the <span class="in">`combined_data`</span> data frame and your choice of ggplot2 functions to see whether there is a relationship between the expression of the TH &amp; PRLH genes and weight lost by mice.</span>
<span id="cb47-878"><a href="#cb47-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-879"><a href="#cb47-879" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-880"><a href="#cb47-880" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-881"><a href="#cb47-881" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;</span><span class="kw">summary</span><span class="dt">&gt;</span>Solution<span class="dt">&lt;/</span><span class="kw">summary</span><span class="dt">&gt;</span></span>
<span id="cb47-882"><a href="#cb47-882" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-883"><a href="#cb47-883" aria-hidden="true" tabindex="-1"></a>Since both variables are continuous, a scatterplot is a good choice to visualise their relationship. Here is some sample code:</span>
<span id="cb47-884"><a href="#cb47-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-887"><a href="#cb47-887" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-888"><a href="#cb47-888" aria-hidden="true" tabindex="-1"></a><span class="co"># first plot: TH</span></span>
<span id="cb47-889"><a href="#cb47-889" aria-hidden="true" tabindex="-1"></a>th_plot <span class="ot">&lt;-</span> combined_data <span class="sc">%&gt;%</span></span>
<span id="cb47-890"><a href="#cb47-890" aria-hidden="true" tabindex="-1"></a>  <span class="co"># == tests for equality in R </span></span>
<span id="cb47-891"><a href="#cb47-891" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Th"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-892"><a href="#cb47-892" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-893"><a href="#cb47-893" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb47-894"><a href="#cb47-894" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb47-895"><a href="#cb47-895" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"TH gene expression (arbitrary units)"</span>,</span>
<span id="cb47-896"><a href="#cb47-896" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Weight loss (g)"</span>) <span class="sc">+</span></span>
<span id="cb47-897"><a href="#cb47-897" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb47-898"><a href="#cb47-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-899"><a href="#cb47-899" aria-hidden="true" tabindex="-1"></a><span class="co"># second plot: PRLH</span></span>
<span id="cb47-900"><a href="#cb47-900" aria-hidden="true" tabindex="-1"></a>prlh_plot <span class="ot">&lt;-</span> combined_data <span class="sc">%&gt;%</span></span>
<span id="cb47-901"><a href="#cb47-901" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"Prlh"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-902"><a href="#cb47-902" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression, <span class="at">y =</span> weight_lost_g)) <span class="sc">+</span></span>
<span id="cb47-903"><a href="#cb47-903" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb47-904"><a href="#cb47-904" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb47-905"><a href="#cb47-905" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"PRLH gene expression (arbitrary units)"</span>,</span>
<span id="cb47-906"><a href="#cb47-906" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Weight loss (g)"</span>) <span class="sc">+</span></span>
<span id="cb47-907"><a href="#cb47-907" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb47-908"><a href="#cb47-908" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-909"><a href="#cb47-909" aria-hidden="true" tabindex="-1"></a><span class="co"># combine them using patchwork + syntax</span></span>
<span id="cb47-910"><a href="#cb47-910" aria-hidden="true" tabindex="-1"></a>th_plot <span class="sc">+</span> prlh_plot</span>
<span id="cb47-911"><a href="#cb47-911" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-912"><a href="#cb47-912" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-913"><a href="#cb47-913" aria-hidden="true" tabindex="-1"></a>We can se there is no association between the expression of TH and weight lost, and the same is true for PRLH expression and weight lost. This probably makes sense, as there are many complex regulatory pathways involved in hunger.</span>
<span id="cb47-914"><a href="#cb47-914" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-915"><a href="#cb47-915" aria-hidden="true" tabindex="-1"></a><span class="dt">&lt;/</span><span class="kw">details</span><span class="dt">&gt;</span></span>
<span id="cb47-916"><a href="#cb47-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-917"><a href="#cb47-917" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-918"><a href="#cb47-918" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-919"><a href="#cb47-919" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-920"><a href="#cb47-920" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb47-921"><a href="#cb47-921" aria-hidden="true" tabindex="-1"></a>Today, we've tried to put together all of the skills we've learned in this R course and use them to address some sample data analysis questions. We've covered:</span>
<span id="cb47-922"><a href="#cb47-922" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-923"><a href="#cb47-923" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The concept of tidy data, that is stored in a consistent format so that we can easily analyse it with functions from the tidyverse</span>
<span id="cb47-924"><a href="#cb47-924" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-925"><a href="#cb47-925" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How to reshape data frames using the <span class="in">`pivot_wider()`</span>, <span class="in">`pivot_longer()`</span> and <span class="in">`separate()`</span> functions</span>
<span id="cb47-926"><a href="#cb47-926" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-927"><a href="#cb47-927" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How to merge data frames with common columns using the <span class="in">`left_join()`</span> function </span>
<span id="cb47-928"><a href="#cb47-928" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-929"><a href="#cb47-929" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The 'stringr' package for string manipulation, and some of the basic ways we can use it to clean up character columns in our data frames</span>
<span id="cb47-930"><a href="#cb47-930" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-931"><a href="#cb47-931" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The overall process of data science, and how to apply it in practice to break down our research questions into a series of steps/function calls in R </span>
<span id="cb47-932"><a href="#cb47-932" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-933"><a href="#cb47-933" aria-hidden="true" tabindex="-1"></a>Of course, a real data analysis will likely be much longer and more complex than the toy examples we have tackled today. However, we hope that it has provided you with a basic, yet versatile toolkit for transforming the messy data you will find in the wild into a neat, tidy format that you can use to explore your research questions and produce gorgeous data visualisations. </span>
<span id="cb47-934"><a href="#cb47-934" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-935"><a href="#cb47-935" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--</span></span>
<span id="cb47-936"><a href="#cb47-936" aria-hidden="true" tabindex="-1"></a><span class="al">###</span><span class="co"> Practice questions</span></span>
<span id="cb47-937"><a href="#cb47-937" aria-hidden="true" tabindex="-1"></a><span class="co">If you have time, try having a go at addressing these additional research questions. There are many different ways to approach them, so don't worry if yours doesn't match exactly with the solutions.</span></span>
<span id="cb47-938"><a href="#cb47-938" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-939"><a href="#cb47-939" aria-hidden="true" tabindex="-1"></a><span class="co">::: {.callout-important title="Practice exercise: looking at individual replicates"}</span></span>
<span id="cb47-940"><a href="#cb47-940" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-941"><a href="#cb47-941" aria-hidden="true" tabindex="-1"></a><span class="co">ask students to facet the plot by replicate</span></span>
<span id="cb47-942"><a href="#cb47-942" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-943"><a href="#cb47-943" aria-hidden="true" tabindex="-1"></a><span class="co">:::</span></span>
<span id="cb47-944"><a href="#cb47-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-945"><a href="#cb47-945" aria-hidden="true" tabindex="-1"></a><span class="co">Normalisation is a common step in many data analyses, as it allows us to more easily compare data across samples or experimental groups. For example, if we were to plot the expression of all genes across the three replicates:</span></span>
<span id="cb47-946"><a href="#cb47-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-949"><a href="#cb47-949" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-950"><a href="#cb47-950" aria-hidden="true" tabindex="-1"></a><span class="co">genes_tidy %&gt;%</span></span>
<span id="cb47-951"><a href="#cb47-951" aria-hidden="true" tabindex="-1"></a><span class="co">  ggplot(aes(x = replicate, y = expression)) + </span></span>
<span id="cb47-952"><a href="#cb47-952" aria-hidden="true" tabindex="-1"></a><span class="co">    geom_boxplot()</span></span>
<span id="cb47-953"><a href="#cb47-953" aria-hidden="true" tabindex="-1"></a><span class="co">```</span></span>
<span id="cb47-954"><a href="#cb47-954" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-955"><a href="#cb47-955" aria-hidden="true" tabindex="-1"></a><span class="co">We can see that replicate 2 has higher expression on average compared to replicates 1 and 3. </span></span>
<span id="cb47-956"><a href="#cb47-956" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-957"><a href="#cb47-957" aria-hidden="true" tabindex="-1"></a><span class="co">There are many different ways to normalise data, depending on the type, but for the sake of this question we'll:</span></span>
<span id="cb47-958"><a href="#cb47-958" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-959"><a href="#cb47-959" aria-hidden="true" tabindex="-1"></a><span class="co">1. Calculate the average HPRT1 expression in each replicate</span></span>
<span id="cb47-960"><a href="#cb47-960" aria-hidden="true" tabindex="-1"></a><span class="co">2. Divide all TH and PRLH measurements by the average HPRT1 expression calculated above</span></span>
<span id="cb47-961"><a href="#cb47-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-962"><a href="#cb47-962" aria-hidden="true" tabindex="-1"></a><span class="co">Take a moment to think about how you could achieve this using either the </span></span>
<span id="cb47-963"><a href="#cb47-963" aria-hidden="true" tabindex="-1"></a><span class="co">--&gt;</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>