<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Session 2: Working with data – WEHI R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session_1.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-063905fb60d21bb612e30333561bc9be.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-b70e54b6d14b88657c83e2b1b0a40c26.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/downloadthis-1.9.1/downloadthis.css" rel="stylesheet">
<script src="site_libs/quarto-contrib/kbd/kbd.js"></script>
<link href="site_libs/quarto-contrib/kbd/kbd.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session_2.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">WEHI R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this session:</h2>
   
  <ul>
  <li><a href="#sec-pipes" id="toc-sec-pipes" class="nav-link active" data-scroll-target="#sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</a></li>
  <li><a href="#sec-dataManip" id="toc-sec-dataManip" class="nav-link" data-scroll-target="#sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</a>
  <ul class="collapse">
  <li><a href="#sec-sorting" id="toc-sec-sorting" class="nav-link" data-scroll-target="#sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</a></li>
  <li><a href="#sec-filter" id="toc-sec-filter" class="nav-link" data-scroll-target="#sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</a></li>
  <li><a href="#sec-missing" id="toc-sec-missing" class="nav-link" data-scroll-target="#sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</a></li>
  <li><a href="#sec-select" id="toc-sec-select" class="nav-link" data-scroll-target="#sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</a></li>
  <li><a href="#sec-mutate" id="toc-sec-mutate" class="nav-link" data-scroll-target="#sec-mutate"><span class="header-section-number">2.2.5</span> Modifying data</a></li>
  <li><a href="#sec-summarise" id="toc-sec-summarise" class="nav-link" data-scroll-target="#sec-summarise"><span class="header-section-number">2.2.6</span> Summarising data</a></li>
  </ul></li>
  <li><a href="#sec-grouping" id="toc-sec-grouping" class="nav-link" data-scroll-target="#sec-grouping"><span class="header-section-number">2.3</span> Grouping</a></li>
  <li><a href="#sec-reshaping" id="toc-sec-reshaping" class="nav-link" data-scroll-target="#sec-reshaping"><span class="header-section-number">2.4</span> Reshaping and combining data</a>
  <ul class="collapse">
  <li><a href="#sec-pivot" id="toc-sec-pivot" class="nav-link" data-scroll-target="#sec-pivot"><span class="header-section-number">2.4.1</span> Reshaping data with pivot functions</a></li>
  <li><a href="#sec-join" id="toc-sec-join" class="nav-link" data-scroll-target="#sec-join"><span class="header-section-number">2.4.2</span> Combining data with join functions</a></li>
  </ul></li>
  <li><a href="#sec-saving" id="toc-sec-saving" class="nav-link" data-scroll-target="#sec-saving"><span class="header-section-number">2.5</span> Saving data to a file</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">2.6</span> Summary</a>
  <ul class="collapse">
  <li><a href="#practice-questions" id="toc-practice-questions" class="nav-link" data-scroll-target="#practice-questions"><span class="header-section-number">2.6.1</span> Practice questions</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-session02" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Session 2: Working with data</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this session we will learn how to manipulate and summarise data using the <code>dplyr</code> package (with a little help from the <code>tidyr</code> package too).</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this session, learners should be able to:</p>
<ol type="1">
<li><p>Use the pipe (<code>%&gt;%</code>) to chain multiple functions together</p></li>
<li><p>Design chains of dplyr functions to manipulate data frames</p></li>
<li><p>Apply grouping for more complex analysis of data</p></li>
<li><p>Understand how pivot and join functions can be used to reshape and combine entire data frames</p></li>
<li><p>Recall how to save data frames to a file</p></li>
</ol>
</div>
</div>
<p>Both <code>dplyr</code> and <code>tidyr</code> are contained within the <code>tidyverse</code> (along with <code>readr</code>) so we can load all of these packages at once using <code>library(tidyverse)</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load tidyverse!</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
<section id="sec-pipes" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</h2>
<p>Pipes are a powerful feature of the <code>tidyverse</code> that allow you to chain multiple functions together. Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</p>
<p>For example, take the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>What do you think this code does? It calculates the mean of <code>my_vector</code>, rounds the result to the nearest whole number, and then converts the result to a character. But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</p>
<p>Instead, we can use the pipe operator (<code>%&gt;%</code>) to chain these functions together in a more readable way:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>See how the code reads naturally from left to right? You can think of the pipe as being like the phrase “and then”. Here, we’re telling R: “Take <code>my_vector</code>, and then calculate the mean, and then round the result, and then convert it to a character.”</p>
<p>You’ll notice that we didn’t need to specify the input to each function. That’s because the pipe automatically passes the output of the previous function as the first input to the next function. We can still specify additional arguments to each function if we need to. For example, if we wanted to round the mean to 2 decimal places, we could do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>R is clever enough to know that the first argument to <code>round()</code> is still the output of <code>mean()</code>, even though we’ve now specified the <code>digits</code> argument.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Plenty of pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Plenty of pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>There is another style of pipe in R, called the ‘base R pipe’ <code>|&gt;</code>, which is available in R version 4.1.0 and later. The base R pipe works in a similar way to the <code>magrittr</code> pipe (<code>%&gt;%</code>) that we use in this course, but it is not as flexible. We recommend using the <code>magrittr</code> pipe for now.</p>
<p>Fun fact: the <code>magrittr</code> package is named after the <a href="https://en.wikipedia.org/wiki/The_Treachery_of_Images">artist René Magritte, who made a famous painting of a pipe</a>.</p>
</div>
</div>
<p>To type the pipe operator more easily, you can use the keyboard shortcut <kbd aria-hidden="true">Cmd-shift-M</kbd><span class="visually-hidden">Cmd-shift-M</span> (although once you get used to it, you might find it easier to type <code>%&gt;%</code> manually).</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What is NOT a valid way to re-write the following code using the pipe operator: <code>round(sqrt(sum(1:10)), 1)</code>. If you’re not sure, try running the different options in the console to see which one gives the same answer.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is the output of the following code? <code>letters %&gt;% head() %&gt;% toupper()</code> Try to guess it before copy-pasting into R.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>"a" "b" "c" "d" "e" "f"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>"A" "B" "C" "D" "E" "F"</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The invalid option is <code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code>. This is because the <code>sqrt()</code> function only takes one argument, so you can’t specify <code>1</code> as an argument in addition to what is being piped in from <code>sum(1:10)</code>. Note that some options used the pipe to send <code>1:10</code> to <code>sum()</code> (like <code>1:10 %&gt;% sum()</code>), and others just used <code>sum(1:10)</code> directly. Both are valid ways to use the pipe, it’s just a matter of personal preference.</li>
<li>The output of the code <code>letters %&gt;% head() %&gt;% toupper()</code> is <code>"A" "B" "C" "D" "E" "F"</code>. The <code>letters</code> vector contains the lowercase alphabet, and the <code>head()</code> function returns the first 6 elements of the vector. Finally, the <code>toupper()</code> function then converts these elements to uppercase.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-dataManip" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</h2>
<p>To really see the power of the pipe, we will use it together with the <code>dplyr</code> package that provides a set of functions to easily filter, sort, select, and summarise data frames. These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</p>
<p>For example, even though we haven’t covered the <code>dplyr</code> functions yet, you can probably guess what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code filters the <code>mousezempic_dosage_data</code> data frame to only include data from cage 3E, then pulls out the <code>weight_lost_g</code> column, and finally calculates the mean of the values in that column. The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <code>pull()</code>) are specified in the brackets (like <code>round(digits = 2)</code> from the previous example).</p>
<p>We also used the enter key after each pipe <code>%&gt;%</code> to break up the code into multiple lines to make it easier to read. This isn’t required, but is a popular style in the R community, so all the code examples in this session will follow this format.</p>
<p>We will now introduce some of the most commonly used <code>dplyr</code> functions for manipulating data frames. To showcase these, we will use the <code>mousezempic_dosage_data</code> that we practiced reading in last session. This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in session 1</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"~/Desktop/R_course/data/mousezempic_dosage_data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 344 Columns: 9
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): mouse_strain, cage_number, replicate, sex
dbl (5): weight_lost_g, drug_dose_g, tail_length_mm, initial_weight_g, id_num

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>Before we start, let’s use what we learned in the previous session to take a look at <code>mousezempic_dosage_data</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>You might also like to use <code>View()</code> to open the data in a separate window and get a closer look.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Using RStudio autocomplete">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Using RStudio autocomplete
</div>
</div>
<div class="callout-body-container callout-body">
<p>Although it’s great to give our data a descriptive name like <code>mousezempic_dosage_data</code>, it can be a bit of a pain to type out every time. Luckily, RStudio has a handy autocomplete feature that can solve this problem. Just start typing the name of the object, and you’ll see it will popup:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F01_rstudio_autocomplete.png" class="img-fluid figure-img"></p>
<figcaption>RStudio autocomplete</figcaption>
</figure>
</div>
<p>You can then press <kbd aria-hidden="true">Tab</kbd><span class="visually-hidden">Tab</span> to autocomplete it. If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</p>
<p>Try using autocomplete this session to save yourself some typing!</p>
</div>
</div>
<section id="sec-sorting" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</h3>
<p>Often, one of the first things you might want to do with a dataset is sort it. In <code>dplyr</code>, this is called ‘arranging’ and is done with the <code>arrange()</code> function.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F02_arrange.png" class="img-fluid figure-img"></p>
<figcaption>Arrange orders rows by their values in one or more columns</figcaption>
</figure>
</div>
<p>By default, <code>arrange()</code> sorts in ascending order (smallest values first). For example, let’s sort the <code>mousezempic_dosage_data</code> data frame by the <code>weight_lost_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 CD-1         3E                   2.85 rep2      female     0.00181
 3 CD-1         3E                   2.85 rep2      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         2B                   2.9  rep2      female     0.00178
 6 CD-1         1A                   2.9  rep3      female     0.00188
 7 BALB C       2B                   2.9  rep1      female     0.00187
 8 CD-1         3E                   2.92 rep3      female     0.00193
 9 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
10 CD-1         2B                   3    rep1      female     0.00185
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</p>
<p>Sometimes you might want to sort in descending order instead (largest values first). You can do this by putting the <code>desc()</code> function around your column name, inside <code>arrange()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Now we can see the mice that lost the most weight are at the top.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Comments and pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Comments and pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Notice how in the previous example we have written a comment in the middle of the pipe chain. This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won’t cause any errors as long as you make sure that the comment is on its own line.</p>
<p>You can also write comments at the end of the line, just make sure it’s after the pipe operator <code>%&gt;%</code>.</p>
<p>For example, these comments are allowed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But this will cause an error, because the <code>#</code> is before the pipe, so R treats it as part of the comment (notice how the <code>%&gt;%</code> has changed colour?) and doesn’t know how the two lines relate to each other. It tries to run them separately, which for the first line is ok (it will just print <code>mousezempic_dosage_data</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># this comment will cause an error %&gt;%</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in eval(expr, envir, enclos): object 'weight_lost_g' not found</code></pre>
</div>
</div>
<p>But for the second line, there is an error that R doesn’t know what the <code>weight_lost_g</code> object is. That’s because it’s a column in the <code>mousezempic_dosage_data</code> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</p>
</div>
</div>
<p>You can also sort by multiple columns by passing multiple column names to <code>arrange()</code>. For example, to sort by the strain first and then by the amount of weight lost:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 BALB C       2B                   2.9  rep1      female     0.00187
 3 BALB C       2B                   3.2  rep2      female     0.00187
 4 BALB C       2B                   3.25 rep1      female     0.00178
 5 BALB C       2B                   3.25 rep3      male       0.00187
 6 BALB C       2B                   3.25 rep3      female     0.00191
 7 BALB C       2B                   3.3  rep1      male       0.00197
 8 BALB C       2B                   3.3  rep1      female     0.00195
 9 BALB C       2B                   3.32 rep3      female     0.00199
10 BALB C       2B                   3.35 rep2      female     0.00187
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Piping into View()">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Piping into View()
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn’t allow us to see the full effect of the sorting.</p>
<p>One handy trick you can use with pipes is to add <code>View()</code> at the end of your chain to open the data in a separate window. Try running this code, and you’ll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is a great way to check that your code has actually done what you intended!</p>
</div>
</div>
<section id="sec-sliceMinMax" class="level4" data-number="2.2.1.1">
<h4 data-number="2.2.1.1" class="anchored" data-anchor-id="sec-sliceMinMax"><span class="header-section-number">2.2.1.1</span> Extracting rows with the smallest or largest values</h4>
<p>Slice functions are used to select rows based on their position in the data frame. The <code>slice_min()</code> and <code>slice_max()</code> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</p>
<p>This is equivalent to using <code>arrange()</code> followed by <code>head()</code>, but is more concise:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.15 rep1      female     0.00172
 2 CD-1         3E                   3.4  rep1      female     0.00174
 3 CD-1         1A                   3.45 rep3      female     0.00176
 4 CD-1         2B                   3.25 rep1      female     0.00178
 5 CD-1         2B                   3.9  rep1      male       0.00178
 6 CD-1         2B                   2.9  rep2      female     0.00178
 7 BALB C       2B                   3.25 rep1      female     0.00178
 8 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
 9 CD-1         1A                   3.7  rep1      &lt;NA&gt;       0.0018 
10 CD-1         3E                   3.6  rep1      male       0.0018 
11 CD-1         3E                   3.8  rep1      male       0.0018 
12 CD-1         3E                   3.95 rep1      male       0.0018 
13 CD-1         2B                   3.55 rep1      female     0.0018 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
6 Black 6      3E                   5.95 rep3      male      0.00229
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But wait— neither of those pieces of code actually gave the number of rows we asked for! In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13. And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6. Why aren’t the <code>slice_</code> functions behaving as expected?</p>
<p>If we take a look at the help page (type <code>?slice_min</code> in the console), we learn that <code>slice_min()</code> and <code>slice_max()</code> have an argument called <code>with_ties</code> that is set to <code>TRUE</code> by default. If we want to make sure we only get the number of rows we asked for, we would have to set it to <code>FALSE</code>, like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to sort the <code>mousezempic_dosage_data</code> data frame from biggest to smallest initial weight?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% sort(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% sort(descending(initial_weight_g))</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which code would you use to extract the 3 mice with the highest initial weight from the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% slice_min(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. I’ve written the below code, but one of the comments is messing it up! Which one?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># comment B %&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Comment A</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Comment B</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Comment C</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Comment D</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to sort the <code>mousezempic_dosage_data</code> data frame from biggest to smallest initial weight is <code>mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))</code>. The <code>arrange()</code> function is used to sort the data frame (although there is a <code>sort()</code> function in R, that’s not part of dplyr and won’t work the same way), and the <code>desc()</code> function is used to sort in descending order.</p></li>
<li><p>The correct code to extract the 3 mice with the highest initial weight from the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)</code>. The <code>slice_max()</code> function is used to select the rows with the largest values in the <code>initial_weight_g</code> column, and the <code>n = 3</code> argument specifies that we want to keep 3 rows. The <code>arrange()</code> function is not needed in this case, because <code>slice_max()</code> will automatically sort the data frame by the specified column.</p></li>
<li><p>The comment that is messing up the code is Comment B. The <code>#</code> symbol is before the pipe operator <code>%&gt;%</code>, so R treats it as part of the comment and this breaks our chain of pipes. The other comments are fine, because they are either at the end of the line or on their own line. Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it’s in the wrong place!</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-filter" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F03_filter.png" class="img-fluid figure-img"></p>
<figcaption>Filter allows you to filter rows using a logical test</figcaption>
</figure>
</div>
<p>In <code>dplyr</code>, the <code>filter()</code> function is used to subset rows based on their values. You provide a logical test, and <code>filter()</code> will keep the rows where the test is <code>TRUE</code>. We can write these tests using the comparison operators we learned in the previous session (e.g.&nbsp;<code>==</code>, <code>&lt;</code> and <code>!=</code>, see <a href="session_1.html#sec-comparisons" class="quarto-xref">Section&nbsp;<span>1.3</span></a>).</p>
<p>For example, to filter the <code>mousezempic_dosage_data</code> data frame to only include mice that lost more than 6g:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or to only include mice from cage 3E:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 168 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.4  rep1      female     0.00174
 2 CD-1         3E                   3.6  rep1      male       0.0018 
 3 CD-1         3E                   3.8  rep1      female     0.00189
 4 CD-1         3E                   3.95 rep1      male       0.00185
 5 CD-1         3E                   3.8  rep1      male       0.0018 
 6 CD-1         3E                   3.8  rep1      female     0.00187
 7 CD-1         3E                   3.55 rep1      male       0.00183
 8 CD-1         3E                   3.2  rep1      female     0.00187
 9 CD-1         3E                   3.15 rep1      female     0.00172
10 CD-1         3E                   3.95 rep1      male       0.0018 
# ℹ 158 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<section id="combining-logical-tests" class="level4" data-number="2.2.2.1">
<h4 data-number="2.2.2.1" class="anchored" data-anchor-id="combining-logical-tests"><span class="header-section-number">2.2.2.1</span> Combining logical tests</h4>
<p>Sometimes we want to filter based on multiple conditions. Here we will show some more advanced operators that can be used to combine logical tests.</p>
<p>The <code>&amp;</code> operator is used to combine two logical tests with an ‘and’ condition. For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
1 CD-1         1A                   3.45 rep1      female     0.00193
2 CD-1         3E                   3.8  rep1      female     0.00189
3 CD-1         2B                   3.3  rep1      female     0.00181
4 CD-1         3E                   3.9  rep3      female     0.00191
5 BALB C       2B                   3.52 rep3      female     0.00194
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>|</code> operator is used to combine two logical tests with an ‘or’ condition. For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 2 CD-1         1A                   4.4  rep1      male       0.00198
 3 CD-1         1A                   3.32 rep1      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         1A                   3.6  rep2      female     0.0019 
 6 CD-1         1A                   3.2  rep2      female     0.00189
 7 CD-1         2B                   3.4  rep2      female     0.00185
 8 CD-1         2B                   2.9  rep2      female     0.00178
 9 CD-1         2B                   3.05 rep3      female     0.00188
10 Black 6      3E                   4.5  rep1      female     0.00211
# ℹ 20 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>%in%</code> operator can be used to filter based on a vector of multiple values (<code>c(x, y)</code>). It’s particularly useful when you have a few character values you want to filter on, as it is shorter to type than <code>|</code> (or).</p>
<p>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <code>|</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or we could use <code>%in%</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to filter the <code>mousezempic_dosage_data</code> data frame to only include mice from replicate 2?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate == 2)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate == rep2)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(replicate == "rep2")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(replicate = "rep2")</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is NOT a valid way to filter the <code>mousezempic_dosage_data</code> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. Which option correctly describes what the following code is doing?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to filter the <code>mousezempic_dosage_data</code> data frame to only include mice from replicate 2 is <code>mousezempic_dosage_data %&gt;% filter(replicate == "rep2")</code>. Option A is incorrect because <code>2</code> is not a value of <code>replicate</code> (when filtering you need to know what values are actually in your columns! So make sure to <code>View()</code> your data first). Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on. Option D is incorrect because <code>=</code> is not the correct way to test for equality, you need to use <code>==</code>.</p></li>
<li><p>The invalid option is <code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code>. This is because the second filtering step is missing the name of the filter function, so R doesn’t know what to do with <code>(initial_weight_g &lt; 40)</code>. The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <code>filter()</code> functions in a row to apply multiple conditions, or the <code>&amp;</code> operator to combine them into a single <code>filter()</code> function. It’s just a matter of personal preference.</p></li>
<li><p>The correct description of the code is that it filters the data frame to only include mice from the “BALB C” and “Black 6” strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-missing" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</h3>
<p>Missing values are a common problem in real-world datasets. In R, missing values are represented by <code>NA</code>. In fact, if you look at the <code>mousezempic_dosage_data</code> data frame we’ve been using, you’ll see that some of the cells contain <code>NA</code>: try spotting them with the <code>View()</code> function.</p>
<p>You can also find missing values in a data frame using the <code>is.na()</code> function in combination with <code>filter()</code>. For example, to find all the rows in the <code>mousezempic_dosage_data</code> data frame that have a missing value for the <code>drug_dose_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 CD-1         1A                     NA rep1      &lt;NA&gt;           NA
2 Black 6      3E                     NA rep3      &lt;NA&gt;           NA
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The problem with missing values is that they can cause problems when you try to perform calculations on your data. For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <code>NA</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p><code>NA</code> values in R are therefore referred to as ‘contagious’: if you put an <code>NA</code> in you usually get an <code>NA</code> out. If you think about it, that makes sense— when we don’t know the value of a particular mouse’s drug dose, how can we calculate the average? That missing value could be anything.</p>
<p>For this reason, it’s important to deal with missing values before performing calculations. Many functions in R will have an argument called <code>na.rm</code> that you can set to <code>TRUE</code> to remove missing values before performing the calculation. For example, to calculate the mean of the <code>drug_dose_g</code> column with the missing values excluded:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.002009152</code></pre>
</div>
</div>
<p>This time, the result is a number, because the missing values have been removed before the calculation.</p>
<p>But not all functions have an <code>na.rm</code> argument. In these cases, you can remove rows with missing values. This can be done for a single column, using the <code>filter()</code> function together with <code>is.na()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 342 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 9 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
10 CD-1         1A                   3.3  rep1      &lt;NA&gt;       0.00186
# ℹ 332 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or, you can remove rows with missing values in any column using the <code>na.omit()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.2  rep1      female     0.00182
 9 CD-1         1A                   3.8  rep1      male       0.00191
10 CD-1         1A                   4.4  rep1      male       0.00198
# ℹ 323 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value. This can be done using the <code>replace_na()</code> function from the <code>tidyr</code> package. <code>replace_na()</code> takes a <code>list()</code> which contains each of the column names you want to edit, and the value that should be used.</p>
<p>For example, to replace missing values in the <code>weight_lost_g</code> columns with 0, replace missing values in the <code>sex</code> column with ‘unknown’ and leave the rest of the data frame unchanged:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex     drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;         &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male        0.00181
 2 CD-1         1A                   3.8  rep1      female      0.00186
 3 CD-1         1A                   3.25 rep1      female      0.00195
 4 CD-1         1A                   0    rep1      unknown    NA      
 5 CD-1         1A                   3.45 rep1      female      0.00193
 6 CD-1         1A                   3.65 rep1      male        0.0019 
 7 CD-1         1A                   3.62 rep1      female      0.00181
 8 CD-1         1A                   4.68 rep1      male        0.00195
 9 CD-1         1A                   3.48 rep1      unknown     0.00193
10 CD-1         1A                   4.25 rep1      unknown     0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>When deciding how to handle missing values, you might have prior knowledge that <code>NA</code> should be replaced with a specific value, or you might decide that removing rows with <code>NA</code> is the best approach for your analysis.</p>
<p>For example, maybe we knew that the mice were given a <code>weight_lost_g</code> of <code>NA</code> if they didn’t lose any weight, it would then make sense to replace those with 0 (as we did in the code above). However, if the <code>drug_dose_g</code> column was missing simply because the data was lost, we might choose to remove those rows entirely.</p>
<p>It’s important to think carefully about how missing values should be handled in your analysis.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What would be the result of running this R code: <code>mean(c(1, 2, 4, NA))</code></p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>2.333333</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>0</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>NA</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which line of code would you use to filter the <code>mousezempic_dosage_data</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(tail_length_mm != NA)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(is.na(tail_length_mm))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% na.omit()</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you replace missing values in the <code>initial_weight_g</code> column with the value 35?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(initial_weight_g = 35)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g == 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% replace_na(35)</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The result of running the code <code>mean(c(1, 2, 4, NA))</code> is <code>NA</code>. This is because the <code>NA</code> value is ‘contagious’, so when you try to calculate the mean of a vector that contains an <code>NA</code>, the result will also be <code>NA</code>. If we wanted to calculate the mean of the vector without the <code>NA</code>, we would need to use the <code>na.rm = TRUE</code> argument.</li>
<li>The correct line of code to filter the <code>mousezempic_dosage_data</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column is <code>mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))</code>. The <code>!</code> symbol is used to negate the result of <code>is.na()</code>, so we are filtering to keep the rows where <code>tail_length_mm</code> is not <code>NA</code>. We can’t use the first option with the <code>!= NA</code> because <code>NA</code> is a special value in R that represents missing data, and it can’t be compared to anything, and the third option is incorrect because <code>na.omit()</code> removes entire rows with missing values, rather than just filtering based on a single column.</li>
<li>The correct line of code to replace missing values in the <code>initial_weight_g</code> column with the value 35 is <code>mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))</code>. The <code>replace_na()</code> function takes a <code>list()</code> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we’re not testing for equality, we’re assigning a value.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-select" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F04_select.png" class="img-fluid figure-img"></p>
<figcaption>Select allows you to select only certain columns</figcaption>
</figure>
</div>
<p>While <code>filter()</code> is used to subset rows, <code>select()</code> is used to subset columns. You can use <code>select()</code> to keep only the columns you’re interested in, or to drop columns you don’t need.</p>
<p>The <code>select()</code> function takes the names of the columns that you want to keep/remove (no vector notation <code>c()</code> or quotation marks <code>""</code> necessary). For example, to select only the <code>mouse_strain</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   mouse_strain initial_weight_g weight_lost_g
   &lt;chr&gt;                   &lt;dbl&gt;         &lt;dbl&gt;
 1 CD-1                     39.1          3.75
 2 CD-1                     39.5          3.8 
 3 CD-1                     40.3          3.25
 4 CD-1                     NA           NA   
 5 CD-1                     36.7          3.45
 6 CD-1                     39.3          3.65
 7 CD-1                     38.9          3.62
 8 CD-1                     39.2          4.68
 9 CD-1                     34.1          3.48
10 CD-1                     42            4.25
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>We can see that all the other columns have been removed from the data frame.</p>
<p>If you want to keep all columns except for a few, you can use <code>-</code> to drop columns. For example, to keep all columns except for <code>cage_number</code> and <code>sex</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 7
   mouse_strain weight_lost_g replicate drug_dose_g tail_length_mm
   &lt;chr&gt;                &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt;
 1 CD-1                  3.75 rep1          0.00181           18.7
 2 CD-1                  3.8  rep1          0.00186           17.4
 3 CD-1                  3.25 rep1          0.00195           18  
 4 CD-1                 NA    rep1         NA                 NA  
 5 CD-1                  3.45 rep1          0.00193           19.3
 6 CD-1                  3.65 rep1          0.0019            20.6
 7 CD-1                  3.62 rep1          0.00181           17.8
 8 CD-1                  4.68 rep1          0.00195           19.6
 9 CD-1                  3.48 rep1          0.00193           18.1
10 CD-1                  4.25 rep1          0.0019            20.2
# ℹ 334 more rows
# ℹ 2 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>There are also some helper functions that can be used to select columns based on their names :</p>
<table class="caption-top table">
<caption>There are several helper functions that can be used with the select function</caption>
<colgroup>
<col style="width: 14%">
<col style="width: 42%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>starts_with()</code></td>
<td>select column(s) that start with a certain string</td>
<td><p>select all columns starting with the letter i</p>
<p><code>select(starts_with("i"))</code></p></td>
</tr>
<tr class="even">
<td><code>ends_with()</code></td>
<td>select column(s) that end with a certain string</td>
<td><p>select all columns ending with _g</p>
<p><code>select(ends_with("_g"))</code></p></td>
</tr>
<tr class="odd">
<td><code>contains()</code></td>
<td>select column(s) that contain a certain string</td>
<td><p>select all columns containing the word ‘weight’</p>
<p><code>select(contains("weight"))</code></p></td>
</tr>
</tbody>
</table>
<p>You need to use quotation marks around the arguments in these helper functions, as they aren’t full column names, just strings of characters.</p>
<p>Try using these helper functions to select columns from the <code>mousezempic_dosage_data</code> data frame!</p>
<div class="callout callout-style-default callout-note callout-titled" title="Reordering columns">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Reordering columns
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/S02F05_reorder.png" class="img-fluid" alt="Relocate allows you to move columns around"> We can reorder columns using the <code>relocate()</code> function, which works similarly to <code>select()</code> (except it just moves columns around rather than dropping/keeping them). For example, to move the <code>sex</code> column to before the <code>cage_number</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain sex    cage_number weight_lost_g replicate drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 CD-1         male   1A                   3.75 rep1          0.00181
 2 CD-1         female 1A                   3.8  rep1          0.00186
 3 CD-1         female 1A                   3.25 rep1          0.00195
 4 CD-1         &lt;NA&gt;   1A                  NA    rep1         NA      
 5 CD-1         female 1A                   3.45 rep1          0.00193
 6 CD-1         male   1A                   3.65 rep1          0.0019 
 7 CD-1         female 1A                   3.62 rep1          0.00181
 8 CD-1         male   1A                   4.68 rep1          0.00195
 9 CD-1         &lt;NA&gt;   1A                   3.48 rep1          0.00193
10 CD-1         &lt;NA&gt;   1A                   4.25 rep1          0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Two useful helper functions here are the <code>everything()</code> and <code>last_col()</code> functions, which can be used to move columns to the start/end of the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   id_num mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
    &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1      1 CD-1         1A                   3.75 rep1      male       0.00181
 2      2 CD-1         1A                   3.8  rep1      female     0.00186
 3      3 CD-1         1A                   3.25 rep1      female     0.00195
 4      4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5      5 CD-1         1A                   3.45 rep1      female     0.00193
 6      6 CD-1         1A                   3.65 rep1      male       0.0019 
 7      7 CD-1         1A                   3.62 rep1      female     0.00181
 8      8 CD-1         1A                   4.68 rep1      male       0.00195
 9      9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10     10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 2 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   cage_number weight_lost_g replicate sex    drug_dose_g tail_length_mm
   &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;
 1 1A                   3.75 rep1      male       0.00181           18.7
 2 1A                   3.8  rep1      female     0.00186           17.4
 3 1A                   3.25 rep1      female     0.00195           18  
 4 1A                  NA    rep1      &lt;NA&gt;      NA                 NA  
 5 1A                   3.45 rep1      female     0.00193           19.3
 6 1A                   3.65 rep1      male       0.0019            20.6
 7 1A                   3.62 rep1      female     0.00181           17.8
 8 1A                   4.68 rep1      male       0.00195           19.6
 9 1A                   3.48 rep1      &lt;NA&gt;       0.00193           18.1
10 1A                   4.25 rep1      &lt;NA&gt;       0.0019            20.2
# ℹ 334 more rows
# ℹ 3 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;, mouse_strain &lt;chr&gt;</code></pre>
</div>
</div>
<p>Re-ordering columns isn’t necessary, but it makes it easier to see the data you’re most interested in within the console (since often not all of the columns will fit on the screen at once). For example, if we are doing a lot of computation on the <code>initial_weight_g</code> column, we’d probably like to have that near the start so we can easily check it.</p>
</div>
</div>
<p>Note that the output of the <code>select()</code> function is a new data frame, even if you only select a single column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from session 1 that class() tells us the type of an object</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
</div>
</div>
<p>Sometimes, we instead want to get the values of a column as a vector.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F06_pull.png" class="img-fluid figure-img"></p>
<figcaption>Pull allows you to pull acolumn out of a data frame as a vector</figcaption>
</figure>
</div>
<p>We can do this by using the <code>pull()</code> function, which extracts a single column from a data frame as a vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
</div>
<p>We can see that the class of the output is now a vector, rather than a data frame. This is important because some functions only accept vectors, not data frames, like <code>mean()</code> for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in mean.default(., na.rm = TRUE): argument is not numeric or logical:
returning NA</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 43.92193</code></pre>
</div>
</div>
<p>Note how both times we used <code>na.rm = TRUE</code> to remove missing values before calculating the mean.</p>
<p>You might remember that we used the <code>$</code> operator in the previous session to extract a single column from a data frame, so why use <code>pull()</code> instead? The main reason is that <code>pull()</code> works within a chain of pipes, whereas <code>$</code> doesn’t.</p>
<p>For example, let’s say we want to know the average initial weight of mice that lost at least 4g. We can do this by chaining <code>filter()</code> and <code>pull()</code> together:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 46.48023</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which line of code would NOT be a valid way to select the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% select(contains("g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(ends_with("_g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. How would I extract the <code>initial_weight_g</code> column from the <code>mousezempic_dosage_data</code> data frame as a vector?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% filter(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% $initial_weight_g</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% select(initial_weight_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% pull(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you move the <code>sex</code> column to the end of the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% relocate(sex)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% reorder(sex, .after = last_col())</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The line of code that would NOT be a valid way to select the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% select(contains("g"))</code>. This line of code would select all columns that contain the letter ‘g’, which would include columns like <code>cage_number</code> and <code>tail_length_mm</code>. We need to specify either <code>ends_with("g")</code> or <code>contains("_g")</code> to only get those with <code>_g</code> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</li>
<li>The correct way to extract the <code>initial_weight_g</code> column from the <code>mousezempic_dosage_data</code> data frame as a vector is <code>mousezempic_dosage_data %&gt;% pull(initial_weight_g)</code>. The <code>pull()</code> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <code>filter()</code> is used to subset rows, <code>$</code> is not used in a pipe chain, and <code>select()</code> is outputs a data frame, not extract them as vectors.</li>
<li>The correct way to move the <code>sex</code> column to the end of the <code>mousezempic_dosage_data</code> data frame is using the <code>relocate()</code> function like this: <code>mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())</code>. The <code>last_col()</code> function is used to refer to the last column in the data frame. The other options are incorrect because <code>reorder()</code> is not a valid function, and you need to remember to include the brackets <code>()</code> when using <code>last_col()</code>.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-mutate" class="level3" data-number="2.2.5">
<h3 data-number="2.2.5" class="anchored" data-anchor-id="sec-mutate"><span class="header-section-number">2.2.5</span> Modifying data</h3>
<p>So far, we’ve learned how to filter rows and select columns from a data frame. But what if we want to change the data itself? This is where the <code>mutate()</code> function comes in.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F07_mutate.png" class="img-fluid figure-img"></p>
<figcaption>Mutate allows you to add new columns to a data frame</figcaption>
</figure>
</div>
<p>The <code>mutate()</code> function is used to add new columns to a data frame, or modify existing columns, often by performing some sort of calculation. For example, we can add a new column to <code>mousezempic_dosage_data</code> that contains the drug dose in mg (rather than g):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a new column called drug_dose_mg</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert drug_dose_g to mg by multiplying by 1000</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># just select the drug dose columns so we can compare them</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(drug_dose_g, drug_dose_mg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   drug_dose_g drug_dose_mg
         &lt;dbl&gt;        &lt;dbl&gt;
 1     0.00181         1.81
 2     0.00186         1.86
 3     0.00195         1.95
 4    NA              NA   
 5     0.00193         1.93
 6     0.0019          1.9 
 7     0.00181         1.81
 8     0.00195         1.95
 9     0.00193         1.93
10     0.0019          1.9 
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>You can see that the <code>drug_dose_mg</code> column has been added to the data frame, and it contains, for each row, the value of the <code>drug_dose_g</code> column multiplied by 1000 (<code>NA</code> values are preserved).</p>
<p>These calculations can be as complex as you like, and involve multiple different columns. For example, to add a new column to the <code>mousezempic_dosage_data</code> data frame that calculates the weight lost as a percentage of the initial weight:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate the % of initial weight that was lost</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_percent =</span> (weight_lost_g <span class="sc">/</span> initial_weight_g) <span class="sc">*</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 10
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 4 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, weight_lost_percent &lt;dbl&gt;</code></pre>
</div>
</div>
<p>A useful helper function for <code>mutate()</code> is the <code>case_when()</code> function, which allows you to create new columns based on multiple conditions. We do this with the notation <code>case_when(condition1 ~ value1, condition2 ~ value2, ...)</code>.</p>
<p>For example, to add a new column to the <code>mousezempic_dosage_data</code> data frame that categorises the mice based on how much weight they lost:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   weight_lost_g weight_loss_category
           &lt;dbl&gt; &lt;chr&gt;               
 1          3.75 Low                 
 2          3.8  Low                 
 3          3.25 Low                 
 4         NA    &lt;NA&gt;                
 5          3.45 Low                 
 6          3.65 Low                 
 7          3.62 Low                 
 8          4.68 Medium              
 9          3.48 Low                 
10          4.25 Medium              
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Note that the conditions are evaluated in order, and the first condition that is <code>TRUE</code> is the one that is used. So if a mouse lost 4.5g, it <code>case_when()</code> would first test if it fits the ‘Low’ category (by checking if 4.5 is less than 4, which it isn’t), and then if it fits the ‘Medium’ category (by checking if 4.5 is less than or equal to 5). Since it is, the mouse would be categorised as ‘Medium’.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Fallback with default value(s)">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Fallback with default value(s)
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above example, what would happen if a mouse lost -1g (gained weight)? It wouldn’t fit any of the conditions, so it would get an <code>NA</code> in the <code>weight_loss_category</code> column. Sometimes you might want this behaviour, but other times you would prefer to specify a ‘fallback’ category that will be assigned to everything that doesn’t fit in the other categories. You can do this by including a <code>.default =</code> argument at the end of the <code>case_when()</code> function. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span>,</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"Unknown"</span></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   weight_lost_g weight_loss_category
           &lt;dbl&gt; &lt;chr&gt;               
 1          3.75 Low                 
 2          3.8  Low                 
 3          3.25 Low                 
 4         NA    Unknown             
 5          3.45 Low                 
 6          3.65 Low                 
 7          3.62 Low                 
 8          4.68 Medium              
 9          3.48 Low                 
10          4.25 Medium              
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Notice how the <code>NA</code> value in the fourth row is now categorised as ‘Unknown’.</p>
</div>
</div>
<p>One final thing to note is that <code>mutate()</code> can be used to modify existing columns as well as add new ones. To do this, just use the name of the existing column as the ‘new’ one.</p>
<p>For example, let’s use <code>mutate()</code> together with <code>case_when()</code> to modify the <code>sex</code> column so that it uses <code>M</code> and <code>F</code> instead <code>male</code> and <code>female</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># modify sex column</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">case_when</span>(</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">~</span> <span class="st">"F"</span>,</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"male"</span> <span class="sc">~</span> <span class="st">"M"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if neither, code it as 'X'</span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"X"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      M         0.00181
 2 CD-1         1A                   3.8  rep1      F         0.00186
 3 CD-1         1A                   3.25 rep1      F         0.00195
 4 CD-1         1A                  NA    rep1      X        NA      
 5 CD-1         1A                   3.45 rep1      F         0.00193
 6 CD-1         1A                   3.65 rep1      M         0.0019 
 7 CD-1         1A                   3.62 rep1      F         0.00181
 8 CD-1         1A                   4.68 rep1      M         0.00195
 9 CD-1         1A                   3.48 rep1      X         0.00193
10 CD-1         1A                   4.25 rep1      X         0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What line of code would you use to add a new column to the <code>mousezempic_dosage_data</code> data frame that converts the <code>tail_length_mm</code> column to cm?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% create(tail_length_cm = tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm == tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% tail_length_cm = tail_length_mm / 10</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_rank =</span> <span class="fu">row_number</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hint: the row_number() function returns the number of each row in the data frame (1 being the first row and so on).</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the least weight.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Adds a new column to the data frame that ranks the mice</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Does nothing, because the <code>row_number()</code> function has no arguments</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. What is wrong with this R code?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: &lt;text&gt;:4:5: unexpected symbol
3:     weight_lost_g &lt; 4 ~ "Low"
4:     weight_lost_g
       ^</code></pre>
</div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>You didn’t include a <code>.default =</code> condition at the end of the <code>case_when()</code> function to act as a fallback</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>You can’t use the <code>case_when()</code> function with the <code>mutate()</code> function</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>weight_lost_g</code> is not a valid column name</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>You need to separate the conditions in the <code>case_when()</code> function with a comma</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>4. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_strain =</span> <span class="fu">case_when</span>(</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    mouse_strain <span class="sc">==</span> <span class="st">"Black 6"</span> <span class="sc">~</span> <span class="st">"B6"</span>,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> mouse_strain</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hint: if you’re not sure, try running the code, but pipe it into <code>View()</code> so that you can take a good look at what’s happening in the <code>mouse_strain</code> column.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Renames the strains of all the mice to “B6”, regardless of their original strain</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>This code will produce an error</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Adds a new column that categorises the mice based on their strain, so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Modifies the <code>mouse_strain</code> column so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged.</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The correct line of code to add a new column to the <code>mousezempic_dosage_data</code> data frame that converts the <code>tail_length_mm</code> column to cm is <code>mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)</code>.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% arrange(desc(weight_lost_g)) %&gt;% mutate(weight_lost_rank = row_number())</code> adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight. First, the <code>arrange(desc(weight_lost_g))</code> function sorts the data frame by the <code>weight_lost_g</code> column in descending order, and then the <code>mutate(weight_lost_rank = row_number())</code> function adds a new column that assigns a rank to each row based on its position (row number) in the sorted data frame.</li>
<li>The error is that the conditions in the <code>case_when()</code> function are not separated by commas. Each condition should be followed by a comma because these are like the arguments in a function. Remeber that it’s optional to include the <code>.default =</code> condition at the end of the <code>case_when()</code> function.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% mutate(mouse_strain = case_when(mouse_strain == "Black 6" ~ "B6", .default = mouse_strain))</code> modifies the <code>mouse_strain</code> column so that any mice from the “Black 6” strain are now called “B6”, and all other strains are left unchanged. As we are calling our column <code>mouse_strain</code>, no new column is being created (we are modifying the existing one) and the <code>.default = mouse_strain</code> condition acts as a fallback to keep the original values (that already exist in the <code>mouse_strain</code> column) for any rows that don’t match our first condition (strain being “Black 6”).</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-summarise" class="level3" data-number="2.2.6">
<h3 data-number="2.2.6" class="anchored" data-anchor-id="sec-summarise"><span class="header-section-number">2.2.6</span> Summarising data</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F08_summarise.png" class="img-fluid figure-img"></p>
<figcaption>Summarise allows you to calculate summary statistics that collapse many rows into one</figcaption>
</figure>
</div>
<p>The <code>summarise()</code> (or <code>summarize()</code>, if you prefer US spelling) function is used to calculate summary statistics on your data. It takes similar arguments to <code>mutate()</code>, but instead of adding a new column to the data frame, it returns a new data frame with a single row and one column for each summary statistic you calculate.</p>
<p>For example, to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  mean_weight_lost
             &lt;dbl&gt;
1             4.20</code></pre>
</div>
</div>
<p>We can also calculate multiple summary statistics at once. For example, to calculate the mean, median, and standard deviation of the weight lost by the mice:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  mean_weight_lost median_weight_lost sd_weight_lost
             &lt;dbl&gt;              &lt;dbl&gt;          &lt;dbl&gt;
1             4.20               4.05          0.802</code></pre>
</div>
</div>
<p>The power of summarising data is really seen when combined with grouping, which we will cover in the next section.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_tail =</span> <span class="fu">mean</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_tail =</span> <span class="fu">min</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_tail =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Calculates the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Produces a data frame containing one column for each of the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Finds the average tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Produces a vector containing the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is NOT a valid way to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% pull(weight_lost_g) %&gt;% mean(na.rm = TRUE)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarize(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>1. The code <strong>produces a data frame</strong> containing one column for each of the average, minimum, and maximum tail length of the mice in the <code>mousezempic_dosage_data</code> data frame.</p>
<p>2. The line of code that is NOT a valid way to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)</code>. This line of code is incorrect because the <code>mean()</code> function is being used directly on the data frame, rather than within a <code>summarise()</code> function. The other options are valid ways to calculate the mean weight lost by the mice in the <code>mousezempic_dosage_data</code> data frame (although note that the second option uses <code>pull()</code> to extract the <code>weight_lost_g</code> column as a vector before calculating the mean, so the mean value is stored in a vector rather than in a data frame).</p>
</details>
</div>
</div>
</section>
</section>
<section id="sec-grouping" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec-grouping"><span class="header-section-number">2.3</span> Grouping</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F09_grouped_summarise.png" class="img-fluid figure-img"></p>
<figcaption>Using summarise with the .by option allows you to create separate summaries for different ‘groups’ of data</figcaption>
</figure>
</div>
<p>Grouping is a powerful concept in in <code>dplyr</code> that allows you to perform operations on subsets of your data. For example, you might want to calculate the mean weight lost by mice in each cage, or find the mouse with the longest tail in each strain.</p>
<p>We can group data using the <code>.by</code> argument that exists in many dplyr functions, like <code>summarise()</code> and <code>mutate()</code>, and passing it the name(s) of column(s) to group by. For example, to calculate the mean weight lost by mice in each cage:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># don't forget it's .by, not by!</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 3E                      4.72
3 2B                      3.71</code></pre>
</div>
</div>
<p>Like when we first learned the summarise function above, we give our new column a name (<code>mean_weight_lost</code>), and then we assign its value to be the mean of the <code>weight_lost_g</code> column (with <code>NA</code>s removed). But this time, we also added the <code>.by</code> argument to specify the column we want to group by (<code>cage_number</code>, in this case). This will return a data frame with the mean weight lost by mice in each cage.</p>
<p>Grouping is a powerful tool for exploring your data and can help you identify patterns that might not be obvious when looking at the data as a whole. For example, notice how this grouped summary reveals that mice in cage 3E lost more weight than those in the other two cages.</p>
<p>It’s also possible to group by multiple columns by passing a vector of column names to the <code>.by</code> argument. For example, to calculate the mean weight lost by mice in each cage and strain:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># group by both cage_number and mouse_strain</span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain mean_weight_lost
  &lt;chr&gt;       &lt;chr&gt;                   &lt;dbl&gt;
1 1A          CD-1                     3.71
2 3E          CD-1                     3.71
3 2B          CD-1                     3.69
4 3E          Black 6                  5.08
5 2B          BALB C                   3.73</code></pre>
</div>
</div>
<p>Of course, <code>mean()</code> is not the only function that we can use within <code>summarise()</code>. We can use any function that takes a vector of values and returns a single value, like <code>median()</code>, <code>sd()</code>, or <code>max()</code>. We can also use multiple functions at once, by giving each column a name and specifying the function we want to use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n =</span> <span class="fu">n</span>(),</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_weight_lost =</span> <span class="fu">max</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_weight_lost =</span> <span class="fu">min</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 7
  cage_number     n mean_weight_lost median_weight_lost sd_weight_lost
  &lt;chr&gt;       &lt;int&gt;            &lt;dbl&gt;              &lt;dbl&gt;          &lt;dbl&gt;
1 1A             52             3.71               3.7           0.445
2 3E            168             4.72               4.78          0.783
3 2B            124             3.71               3.69          0.417
# ℹ 2 more variables: max_weight_lost &lt;dbl&gt;, min_weight_lost &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Here, we also used the <code>n()</code> function to calculate the number of mice in each cage. This is a special helper function that works within <code>summarise</code> to count the number of rows in each group.</p>
<div class="callout callout-style-default callout-note callout-titled" title="To `.by` or not to `.by`?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
To <code>.by</code> or not to <code>.by</code>?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the <code>dplyr</code> package, there are two ways to group data: using the <code>.by</code> argument within various functions (as we have covered so far), or using the <code>group_by()</code> function, then performing your operations and ungrouping with <code>ungroup()</code>.</p>
<p>For example, we’ve seen above how to calculate the mean weight lost by mice in each cage using the <code>.by</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 3E                      4.72
3 2B                      3.71</code></pre>
</div>
</div>
<p>But we can also do the same using <code>group_by()</code> and <code>ungroup()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cage_number) <span class="sc">%&gt;%</span></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  cage_number mean_weight_lost
  &lt;chr&gt;                  &lt;dbl&gt;
1 1A                      3.71
2 2B                      3.71
3 3E                      4.72</code></pre>
</div>
</div>
<p>The two methods are equivalent, but using the <code>.by</code> argument within functions can be more concise and easier to read. Still, it’s good to be aware of <code>group_by()</code> and <code>ungroup()</code> as they are widely used, particularly in older code.</p>
</div>
</div>
<p>Although grouping is most often used with <code>summarise()</code>, it can be used with <code>dplyr</code> functions too. For example mutate() function can also be used with grouping to add new columns to the data frame based on group-specific calculations. Let’s say we wanted to calculate the Z-score (also known as the <a href="https://en.wikipedia.org/wiki/Standard_score">standard score</a>) to standardise the weight lost by each mouse within each strain.</p>
<p>As a reminder, the formula for calculating the Z-score is <span class="math inline">\(\frac{x - \mu}{\sigma}\)</span>, where <span class="math inline">\(x\)</span> is the value (in our case the <code>weight_lost_g</code> column), <span class="math inline">\(\mu\)</span> is the mean, and <span class="math inline">\(\sigma\)</span> is the standard deviation.</p>
<p>We can calculate this for each mouse in each strain using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove NAs before calculating the mean and SD</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_z =</span> (weight_lost_g <span class="sc">-</span> <span class="fu">mean</span>(weight_lost_g)) <span class="sc">/</span> <span class="fu">sd</span>(weight_lost_g), <span class="at">.by =</span> mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns</span></span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, weight_lost_g, weight_lost_z)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 342 × 3
   mouse_strain weight_lost_g weight_lost_z
   &lt;chr&gt;                &lt;dbl&gt;         &lt;dbl&gt;
 1 CD-1                  3.75         0.108
 2 CD-1                  3.8          0.217
 3 CD-1                  3.25        -0.983
 4 CD-1                  3.45        -0.547
 5 CD-1                  3.65        -0.110
 6 CD-1                  3.62        -0.165
 7 CD-1                  4.68         2.12 
 8 CD-1                  3.48        -0.492
 9 CD-1                  4.25         1.20 
10 CD-1                  3.3         -0.874
# ℹ 332 more rows</code></pre>
</div>
</div>
<p>Unlike when we used <code>.by</code> with <code>summarise()</code>, we still get the same number of rows as the original data frame, but now we have a new column <code>weight_lost_z</code> that contains the Z-score for each mouse within each strain. This could be useful for identifying outliers or comparing the weight lost by each mouse to the average for its strain.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which line of code would you use to calculate the median tail length of mice belonging to each strain in the <code>mousezempic_dosage_data</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), by = mouse_strain)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Explain in words what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">max_tail_len =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, replicate))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice for each strain in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice for each replicate in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Calculates the maximum tail length of all mice in the <code>mousezempic_dosage_data</code> data frame</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Calculates the maximum tail length of mice in each unique combination of strain and replicate in the <code>mousezempic_dosage_data</code> data frame.</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. I want to count how many male and how many female mice there are for each strain in the <code>mousezempic_dosage_data</code> data frame. Which line of code would I use?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = sex)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain, sex)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>4. I want to find the proportion of weight lost <strong>by each mouse</strong> in each cage in the <code>mousezempic_dosage_data</code> data frame. Which line of code would I use?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% summarise(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE, .by = cage_number))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE))</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The correct line of code to calculate the median tail length of mice belonging to each strain in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)</code>. Remember to use <code>na.rm = TRUE</code> to remove any missing values before calculating the median, and to use <code>.by</code> to specify the column to group by (not <code>by</code>). Seeing as we want to calculate the median (collapse down to a single value per group), we need to use <code>summarise()</code> rather than <code>mutate()</code>.</li>
<li>The code <code>mousezempic_dosage_data %&gt;% summarise(max_tail_len = max(tail_length_mm, na.rm = TRUE), .by = c(mouse_strain, replicate))</code> calculates the maximum tail length of mice in each <strong>unique combination</strong> of strain and replicate in the <code>mousezempic_dosage_data</code> data frame.</li>
<li>The correct line of code to count how many male and how many female mice there are for each strain in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))</code>. We need to group by both <code>mouse_strain</code> and <code>sex</code> to get the count for each unique combination of strain and sex. Don’t forget that we specify the column names as a vector when grouping by multiple columns.</li>
<li>The correct line of code to find the proportion of weight lost <strong>by each mouse</strong> in each cage in the <code>mousezempic_dosage_data</code> data frame is <code>mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)</code>. We use <code>mutate()</code> because we want a value for each mouse (each row in our data), rather than to collapse down to a single value for each group (cage number in this case). Be careful that you use the <code>.by</code> argument within the <code>mutate()</code> function call, not within the <code>sum()</code> function by mistake (this is what is wrong with the third option).</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-reshaping" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="sec-reshaping"><span class="header-section-number">2.4</span> Reshaping and combining data</h2>
<p>In the previous section, we learned how to filter, select, mutate, and summarise data. These are the most common operations you’ll perform on your data, and all generally revolve around working with specific columns or rows of your data frame. But what if you need to change the structure of your entire data frame? Re-shaping (with <code>pivot_</code> functions) and combining (with <code>join</code> functions) data are two more advanced operations that can help you do this.</p>
<section id="sec-pivot" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="sec-pivot"><span class="header-section-number">2.4.1</span> Reshaping data with pivot functions</h3>
<p>Pivoting is a way to change the structure of your data frame by rotating it so that rows become columns and vice versa. This can be useful when you want to change the way your data is organised, for example, to make it easier to plot or analyse.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F10_pivot.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption>Pivot functions allow you to change the structure of your data frame</figcaption>
</figure>
</div>
<p>The <code>pivot_longer()</code> function is used to pivot data from wide to long format, and the <code>pivot_wider()</code> function is used to pivot data from long to wide format.</p>
<section id="pivot-wider" class="level4" data-number="2.4.1.1">
<h4 data-number="2.4.1.1" class="anchored" data-anchor-id="pivot-wider"><span class="header-section-number">2.4.1.1</span> Pivot wider</h4>
<p>A common use case for <code>pivot_wider()</code> is to make a <a href="https://en.wikipedia.org/wiki/Contingency_table">contingency table</a>, which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</p>
<p>For example, let’s say we want to create a table that shows how many mice there are of each strain, in each cage number. We can achieve this in a long format using <code>summarise()</code> as we learned in the previous section:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain n_mice
  &lt;chr&gt;       &lt;chr&gt;         &lt;int&gt;
1 1A          CD-1             52
2 3E          CD-1             44
3 2B          CD-1             56
4 3E          Black 6         124
5 2B          BALB C           68</code></pre>
</div>
</div>
<p>Notice how we have a row for each unique combination of <code>cage_number</code> and <code>mouse_strain</code>. Instead, we can use <code>pivot_wider()</code> to create a table where each row represents a unique <code>cage_number</code> and each column represents a unique <code>mouse_strain</code>.</p>
<p>To use a <code>pivot_wider()</code> you need to tell R two bits of information:</p>
<ol type="1">
<li>What the new columns should be (with <code>names_from</code>)</li>
<li>Where the values in these columns should come from (with <code>values_from</code>).</li>
</ol>
<p>In this case, we want the mouse strains to be our new columns (so <code>names_from = mouse_strain</code>), and the average weight lost to be the values in the table (<code>values_from = n_mice</code>). Putting it all together, here is the code to make our summary table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  cage_number `CD-1` `Black 6` `BALB C`
  &lt;chr&gt;        &lt;int&gt;     &lt;int&gt;    &lt;int&gt;
1 1A              52        NA       NA
2 3E              44       124       NA
3 2B              56        NA       68</code></pre>
</div>
</div>
<p>Notice how <code>NA</code> values are used where there is no data for a particular combination of <code>cage_number</code> and <code>mouse_strain</code>.</p>
</section>
<section id="pivot-longer" class="level4" data-number="2.4.1.2">
<h4 data-number="2.4.1.2" class="anchored" data-anchor-id="pivot-longer"><span class="header-section-number">2.4.1.2</span> Pivot longer</h4>
<p>While wide data can be easier to read, long data is often easier to work with. That’s because a lot of the tidyverse functions we’ve learned so far (like <code>filter()</code>, <code>mutate()</code>, and <code>summarise()</code>) work in a row-wise fashion, which is more natural with long data.</p>
<p>To demonstrate <code>pivot_longer()</code>, we will introduce a new data frame called <code>mousezempic_expression_data</code>. This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate. You can download it here:</p>
<p><a href="#expr_data" onclick="fetch('data:text/plain;base64,id_num	TH_rep1	TH_rep2	TH_rep3	PRLH_rep1	PRLH_rep2	PRLH_rep3	HPRT1_rep1	HPRT1_rep2	HPRT1_rep3	group
1	1431.1093720602153	2736.785265345733	2036.3202942858027	296.7441794226677	427.810089156969	320.34926030753473	740.5958429164343	1300.3719827586106	988.2622626407614	treatment
2	2698.8123375791315	4179.715280619704	3859.598566651367	387.542300344862	679.7553511912523	500.4196542194733	842.7161132773449	1569.918432913882	1256.192799853172	treatment
3	3595.7474691009215	5194.217882019865	4128.292348709684	201.07576109923752	346.1541702276	265.6814756382276	1047.1492082902296	1836.7502073977273	1412.7124095319923	treatment
4	2722.5943845817865	4442.164395102655	3727.658653882997	317.4196109742109	575.9318567216695	477.60045319689704	541.3816408390292	937.4239509772243	734.5309707544637	treatment
5	2196.6961468931177	3656.189987096028	2886.794235552652	114.11928269285484	204.591279829647	172.6728085920942	798.0724258965927	1366.354268029223	994.7015553142517	treatment
6	2315.553342480227	3948.823687948957	3355.285792337175	241.55368829282503	383.1822738058106	305.4358892483079	762.836634380237	1208.3035976627737	903.5523689037086	treatment
7	1570.0384053870253	2740.244014900523	2129.3165542521665	294.25644473028194	489.2636596667801	427.46182614173506	796.175797450541	1355.5732090509186	1186.7852431233214	treatment
8	2683.1784589352524	4749.31552638088	3563.340155663348	230.71640700305227	399.54806586257854	308.5662785118385	921.9222771906581	1722.1296640296948	1209.946676223021	treatment
9	3022.5300972452487	5169.6349413799635	3829.510595001945	289.4419064552762	464.21346387726294	389.84623231091484	920.6486130797574	1644.1153528247905	1216.702630038169	treatment
10	3254.4423441866497	5979.9985760029995	4250.2422130096165	336.95792095748874	582.9821483755904	436.099463594029	740.1570420244235	1199.800879443731	908.1493762587955	treatment
11	2622.933492219875	4097.117023948522	3158.048059584204	203.48495979257333	416.99664602057044	301.2566037347214	674.614275595239	1119.5776759334653	838.6418280429436	treatment
12	2731.9670386014304	4799.954369257564	3981.1417573198937	258.44739450074627	410.50718599725275	338.67059197984514	724.8963382466953	1242.4239596595087	923.0039857538986	treatment
13	542.6051666028301	909.51521920318	699.265312181315	287.21521690448685	511.72052430032727	360.8721947121072	650.8790883377237	1081.7149485344382	880.9240705266633	treatment
14	4262.812863674237	7348.993063835531	6319.893761693676	211.71852735218832	381.4008463835529	312.43249534567275	1042.8294793801056	1724.959264519309	1381.3259424352382	treatment
15	3078.506148005289	5890.291960816134	4684.921763399016	413.27081409734103	628.9343052128197	485.14532130021007	745.212101842568	1418.6696415882295	1167.3590929261434	treatment
16	1689.6228312695741	2538.8107094574298	1967.5914518694337	224.33452182866415	345.9898659137374	266.975580411255	743.9816354924799	1190.709094314654	963.4840492410527	treatment
17	2805.621306032408	5205.637549591034	3834.5648450743734	290.6681919407355	470.4786715523202	367.7762033452756	611.001471234083	981.5152651444957	743.3894407144046	treatment
18	4283.366002759157	7923.330121595857	6889.253192812651	301.0191964852899	493.058071146218	365.5447488765637	717.5821821652148	1202.3627277049293	962.6841332668369	treatment
19	2916.234538415369	4387.241579740278	3984.285989289568	222.06793310523406	351.83424465181037	301.0330362049308	844.1623412763415	1334.3513762913722	1031.683329207775	treatment
20	3091.3111464006197	5508.755445245894	4266.925254983622	453.33805349900064	687.153672561062	526.9528482909639	730.0987516465474	1233.911016985175	897.6060684366503	treatment
21	2095.372366420511	3219.5611896322507	2486.062516263903	232.35216109362398	437.2176072217773	318.10617587814505	682.2720305785303	1163.1985940924546	997.1650082683241	treatment
22	2401.52743284353	3312.222871653619	2569.5368668247193	368.942444252135	563.3355249461183	434.11175197333193	695.3003611800881	1025.2059244077625	852.7083063713696	treatment
23	1536.4811745908185	2501.4265651107894	2054.257543218772	367.0828715255788	552.1535242437224	448.26132157121367	975.1714575031991	1596.953543625152	1329.786652428477	treatment
24	2180.2836371643393	3859.1829249787124	3012.418032675163	363.8336928917229	662.9996401215244	490.3891737391876	614.9581153821828	1122.3355953867813	836.9486197063349	treatment
25	3827.1306745710117	6553.700302104697	5191.995040200098	171.14704382866398	283.80029433600066	210.46963154902642	596.0626648723002	1213.907334087397	943.1510076229746	treatment
26	2328.3423678154486	3756.2560468714983	3463.417676226422	152.42184419949692	255.89193147900986	199.08004779298153	847.7398300185733	1348.0280715335819	1072.670718848024	treatment
27	3150.701105820609	4773.7354186348575	3680.484650729412	243.65913684814834	453.45793093691583	304.2487911949715	706.1976963977578	1204.9230751828607	1004.6420635257873	treatment
28	2203.0066369473116	3395.9341511745747	2829.87150609335	350.8481490595394	520.8119386802817	411.44951082137345	629.8133351763917	1105.2678440952955	836.6220044802694	treatment
29	3402.3810258978565	5554.063070894495	3964.4365732832493	182.12368444630889	283.3115150671031	232.51108542992418	822.3294084218179	1217.678678132122	1029.6936606838237	treatment
30	3923.4219235629107	6219.5100299176875	5136.428387452731	229.79881264613346	351.50143000535536	291.30973149806715	787.9597936973635	1202.2380787336756	923.2567571616615	treatment
31	2865.4996954001767	4641.797170103962	3788.450150757315	285.4070578300315	507.1845367549771	417.4561352924738	921.0277890133773	1709.806916868349	1186.0469302964593	treatment
32	2557.5519964854934	4613.05145132239	3584.379722247227	248.1650519135675	405.2150479305774	318.69737390441605	595.5910866303412	1120.0342398556434	933.173215420665	treatment
33	1852.941085504514	3183.154754805304	2536.845443006048	115.40970723541308	210.5442642059359	167.84374634549388	895.7552084426395	1444.0237326512465	1099.4468887256119	treatment
34	1106.879711133952	1913.2338393831863	1575.3303747679186	233.8136062437639	382.942809813083	301.2573871359615	787.6856654105496	1411.903286030101	975.2216170345984	treatment
35	4607.418577412433	6651.954560981832	5611.1686851321565	200.6796157755884	366.8494630073261	253.6054337058921	703.558483405825	1289.0957082865705	1049.2978266077812	treatment
36	2041.5166746539605	3141.056081759433	2600.4359198536217	300.7845891026789	560.7243403500665	397.67626145944354	585.8999567613329	1110.672044424153	878.5286571482575	treatment
37	3618.209100425721	6210.869092139427	4589.584402463483	249.1828260618108	433.42249185691975	333.08239755245063	696.9779005002904	1163.5600244020998	917.089023654707	treatment
38	3754.3061088784348	5879.830723642607	4668.850954423848	372.28309602980005	556.1061460202037	403.44065911080446	785.7955541529275	1349.7255712792237	950.369388066898	treatment
39	2630.155927923735	5477.652683553825	4212.279267652231	254.50455827384317	453.2651007850621	347.2632772143642	452.80238730862493	731.370661044236	713.5004405009429	treatment
40	3820.0138816642725	6622.218400137144	5296.701190765536	362.6476532105271	600.4702959420174	483.72546345233303	835.1971701810244	1252.0758444761902	931.7692210871866	treatment
41	3317.444911388341	4819.831108537876	4011.6411081881324	338.74322223493823	541.1639117210664	458.72413203656595	768.5307207223144	1035.7974065803019	905.6894384158402	treatment
42	1752.5831934600747	3535.017492278785	2709.7714027477045	312.8655934808979	451.1025378845722	376.1897407619868	957.8169105869481	1376.1462446510404	1180.2479301552255	treatment
43	2292.1704014274824	4573.4585147437165	3327.6997053240593	228.91895335949013	389.37861221006125	318.3257197452914	796.9476002679129	1424.7756993390005	1083.5303963019387	treatment
44	2913.5278585473925	5407.7312999923315	4304.198115546382	274.75872248870587	465.87861972100376	343.0135000187401	938.5176989976567	1569.6585290926123	1318.5021396069494	treatment
45	3538.062357103726	5375.387470093297	4142.398785866782	314.7564137260585	493.47014169370857	384.1934667881626	822.4472090594417	1444.096552563819	1089.1671853404885	treatment
46	3905.0381128439444	6286.931427141668	4906.18397927392	268.47964424572234	453.5806016489755	358.90297395874893	668.4292669722594	1049.6270700057134	850.8365437402277	treatment
47	3070.786637853136	5088.885996536528	4021.286015647545	169.66619026152054	299.1882530021367	253.4848431032774	714.2739864745827	1240.5713555544976	1003.7537148033296	treatment
48	1751.5991350305662	3079.2998902876884	2356.8283342120135	237.17833569064948	404.47153642707644	305.2392840740477	823.742825722996	1362.6658488729126	951.8843224643307	treatment
49	1666.6053267688442	2732.281297023989	2278.096431895711	264.9695063151044	445.9647762171061	380.40721093519613	736.7941284494266	1271.7461788656321	974.7246875838453	treatment
50	2448.588566584274	4115.418698959213	3077.222506705183	358.11682701658134	539.9975852237641	449.3053308196942	848.0842698130446	1465.9357620275557	1131.0248525282996	treatment
51	2395.2851569237405	4292.942949045696	3606.5783420756984	201.98232909275214	337.0588643495265	297.3537900132228	737.9560192844364	1120.3630889494825	926.5925510168917	treatment
52	1817.0167623522811	2990.4329885579905	2393.9575874798707	246.2441301185262	375.6049857155472	311.116608532017	697.2967098842774	975.8122852237492	833.9655615375813	treatment
53	938.5393602378407	1462.1737425022152	1300.0822565197946	392.8904009440824	665.9456527077358	489.34739927476437	608.0559950189999	1187.3598520652174	893.0595499490021	treatment
54	1967.51739284738	3005.480809985654	2818.06836060541	291.84990033492346	487.82879093825284	367.10873337045155	749.862988048302	1272.1710356934877	1077.7248245594812	treatment
55	1720.295621076275	2989.1040894053117	2557.9079705441823	343.11828816002566	551.5439967734334	504.82033845399116	726.6553367635336	1192.034266108541	1051.1394101325197	treatment
56	1171.6815436231743	2001.924217944249	1667.9314894466663	325.8813242433968	487.17491891878086	392.6622443369318	714.8645569140889	1338.9508899648959	1101.1758885552165	treatment
57	3101.3433668038383	4771.427205950882	3633.0692698410007	222.3376030939682	386.43769708272833	273.94560510952266	611.3220552385451	1291.0817566495089	1051.1390813920832	treatment
58	1475.4568850895778	2727.996185879896	2262.562487595748	158.35963788318458	291.3044176836351	221.52141092552202	611.0737089114839	1029.7323424583556	782.4235639002304	treatment
59	2389.5698053197507	3877.4441275985646	3219.8742780246616	270.97711597874206	427.83982298950866	388.088934325081	861.8633239960225	1343.6426564197156	1222.6809856885222	treatment
60	3731.0440930797413	5813.027268494195	5349.645437865403	326.3671024082803	590.5488124508207	410.40013401789145	545.6167672009876	964.3272653042593	759.1240740186988	treatment
61	2916.7358024170703	5313.838640552356	4195.205496509589	304.4237794960036	453.83830941121505	346.0513064992218	592.4793780639477	1102.263465648559	871.6426131378236	treatment
62	3340.277492888499	5412.555005135462	4641.9239371565445	282.11153577329554	445.5176429378855	339.54540278986826	633.5790902988344	971.3058898490227	793.8530254418926	treatment
63	3180.496438004197	5432.841903447049	4469.650963610329	330.8256556905929	526.9930784410303	422.49304385746126	605.3783777921822	1158.637349689302	897.8550072468923	treatment
64	738.1745849861917	1346.786764711306	1080.8855096005173	281.9528712761793	552.109604790435	443.7779383956466	654.1284032461742	1043.0698807133845	858.9432134943312	treatment
65	1749.2917992781775	2990.0575463788628	2571.5803769006625	485.2238218761115	805.6216341185185	641.4226133442387	722.5397981299491	1229.9890704398508	955.7622414269567	treatment
66	2463.3011392135772	3661.6792775970002	3051.738942539642	265.4041935303847	449.4913437513679	306.75077638862825	777.6923653477719	1311.202858454924	1053.7371824625568	treatment
67	1880.1050069445446	3054.57154100925	2380.3912517928966	171.51249359235524	286.93197671527247	226.36363794591043	920.133825671762	1549.6127225480475	1297.8435244780512	treatment
68	2925.7446267487835	5402.3555521286235	4567.010243342454	121.76063002288805	184.8161589675075	151.43931622340637	656.8407453563154	1312.6533814044196	900.1026191506539	treatment
69	2682.830410755193	4541.2302825814495	3822.0338052234215	176.5504724253199	306.1721908592055	234.10853328602957	790.9678138128698	1395.230530391891	1129.0883332865471	treatment
70	4745.511040965952	6995.959095071418	5566.771102671899	305.5343341399925	461.93197272907497	369.5927485588191	948.6762330361512	1447.6110110479758	1169.342357778443	treatment
71	4140.278369602492	6298.290673456161	4971.474214835077	355.7985570319953	577.2820512536515	474.1096699354619	958.7306534451866	1457.4583271426693	1188.005458198724	treatment
72	1884.394783702158	3368.626640902926	2708.783632395904	270.74476902961726	456.8534364868011	336.0395724494072	764.9260674881273	1189.945181932799	991.4287951203316	treatment
73	2595.4512564628744	4593.745891557875	3480.1454927918144	128.46958085572865	208.6574354398531	170.60267966729432	601.7853982651012	1134.5970130624755	945.5127018111517	treatment
74	4189.052705262475	5952.896918274161	4898.467040986155	405.2743284489554	706.8717552079631	567.760003475097	535.7539454912559	915.0789977935664	745.5876398779735	treatment
75	3097.199068101482	5263.789182026348	4334.11019595239	282.03060350357964	407.24039780476704	365.03085966614	803.5304552943966	1349.1794818523185	1057.84658734574	treatment
76	4850.6436686541165	8891.116487465988	6927.86609379311	263.1547272609346	495.9735959583356	395.0025808738905	784.2288233779177	1180.0122326993803	972.1147892178716	treatment
77	2213.01819929253	3508.0327480431183	2898.1546414417858	345.59020886271327	537.2904115933362	411.23296749598535	742.4346661769029	1346.103121607763	1069.7952385310575	treatment
78	2521.312903436817	3878.80190591254	3442.608466750794	283.34632828769713	533.9737510777657	433.3234236141036	745.5539691187217	1172.2127595015602	945.6901069718732	treatment
79	4776.018527211103	7126.865831840239	5773.368864382195	316.98736585832006	512.6878224127573	433.65089677932406	784.2664578818271	1468.3159410004343	1095.089435216519	treatment
80	3471.260724261353	6014.493894333743	4874.619854602394	387.49668146422357	579.920085073778	450.7998335151751	643.7176349348894	1117.0611384068993	922.2807394768971	treatment
81	4165.196096134203	6514.304738451957	5371.323928871316	218.7158225441569	327.4456867671096	270.24497644319655	837.0590950751874	1288.161506521298	1024.9271717476558	treatment
82	3449.784752191189	5528.46063242446	4791.6585843802395	269.69657680024847	461.07396015496215	333.6379970738681	766.3184016313809	1270.6209269861497	1046.9027685131264	treatment
83	2350.7501647518598	3845.5711986080787	3163.8799439496506	325.2982648227862	534.4631433205003	433.92701976760355	769.567505963163	1347.1370770358717	1003.4682842974129	treatment
84	2393.1706273213317	3726.70394839869	3287.329335870476	346.10363950851826	587.5034567224502	506.51141906939955	817.0701345389531	1306.6710880536345	961.9037843892017	treatment
85	2966.2659540842974	5002.255145688521	3981.2783456413185	266.58255596925414	424.1435377499269	323.3832205640167	769.6002915530894	1225.228640795811	905.2028757791671	treatment
86	2754.9287757873703	4116.712200420984	3172.3079387869975	74.65276648228559	121.43074662277378	99.4675110965198	724.0700668908939	1200.4394324354535	970.34458960104	treatment
87	2972.3471109789216	4880.207848013685	4027.8295695322554	303.48980654780627	455.702768203201	377.2071213732583	865.2940688261732	1595.1601720696556	1289.5991817109875	treatment
88	2777.867861027632	4965.288178075193	3575.710890227176	264.201925264385	426.0829069707306	336.87367571960925	789.6734150018282	1326.6716408239522	1051.7844848506418	treatment
89	2733.782304795157	4365.589261907352	3227.8683955037745	289.3648161513095	469.9168440552683	355.7976686283159	679.2101161923318	1046.0529511013797	827.8808012618597	treatment
90	2576.334526857596	4820.165489959923	3769.3099098807984	431.38837375022433	628.8770190634293	544.5020289164352	771.1137757855994	1446.7277253873424	1091.2756687416513	treatment
91	2876.6549442756245	4557.447561242684	3979.330270529862	370.6091048830341	559.4086654477355	458.7671437178499	928.677085181722	1317.12148392527	1150.606890380244	treatment
92	2281.3799086067406	3855.814226064159	3018.0973924428613	191.40301741500375	339.8506813736784	270.9872192890809	877.2142363956569	1265.9641740930037	916.434771745593	treatment
93	1875.9024195925576	2737.712775479284	2421.5046325517133	235.89409791852705	374.1998990689471	303.27750056542703	786.7440978121994	1327.5104298577035	1049.3777298534367	treatment
94	2804.6073290023	4108.635368605844	3161.5126578840586	300.4385302734163	521.0622510514357	440.54131742737445	640.1211239208048	1056.6848976331196	809.2821952129442	treatment
95	1440.060045140079	2067.144454403515	1773.8293638654497	367.70785085317596	641.9221699777148	505.63281405475664	859.1560437932121	1459.5631246861972	1191.8201069591748	treatment
96	4350.2818245637245	7715.860214943853	6325.730779140983	228.47308037428706	463.93989359468327	340.59165144709806	950.1415586421441	1611.874789308926	1397.0403979378862	treatment
97	3041.5838074881362	5017.368770998141	3346.0120024841694	126.211335207203	209.18627735033022	169.7190230285009	948.8410383232729	1493.559441971886	1104.0962129249986	treatment
98	2587.1194534577926	4270.6915137186115	2936.7379150265215	224.04700697682733	432.8612362840743	322.0475597740687	633.2234843073334	990.0436602114964	793.0076827965822	treatment
99	2250.6843425711386	3679.6042324770065	2863.743043549651	147.00135062488442	250.92271123083202	193.47448808897127	783.4356232498213	1243.4530708557022	1054.389958851894	treatment
100	2359.5615800785017	4035.8916291301766	3130.828874550296	282.76557555029774	447.62521113915903	333.87295046710483	781.2950622764157	1343.408707336175	1055.7961726267206	treatment
101	4701.267104265655	7456.1073311902655	6166.0895963370685	208.5319793927896	349.6453771734236	298.9550201567339	775.9713215980286	1258.5793304312049	1025.3692541275643	treatment
102	2766.6385582561547	4352.470392281443	3513.4102477814613	316.61237402955516	600.9178128738884	436.008945287397	737.1830504711413	1195.7213108637504	1011.9678536143429	treatment
103	2679.8283713003552	4029.0184736236934	3399.1779538020937	311.5649972379724	472.77504521800137	370.86456291517834	871.6427569225616	1548.5666475061837	1265.889138537164	treatment
104	3194.55557296735	5399.942300844081	4368.555106155623	261.17727669170193	358.4965249584646	318.10984129638933	675.4262721511037	1040.9902231418241	869.071103149673	treatment
105	2699.9950714710726	4955.6528901635875	3869.6757903475914	173.9048948664786	290.66974485708516	235.9509204556836	667.6882733694895	1238.8764227686936	982.7380020140298	treatment
106	2773.368949061404	5415.1854758028785	4321.091623905225	286.2018387815785	464.15352005013096	382.194565373236	805.794106749135	1246.5789572750737	1038.4327133786262	treatment
107	1441.9795872465386	2732.08603990223	2112.8789841436314	141.86889884725386	293.23088520990103	220.45098366776108	459.1320810850905	921.8122068062731	697.4507047856489	treatment
108	2337.1785189893008	3829.711054997179	3166.30858174962	259.7467344594302	457.1890859030027	369.52305446578265	890.4057075227284	1420.8632420297783	1169.880624115584	treatment
109	4154.735114591685	6309.069210894991	5204.489760480696	126.3608395425695	247.04349878613854	182.41083100896202	888.0311497597235	1352.4729998049272	1086.4319161283915	treatment
110	2971.780765885467	5561.165072952215	3696.0643167680423	211.78972628093445	351.8646488084047	285.6979095242058	785.5445565858455	1306.4071967835957	1077.0337733714443	treatment
111	1981.6969577068576	3131.4865906327377	2307.4916163653725	254.26591753340747	397.2838402577943	372.9835831622139	721.5965549758339	1057.3971180106232	883.26679650131	treatment
112	2789.3003179618513	4618.911692949168	3934.51438888081	184.45730607331012	312.60240503780244	253.33936847352356	693.551813634895	1137.4084683334445	914.5689380009505	treatment
113	3501.0352593564066	6600.714653166506	4485.363642047199	296.7281758589603	490.3596458563445	396.6749538509565	776.4433498081935	1260.0812758159655	952.9313534905784	treatment
114	3071.1406976229537	4788.366866792436	4594.630547173657	209.0736870043439	351.4860986858513	299.2593845413677	696.3168486843829	1083.2623843119609	873.5945050744828	treatment
115	3480.737094460424	5038.378493783125	4107.8973322729335	325.8584335697911	580.7344732341498	482.13630076150577	823.2698220509596	1539.7541966398464	1259.2029129973932	treatment
116	2477.297523483397	3948.7603863850013	3227.8214751344285	268.344158320307	466.67736904747636	384.8192737124606	759.2551197561769	1303.7005443844666	1105.830533802183	treatment
117	3122.2339316857624	4892.529068106065	4218.150781290914	292.3457004932575	485.4114435085184	341.1159755137587	815.814194105152	1289.83183377393	1112.4602099894546	treatment
118	3331.6300327872186	5716.057328310132	4622.492969183419	388.5223737889639	640.0057482878493	522.5211424836323	904.5926440393679	1419.8205281595583	1089.2492703199355	treatment
119	2627.5920641523944	4531.982792484903	3397.920167759747	267.6781506401361	549.3173581183964	393.6018815275862	912.5936135395592	1543.8380376429407	1101.6145683372822	treatment
120	1038.8527787042196	2014.1416740824689	1518.623314261295	290.94571470036726	482.2094252359859	331.36971073889197	517.5275159652876	872.5875798921843	705.9455536099963	treatment
121	1434.586626326041	2484.4908798800025	1941.7894551749587	378.33190482578595	528.8475290842599	481.271079219115	806.4315629708226	1429.5308896695942	1117.576827256733	treatment
122	1999.2400136657427	3547.7633936992206	2729.80965050071	402.15444951113767	681.2954035313547	497.7097404643701	798.6229591616659	1237.2881300567255	1019.284394367559	treatment
123	3239.32707365119	5963.574807021373	4123.485594854615	225.34634954289032	358.78106938788954	314.1712498859363	858.2661371801362	1292.340261657471	987.5890599369791	treatment
124	3415.7997172136534	6385.642503069374	4909.735267087163	308.5772737543231	508.69282415704146	431.0703960700505	809.5239630739782	1334.936989094967	891.0894403430123	treatment
125	3047.8487053152667	4425.798651719017	3637.5914286816787	264.44504836933453	431.9127235437203	331.2233631285507	760.9108373796073	1351.8815275047289	1095.6077106411612	treatment
126	3963.183850218377	6383.3283092189	5055.784683402852	199.84933314175956	339.0758014535198	281.51163888996706	560.463151749945	809.9252590837538	668.7648578138037	treatment
127	1485.2345128532045	2374.571269313785	1942.5581441203672	212.30882438469538	344.8124493966279	283.1473523526593	823.0823786230106	1362.7386798310326	1056.0960457133226	treatment
128	2309.7434844075356	4006.158440387534	2942.1425506441255	296.8257062699077	469.65864476889027	347.6580041754423	739.8883975178096	1255.7797783450335	1008.7831474344919	treatment
129	3900.2151741963985	6594.753577286295	4866.567984518977	328.04445516930605	543.5490069446354	440.253087027539	744.4825978895661	1306.4259140342913	971.9308320311425	treatment
130	2112.692316622329	3113.175080393346	2435.983361192045	387.8865809956042	637.20752906381006	490.2640928968164	827.6235921101866	1421.4219400397187	1149.1897297548223	treatment
131	2740.383773135589	4249.359156993929	3846.228912459982	325.9173246777774	554.7969779687431	455.80229062719314	657.8653737258044	1075.2568603084187	888.0449578183407	treatment
132	3264.282902494544	4777.627553390432	4008.7059348404273	184.94627140921523	317.63080816368904	281.4789992998987	849.8615275201458	1379.0385535019093	973.6564075945314	treatment
133	3231.680176362815	5228.656316185845	3608.200936856974	226.82889647226895	308.3138609117283	249.34395401196684	708.7658996557174	1197.3079609080678	922.2981680157164	treatment
134	2146.404477732357	3300.3296118778026	2617.991314591962	267.6725824640255	422.12477108938003	337.4538282719254	596.2606164248414	1065.4899516489616	861.512323917538	treatment
135	1935.5702428655973	3666.4830248790036	2963.8541481922152	301.92488820816556	440.94095298492533	354.5375298091997	625.1376512838369	1082.6615163807069	902.8516146748065	treatment
136	2822.0546434925222	4947.376317642514	3840.1102830461264	400.55532463741463	778.7654553990083	583.0149307007807	925.8478493691415	1529.7120877741115	1279.8625832426453	treatment
137	1753.802012782444	2955.723061228919	2041.0233424799062	329.9835437686991	578.9135100137463	463.3575881447996	787.6708724131983	1283.8136477545868	1061.7581931267794	treatment
138	2431.871142614543	4064.8893128185764	3304.9220897562604	359.89548182211274	591.2767472549357	474.8822902385791	779.6575156603135	1321.1768726662383	1040.4803632127637	treatment
139	4380.105664725823	7293.58151825399	5624.901840990561	352.33870273611313	541.4109007164631	427.7143122056421	796.1131867490399	1274.238594133512	1093.1030873894658	treatment
140	4153.795479712162	5981.865992063436	5666.778939794403	347.9079691545323	568.0353853035316	456.63826001291386	674.3958365003123	1100.7750883913866	930.9862068517924	treatment
141	2106.273530238951	3121.0428155851205	2884.3062011585216	287.0556755934974	486.95805645120794	360.21306247197157	893.0220064291002	1369.9894789253783	1040.549005465007	treatment
142	3010.8834941737414	5075.600071030867	3927.7727893218957	311.4349824038862	529.9568328437964	377.02764769683904	637.6057202266596	1230.2433282164304	803.1988789707947	treatment
143	2697.8675620643257	4359.786133262855	3553.998235996423	244.25534362134783	380.07942232335614	303.1597003883619	718.1476813237475	1101.0718623270889	884.199120474213	treatment
144	4413.589829518077	6192.748982413661	5072.942167658403	282.8563398497751	406.4215190665667	382.56773381600783	774.8397891348616	1341.0507129226319	1098.147312719395	treatment
145	2671.6194498147106	4479.28895534636	3666.231730352508	259.98264639251187	470.6117701589679	392.2721736768241	1059.8655653545575	1605.5139451674354	1342.1875337316076	treatment
146	3528.32020057895	6292.180973114843	5068.09137298085	368.4817940277569	547.5749214259396	437.7974327072596	679.3325134224218	1114.3235101061232	848.9423382030583	treatment
147	1355.5240259201323	2462.298011361512	1895.8141380462873	241.1702895296745	447.2066374970535	340.53545790087304	842.281158403119	1341.6372002756282	1036.681516670354	treatment
148	2545.36794573656	4441.053886654437	3496.8174482545387	363.33928346488517	588.3592235333441	464.53402179620235	885.7607255559308	1366.404728133559	1156.6319942547404	treatment
149	3224.9970011716264	4999.287654984887	4094.3267833318196	469.9007550770525	748.5349642032717	573.9688243442544	730.6002183477963	1267.3550005034215	1034.314553850682	treatment
150	3162.8541547118384	5671.718103336311	4627.613277851763	239.3983681142161	399.70216195729364	324.53452826884194	986.880592287596	1579.7940250459444	1369.7100529963418	treatment
151	4477.159442955364	7264.162536919994	6592.069223974882	358.24282245509977	591.4013346673189	460.5397967954705	778.1167732801265	1264.2083891000639	1035.3467205865866	treatment
152	3098.889947628785	4737.504124032651	4052.2001048227166	417.4962212193127	703.1125582254064	487.32970050444584	807.7494657199633	1307.614758164845	1066.8766572832749	treatment
153	2167.6844968531977	3538.4482728276007	2692.017423215653	315.513856737981	485.4023617808056	443.580894185211	562.265550079047	913.4927308154548	764.8045174016364	treatment
154	1776.9379682879603	2888.8427784603477	2203.0666613490603	189.41056202028113	327.8226035774551	219.678659244655	712.7557708669051	1167.3313133290055	1017.1773429619655	treatment
155	1475.414448856222	2858.943844216804	2004.1716292121002	314.03848058030536	517.3397474858006	421.95115896959015	713.0893990433382	1168.021076339278	989.215684014689	treatment
156	2009.4543063569533	3355.380602415072	2643.4453710712182	348.52169533496266	603.5119739342795	469.17302339701405	988.1478520572682	1489.15688782919	1242.2057506423719	treatment
157	2790.382196287239	4292.166848157649	3372.08587432366	417.68430306352735	779.0385544425471	627.7934019514194	693.8709399993915	1164.085002566093	893.212862020942	treatment
158	623.6654415789983	1074.1525224548552	825.5498138441743	240.84400557620762	401.0293033078913	271.48843917326724	904.6520742636387	1402.159656953486	1189.5506245099098	treatment
159	4215.413840532042	6243.056889057381	5320.776598750984	338.0125343608948	586.6333799570606	449.3951172426282	820.5307311017028	1433.832880944834	1122.7202131700735	treatment
160	2141.6299336245634	3626.3112373984454	2948.5961688883117	362.0298321493772	610.3572150586263	469.734340543666	604.802946951056	1192.9894740558234	940.9203943712142	treatment
161	3123.8023735196366	5318.10724599163	4306.576203949726	310.3407544847942	459.1035453287076	368.5655630567443	738.5436321285335	1233.199650823379	1009.4798344461441	treatment
162	2329.8421191025345	4568.537039779587	3703.5651765201214	339.94063138291585	567.3240916543697	429.65284728296587	686.5310710281686	1124.5507967632157	864.0793397376574	treatment
163	2256.925162985053	3722.7485228173045	2691.954366663035	249.4783939891402	428.677042578735	307.6380825468514	753.4619698846276	1198.0283481165436	1026.7555449702236	treatment
164	3855.5856518724477	6765.168529557027	5090.989512527458	339.32475590859576	520.8750427855028	414.88961037450673	966.0769878563704	1591.957213465314	1207.6517814505216	treatment
165	3349.7013017531926	5194.895934366841	3858.9871897485655	217.70501911275517	376.02483143805824	283.22050193674636	739.8390523138758	1231.3918153924915	927.5520130173123	treatment
166	1527.5380546416593	2543.555628618371	2183.5676976402538	384.49755423948375	623.5156358207673	432.7232139280436	738.1912046510834	1225.194011973163	1085.5860114812187	treatment
167	3652.4671022627695	6042.442690454258	4772.322432351161	486.6987320795758	728.2954334925165	598.3553326939124	732.8969661607111	1327.5812547944433	1121.10161349493	treatment
168	3275.085898131942	6855.725663618429	5299.363039805023	256.9110560820893	422.2747754374006	337.042277219508	617.4657158241175	1007.15490009973	867.3069726182392	treatment
169	2460.699701673644	4144.6557134187715	3240.7451247644117	285.2341458201342	487.7072368607707	407.93100989852434	892.1422187728375	1410.979946662188	1134.7319307049768	treatment
170	1721.1425638507085	2475.245980646392	1979.939421868424	224.1425820127355	452.93711143757935	365.40733041102027	747.0641121790364	1366.9851814973117	1091.7641280475812	treatment
171	3409.928727343928	5480.420534912251	4417.071287233253	245.3803098426146	411.4512086172711	364.95769650213504	723.8743465926128	1271.6820705118132	940.1677202673002	treatment
172	1778.8331203135617	3144.887387942501	2651.773209536827	241.89336168469444	438.8975392632435	302.3319237534356	986.977177597423	1613.8627108368246	1424.8928267068684	treatment
173	2470.3675389751897	4255.935370287151	3687.6107258843504	88.36072438926178	143.61218263244407	111.26969861952036	500.74960067459267	863.1221386124616	734.708325806658	treatment
174	555.3683585847775	843.6121904274872	706.6820939452044	305.4623068883679	519.470605748654	370.5310189649527	776.7055383082821	1421.3720760732544	1213.6345713995725	treatment
175	2489.6303802764082	3996.950520023257	3167.4140247444	219.68982839343138	378.3283200260975	294.246075424143	772.3032166239185	1212.022239177245	971.3232501278538	treatment
176	2367.989855708655	3999.3975215899463	3339.503771029508	218.28285833526897	356.5795874786934	272.55592702161556	766.6141696689817	1204.54096043319	1070.4469169826652	treatment
177	2784.483659495619	4720.43600671865	3804.4249177969095	276.5317763343602	420.47197820906683	347.77410636706105	716.2433783819556	1259.2707169995026	934.8967424736755	treatment
178	2957.803061792118	4429.836203605202	3411.761121329149	294.7572483576894	454.8227121678262	362.8614572737252	564.1669591111978	1121.4983767243702	753.8430874999084	treatment
179	3959.387024540418	6229.678380598581	4923.555472259655	250.162272812998	389.96142244087025	285.51090529580625	597.009616870199	1027.615783834395	861.6432678926284	treatment
180	2943.556747321849	4790.38522728985	4358.806187079195	347.63711552510205	563.922872939358	471.7812428690191	801.1640351777149	1464.2836850965984	1058.1189648308298	treatment
181	5623.965541990376	8802.396236227238	6273.659879045796	300.8418743995285	461.9767431794199	359.18660980142437	952.8037105203673	1396.7384715331484	1113.6307363954295	treatment
182	3007.0338822047497	5927.697947220317	4771.399562830302	398.79053719201687	596.820320830666	495.7079166227146	675.0130642930526	1154.0930802442763	900.064378483013	treatment
183	4479.583776730593	8059.721990838022	6749.679501506878	315.85853928741693	510.8964437272861	418.2053117984103	983.7204607717882	1509.137007651956	1064.952812377249	treatment
184	2410.847374647987	3853.030914753757	2912.3362942175922	350.1350255285699	559.2279639007529	440.365568226411	807.5234991706308	1267.2028215108885	977.3526416547854	treatment
185	4055.2055581111117	6455.480362786697	5654.7179013940695	96.06368486433611	165.322350609766	125.57941161186197	683.1533343027392	1236.785528757434	959.8890182875505	treatment
186	2529.888990490864	4719.120194252577	3725.5761578184806	190.25949399421754	328.25870114131794	286.0763173633484	694.2364640908528	1164.9666118956263	1014.2804958800025	treatment
187	3631.3281152621325	6576.31590131629	5190.606293581381	235.18310141122754	435.26717413565524	299.2959247392728	568.5603620347015	890.6876663415875	687.8247212474109	treatment
188	3721.3341447539347	6238.000253584462	5005.285171665315	244.60879662689672	391.36584534973724	326.53992095186203	786.4887006612613	1265.589262225645	934.6079089751113	treatment
189	2665.395144172634	3819.5535655067315	2958.292298858535	230.4163906710621	391.48013273172836	300.6778351329294	740.3405795677082	1527.4627940938017	1188.1301702535857	treatment
190	2801.843130258892	4833.635634312886	3768.0175447873385	321.9833154950182	523.5807832969265	405.97948026037807	770.1231635355801	1074.2026510473536	876.7854663832577	treatment
191	3922.4408115857304	6327.61455157732	5428.447133496582	274.2279181751988	521.3203614307714	391.1641668146398	811.327375514444	1336.8856189780095	1024.4600892956132	treatment
192	4464.544144534952	6721.118625146547	5964.020468082625	142.4932208786597	249.26059311986697	187.93236641678365	707.2974484064015	1246.8966847031322	945.5599742753493	treatment
193	3602.070058284851	6295.802576467068	5096.864196216055	279.9945000906283	468.72095050911497	364.1263338792111	778.9659072092605	1266.5377668341216	968.7098963791102	treatment
194	3285.941523127866	5876.63704441218	3926.7656097974996	225.60976975159159	405.46117412948684	317.43750425259196	923.7767704632083	1502.9193062007728	1266.1442948322904	treatment
195	1935.8732979637568	3346.892184543983	2950.5191109406683	386.11607351924414	531.3681087314482	434.81929217959464	791.8474546262589	1219.0421573573813	1070.6008902849767	treatment
196	1535.1796115082593	3092.1294283026946	2436.9982330550356	226.50076755533107	380.13098113843495	294.0356253711002	1047.9969576981803	1667.9911335779864	1349.5305476970104	treatment
197	3622.029245624157	5772.700448467904	5093.487097638908	282.62711195022473	437.1285019645794	333.61366536654765	618.9541043006304	1232.5620359458442	912.8795823936764	treatment
198	3356.203447809158	5511.2991711348795	4577.978264555206	303.98273573388605	528.696748767576	489.6348920106746	716.6484833327783	1415.9517915568274	1090.0253575669865	treatment
199	3076.168316621247	5048.373401054432	4209.804411894194	303.20575176873837	481.9087139727381	404.98715435974646	755.6735217478813	1290.597897747038	1025.4891257541135	treatment
200	2799.2374583032174	5149.191122159922	3748.7909350113728	389.504804450116	610.9673014808882	537.2757778936965	840.9954106755877	1355.8437432476867	1172.5498610772368	treatment
201	2866.6162354523444	4998.179273512792	4032.3331539745914	306.52578830597764	487.38042223890744	365.6099935055561	792.7643484135247	1209.9876460466187	1143.9881492585323	treatment
202	2849.97560559462	4258.803857308354	3755.6685450960954	261.3757997735775	481.2765146830031	361.8797518473694	836.4693114569579	1344.818447573874	1067.0965821206303	treatment
203	2616.9454936145385	4251.704600892785	3685.8720115452556	331.35977266984946	519.7776946305527	386.324610485541	690.3899716289151	1130.782767078567	925.150568079756	treatment
204	2387.891800263277	4332.973525101137	3586.9161675548276	378.4219199566765	673.7263639998607	473.25568391793536	812.6577707495991	1357.511974243634	1134.7022989421262	treatment
205	1459.2998602984514	2423.407689787165	2073.7564719129546	252.03147772822155	403.64299946350906	309.8743181822997	925.9342332202898	1298.789586276744	969.2236555593171	treatment
206	2216.714170473147	4205.258216091621	3109.4900711522528	260.86514167678575	424.4545351700101	300.57524967119485	576.9271048861868	987.4062943826294	770.3347261162804	treatment
207	3239.4810926448317	5197.575435226473	3930.101864325455	385.4117230757281	654.1425532767231	507.7464767523896	623.3138116914736	992.7595038979235	818.2024882166289	treatment
208	2113.3944599433116	3806.600062460747	2917.9039600025153	227.7195378240042	409.77003612106364	321.64658727819995	773.7525757590863	1502.3398922200993	1106.93898108435	treatment
209	1688.3777502996472	2849.8712378651026	2265.8311827777115	408.5890603888181	712.1884725399001	510.12310532448726	983.7612463093553	1463.5076472084834	1083.986864904413	treatment
210	1567.8577546109402	2639.909006090477	1853.4695043840875	291.15543559807185	453.1281035351412	409.0807041854564	548.4179035337287	908.7230565138904	719.4496675183719	treatment
211	2740.426673039343	4863.203050232614	4497.398844201113	233.96187803795036	383.44990014689984	288.2788501422748	701.3270557156443	1275.702545278885	1102.7184838409112	treatment
212	2244.6155039528153	3600.3606697859245	2976.3336721847	282.63092963157567	435.28153006939056	370.5696282694735	712.5360359383106	1226.12429375875	1037.3600139572222	treatment
213	2076.4102690417753	3323.863287187339	2635.4550904580265	301.1493381996863	540.4272771166413	417.75923060840495	791.9166631063862	1303.7900757617908	997.7137180670887	treatment
214	3132.9554255516127	4970.143034588721	3982.7827913337424	211.85808327412173	329.6865391388349	270.78605639225543	675.8984960715902	1144.0535736770635	911.9686132532353	treatment
215	3160.144357001997	5067.325977858288	3784.003111591749	361.24311835542005	602.9701290178442	447.90747006941575	769.4316276416399	1389.471006407216	1164.7050145922788	treatment
216	4222.743548785217	7065.726467159759	5250.894922661728	314.7183686252282	487.5438442468327	395.89793167519946	796.0974406391654	1305.8643161221542	951.0463343945906	treatment
217	1617.8615904430017	2224.929379634833	1820.8884047138824	227.37193603140975	417.69582312846495	358.1444034505086	838.0821042701907	1428.632675595589	1177.0992422435809	treatment
218	3113.966535896698	5245.83222227022	4251.562461518445	253.3183674430895	461.93761111689514	364.9047374216178	690.7781919526311	1231.5647011371548	1075.3882774027277	treatment
219	3190.771668983332	5759.604901039559	4474.805409221926	325.1661826023398	533.8311555123719	450.54253157175765	717.0421477260508	1313.2228732680937	1047.3667455140398	treatment
220	2371.4393111845625	3391.332911939652	2669.779943191821	306.465584561715	514.7731760635893	403.94682869756906	758.33499120816	1296.213751039796	1006.4500468415519	treatment
221	2486.805551899376	3766.2984691090874	2960.7496635368	349.6008361659723	542.2058809811125	490.04259664560624	654.1580784560905	1119.3715306242038	814.6591002031512	treatment
222	4098.779920595117	6345.831027889193	5025.2806000015835	263.5353235289874	440.16856597747585	322.8853523949542	561.7264274974161	910.4600598450827	765.2569127330529	treatment
223	3796.232778195434	6112.977270301277	4890.383951279249	289.82858516698184	454.45229927713024	376.7902660643015	704.7169814267636	1196.292787715952	1017.1715343059061	treatment
224	3782.207067936628	6009.115165432223	4652.735359916334	244.5507291163082	407.30534633242104	342.4669455683641	674.2429074932727	1111.8600863667646	872.3828965059018	treatment
225	3610.200370485655	6479.9243861934765	4924.1672187652885	287.3867712569881	503.46803055685666	358.7204179376432	654.2724750171988	1148.9554705590651	878.181533121926	treatment
226	1787.3040251124808	2725.4551194211595	2431.507798051132	296.31419937012987	479.48542305062585	399.9519917698868	747.0523231615297	1236.1236349240812	895.1104816671501	treatment
227	3091.2656922704036	5364.668070431964	4684.624182905844	320.30172471023036	588.7345998196583	466.1946650482704	815.1887016066643	1416.4832371526386	1126.5105068913542	treatment
228	1072.6470921457126	1747.9338528973162	1372.802934423189	305.527285900214	476.15807531412696	416.5943626717294	616.7243233954624	1037.6788995212983	769.337525199751	treatment
229	2617.261581125033	4312.116712219707	3364.941151622474	321.8201491299365	510.9427735589387	387.20806046072863	944.6007467498916	1212.6054128959602	1089.4627645544915	treatment
230	4152.891962371304	6421.30291756914	6102.436319988776	304.8948596736201	478.19471249388096	390.59507320674464	836.4014283142371	1329.6319937838293	970.0805668047296	treatment
231	1823.9197090824177	3163.7495263273195	2758.6797145529663	320.6651006080685	650.0574255023186	475.9840314340549	903.65447095745	1491.672264571899	1104.6166790210118	treatment
232	1258.7114605685033	2155.731906066282	1801.8652883169082	220.65270310037704	339.54897224772515	306.85989580759815	662.3725554475892	1134.3135256991375	892.4553126676645	treatment
233	3043.2898415128648	5035.147790407218	3666.8243077668467	138.00892795724994	247.41075204729938	216.4472135659884	645.0624515776648	1088.7827066992227	864.4723747196728	treatment
234	1591.7904704597584	2570.663024959089	1988.6725356810607	377.6765504651576	654.820170157311	522.3162992313072	701.738352014944	1130.2756695280686	895.8544329770095	treatment
235	1904.344988090119	2866.2386304686247	2541.667280839009	199.83467907465734	318.4477813522646	264.0981766255478	932.4409916108929	1529.6841801248793	1260.9405210267346	treatment
236	3080.8642747180897	5582.225343351895	4382.070643540719	384.98300864479484	637.0787149481828	473.44621401709753	724.5116662355679	1237.3697200211886	947.4214676109824	treatment
237	1946.7284913439087	2991.4934350935478	2670.079094794061	264.4858378169146	415.3452837328491	352.71810785848936	781.4406062096544	1145.3348231069353	1047.9787093267241	treatment
238	2302.0938118805625	3562.596171188779	2939.3963379728534	202.91638661227586	377.53131249101915	303.3798608791365	848.2285089714809	1273.6613917634259	1168.6067917784992	treatment
239	4023.7779642169485	6502.498337701992	5299.343863546569	222.48773305607457	393.9984924900216	342.4187153471388	898.7831598071716	1389.6700601734335	1090.5893650838955	treatment
240	1446.1135067333757	2482.617806568394	1937.0243242693298	357.61107678356524	556.4610933672377	488.61794553875063	873.8636859794266	1317.4445347475087	1167.9488445515258	treatment
241	3621.0054038312496	6022.948336022616	4740.235898470944	203.66869467999413	365.0747527826954	291.7054534526733	786.1926601023504	1240.2784841055834	1001.6157446434094	treatment
242	909.2867999532464	1794.055673345111	1359.7457506440855	223.61381264749613	418.3565525305698	282.5978914274172	711.1361779946305	1238.339055066618	961.5202978452204	treatment
243	4400.096186136479	7444.518050256134	6139.411637272558	238.45933125442357	456.2172040825925	368.8712755159863	756.4211479090247	1279.5867438421478	953.37832799255375	treatment
244	2537.365465401056	3923.4189411961524	3182.6456714821074	352.1190175475821	552.8609160139526	466.53886571546315	738.7723977988057	1252.7056650148927	1078.7596210759614	treatment
245	4857.995250141132	6975.288787682207	5138.289581707573	266.1181119694924	425.531047872998	367.6848087189949	695.4095170366755	1221.8715227366963	924.8778134050974	treatment
246	2609.94163375572	4484.3523243623295	3258.7461764292266	165.03843410560577	293.917907418397	233.1799215917459	578.8937892264668	1052.0807803531375	777.5949855881854	treatment
247	3661.4967328121083	7015.209100385405	5258.361164504546	201.70861808651642	347.53019865016245	273.6154726266211	686.4028164229483	1077.069318613897	887.5109310514847	treatment
248	2929.32378301621	4755.474421098448	3745.558251645052	320.8412852206709	604.5878210108002	502.7931301822381	660.2831373265249	1069.4855761311287	845.6635803447795	treatment
249	2659.8322529825396	4128.7662755527635	3569.64176857487	143.25375200434635	254.3796510516312	180.1992672991325	767.6183581579546	1259.63506282111	1062.5281698544447	treatment
250	4146.855608028452	7935.622439851654	6083.690145670506	272.04150137612964	489.069686385688	402.7130498744447	531.2010645345073	946.2252970134397	722.75932285662	treatment
251	3770.015566396504	7063.143326444022	5271.360981257405	356.43215361904817	549.1894741558639	495.9147547728837	861.7892314397777	1402.0781158938373	1063.3120298212486	treatment
252	3442.639288291342	5831.030156657163	4663.336273466703	408.18997772217824	686.6611558907649	550.6873982184777	575.8975479297392	1022.6675027449731	775.6386776282233	treatment
253	1322.7340929115755	2166.737096725166	1756.8563432358321	372.4547610237133	536.0897632992192	463.65287015534744	576.855151028014	968.506596505262	745.4624537683627	treatment
254	1888.8644754841123	3162.6379120725796	2678.5930631865053	234.6296485945882	326.42704060288264	297.2062848335735	727.6331600366041	1137.797719711972	919.6318988568436	treatment
255	3191.070538885123	5662.23633116582	4290.02003455456	256.899352208817	444.6672232622963	352.9605708593319	828.50164714529	1259.2552081605095	977.6818852554536	treatment
256	2446.3546708765966	4668.458768201255	3717.712805137349	159.43692027287017	297.1119365239532	232.86375815608227	767.7343123586082	1220.0223772552104	912.3740169830386	treatment
257	3526.663154218999	5984.561282808464	4721.093212249772	539.5260982026215	820.5990565607938	636.6731824255055	682.5774683176388	1075.4126708990605	780.2028251742836	treatment
258	1143.0782984460898	1907.0367075466354	1661.933002697519	341.9456969567126	576.1633373617461	461.41934056750745	564.1594930017582	1092.5801720680936	805.8565808052845	treatment
259	3037.626944616022	5344.535893704566	4194.811569263921	258.693954173264	452.47122368958185	352.7301064945715	694.5301052048848	1074.725994120922	909.6183836622522	treatment
260	3127.900933698892	5635.72634784715	4005.9557153862343	298.63206505403576	533.0530131271106	396.34674089462925	815.6363004674016	1360.949108239829	1174.7438018585806	treatment
261	1971.9010819943662	2911.897397903227	2274.476615014033	352.72728703095856	564.7477588611863	450.5807537285774	777.4406478244331	1394.5351046134692	1166.7992089446861	treatment
262	3099.681968652359	5353.112645774042	4347.999220788081	302.25277841288164	506.8308423217661	415.58881662693904	906.4543871865712	1479.0225376919689	1223.148513822026	treatment
263	951.2064886305401	1604.1735687126022	1256.9722398145107	270.87246177495183	439.6886784887767	337.73382819476035	990.7610144928336	1534.1586865937927	1206.9847560891183	treatment
264	3817.3623963578398	5819.309996891225	4999.288384056986	184.72126117565531	327.4246766161617	280.9214595228489	503.87925309232656	897.2894924560879	741.9172416211862	treatment
265	2592.8854706815146	3817.703609267675	2900.9750292129575	448.5671595293399	642.1265057727849	530.7507920800808	825.7307034296467	1434.5442676253788	1233.7079139318666	treatment
266	2666.0898572945966	4800.180373540009	3958.9853189801897	207.59142046211372	356.8911772991821	273.7617434067891	603.9881962058386	1212.5686600837757	901.4589386525081	treatment
267	3260.709756784345	5733.760364468802	4227.669367623305	339.25623131422026	540.6497291398455	464.3907793654835	762.5159911997022	1181.9841349121277	994.3176666377022	treatment
268	2682.09201916726	4549.723437861066	3738.2452602389544	263.23060125344125	453.12687486432634	372.05861719124175	879.5909052950144	1337.4904119357311	1155.9455235817081	treatment
269	2183.1859450817883	3734.889283466126	2979.500167897146	418.449498254356	627.2733684049327	516.2062664966933	922.6324558300802	1614.6867992688092	1264.9647416556227	treatment
270	1238.9210134487287	2166.2561426881452	1808.2581374360343	310.9457557363871	488.8151839137358	437.09264999812086	813.0518416397226	1204.2198477128147	983.2663298184596	treatment
271	3241.7542531259546	6088.200375532264	4804.571846552475	264.5592606796639	437.86158102701376	327.16093389078713	728.5254092683853	1128.688888226	945.439081096306	treatment
272	2089.6156600313084	3431.3640261907	2623.826125445407	232.51704072093438	379.0620345763159	318.7396337302124	707.3473526075923	1295.2095139614169	1037.797777486168	treatment
273	3829.887160068654	6603.2048471935	6045.890065359027	362.18389920305844	586.9560462666985	466.20270105887414	944.8531138965134	1557.2888263806444	1268.7496755742022	treatment
274	3632.383965551818	6078.257670523448	4417.3594777260905	362.01842593275825	615.383522746349	527.8369375548525	658.2706843571958	1142.5158143567458	914.5715723764953	treatment
275	1996.4305025394506	3164.612173763971	2358.076640761347	214.60311723246673	322.44063561545426	279.1900185040621	544.436822935323	1067.5693927631396	879.1904057127687	treatment
276	2168.196170395621	3695.6811537461203	3199.4827416503854	341.79691041928965	494.8696349929391	394.560809089578	1060.8425450733025	1582.8423811850325	1364.9210186442729	treatment
277	1712.3660881901924	3055.0852589805672	2574.7736109286916	271.71003982913743	443.79398966355535	349.5730210369509	1079.016675173225	1641.310465650556	1349.8173051508206	treatment
278	1796.315175734072	3086.6975117200755	2759.0832270832975	345.79335667678754	563.4157104985319	442.3989217935145	673.8427540023157	1073.6222337938211	777.8009323000208	treatment
279	3138.5191893462725	5322.8650255245775	4368.53744271	275.6663496788407	460.6772650800282	340.0857624409859	764.1319108618152	1316.142406647382	972.1775233060254	treatment
280	3714.606339490174	6114.492978063054	4836.975397667253	234.12772404488422	386.7830598497221	311.6247246724345	555.4899168380634	950.3817891415217	725.2466977568621	treatment
281	1837.2506805912935	3001.5963754059226	2495.953395919578	168.85244853223324	295.6905198495561	218.50284630539903	819.198613618361	1266.7017079904442	976.4979141955589	treatment
282	2480.5344220069746	4340.2574901185	3689.4802394168814	322.02536683326514	541.8446663113882	486.4244568967659	792.3930117980298	1301.802657451245	983.0345745161064	treatment
283	868.8389145641622	1407.830815223253	1152.7738541564509	318.6745442071871	500.3039918163874	383.8038587789528	834.0300369745349	1370.4351074190627	1226.5936275652602	treatment
284	2395.156102147767	4071.3321338566648	3510.0384316483633	330.15350421276264	542.2026473074096	426.6209942120444	849.9363527936035	1338.9387892574127	1144.0357379721331	treatment
285	1885.0572991839917	2880.054840002317	2335.4705813681076	249.46626584735594	416.9201490005571	341.1288266370984	548.2408301750904	919.1568748608966	733.5815619176243	treatment
286	2758.339728980365	5167.9006576517195	3918.208108441972	328.69323775056824	521.8314448488321	450.8080541949615	757.6156482491513	1218.9293633509658	907.4977825861184	treatment
287	2901.2301504410507	5174.549985268954	4427.5184846893535	319.400450490406	503.1575472669748	389.4637703637854	839.7308962627224	1262.2700671081375	1065.0150766929423	treatment
288	3382.6626499522213	5602.947343363269	4253.854143947675	327.7576131438132	543.5620754116476	467.17319796178134	754.5817375696836	1247.2676790197124	982.9710046522123	treatment
289	2246.7749756459875	4041.782460006389	3222.1172176967143	364.9032390019333	624.9113553034522	566.7983131498065	532.8973806451324	935.4731320455114	762.4163818717153	treatment
290	3183.9874033301617	5099.940627364704	4583.688491983889	120.44979120411307	201.9598977457137	162.78526242842705	600.107247907633	951.617895739501	779.54350859604	treatment
291	1442.3366510786952	2464.375749836978	1897.152413621843	385.16425859374414	614.634743943348	477.4764803085005	647.46191083824306	1202.6176714863661	894.7690845099219	treatment
292	3834.330558489601	5695.8606595125	4836.398111387906	235.5363710946496	423.6851659227645	326.31036548762773	720.2302506208172	1325.0741423467339	1063.190334827777	treatment
293	1614.716853548663	2784.4824560428997	2287.450669410783	283.62637610670737	514.6315866317619	400.00795615868674	773.3044160992578	1280.7179182172788	1020.1351870630727	treatment
294	2383.978466125026	3460.4745597903543	3114.8274620245	347.3358712424549	651.262629057642	490.90137327663456	896.4648305364375	1329.929392982824	1119.3700414637024	treatment
295	3756.2589766041488	6133.796065763803	4773.498348164357	333.2684223646349	508.48222972368677	402.6501959777005	571.014284853911	909.9823058618962	672.6330732037072	treatment
296	2872.27997233764	4903.6850029730995	3506.865822837983	260.2467939378215	474.9410554988091	354.4542516499022	757.726990736034	1372.0965377779487	986.1127029484719	treatment
297	2720.1557265259485	4619.949851439949	3668.125243453125	291.7614121181122	542.6520532071629	425.87931954379343	967.275120167995	1527.874438934329	1336.2032205936077	treatment
298	2195.811074292222	3122.8950965397134	2710.310599820717	240.01210602908083	409.7795789148043	393.4042608205925	737.9513298229194	1377.4407880809408	1071.0392015457212	treatment
299	1770.550077617053	3139.0017860775815	2472.8043332824827	265.2389876605066	425.17204637744203	400.49628328052466	895.5131859801552	1589.310120632645	1259.0402717568347	treatment
300	2942.3418129612382	5116.669934613684	3972.6104630025043	274.4865644508331	446.6828197658102	347.793154839706	554.6014539439014	953.2667591962024	749.6486876230739	treatment
301	1382.2961690783857	2905.8310765390925	2006.860135396557	398.4446477584182	679.0747760869929	586.03354233791	838.6093735042442	1325.2082769390522	1038.0638533141575	treatment
302	3462.98825711719	6067.885761625371	4729.739171653583	169.09837373469114	315.6265128318119	273.5771975244164	729.1031808578293	1217.323505636689	985.9533632715365	treatment
303	3573.6046689784607	6557.491166910402	5283.164078533555	252.83562535385062	432.2862820934896	363.9282933926951	865.4043363212102	1349.145734181746	1044.43355561336	treatment
304	3513.937265917433	5385.622224406558	4377.890401929289	323.1354487867987	496.5603383490433	401.12320672960635	780.8445687377722	1364.535934615052	1044.782083205603	treatment
305	3326.3305001107606	6039.714479273237	4289.026182040943	289.4953275728702	542.8169398832937	428.3278294549843	854.4768986333341	1415.1571397324537	1058.2730180695241	treatment
306	3457.6119108104253	5396.193937476062	3896.5483118515726	179.94125331656193	326.35073503789096	262.18097508033543	751.0694069411389	1271.5250167338245	1045.5100019402785	treatment
307	1491.527791817128	2979.7469325910442	2298.108167309533	303.53038872652627	545.8028222361822	429.6959914015241	1145.0761291793453	1976.333292488769	1470.2030888712063	treatment
308	3247.5297409191426	5665.5294638411715	4408.153016408479	327.5246981098204	511.9800927917918	433.28175454534255	862.5069492946029	1425.9073934521343	1092.4585216282667	treatment
309	3927.582113913688	5921.136434253222	4212.58160784479	277.29115524997184	497.8053945310771	414.44443722957806	745.5818835915048	1488.3052663520984	1078.461188420195	treatment
310	1441.4809474258836	2364.443669971062	1964.9980218696894	230.02621927593495	412.3747134680483	326.35285942615275	818.8531057734784	1303.3807231551464	992.5392212715525	treatment
311	3296.217875260893	5681.531959194044	4751.06487857187	318.7379594780169	480.94603432184016	383.40769319404154	833.4913549848028	1417.130857347432	1105.5572941020773	treatment
312	2641.1187557875796	4036.586528591119	3237.291927054555	285.72030747434303	483.320363298108	365.9472095850419	536.7804975100167	995.263739650496	745.8875889703339	treatment
313	1271.6578867842575	2367.1469477315454	1832.8158045983912	347.7443025021261	552.7044480688229	481.78592609826865	804.4236481119041	1344.8185151304378	1154.3976767687477	treatment
314	2762.0886602958853	5198.982734723641	4320.090218415762	379.1615020974273	635.7910697018405	519.889506317109	717.942624838172	1228.2737321013576	926.2375912992233	treatment
315	2634.775611615633	4351.284591707857	3755.7738946634354	150.79310416404414	270.1274730001093	199.872060795701	743.853106699062	1347.7100071537698	1020.0513716697343	treatment
316	2359.163028255525	3647.7711891032177	2854.8236958478187	375.5575338833621	646.6977848683904	564.389149299167	702.4997016723248	1043.1245481917897	894.4129012975699	treatment
317	1575.4169660549026	2402.986350317597	2014.5597737404698	306.50876357627783	592.2008177298281	465.9714791124779	682.3325521426037	1283.7742313762828	951.389647209951	treatment
318	3909.5760148562995	7052.137914520147	5644.128602312304	242.58243359023928	440.66495791209474	332.55667578102464	830.2395266674406	1320.838902387328	1166.466947785138	treatment
319	4878.712752142218	7498.255709632916	6272.522760761215	172.68679140439852	304.36672539980987	237.49119793946983	645.8112339851289	1015.4732477920234	866.4731701769812	treatment
320	2648.2547242999863	4792.088563865667	3768.041050892706	316.95444251677475	508.1834173864189	394.4618737860198	797.9113339819379	1117.7127226939015	986.6997400099353	treatment
321	2928.9729488938806	4407.617272916727	3851.726045192354	259.6790117839683	436.3926698546412	362.0914734213279	585.7636962422893	973.9098317583808	774.4636662107314	treatment
322	3753.85334669345	6055.503822792459	5115.731886774752	231.78800674753143	426.59681037829864	370.89876392502185	788.9541182288484	1255.0229820987752	1081.211888709915	treatment
323	2641.7414729661405	4670.546196405593	3413.021603338099	249.87838271573966	401.0403832760515	316.41209602012816	457.03735908409516	775.7401480056823	621.6171701553743	treatment
324	3070.302939110467	6065.501356186503	4790.39116836391	294.3760152880126	536.3462147110735	415.5146364544575	765.6261246556701	1192.2125329646565	882.8897197073695	treatment
325	3160.8770360554417	4805.815670580079	3617.1320487666844	336.32715311361414	534.5228561008996	449.6552611788843	734.7335706483406	1213.8511426533735	954.3974848564282	treatment
326	1674.9613499160073	2890.7464362421774	2312.051940930138	197.75488550915796	350.35587285179867	283.47294362258737	808.0753099038874	1291.1251780689936	1050.7253431807183	treatment
327	141.31335873857185	227.21278784483732	181.71511584896828	295.3415984714646	484.6996119999628	338.5211677070437	729.0489927227537	1351.4484594879302	1043.1946367284165	treatment
328	1709.686267484753	3226.3201072764914	2619.023225601558	266.62914148958976	459.3554300108221	345.3492522214491	724.9158802049868	1317.6448793245195	1094.6788494230962	treatment
329	2873.251542293508	4923.709713110213	4287.389545356264	343.3326426285266	538.707404793254	460.5955479922417	879.6845898202295	1547.4805491383968	1257.2977537003014	treatment
330	2664.5719810180617	4866.567979800444	4240.358368735494	252.65546988393288	460.239827317221	357.819813181129	709.1421179695905	1212.3968713444963	918.3605463814	treatment
331	866.2447776521632	1622.1171614671487	1185.618377820612	268.7504470509996	461.29183909848103	340.54473739454335	975.42621310518	1640.3236932182858	1363.1147681796817	treatment
332	3587.353057673548	6402.9553390518895	5102.403650378114	243.92372087556987	410.5928643183428	282.92682743402503	765.9061524050301	1414.9449174650338	1084.1421266309464	treatment
333	2176.7870171813715	4202.374123204835	3399.5832924251044	170.46636525313872	262.03232842856323	226.61850675864514	737.6303359199494	1181.925903754466	888.6012899744302	treatment
334	2767.387112664201	5782.834093463219	4047.807478609967	250.77546193627794	455.84535797843864	346.43907495149097	878.7996219460991	1674.2314677886711	1244.4485154113495	treatment
335	1737.2114862222113	2684.572913862717	2197.6809867460197	296.1152767835152	467.46184716646155	393.117969671609	580.7595138714578	931.9763019522228	801.8839751381415	treatment
336	1770.8337133281054	2730.302085841173	2346.591692607711	245.07559136603808	436.2082198602967	305.26388207994404	477.7068148820634	809.6408582833855	578.3485170401968	treatment
337	2274.0331681819575	3965.5782289778876	3007.0800680206703	287.1253049278975	405.54136784639115	332.9178550411764	635.916919523798	1208.122870986091	947.0421896996521	treatment
338	2801.2683076290864	4976.941021049555	3795.1098945745803	246.34484150876543	424.29725184099607	321.39362716041666	856.0576496102566	1496.2957349579015	1209.5494770392552	treatment
339	1413.4841876377866	2180.475747436835	1607.803247559102	219.709261006416	400.35483582568924	301.08561324625566	707.1699061083631	1311.687324417272	1008.7601820412739	treatment
340	2782.5032597076006	4935.72902688926	3927.858859815593	235.4784360803317	395.221949558554	307.1440943516	777.3573466415921	1140.8963594551349	943.1430418955323	treatment
341	1693.47248229007	2576.015354979953	2255.4918279200187	367.7026145585234	499.22367877636214	445.0777824173194	668.5329860095702	1133.252273635324	942.2605981338717	treatment
342	2483.847237414346	4369.170498504262	3088.719756396661	336.4811562543499	533.3924516148418	476.96925691230416	711.307750018119	1272.2615509668974	975.3699919395555	treatment
343	1830.8502104593388	3487.606222918581	2802.0869807855543	254.25870921775783	405.07313677891307	310.9540993903825	691.0424081112712	1293.0866517744814	1020.3484241279044	treatment
344	3977.7113730164174	6762.821937912409	4983.051010250354	284.8896749956274	455.740411075851	369.58875866857755	663.965950492047	1025.6351088971883	817.8396985033231	treatment
345	2389.0102039244643	3798.7012730926294	2819.266708723055	202.13614959601935	394.17511777294555	311.527899543279	958.774757875351	1681.5785599907365	1294.87505415344	control
346	3244.281486102799	4862.64825637294	4080.1763543396164	210.4119660564334	398.69857475030403	282.35453711370093	770.5057074977846	1252.6514635612077	1002.5915306069871	control
347	587.4600109885937	1004.8906252776417	815.6443252713295	198.56495306675401	386.1003171901096	326.6183309532847	911.9666354267541	1500.1429802466585	1223.3381347625352	control
348	2054.2825218209055	3124.426558939565	2753.666222713855	248.75736533185804	418.5896778327982	367.82322260280046	671.9352557275934	1072.773389186488	882.014316331835	control
349	3369.47514332738	4920.446866095723	4178.522461228587	174.7674049016795	267.2151495339175	226.0755713441546	595.9136854044383	1136.5929253405418	962.1693603458735	control
350	2334.2503741029186	4505.733997760311	3522.1433888424135	277.6218069886729	432.79835331349454	348.7802415175686	706.9325813087141	1272.6101857158105	988.8717711880696	control
351	1301.5410157825747	2481.910111172318	2049.782216796468	218.6281371534675	315.9553548671123	261.6052029410885	920.5803958039601	1375.0379682333487	1248.767384111455	control
352	1948.383739797403	3339.776051211277	2751.889060918833	225.03546262126176	369.95402936841106	330.5176266856468	636.8658587530276	1061.299164861404	902.801667148285	control
353	2419.8075456897986	3906.206715172017	3283.7722744876837	228.465324631526	390.97879364713265	328.242852629203	614.4730175635883	958.07522244429	752.4379053828958	control
701	2015.1182751299154	3284.41787517618	2831.9149159760013	235.9820041731285	395.3460050077633	335.2483880579183	658.7698302611711	1223.9672939352936	1009.8825717924634	control
702	2823.9301515216966	4859.728962193689	4037.545309502784	216.62930596887102	401.5711368975614	323.0601965447758	870.5418528241465	1327.0011708796171	1042.7391593954198	control
703	1176.6841266539543	2052.866336340319	1599.3884817163475	174.2875027163055	286.26516325320256	237.17912518473878	673.3231628801097	1089.9727117139134	917.9856901741772	control
704	1084.331218689796	1866.2237086473792	1418.9531036863148	299.8630248290024	462.6522707200744	401.3926258732863	1057.349621095431	1562.2620720104574	1382.7099721678537	control
705	1877.3302798227026	3176.207595533262	2496.994843886734	161.0636896710535	238.47352985183454	221.12450170347546	839.2995597319571	1617.1248297602883	1266.4656020105172	control
706	2448.770591601197	3973.045827323002	2970.7920324087445	283.73723667193235	457.0956118301232	424.0666601196063	732.5040088468429	1197.1201243327982	929.5180388257953	control
707	2789.876753665831	4204.083867666073	3693.820810315038	242.03897783026136	456.9615448884857	346.7686950384486	840.7180709099588	1474.6610678451552	1161.3576488135084	control
708	1952.2250045479977	3312.2581730887714	2675.904359182051	252.89360737742436	375.5114488990306	327.18268529955327	776.8582452447816	1249.9280357892555	1018.0357000746983	control
709	2759.8504907410256	4471.467589068388	4128.778478243655	251.2378223705395	420.63793263777643	360.83454955312305	606.4265718760489	923.8495582383811	751.878836789187	control
710	1508.5618802111112	2575.405125882996	2019.8039181753047	252.34482344046592	414.9307066542737	328.2463970248622	949.0504492130493	1459.916604613713	1165.6018389125502	control
711	2177.1464868441076	3842.4872280593204	3005.053064643935	240.20988796303547	445.49009638234156	348.52931623544873	676.3789409852299	1173.474302397297	934.084635323525	control
712	2260.8163518137007	3892.2299387809903	3625.9549526440137	254.27508475581936	469.40888597904933	389.4476966578205	772.5459715617634	1310.3421731573021	941.6238288916827	control
713	2591.931690657268	4534.710311096555	3694.7951761368136	165.43067242192961	280.9906506735066	210.38280108850552	1001.7523453134455	1696.0820207210293	1347.2686961166203	control
714	3221.3978642083134	5761.795708656671	4396.038385828289	241.68080702566508	397.43179031033543	336.0387780709185	621.2953954441834	1088.6843455003834	844.1412069770315	control
715	1221.8149183826342	2140.975410534185	1622.4457471731223	275.5175461749865	431.33472721861943	378.1428345075016	707.4188388970786	1182.0910004404136	1017.8300690163834	control
716	2247.091523568617	3974.8727638917258	3213.488500089081	207.83624797180113	388.3741107892201	276.38345130036123	836.7806819365488	1453.6359488494695	1162.8496405745682	control
717	3059.4804555076776	5357.963577364287	4123.748501640518	173.5847052183539	287.4361053596048	230.09168986304078	759.5567194336938	1492.5174906089583	1125.0603986523302	control
718	1055.6571622645044	1836.293599939875	1422.9915568836236	232.1419080486298	356.19309549818746	343.1660790591211	721.7575009991925	1170.9896580188652	992.6622419511203	control
719	2669.9736648628254	5784.583794932432	4457.806044433606	221.6912146565042	360.1922631108991	288.255938722085	926.7247439188731	1735.5419632425442	1263.6656570812552	control
720	3250.6971758031245	4979.36762380073	4095.186614852603	314.47668057640794	458.8571934572477	419.79608441651993	631.2504200233224	965.1075121886277	891.8300084274744	control
721	2561.668773544801	4169.4883088920415	3017.0073215704138	302.9877137220883	526.2004571011126	443.59082387806393	534.8663888632079	989.1733010842345	725.2744642797106	control
722	2643.9910187090445	4452.341712774285	3346.5371011213056	251.9359284968713	342.98243010833784	312.2212719966681	855.684327716433	1538.788955752123	1099.1777945255205	control
723	1801.6936926674302	2606.160110261837	2261.138690966079	198.02273633373244	333.86682008105055	283.0925047007879	759.8908401700895	1088.529502397928	802.1662577571497	control
724	2945.0944464975814	4545.294358267197	3857.181586412775	189.99942501022815	316.164371461136	249.6014109224453	586.3281082636514	903.9203625519293	745.5517100803129	control
725	2093.882370558672	3328.0710708096753	2446.416388540739	247.93151686463207	374.76116770808295	311.45173681524307	974.2672452282344	1504.3320992123963	1232.535680523138	control
726	2371.544126206789	3759.2497830194466	3181.5438852494694	231.99571601894678	361.39361625561554	260.9758618391176	616.9436087031552	1104.339560373419	861.1004614550059	control
727	2341.5277828155954	3681.9839050358355	2883.6939700621315	301.9422221016533	434.31567389616083	364.4691264102309	877.8660019158066	1373.6166947826612	1073.3429386172882	control
728	1771.0060805230385	2723.7780737997173	2298.900442816493	262.4295010588375	482.3778010825804	403.36491201261026	634.7484695469095	1268.1542777196798	977.6319260298636	control
729	1530.8423009638504	2603.0099954999473	2177.3598918233824	221.47750826818802	355.7116876525897	272.08417578318495	804.5797823414059	1377.281442808765	1002.285064711648	control
730	3803.922074933568	6901.846811089926	5379.295167247046	225.4795686170737	352.06981294238636	276.36791965222494	596.5175892634023	979.7113167621229	760.9202609455552	control
731	2463.6354741881237	3689.721645059899	3056.9457005215613	175.41432735208835	327.76674602215905	233.08127842997533	667.7218949890224	1022.0685563773593	775.2676400854223	control
732	3187.088761730298	4641.836724885764	3826.8656266573744	236.9878519630222	401.5658404526182	319.5475910024599	581.8848477037476	1053.686664678594	916.6673927408133	control
733	2157.880331396009	3334.243022762464	2718.568696885821	191.7448815666057	364.29719103709664	293.04979722079383	721.8592256498303	1209.733055225454	943.107827118729	control
734	2340.291553199021	4005.641700115059	3048.1782687296354	275.44843167131614	468.3544373434715	365.8176697879364	602.4422584965298	978.8391204453584	789.8410383592122	control
735	2422.3569968884967	3812.6149504762834	3311.586672936257	281.9317945248115	421.69476404721456	337.77584201676854	465.51662133035484	861.6485578428715	662.8907050640798	control
736	2524.281171997879	3750.5518490055806	3084.274323434772	314.18570436977427	472.8495245653308	429.88342165915475	713.545675815611	1251.8965329331063	998.5485034613763	control
737	1376.439706257162	2506.86499391669	1927.9524840073175	203.307175668448	351.1851948204462	271.6731627084085	688.5751789101711	1178.737066939637	976.8291338062288	control
738	2690.1213827326087	4358.350120072783	3647.3121293236327	255.4693480556789	410.22227639274513	322.2320551979999	607.1783068915902	907.831053896696	770.9555865261357	control
739	1834.4467430962745	3151.4316573230994	2642.6583466043908	262.7623525317023	383.40225961198155	340.21294625252847	929.4581438155425	1651.5013655904868	1355.5446789125106	control
740	2988.436079547069	5626.311461748352	3917.6197895234386	179.91522186907568	298.17517522625116	214.28699016291853	887.4118414149315	1520.0245205906863	1267.768343162141	control
741	3227.7387932441648	5632.23378069279	4207.265624875781	182.22665621378883	298.19735040959364	245.16799298922814	567.8544648589809	989.7725427240819	786.9084298605586	control
742	2181.5709975216864	3509.5239354381074	3199.3395472317916	183.50945195015905	288.3706341828948	234.59718735386224	759.1818646448138	1177.2516983596665	973.2929608125337	control
743	1806.285775575974	3590.320357758408	3035.972871480445	296.38479999635604	484.24665736416733	373.81738218637594	791.9767156690361	1226.5797123161688	956.0045094145298	control
744	2622.5303894862946	4488.4860282598	3539.4590325864133	252.35516888076896	423.0717098564153	335.3127123100376	517.7496077251728	833.3777450243932	777.4439171789392	control
745	4226.965984135882	5930.281669209752	4577.255842243892	230.88361133666015	320.3104254847113	268.05318818229495	602.8692674412687	958.3358511680216	727.3381834032303	control
746	1799.5939866115793	2924.8669259876715	2426.437799794752	216.82886227983263	404.2435807975879	318.15197461993216	935.2549366418663	1545.2820871807182	1245.9281185759346	control
747	2315.816305698798	4178.65494422875	3629.1729795854467	204.05954519066887	352.2825050211099	298.9585807018038	826.3269249418711	1339.424389171994	1033.1127944765958	control
748	2635.8291459689335	4412.6310001014	3481.406977739959	193.68420549270255	351.77685516987475	279.7415454194125	754.927445328078	1284.0356493926479	1030.4240559599673	control
749	2911.9586840284987	4744.072302231943	3468.3740745915884	292.00344505021894	492.61798850244514	362.2766846797735	824.9945859941045	1244.640937589317	981.4761783850746	control
750	1793.478880934039	2903.59720506881	2403.644679012863	202.0860742255983	362.9963868550959	254.3051815046833	814.4170207672327	1327.1554052252825	1131.1152040359302	control
751	1554.594669937884	2732.030037319756	2056.1682758154766	263.5750759833888	482.76088274312036	353.8430537989948	576.970799332291	918.9628821347517	745.7247267321186	control
752	2392.879060198346	4138.278743666322	2910.956284483826	54.94697650596643	102.16303352936724	79.71393514049444	591.8667808131601	1062.306032649635	868.1449767440278	control
753	2722.867015663962	4271.3823918273365	3449.6818139611514	142.81251137061187	300.75666544258075	227.71710966412905	800.1458043239797	1357.6581259902068	1074.466061663058	control
754	3523.5014334058033	5751.038824171586	4837.578144748061	254.16916308966447	393.9572535855385	311.7129937494443	818.00784851896	1199.435295027133	953.085906145617	control
755	1339.9639661380556	2356.8691108825806	1955.2603882259475	299.0783506404707	445.74317232461107	358.8794110473283	849.7350853580242	1354.779034638416	1059.1546269670719	control
756	2904.979524111538	4875.267530545923	3842.9481328299366	190.59225183621456	340.3677851381231	275.3703255489808	806.6256809591687	1279.1825015813517	1112.9196045638855	control
757	1928.2502582222844	3213.764934113309	2555.306418638339	214.1670359896676	310.7750052002745	265.99417195650483	819.4277814331433	1344.443500461039	979.5451215329065	control
758	2879.532511098483	5060.958479697254	3970.1504323683484	282.8014693246637	472.41107787116766	385.7964266720868	924.9158484599601	1472.4624073192037	1165.9297728331626	control
759	3469.4369827964215	4975.39296026624	4692.912900133531	300.69382390167	532.3591551588222	387.0699584872605	823.4211480023538	1400.824839951915	1008.7381195737138	control
760	2703.36511834997	4420.4365637213	3659.1561952735597	255.8929186270646	447.6386267397175	341.78588845668196	624.392586738956	1092.7736463109939	842.6053868543308	control
761	2760.57057345916	5131.375175999033	4413.024838393147	168.1998769295567	279.4838923221065	242.76451157427908	611.6189226013026	1048.319460623319	806.7763186252246	control
762	3199.2826660211554	5731.688790303953	4608.923405703855	203.5846161046918	362.59976822712434	316.0459912755512	759.2377289033956	1346.7917765093882	1140.1869120347708	control
763	1598.9200644325165	2522.357996741745	2086.15480507047	193.96335602554223	355.98267877396006	290.7998190190817	849.8617312002843	1392.9509372295483	1224.3867127189676	control
764	1043.0722301820772	1745.9139784501679	1337.5108177591333	262.6617735677317	529.4544995124054	386.4263418978787	724.0723715665424	1171.2561279451245	997.3045530481638	control
765	1870.779159236243	3123.849125953832	2590.3073545765033	301.74268570987795	443.7564484329417	385.10026451822796	634.5967416057722	1163.2699720823782	889.9587578594798	control
766	2712.0485745352653	5207.595722528408	3369.1547349722005	163.95726859811265	280.56842252622704	220.5943279600962	743.6195043628763	1212.4516092924493	990.4786897710128	control
767	1375.6936209456373	2400.020330891372	1794.157250084868	262.27738097651	388.54265861286757	317.82327601361743	780.5767728112826	1295.582798365117	1051.7416171204734	control
768	4100.543583591004	5825.539652285268	5130.389441467253	302.8732157439316	489.23089872776336	363.5368493847078	735.1213910815587	1333.3868991011573	1032.1635566034374	control
769	3461.4636674045555	6834.637209772043	4625.151184931265	312.86240460439336	470.87929814080604	383.7681458990444	819.1419251924935	1390.1935306293067	1065.9919134152442	control
770	2511.6238742350374	4414.802025184977	3718.067266858722	110.38139218928434	152.27200749334355	134.30682380686815	766.7612973880547	1280.3556721890286	1055.9835528583276	control
771	91.16119171222077	153.93830825142555	130.55694525994468	202.44691305953077	298.0984568904791	231.71609656554222	776.0884893089415	1159.362123294038	982.0880912803499	control
772	1193.8721755740357	2306.2236040110765	1891.0278927791658	205.68246871467244	314.84109544260275	245.06530997285273	583.8349588152925	1030.6868906277841	743.2975144187212	control
773	1462.7137918970477	2197.6531133811454	1967.6309109517638	261.20006453483614	427.51300510498186	349.0769176348648	704.3703691263789	1115.9287672080627	964.8867719732435	control
774	2529.9680089915773	4078.9859710492674	2850.896947305293	195.00732050942895	327.8581366330621	275.6804576937408	655.4629741913726	1162.0742880716796	898.501866606089	control
775	1154.9329124349533	2107.3654312279255	1526.1728060683738	272.61805957395507	413.56431136426124	341.41522045433874	690.6402721867696	1388.7552081604438	1128.477898667677	control
776	2504.0490691563928	4007.792704972531	3075.515144188602	236.55768179942459	395.44499002082244	331.42887838194923	749.0669789084396	1194.8879643729138	892.3798372426987	control
777	2315.3972843117267	3634.059801892802	3491.299659199749	283.55199644649116	464.4870719449667	361.86588154772534	626.4144178885139	1009.7025195001447	778.1054361063599	control
778	2247.372777989452	3716.2256774174466	2902.496245257891	264.8152252453355	428.19779377565743	353.482921746331	961.4922451956901	1400.8136956911044	1131.6520758825243	control
779	2652.8502262976613	4210.444777139047	3573.5552133898846	194.08432019358565	334.73694588340896	281.742973434593	750.4899542414752	1233.962844364492	976.5332433982785	control
780	1122.9978478438672	1730.2193811163913	1439.502084430527	201.35373313472874	360.11228615680386	279.3799975940225	685.142333925882	1348.7478226707954	1020.4949567383856	control
781	3430.3006781704353	5639.783050401233	4381.388031332473	253.37388448632103	397.28319889374023	299.623054361652	704.7752667601923	1151.9785312658696	986.8812551400215	control
782	3213.343837184433	4464.171144635509	3695.1899826530935	259.02976087609164	355.61356823163675	289.91399542293885	599.7551243342759	1082.3609998635782	827.5853680230763	control
783	2405.7610516552786	3541.122893570001	2757.810909263623	115.19506319826738	207.24091843017527	163.54023444265172	704.6128582718459	1154.4633914429278	874.628620528676	control
784	1974.4595081739985	3406.4935372928994	2720.8942378025263	220.2280336405643	381.2618286307246	289.66561132832	620.0604184398761	1058.5911609281816	930.016799982833	control
785	2573.576198954028	4432.3568824718195	3577.5356567811796	212.4455105969902	329.44688668761944	263.2846008633774	709.4269167737843	1229.7304920249628	973.636852044703	control
786	1506.0827992858979	2636.715208335005	2101.517754847028	160.44112666242864	243.87488886127554	211.44995671126486	748.9098013664478	1174.0285289922983	991.9847396979541	control
787	1855.4472512343418	3306.8485752626475	2805.9738248921694	211.49887491539275	375.2713639812561	293.451509914423	719.8244936878131	1165.7817966345326	904.5806471313039	control
788	2596.6995667001147	4370.6060668169985	3609.535973419835	286.44718101476684	505.67616235789967	377.20445440334237	609.5137640594552	999.8620253736002	796.5064924986959	control
789	3196.7040905994722	5748.138372504782	4557.766263087927	220.30118641081864	346.6402882642678	290.19196882093973	714.451264962354	1152.4806932445497	942.3789347333433	control
790	1575.9489863222316	2496.0912435443797	2083.110857319867	169.60136835916452	326.50996601020694	234.04194021853337	934.4475043148544	1477.0682929026534	1079.3219092013192	control
791	2287.879168241887	3855.8165567044334	3117.336039364406	214.47682052164458	308.16138422025193	280.0365091605405	842.9577492862655	1321.1766864278536	1069.0307841002189	control
792	2939.3792084093116	4670.773035466702	4172.864568818619	191.0057537229516	341.4566543166412	259.48180047089426	812.5446496218564	1357.8691931446585	1005.1523747941335	control
793	2383.24861831435	3562.7043398543537	3068.489404703075	292.99676099418656	460.843272897467	340.47081975729645	598.8508962789489	1030.327338987813	882.7354833583928	control
794	2185.2485499040054	4135.262456101866	3387.862729723933	199.0166056169673	312.0027612361979	217.95808468272796	569.3991986074203	979.285074946123	768.2935614054174	control
795	1778.8646266363626	3178.3673549786186	2636.466861092093	252.23181678546533	465.84819443312614	348.82701706765664	797.4391612396438	1275.1541086964073	1009.564343914638	control
796	2524.238210594596	3872.3093273752315	3303.444810072743	150.74528758950146	247.61176328120882	196.36996868441904	769.7675694256642	1179.891397953324	950.7756800294721	control
797	595.4913663982136	1098.9519482437954	883.7775880450694	143.08792210961658	247.85088202947935	190.35084285399452	567.6800502351896	929.6627691377937	739.3299983027662	control
798	2377.8211554282593	4294.890825697262	3274.6731376482785	218.745867380814	352.90334674849487	263.5406348455474	712.9332803350901	1159.3582600696325	1016.4749851209073	control
799	1824.4422573661416	3177.805719861113	2664.1168252140646	179.60463011441735	311.87682232461117	261.64417932813274	637.1093097675982	1408.0529688919419	1132.0868618255552	control
800	1973.485483451019	3666.6756119411702	2954.595069657003	228.2530823870115	353.7675446183424	292.5761661518402	756.1106853302756	1190.6727441763303	1017.8967831391626	control
').then(res => res.blob()).then(blob => {
      const downloadURL = window.URL.createObjectURL(blob);
      const a = document.createElement('a');
      document.body.appendChild(a);
      a.href = downloadURL;
      a.download = 'mousezempic_expression_data.tsv'; a.click();
      window.URL.revokeObjectURL(downloadURL);
      document.body.removeChild(a);
        });"><button class="btn btn-primary downloadthis  data-button" id="expr_data"><i class="bi bi-file-earmark-arrow-down"></i> Download mousezempic_expression_data.tsv </button></a></p>
<p>Or run the following code in your R console to download it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://raw.githubusercontent.com/kzeglinski/new_wehi_r_course/refs/heads/main/data/mousezempic_expression_data.tsv"</span>, <span class="st">"data/mousezempic_expression_data.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once you’ve done that, read it into R (notice that it’s a <code>.tsv</code> file, so we need to use either <code>read_tsv()</code> or <code>read_delim()</code>, not <code>read_csv()</code>)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 453 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (1): group
dbl (10): id_num, TH_rep1, TH_rep2, TH_rep3, PRLH_rep1, PRLH_rep2, PRLH_rep3...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print it out</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 453 × 11
   id_num TH_rep1 TH_rep2 TH_rep3 PRLH_rep1 PRLH_rep2 PRLH_rep3 HPRT1_rep1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1      1   1431.   2737.   2036.      297.      428.      320.       741.
 2      2   2699.   4180.   3860.      388.      680.      500.       843.
 3      3   3596.   5194.   4128.      201.      346.      266.      1047.
 4      4   2723.   4442.   3728.      317.      576.      478.       541.
 5      5   2197.   3656.   2887.      114.      205.      173.       798.
 6      6   2316.   3949.   3355.      242.      383.      305.       763.
 7      7   1570.   2740.   2129.      294.      489.      427.       796.
 8      8   2683.   4749.   3563.      231.      400.      309.       922.
 9      9   3023.   5170.   3830.      289.      464.      390.       921.
10     10   3254.   5980.   4250.      337.      583.      436.       740.
# ℹ 443 more rows
# ℹ 3 more variables: HPRT1_rep2 &lt;dbl&gt;, HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<p>As you can see, the data is in wide format, with each row representing a different mouse (identified by its <code>id_num</code>) and each column representing a different measurement of a gene. To put this data into a long format (where each measurement is contained on a separate row), we can use <code>pivot_longer()</code>, specifying three arguments:</p>
<ol type="1">
<li><code>cols</code>: the columns to pivot from. You can use selection helpers like <code>contains()</code> or <code>starts_with()</code> to easily select multiple columns at once.</li>
<li><code>names_to</code>: the name of a new column that will contain the old column names.</li>
<li><code>values_to</code>: the name of a new column that will contain the values from the old columns.</li>
</ol>
<p>In this particular case here’s what the code would look like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="sc">%&gt;%</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>), <span class="at">names_to =</span> <span class="st">"measurement"</span>, <span class="at">values_to =</span> <span class="st">"expression_level"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,077 × 4
   id_num group     measurement expression_level
    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;                  &lt;dbl&gt;
 1      1 treatment TH_rep1                1431.
 2      1 treatment TH_rep2                2737.
 3      1 treatment TH_rep3                2036.
 4      1 treatment PRLH_rep1               297.
 5      1 treatment PRLH_rep2               428.
 6      1 treatment PRLH_rep3               320.
 7      1 treatment HPRT1_rep1              741.
 8      1 treatment HPRT1_rep2             1300.
 9      1 treatment HPRT1_rep3              988.
10      2 treatment TH_rep1                2699.
# ℹ 4,067 more rows</code></pre>
</div>
</div>
<p>Let’s go through those three arguments:</p>
<ol type="1">
<li><code>cols = contains("_rep")</code>: This tells R to pivot all columns that contain the string “_rep”. In this case, that’s the columns <code>TH_rep1</code>, <code>TH_rep2</code>, <code>TH_rep3</code>, <code>PRLH_rep1</code>, <code>PRLH_rep2</code>, <code>PRLH_rep3</code>, <code>HPRT1_rep1</code>, <code>HPRT1_rep2</code>, and <code>HPRT1_rep3</code>.</li>
<li><code>names_to = "measurement"</code>: This tells R to create a new column called ‘measurement’ that contains the names of the old columns. So, for example, the value of the ‘measurement’ column for the first row will be ‘TH_rep1’.</li>
<li><code>values_to = "expression_level"</code>: This tells R to create a new column called ‘expression_level’ that contains the values from the old columns. So, for example, the value of the ‘expression_level’ column for the first row will be the value of the <code>TH_rep1</code> column.</li>
</ol>
<p>After running this code, you’ll see that the data frame is now in long format, with each row representing a different measurement of a gene for a different mouse.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Perplexed by pivoting?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Perplexed by pivoting?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Pivoting can be a bit tricky to get your head around! Often when you’re doing analysis, you’ll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <a href="https://r4ds.hadley.nz/data-tidy.html#sec-pivoting">like those in the R for Data Science book</a>, or to just experiment with different arguments until you get the result you want.</p>
</div>
</div>
</section>
</section>
<section id="sec-join" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="sec-join"><span class="header-section-number">2.4.2</span> Combining data with join functions</h3>
<p>Another way you might need to change the structure of your data is by combining it with other data frames. This is where the <code>join</code> functions come in. There are many different types of joins in <code>dplyr</code> (which you can read about here if you’re interested), but the most useful one is the <code>left_join()</code> function.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F11_left_join.png" class="img-fluid figure-img"></p>
<figcaption>The left_join() function allows you to combine two data frames</figcaption>
</figure>
</div>
<p><code>left_join()</code> is used to combine two data frames based on a common column. It takes three arguments - the first data frame, the second data frame, and the common column to join on. The reason it’s called a ‘left join’ is because it keeps all the rows from the first data frame (the ‘left’ data frame), and only adds rows from the second data frame (the ‘right’ data frame) if they match the common column.</p>
<p>For example, we can join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column, which is common to both data frames:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># example of using it in a pipe</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here mousezempic_dosage_data is the left data frame (first) data frame passed on by the pipe</span></span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="st">"id_num"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 19
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 13 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, TH_rep1 &lt;dbl&gt;, TH_rep2 &lt;dbl&gt;, TH_rep3 &lt;dbl&gt;, PRLH_rep1 &lt;dbl&gt;,
#   PRLH_rep2 &lt;dbl&gt;, PRLH_rep3 &lt;dbl&gt;, HPRT1_rep1 &lt;dbl&gt;, HPRT1_rep2 &lt;dbl&gt;,
#   HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<p>This code will return a new data frame that contains all the columns from <code>mousezempic_dosage_data</code> and all the columns from <code>mousezempic_expression_data</code>, with the rows matched based on the <code>id_num</code> column. If there is no match for a particular <code>id_num</code> in the <code>mousezempic_expression_data</code> data frame, the columns from that data frame will be filled with <code>NA</code>. Rows that only appear in the <code>mousezempic_expression_data</code> data frame will not be included in the final data frame.</p>
<p>We can also use <code>left_join()</code> when our columns have different names (but contain the same value) using a named vector the format <code>c("name1" = "name2")</code>, where ‘name1’ is the name of the column in the left/first data frame (the one being piped in) and ‘name2’ is the name of the corresponding column in the right/second data frame.</p>
<p>For example, if the <code>id_num</code> column in the <code>mousezempic_dosage_data</code> data frame was called <code>mouse_id</code> instead, we could still join the two data frames using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># for the sake of this example, make a new column called mouse_id</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_id =</span> id_num) <span class="sc">%&gt;%</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join the two data frames where mouse_id in the first data frame matches id_num in the second data frame</span></span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"mouse_id"</span> <span class="ot">=</span> <span class="st">"id_num"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 20
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 14 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, mouse_id &lt;dbl&gt;, TH_rep1 &lt;dbl&gt;, TH_rep2 &lt;dbl&gt;, TH_rep3 &lt;dbl&gt;,
#   PRLH_rep1 &lt;dbl&gt;, PRLH_rep2 &lt;dbl&gt;, PRLH_rep3 &lt;dbl&gt;, HPRT1_rep1 &lt;dbl&gt;,
#   HPRT1_rep2 &lt;dbl&gt;, HPRT1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What does the following code do?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Pivots data into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots into a wide format where there is a column for each sex.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots into a wide format where there is a column for each mouse strain.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>It just gives an error</p><p></p>
</div>
</div>

</div>
</div>
<p>2. I have run the following code to create a new column in the <code>mousezempic_dosage_data</code> data frame that gives the weight of the mice at the end of the experiment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   id_num initial_weight_g final_weight_g
    &lt;dbl&gt;            &lt;dbl&gt;          &lt;dbl&gt;
 1      1             39.1           35.4
 2      2             39.5           35.7
 3      3             40.3           37.0
 4      4             NA             NA  
 5      5             36.7           33.2
 6      6             39.3           35.6
 7      7             38.9           35.3
 8      8             39.2           34.5
 9      9             34.1           30.6
10     10             42             37.8
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</p>
<div class="choices">
<div class="choice">
<p><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")</code></p>
</div>
<div class="choice">
<p><code>pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")</code></p>
</div>
<div class="choice">
<p><code>pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)</code></p>
</div>
<div class="choice correct-choice">
<p><code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code></p>
</div>
</div>
<div class="question">
<p>3. Which of the following is NOT a valid way to join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>left_join(mousezempic_dosage_data, mousezempic_expression_data, by = "id_num")</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = ("id_num" = "id_num"))</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>1. The code first calculates the median tail length for each unique combination of <code>mouse_strain</code> and <code>sex</code> in the <code>mousezempic_dosage_data</code> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <code>names_from = sex</code> )</p>
<p>2. The correct pivot function call to take the data from a wide format to a long format is <code>pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")</code>. This code tells R to pivot the <code>initial_weight_g</code> and <code>final_weight_g</code> columns into a long format, where there is a row for each mouse and each weight measurement. The <code>names_to</code> argument specifies to make a column called ‘timepoint’ that tells us whether the measurement is initial or final, and the <code>values_to</code> argument specifies the name of the new column that will contain these measurements.</p>
<p>3. The line of code that is NOT a valid way to join the <code>mousezempic_dosage_data</code> data frame with the <code>mousezempic_expression_data</code> data frame based on the <code>id_num</code> column is <code>mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")</code>. This line of code is incorrect because the <code>.by</code> argument is not used in the <code>left_join()</code> function (this can be confusing! it’s <code>.by</code> when grouping by <code>by</code> when joining). The other options are valid ways to join the two data frames based on the <code>id_num</code> column: remember that we don’t have to use pipes to join data frames, we can use the <code>left_join()</code> function directly, and we can use a named vector to specify the columns to join on (although here it’s a bit redundant as the columns have the same name).</p>
</details>
</div>
</div>
</section>
</section>
<section id="sec-saving" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="sec-saving"><span class="header-section-number">2.5</span> Saving data to a file</h2>
<p>Once you’ve cleaned and transformed your data, you’ll often want to save it to a file so that you can use it in other programs or share it with others. The <code>write_csv()</code> and <code>write_tsv()</code> functions from the <code>readr</code> package are a great way to do this. They take two arguments - the data frame you want to save and the file path where you want to save it.</p>
<p>For example, let’s say I want to save my contingency table of the number of mice in each cage and strain to a CSV file called <code>cage_strain_counts.csv</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create the contingency table</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>cage_strain_counts <span class="ot">&lt;-</span> mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-8"><a href="#cb130-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the data to a CSV file</span></span>
<span id="cb130-9"><a href="#cb130-9" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(cage_strain_counts, <span class="st">"cage_strain_counts.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>CSV files are particularly great because they can be easily read into other software, like Excel.</p>
<p>It’s also possible to use the <code>write_*()</code> functions along with a pipe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice) <span class="sc">%&gt;%</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_csv</span>(<span class="st">"cage_strain_counts.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remember here that the first argument (the data frame to save) is passed on by the pipe, so the only argument in the brackets is the second one: the file path.</p>
</section>
<section id="summary" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="summary"><span class="header-section-number">2.6</span> Summary</h2>
<p>Here’s what we’ve covered in this session:</p>
<ul>
<li><p>The pipe operator <code>%&gt;%</code> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</p></li>
<li><p>The basic dplyr verbs <code>filter()</code>, <code>select()</code>, <code>mutate()</code>, and <code>arrange()</code> and how they can be used to tidy and analyse data.</p></li>
<li><p>Missing values (<code>NA</code>) and how to remove or replace them</p></li>
<li><p>The <code>summarise()</code> function and how it can be used to calculate summary statistics on your data, as well as the power of grouping data with the <code>.by</code> argument.</p></li>
<li><p>Reshaping data with <code>pivot_longer()</code> and <code>pivot_wider()</code> to change the structure of your data frame.</p></li>
<li><p>Combining data with <code>left_join()</code> to merge two data frames based on a common column.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="Why does data need to be tidy anyway?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why does data need to be tidy anyway?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this session, we’ve been focusing on making our data ‘tidy’: that is, structured in a consistent way that makes it easy to work with. A nice visual illustration of tidy data and its importance can be <a href="https://allisonhorst.com/other-r-fun">found here</a>.</p>
</div>
</div>
<section id="practice-questions" class="level3" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="practice-questions"><span class="header-section-number">2.6.1</span> Practice questions</h3>
<ol type="1">
<li><p>What is the purpose of the pipe operator <code>%&gt;%</code>? Keeping this in mind, re-write the following code to use the pipe.</p>
<ol type="a">
<li><code>round(mean(c(1, 2, 3, 4, 5)))</code></li>
<li><code>print(as.character(1 + 10))</code></li>
</ol></li>
<li><p>What would be the result of evaluating the following expressions? You don’t need to know these off the top of your head, use R to help! (Hint: some expressions might give an error. Try to think about why)</p>
<ol type="a">
<li><code>mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 10)</code></li>
<li><code>mousezempic_dosage_data %&gt;% select(tail_length_mm, weight_lost_g)</code></li>
<li><code>mousezempic_dosage_data %&gt;% mutate(weight_lost_kg = weight_lost_g / 1000)</code></li>
<li><code>mousezempic_dosage_data %&gt;% arrange(tail_length_mm)</code></li>
<li><code>mousezempic_dosage_data %&gt;% filter(initial_Weight_g &gt; 10) %&gt;% arrange(mouse_strain)</code></li>
<li><code>mousezempic_dosage_data %&gt;% relocate(mouse_strain, .after = cage_number)</code></li>
<li><code>mousezempic_dosage_data %&gt;% pull(weight_lost_g)</code></li>
<li><code>mousezempic_dosage_data %&gt;% filter(!is.na(weight_lost_g))</code></li>
<li><code>mousezempic_dosage_data %&gt;% replace_na(list(weight_lost_g = 0))</code></li>
<li><code>mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))</code></li>
<li><code>mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE), .by = cage_number)</code></li>
</ol></li>
<li><p>What is a missing value in R? What are two ways to deal with missing values in a data frame?</p></li>
<li><p>Using the <code>mousezempic_dosage_data</code> data frame, write R code to:</p>
<ol type="a">
<li>Make a data frame that shows the number of mice of each strain, in each replicate.</li>
<li>Pivot this data frame into a wide format to create a contingency table.</li>
<li>Pivot the wide data frame from (b) back into a long format.</li>
</ol></li>
<li><p>Let’s say I have two data frames, <code>df1</code> and <code>df2</code>, that I want to join based a shared ‘key’ column, that is called ‘key’ in <code>df1</code> and ‘item_key’ in <code>df2</code>. Write R code to join these two data frames using the <code>left_join()</code> function.</p></li>
</ol>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The pipe operator <code>%&gt;%</code> is used to chain together multiple function calls, passing the result of one function to the next. Here’s how you could re-write the code to use the pipe:</p>
<ol type="a">
<li><code>c(1, 2, 3, 4, 5) %&gt;% mean() %&gt;% round()</code></li>
<li><code>as.character(1 + 10) %&gt;% print()</code></li>
</ol></li>
<li><p>The result of evaluating the expressions would be:</p>
<ol type="a">
<li>A data frame containing only the rows where <code>weight_lost_g</code> is greater than 10.</li>
<li>A data frame containing only the <code>tail_length_mm</code> and <code>weight_lost_g</code> columns.</li>
<li>A data frame with an additional column <code>weight_lost_kg</code> that contains the weight lost in kilograms.</li>
<li>A data frame sorted by <code>tail_length_mm</code>, in ascending order.</li>
<li>An error because <code>initial_Weight_g</code> is not a column in the data frame.</li>
<li>A data frame with the <code>mouse_strain</code> column moved to be after the <code>cage_number</code> column.</li>
<li>A <strong>vector</strong> containing the values of the <code>weight_lost_g</code> column.</li>
<li>A data frame containing only the rows where <code>weight_lost_g</code> is not <code>NA</code>.</li>
<li>A data frame with missing values in the <code>weight_lost_g</code> column replaced with 0.</li>
<li>A data frame with the mean weight lost by all mice.</li>
<li>A data frame with the mean weight lost by mice in each cage.</li>
</ol></li>
<li><p>A missing value in R is represented by <code>NA</code>. Two ways to deal with missing values in a data frame are to remove them using <code>filter(!is.na(column_name))</code> or to replace them with a specific value using <code>replace_na(list(column_name = value))</code>.</p></li>
<li><p>Here’s how you could write R code to achieve the tasks:</p>
<ol type="a">
<li><code>mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate))</code></li>
<li><code>mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice)</code></li>
<li><code>mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice) %&gt;% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")</code></li>
</ol></li>
<li><p>To join the two data frames you could use <code>df1 %&gt;% left_join(df2, by = c("key" = "item_key"))</code> (with pipe) or <code>left_join(df1, df2, by = c("key" = "item_key"))</code> (without pipe).</p></li>
</ol>
</details>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session_1.html" class="pagination-link" aria-label="Session 1: Introduction to R">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Session 1: Introduction to R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb132" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this session:"</span></span>
<span id="cb132-9"><a href="#cb132-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb132-10"><a href="#cb132-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-11"><a href="#cb132-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Session 2: Working with data {#sec-session02}</span></span>
<span id="cb132-12"><a href="#cb132-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-13"><a href="#cb132-13" aria-hidden="true" tabindex="-1"></a>In this session we will learn how to manipulate and summarise data using the <span class="in">`dplyr`</span> package (with a little help from the <span class="in">`tidyr`</span> package too).</span>
<span id="cb132-14"><a href="#cb132-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-15"><a href="#cb132-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb132-16"><a href="#cb132-16" aria-hidden="true" tabindex="-1"></a>At the end of this session, learners should be able to:</span>
<span id="cb132-17"><a href="#cb132-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-18"><a href="#cb132-18" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Use the pipe (<span class="in">`%&gt;%`</span>) to chain multiple functions together</span>
<span id="cb132-19"><a href="#cb132-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-20"><a href="#cb132-20" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Design chains of dplyr functions to manipulate data frames</span>
<span id="cb132-21"><a href="#cb132-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-22"><a href="#cb132-22" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Apply grouping for more complex analysis of data</span>
<span id="cb132-23"><a href="#cb132-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-24"><a href="#cb132-24" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Understand how pivot and join functions can be used to reshape and combine entire data frames</span>
<span id="cb132-25"><a href="#cb132-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-26"><a href="#cb132-26" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Recall how to save data frames to a file</span>
<span id="cb132-27"><a href="#cb132-27" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-28"><a href="#cb132-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-29"><a href="#cb132-29" aria-hidden="true" tabindex="-1"></a>Both <span class="in">`dplyr`</span> and <span class="in">`tidyr`</span> are contained within the <span class="in">`tidyverse`</span> (along with <span class="in">`readr`</span>) so we can load all of these packages at once using <span class="in">`library(tidyverse)`</span>:</span>
<span id="cb132-30"><a href="#cb132-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-33"><a href="#cb132-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-34"><a href="#cb132-34" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load tidyverse!</span></span>
<span id="cb132-35"><a href="#cb132-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb132-36"><a href="#cb132-36" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-37"><a href="#cb132-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-38"><a href="#cb132-38" aria-hidden="true" tabindex="-1"></a><span class="fu">## Chaining functions together with pipes {#sec-pipes}</span></span>
<span id="cb132-39"><a href="#cb132-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-40"><a href="#cb132-40" aria-hidden="true" tabindex="-1"></a>Pipes are a powerful feature of the <span class="in">`tidyverse`</span> that allow you to chain multiple functions together. Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</span>
<span id="cb132-41"><a href="#cb132-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-42"><a href="#cb132-42" aria-hidden="true" tabindex="-1"></a>For example, take the following code:</span>
<span id="cb132-43"><a href="#cb132-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-46"><a href="#cb132-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-47"><a href="#cb132-47" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb132-48"><a href="#cb132-48" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span>
<span id="cb132-49"><a href="#cb132-49" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-50"><a href="#cb132-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-51"><a href="#cb132-51" aria-hidden="true" tabindex="-1"></a>What do you think this code does? It calculates the mean of <span class="in">`my_vector`</span>, rounds the result to the nearest whole number, and then converts the result to a character. But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</span>
<span id="cb132-52"><a href="#cb132-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-53"><a href="#cb132-53" aria-hidden="true" tabindex="-1"></a>Instead, we can use the pipe operator (<span class="in">`%&gt;%`</span>) to chain these functions together in a more readable way:</span>
<span id="cb132-54"><a href="#cb132-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-57"><a href="#cb132-57" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-58"><a href="#cb132-58" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb132-59"><a href="#cb132-59" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb132-60"><a href="#cb132-60" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-61"><a href="#cb132-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-62"><a href="#cb132-62" aria-hidden="true" tabindex="-1"></a>See how the code reads naturally from left to right? You can think of the pipe as being like the phrase "and then". Here, we're telling R: "Take <span class="in">`my_vector`</span>, and then calculate the mean, and then round the result, and then convert it to a character."</span>
<span id="cb132-63"><a href="#cb132-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-64"><a href="#cb132-64" aria-hidden="true" tabindex="-1"></a>You'll notice that we didn't need to specify the input to each function. That's because the pipe automatically passes the output of the previous function as the first input to the next function. We can still specify additional arguments to each function if we need to. For example, if we wanted to round the mean to 2 decimal places, we could do this:</span>
<span id="cb132-65"><a href="#cb132-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-68"><a href="#cb132-68" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-69"><a href="#cb132-69" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb132-70"><a href="#cb132-70" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-71"><a href="#cb132-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-72"><a href="#cb132-72" aria-hidden="true" tabindex="-1"></a>R is clever enough to know that the first argument to <span class="in">`round()`</span> is still the output of <span class="in">`mean()`</span>, even though we've now specified the <span class="in">`digits`</span> argument.</span>
<span id="cb132-73"><a href="#cb132-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-74"><a href="#cb132-74" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Plenty of pipes"}</span>
<span id="cb132-75"><a href="#cb132-75" aria-hidden="true" tabindex="-1"></a>There is another style of pipe in R, called the 'base R pipe' <span class="in">`|&gt;`</span>, which is available in R version 4.1.0 and later. The base R pipe works in a similar way to the <span class="in">`magrittr`</span> pipe (<span class="in">`%&gt;%`</span>) that we use in this course, but it is not as flexible. We recommend using the <span class="in">`magrittr`</span> pipe for now.</span>
<span id="cb132-76"><a href="#cb132-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-77"><a href="#cb132-77" aria-hidden="true" tabindex="-1"></a>Fun fact: the <span class="in">`magrittr`</span> package is named after the <span class="co">[</span><span class="ot">artist René Magritte, who made a famous painting of a pipe</span><span class="co">](https://en.wikipedia.org/wiki/The_Treachery_of_Images)</span>.</span>
<span id="cb132-78"><a href="#cb132-78" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-79"><a href="#cb132-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-80"><a href="#cb132-80" aria-hidden="true" tabindex="-1"></a>To type the pipe operator more easily, you can use the keyboard shortcut {{&lt; kbd Cmd-shift-M &gt;}} (although once you get used to it, you might find it easier to type <span class="in">`%&gt;%`</span> manually).</span>
<span id="cb132-81"><a href="#cb132-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-82"><a href="#cb132-82" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-83"><a href="#cb132-83" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-84"><a href="#cb132-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-85"><a href="#cb132-85" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-86"><a href="#cb132-86" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What is NOT a valid way to re-write the following code using the pipe operator: <span class="in">`round(sqrt(sum(1:10)), 1)`</span>. If you're not sure, try running the different options in the console to see which one gives the same answer.</span>
<span id="cb132-87"><a href="#cb132-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-88"><a href="#cb132-88" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-89"><a href="#cb132-89" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-90"><a href="#cb132-90" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)`</span></span>
<span id="cb132-91"><a href="#cb132-91" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-92"><a href="#cb132-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-93"><a href="#cb132-93" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-94"><a href="#cb132-94" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span></span>
<span id="cb132-95"><a href="#cb132-95" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-96"><a href="#cb132-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-97"><a href="#cb132-97" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-98"><a href="#cb132-98" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb132-99"><a href="#cb132-99" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-100"><a href="#cb132-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-101"><a href="#cb132-101" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-102"><a href="#cb132-102" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb132-103"><a href="#cb132-103" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-104"><a href="#cb132-104" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-105"><a href="#cb132-105" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-106"><a href="#cb132-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-107"><a href="#cb132-107" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-108"><a href="#cb132-108" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is the output of the following code? <span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> Try to guess it before copy-pasting into R.</span>
<span id="cb132-109"><a href="#cb132-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-110"><a href="#cb132-110" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-111"><a href="#cb132-111" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-112"><a href="#cb132-112" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"`</span></span>
<span id="cb132-113"><a href="#cb132-113" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-114"><a href="#cb132-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-115"><a href="#cb132-115" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-116"><a href="#cb132-116" aria-hidden="true" tabindex="-1"></a><span class="in">`"a" "b" "c" "d" "e" "f"`</span></span>
<span id="cb132-117"><a href="#cb132-117" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-118"><a href="#cb132-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-119"><a href="#cb132-119" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-120"><a href="#cb132-120" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb132-121"><a href="#cb132-121" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-122"><a href="#cb132-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-123"><a href="#cb132-123" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-124"><a href="#cb132-124" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F"`</span></span>
<span id="cb132-125"><a href="#cb132-125" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-126"><a href="#cb132-126" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-127"><a href="#cb132-127" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-128"><a href="#cb132-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-129"><a href="#cb132-129" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-130"><a href="#cb132-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-131"><a href="#cb132-131" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-132"><a href="#cb132-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-133"><a href="#cb132-133" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb132-134"><a href="#cb132-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-135"><a href="#cb132-135" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The invalid option is <span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span>. This is because the <span class="in">`sqrt()`</span> function only takes one argument, so you can't specify <span class="in">`1`</span> as an argument in addition to what is being piped in from <span class="in">`sum(1:10)`</span>. Note that some options used the pipe to send <span class="in">`1:10`</span> to <span class="in">`sum()`</span> (like <span class="in">`1:10 %&gt;% sum()`</span>), and others just used <span class="in">`sum(1:10)`</span> directly. Both are valid ways to use the pipe, it's just a matter of personal preference.</span>
<span id="cb132-136"><a href="#cb132-136" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The output of the code <span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> is <span class="in">`"A" "B" "C" "D" "E" "F"`</span>. The <span class="in">`letters`</span> vector contains the lowercase alphabet, and the <span class="in">`head()`</span> function returns the first 6 elements of the vector. Finally, the <span class="in">`toupper()`</span> function then converts these elements to uppercase.</span>
<span id="cb132-137"><a href="#cb132-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-138"><a href="#cb132-138" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb132-139"><a href="#cb132-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-140"><a href="#cb132-140" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-141"><a href="#cb132-141" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-142"><a href="#cb132-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-143"><a href="#cb132-143" aria-hidden="true" tabindex="-1"></a><span class="fu">## Basic data manipulation {#sec-dataManip}</span></span>
<span id="cb132-144"><a href="#cb132-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-145"><a href="#cb132-145" aria-hidden="true" tabindex="-1"></a>To really see the power of the pipe, we will use it together with the <span class="in">`dplyr`</span> package that provides a set of functions to easily filter, sort, select, and summarise data frames. These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</span>
<span id="cb132-146"><a href="#cb132-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-147"><a href="#cb132-147" aria-hidden="true" tabindex="-1"></a>For example, even though we haven't covered the <span class="in">`dplyr`</span> functions yet, you can probably guess what the following code does:</span>
<span id="cb132-148"><a href="#cb132-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-151"><a href="#cb132-151" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-152"><a href="#cb132-152" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-153"><a href="#cb132-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-154"><a href="#cb132-154" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb132-155"><a href="#cb132-155" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-156"><a href="#cb132-156" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-157"><a href="#cb132-157" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb132-158"><a href="#cb132-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span>
<span id="cb132-159"><a href="#cb132-159" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-160"><a href="#cb132-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-161"><a href="#cb132-161" aria-hidden="true" tabindex="-1"></a>This code filters the <span class="in">`mousezempic_dosage_data`</span> data frame to only include data from cage 3E, then pulls out the <span class="in">`weight_lost_g`</span> column, and finally calculates the mean of the values in that column. The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <span class="in">`pull()`</span>) are specified in the brackets (like <span class="in">`round(digits = 2)`</span> from the previous example).</span>
<span id="cb132-162"><a href="#cb132-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-163"><a href="#cb132-163" aria-hidden="true" tabindex="-1"></a>We also used the enter key after each pipe <span class="in">`%&gt;%`</span> to break up the code into multiple lines to make it easier to read. This isn't required, but is a popular style in the R community, so all the code examples in this session will follow this format.</span>
<span id="cb132-164"><a href="#cb132-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-165"><a href="#cb132-165" aria-hidden="true" tabindex="-1"></a>We will now introduce some of the most commonly used <span class="in">`dplyr`</span> functions for manipulating data frames. To showcase these, we will use the <span class="in">`mousezempic_dosage_data`</span> that we practiced reading in last session. This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</span>
<span id="cb132-166"><a href="#cb132-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-169"><a href="#cb132-169" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-170"><a href="#cb132-170" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-171"><a href="#cb132-171" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in session 1</span></span>
<span id="cb132-172"><a href="#cb132-172" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"~/Desktop/R_course/data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb132-173"><a href="#cb132-173" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-174"><a href="#cb132-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-177"><a href="#cb132-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-178"><a href="#cb132-178" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb132-179"><a href="#cb132-179" aria-hidden="true" tabindex="-1"></a><span class="co"># just for rendering the book not for students to see</span></span>
<span id="cb132-180"><a href="#cb132-180" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb132-181"><a href="#cb132-181" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-182"><a href="#cb132-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-183"><a href="#cb132-183" aria-hidden="true" tabindex="-1"></a>Before we start, let's use what we learned in the previous session to take a look at <span class="in">`mousezempic_dosage_data`</span>:</span>
<span id="cb132-184"><a href="#cb132-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-187"><a href="#cb132-187" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-188"><a href="#cb132-188" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb132-189"><a href="#cb132-189" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data</span>
<span id="cb132-190"><a href="#cb132-190" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-191"><a href="#cb132-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-192"><a href="#cb132-192" aria-hidden="true" tabindex="-1"></a>You might also like to use <span class="in">`View()`</span> to open the data in a separate window and get a closer look.</span>
<span id="cb132-193"><a href="#cb132-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-194"><a href="#cb132-194" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Using RStudio autocomplete"}</span>
<span id="cb132-195"><a href="#cb132-195" aria-hidden="true" tabindex="-1"></a>Although it's great to give our data a descriptive name like <span class="in">`mousezempic_dosage_data`</span>, it can be a bit of a pain to type out every time. Luckily, RStudio has a handy autocomplete feature that can solve this problem. Just start typing the name of the object, and you'll see it will popup:</span>
<span id="cb132-196"><a href="#cb132-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-197"><a href="#cb132-197" aria-hidden="true" tabindex="-1"></a><span class="al">![RStudio autocomplete](images/S02F01_rstudio_autocomplete.png)</span></span>
<span id="cb132-198"><a href="#cb132-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-199"><a href="#cb132-199" aria-hidden="true" tabindex="-1"></a>You can then press {{&lt; kbd Tab &gt;}} to autocomplete it. If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</span>
<span id="cb132-200"><a href="#cb132-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-201"><a href="#cb132-201" aria-hidden="true" tabindex="-1"></a>Try using autocomplete this session to save yourself some typing!</span>
<span id="cb132-202"><a href="#cb132-202" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-203"><a href="#cb132-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-204"><a href="#cb132-204" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sorting data {#sec-sorting}</span></span>
<span id="cb132-205"><a href="#cb132-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-206"><a href="#cb132-206" aria-hidden="true" tabindex="-1"></a>Often, one of the first things you might want to do with a dataset is sort it. In <span class="in">`dplyr`</span>, this is called 'arranging' and is done with the <span class="in">`arrange()`</span> function.</span>
<span id="cb132-207"><a href="#cb132-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-208"><a href="#cb132-208" aria-hidden="true" tabindex="-1"></a><span class="al">![Arrange orders rows by their values in one or more columns](images/S02F02_arrange.png)</span></span>
<span id="cb132-209"><a href="#cb132-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-210"><a href="#cb132-210" aria-hidden="true" tabindex="-1"></a>By default, <span class="in">`arrange()`</span> sorts in ascending order (smallest values first). For example, let's sort the <span class="in">`mousezempic_dosage_data`</span> data frame by the <span class="in">`weight_lost_g`</span> column:</span>
<span id="cb132-211"><a href="#cb132-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-214"><a href="#cb132-214" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-215"><a href="#cb132-215" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-216"><a href="#cb132-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span>
<span id="cb132-217"><a href="#cb132-217" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-218"><a href="#cb132-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-219"><a href="#cb132-219" aria-hidden="true" tabindex="-1"></a>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</span>
<span id="cb132-220"><a href="#cb132-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-221"><a href="#cb132-221" aria-hidden="true" tabindex="-1"></a>Sometimes you might want to sort in descending order instead (largest values first). You can do this by putting the <span class="in">`desc()`</span> function around your column name, inside <span class="in">`arrange()`</span>:</span>
<span id="cb132-222"><a href="#cb132-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-225"><a href="#cb132-225" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-226"><a href="#cb132-226" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-227"><a href="#cb132-227" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb132-228"><a href="#cb132-228" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb132-229"><a href="#cb132-229" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-230"><a href="#cb132-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-231"><a href="#cb132-231" aria-hidden="true" tabindex="-1"></a>Now we can see the mice that lost the most weight are at the top.</span>
<span id="cb132-232"><a href="#cb132-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-233"><a href="#cb132-233" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Comments and pipes"}</span>
<span id="cb132-234"><a href="#cb132-234" aria-hidden="true" tabindex="-1"></a>Notice how in the previous example we have written a comment in the middle of the pipe chain. This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won't cause any errors as long as you make sure that the comment is on its own line.</span>
<span id="cb132-235"><a href="#cb132-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-236"><a href="#cb132-236" aria-hidden="true" tabindex="-1"></a>You can also write comments at the end of the line, just make sure it's after the pipe operator <span class="in">`%&gt;%`</span>.</span>
<span id="cb132-237"><a href="#cb132-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-238"><a href="#cb132-238" aria-hidden="true" tabindex="-1"></a>For example, these comments are allowed:</span>
<span id="cb132-239"><a href="#cb132-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-242"><a href="#cb132-242" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-243"><a href="#cb132-243" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb132-244"><a href="#cb132-244" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb132-245"><a href="#cb132-245" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb132-246"><a href="#cb132-246" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-247"><a href="#cb132-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-248"><a href="#cb132-248" aria-hidden="true" tabindex="-1"></a>But this will cause an error, because the <span class="in">`#`</span> is before the pipe, so R treats it as part of the comment (notice how the <span class="in">`%&gt;%`</span> has changed colour?) and doesn't know how the two lines relate to each other. It tries to run them separately, which for the first line is ok (it will just print <span class="in">`mousezempic_dosage_data`</span>):</span>
<span id="cb132-249"><a href="#cb132-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-252"><a href="#cb132-252" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-253"><a href="#cb132-253" aria-hidden="true" tabindex="-1"></a><span class="co">#| error: true</span></span>
<span id="cb132-254"><a href="#cb132-254" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># this comment will cause an error %&gt;%</span></span>
<span id="cb132-255"><a href="#cb132-255" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb132-256"><a href="#cb132-256" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-257"><a href="#cb132-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-258"><a href="#cb132-258" aria-hidden="true" tabindex="-1"></a>But for the second line, there is an error that R doesn't know what the <span class="in">`weight_lost_g`</span> object is. That's because it's a column in the <span class="in">`mousezempic_dosage_data`</span> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</span>
<span id="cb132-259"><a href="#cb132-259" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-260"><a href="#cb132-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-261"><a href="#cb132-261" aria-hidden="true" tabindex="-1"></a>You can also sort by multiple columns by passing multiple column names to <span class="in">`arrange()`</span>. For example, to sort by the strain first and then by the amount of weight lost:</span>
<span id="cb132-262"><a href="#cb132-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-265"><a href="#cb132-265" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-266"><a href="#cb132-266" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb132-267"><a href="#cb132-267" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-268"><a href="#cb132-268" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span>
<span id="cb132-269"><a href="#cb132-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-270"><a href="#cb132-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-271"><a href="#cb132-271" aria-hidden="true" tabindex="-1"></a>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</span>
<span id="cb132-272"><a href="#cb132-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-273"><a href="#cb132-273" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Piping into View()"}</span>
<span id="cb132-274"><a href="#cb132-274" aria-hidden="true" tabindex="-1"></a>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn't allow us to see the full effect of the sorting.</span>
<span id="cb132-275"><a href="#cb132-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-276"><a href="#cb132-276" aria-hidden="true" tabindex="-1"></a>One handy trick you can use with pipes is to add <span class="in">`View()`</span> at the end of your chain to open the data in a separate window. Try running this code, and you'll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</span>
<span id="cb132-277"><a href="#cb132-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-280"><a href="#cb132-280" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-281"><a href="#cb132-281" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-282"><a href="#cb132-282" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb132-283"><a href="#cb132-283" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-284"><a href="#cb132-284" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb132-285"><a href="#cb132-285" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span>
<span id="cb132-286"><a href="#cb132-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-287"><a href="#cb132-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-288"><a href="#cb132-288" aria-hidden="true" tabindex="-1"></a>This is a great way to check that your code has actually done what you intended!</span>
<span id="cb132-289"><a href="#cb132-289" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-290"><a href="#cb132-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-291"><a href="#cb132-291" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Extracting rows with the smallest or largest values {#sec-sliceMinMax}</span></span>
<span id="cb132-292"><a href="#cb132-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-293"><a href="#cb132-293" aria-hidden="true" tabindex="-1"></a>Slice functions are used to select rows based on their position in the data frame. The <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</span>
<span id="cb132-294"><a href="#cb132-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-295"><a href="#cb132-295" aria-hidden="true" tabindex="-1"></a>This is equivalent to using <span class="in">`arrange()`</span> followed by <span class="in">`head()`</span>, but is more concise:</span>
<span id="cb132-296"><a href="#cb132-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-299"><a href="#cb132-299" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-300"><a href="#cb132-300" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb132-301"><a href="#cb132-301" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-302"><a href="#cb132-302" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb132-303"><a href="#cb132-303" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb132-304"><a href="#cb132-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-305"><a href="#cb132-305" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb132-306"><a href="#cb132-306" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-307"><a href="#cb132-307" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb132-308"><a href="#cb132-308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb132-309"><a href="#cb132-309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-310"><a href="#cb132-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-311"><a href="#cb132-311" aria-hidden="true" tabindex="-1"></a>But wait— neither of those pieces of code actually gave the number of rows we asked for! In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13. And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6. Why aren't the <span class="in">`slice_`</span> functions behaving as expected?</span>
<span id="cb132-312"><a href="#cb132-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-313"><a href="#cb132-313" aria-hidden="true" tabindex="-1"></a>If we take a look at the help page (type <span class="in">`?slice_min`</span> in the console), we learn that <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> have an argument called <span class="in">`with_ties`</span> that is set to <span class="in">`TRUE`</span> by default. If we want to make sure we only get the number of rows we asked for, we would have to set it to <span class="in">`FALSE`</span>, like so:</span>
<span id="cb132-314"><a href="#cb132-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-317"><a href="#cb132-317" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-318"><a href="#cb132-318" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb132-319"><a href="#cb132-319" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-320"><a href="#cb132-320" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb132-321"><a href="#cb132-321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span>
<span id="cb132-322"><a href="#cb132-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-323"><a href="#cb132-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-324"><a href="#cb132-324" aria-hidden="true" tabindex="-1"></a>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</span>
<span id="cb132-325"><a href="#cb132-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-326"><a href="#cb132-326" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-327"><a href="#cb132-327" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-328"><a href="#cb132-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-329"><a href="#cb132-329" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-330"><a href="#cb132-330" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which code would you use to sort the <span class="in">`mousezempic_dosage_data`</span> data frame from biggest to smallest initial weight?</span>
<span id="cb132-331"><a href="#cb132-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-332"><a href="#cb132-332" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-333"><a href="#cb132-333" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-334"><a href="#cb132-334" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% sort(initial_weight_g)`</span></span>
<span id="cb132-335"><a href="#cb132-335" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-336"><a href="#cb132-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-337"><a href="#cb132-337" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-338"><a href="#cb132-338" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb132-339"><a href="#cb132-339" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-340"><a href="#cb132-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-341"><a href="#cb132-341" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-342"><a href="#cb132-342" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% sort(descending(initial_weight_g))`</span></span>
<span id="cb132-343"><a href="#cb132-343" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-344"><a href="#cb132-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-345"><a href="#cb132-345" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-346"><a href="#cb132-346" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb132-347"><a href="#cb132-347" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-348"><a href="#cb132-348" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-349"><a href="#cb132-349" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-350"><a href="#cb132-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-351"><a href="#cb132-351" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-352"><a href="#cb132-352" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Which code would you use to extract the 3 mice with the highest initial weight from the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb132-353"><a href="#cb132-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-354"><a href="#cb132-354" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-355"><a href="#cb132-355" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-356"><a href="#cb132-356" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)`</span></span>
<span id="cb132-357"><a href="#cb132-357" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-358"><a href="#cb132-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-359"><a href="#cb132-359" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-360"><a href="#cb132-360" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb132-361"><a href="#cb132-361" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-362"><a href="#cb132-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-363"><a href="#cb132-363" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-364"><a href="#cb132-364" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% slice_min(initial_weight_g, n = 3)`</span></span>
<span id="cb132-365"><a href="#cb132-365" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-366"><a href="#cb132-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-367"><a href="#cb132-367" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-368"><a href="#cb132-368" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb132-369"><a href="#cb132-369" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-370"><a href="#cb132-370" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-371"><a href="#cb132-371" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-372"><a href="#cb132-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-373"><a href="#cb132-373" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-374"><a href="#cb132-374" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> I've written the below code, but one of the comments is messing it up! Which one?</span>
<span id="cb132-375"><a href="#cb132-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-378"><a href="#cb132-378" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-379"><a href="#cb132-379" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-380"><a href="#cb132-380" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb132-381"><a href="#cb132-381" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="co"># comment B %&gt;%</span></span>
<span id="cb132-382"><a href="#cb132-382" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb132-383"><a href="#cb132-383" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span>
<span id="cb132-384"><a href="#cb132-384" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-385"><a href="#cb132-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-386"><a href="#cb132-386" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-387"><a href="#cb132-387" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-388"><a href="#cb132-388" aria-hidden="true" tabindex="-1"></a>Comment A</span>
<span id="cb132-389"><a href="#cb132-389" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-390"><a href="#cb132-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-391"><a href="#cb132-391" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-392"><a href="#cb132-392" aria-hidden="true" tabindex="-1"></a>Comment B</span>
<span id="cb132-393"><a href="#cb132-393" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-394"><a href="#cb132-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-395"><a href="#cb132-395" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-396"><a href="#cb132-396" aria-hidden="true" tabindex="-1"></a>Comment C</span>
<span id="cb132-397"><a href="#cb132-397" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-398"><a href="#cb132-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-399"><a href="#cb132-399" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-400"><a href="#cb132-400" aria-hidden="true" tabindex="-1"></a>Comment D</span>
<span id="cb132-401"><a href="#cb132-401" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-402"><a href="#cb132-402" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-403"><a href="#cb132-403" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-404"><a href="#cb132-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-405"><a href="#cb132-405" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-406"><a href="#cb132-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-407"><a href="#cb132-407" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-408"><a href="#cb132-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-409"><a href="#cb132-409" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to sort the <span class="in">`mousezempic_dosage_data`</span> data frame from biggest to smallest initial weight is <span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(initial_weight_g))`</span>. The <span class="in">`arrange()`</span> function is used to sort the data frame (although there is a <span class="in">`sort()`</span> function in R, that's not part of dplyr and won't work the same way), and the <span class="in">`desc()`</span> function is used to sort in descending order.</span>
<span id="cb132-410"><a href="#cb132-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-411"><a href="#cb132-411" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct code to extract the 3 mice with the highest initial weight from the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% slice_max(initial_weight_g, n = 3)`</span>. The <span class="in">`slice_max()`</span> function is used to select the rows with the largest values in the <span class="in">`initial_weight_g`</span> column, and the <span class="in">`n = 3`</span> argument specifies that we want to keep 3 rows. The <span class="in">`arrange()`</span> function is not needed in this case, because <span class="in">`slice_max()`</span> will automatically sort the data frame by the specified column.</span>
<span id="cb132-412"><a href="#cb132-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-413"><a href="#cb132-413" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The comment that is messing up the code is Comment B. The <span class="in">`#`</span> symbol is before the pipe operator <span class="in">`%&gt;%`</span>, so R treats it as part of the comment and this breaks our chain of pipes. The other comments are fine, because they are either at the end of the line or on their own line. Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it's in the wrong place!</span>
<span id="cb132-414"><a href="#cb132-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-415"><a href="#cb132-415" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-416"><a href="#cb132-416" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-417"><a href="#cb132-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-418"><a href="#cb132-418" aria-hidden="true" tabindex="-1"></a><span class="fu">### Filtering data (rows) {#sec-filter}</span></span>
<span id="cb132-419"><a href="#cb132-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-420"><a href="#cb132-420" aria-hidden="true" tabindex="-1"></a><span class="al">![Filter allows you to filter rows using a logical test](images/S02F03_filter.png)</span></span>
<span id="cb132-421"><a href="#cb132-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-422"><a href="#cb132-422" aria-hidden="true" tabindex="-1"></a>In <span class="in">`dplyr`</span>, the <span class="in">`filter()`</span> function is used to subset rows based on their values. You provide a logical test, and <span class="in">`filter()`</span> will keep the rows where the test is <span class="in">`TRUE`</span>. We can write these tests using the comparison operators we learned in the previous session (e.g. <span class="in">`==`</span>, <span class="in">`&lt;`</span> and <span class="in">`!=`</span>, see <span class="co">[</span><span class="ot">Section @sec-comparisons</span><span class="co">]</span>).</span>
<span id="cb132-423"><a href="#cb132-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-424"><a href="#cb132-424" aria-hidden="true" tabindex="-1"></a>For example, to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice that lost more than 6g:</span>
<span id="cb132-425"><a href="#cb132-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-428"><a href="#cb132-428" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-429"><a href="#cb132-429" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-430"><a href="#cb132-430" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span>
<span id="cb132-431"><a href="#cb132-431" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-432"><a href="#cb132-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-433"><a href="#cb132-433" aria-hidden="true" tabindex="-1"></a>Or to only include mice from cage 3E:</span>
<span id="cb132-434"><a href="#cb132-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-437"><a href="#cb132-437" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-438"><a href="#cb132-438" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-439"><a href="#cb132-439" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb132-440"><a href="#cb132-440" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span>
<span id="cb132-441"><a href="#cb132-441" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-442"><a href="#cb132-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-443"><a href="#cb132-443" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Combining logical tests</span></span>
<span id="cb132-444"><a href="#cb132-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-445"><a href="#cb132-445" aria-hidden="true" tabindex="-1"></a>Sometimes we want to filter based on multiple conditions. Here we will show some more advanced operators that can be used to combine logical tests.</span>
<span id="cb132-446"><a href="#cb132-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-447"><a href="#cb132-447" aria-hidden="true" tabindex="-1"></a>The <span class="in">`&amp;`</span> operator is used to combine two logical tests with an 'and' condition. For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</span>
<span id="cb132-448"><a href="#cb132-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-451"><a href="#cb132-451" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-452"><a href="#cb132-452" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-453"><a href="#cb132-453" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span>
<span id="cb132-454"><a href="#cb132-454" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-455"><a href="#cb132-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-456"><a href="#cb132-456" aria-hidden="true" tabindex="-1"></a>The <span class="in">`|`</span> operator is used to combine two logical tests with an 'or' condition. For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</span>
<span id="cb132-457"><a href="#cb132-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-460"><a href="#cb132-460" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-461"><a href="#cb132-461" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-462"><a href="#cb132-462" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span>
<span id="cb132-463"><a href="#cb132-463" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-464"><a href="#cb132-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-465"><a href="#cb132-465" aria-hidden="true" tabindex="-1"></a>The <span class="in">`%in%`</span> operator can be used to filter based on a vector of multiple values (<span class="in">`c(x, y)`</span>). It's particularly useful when you have a few character values you want to filter on, as it is shorter to type than <span class="in">`|`</span> (or).</span>
<span id="cb132-466"><a href="#cb132-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-467"><a href="#cb132-467" aria-hidden="true" tabindex="-1"></a>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <span class="in">`|`</span> like this:</span>
<span id="cb132-468"><a href="#cb132-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-471"><a href="#cb132-471" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-472"><a href="#cb132-472" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-473"><a href="#cb132-473" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span>
<span id="cb132-474"><a href="#cb132-474" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-475"><a href="#cb132-475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-476"><a href="#cb132-476" aria-hidden="true" tabindex="-1"></a>Or we could use <span class="in">`%in%`</span> like this:</span>
<span id="cb132-477"><a href="#cb132-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-480"><a href="#cb132-480" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-481"><a href="#cb132-481" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-482"><a href="#cb132-482" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span>
<span id="cb132-483"><a href="#cb132-483" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-484"><a href="#cb132-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-485"><a href="#cb132-485" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-486"><a href="#cb132-486" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-487"><a href="#cb132-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-488"><a href="#cb132-488" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-489"><a href="#cb132-489" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which code would you use to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice from replicate 2?</span>
<span id="cb132-490"><a href="#cb132-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-491"><a href="#cb132-491" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-492"><a href="#cb132-492" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-493"><a href="#cb132-493" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == 2)`</span></span>
<span id="cb132-494"><a href="#cb132-494" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-495"><a href="#cb132-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-496"><a href="#cb132-496" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-497"><a href="#cb132-497" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == rep2)`</span></span>
<span id="cb132-498"><a href="#cb132-498" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-499"><a href="#cb132-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-500"><a href="#cb132-500" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-501"><a href="#cb132-501" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == "rep2")`</span></span>
<span id="cb132-502"><a href="#cb132-502" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-503"><a href="#cb132-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-504"><a href="#cb132-504" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-505"><a href="#cb132-505" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(replicate = "rep2")`</span></span>
<span id="cb132-506"><a href="#cb132-506" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-507"><a href="#cb132-507" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-508"><a href="#cb132-508" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-509"><a href="#cb132-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-510"><a href="#cb132-510" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-511"><a href="#cb132-511" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is NOT a valid way to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</span>
<span id="cb132-512"><a href="#cb132-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-513"><a href="#cb132-513" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-514"><a href="#cb132-514" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-515"><a href="#cb132-515" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)`</span></span>
<span id="cb132-516"><a href="#cb132-516" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-517"><a href="#cb132-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-518"><a href="#cb132-518" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-519"><a href="#cb132-519" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span></span>
<span id="cb132-520"><a href="#cb132-520" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-521"><a href="#cb132-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-522"><a href="#cb132-522" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-523"><a href="#cb132-523" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)`</span></span>
<span id="cb132-524"><a href="#cb132-524" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-525"><a href="#cb132-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-526"><a href="#cb132-526" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-527"><a href="#cb132-527" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)`</span></span>
<span id="cb132-528"><a href="#cb132-528" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-529"><a href="#cb132-529" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-530"><a href="#cb132-530" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-531"><a href="#cb132-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-532"><a href="#cb132-532" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-533"><a href="#cb132-533" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Which option correctly describes what the following code is doing?</span>
<span id="cb132-534"><a href="#cb132-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-537"><a href="#cb132-537" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-538"><a href="#cb132-538" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-539"><a href="#cb132-539" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-540"><a href="#cb132-540" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb132-541"><a href="#cb132-541" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-542"><a href="#cb132-542" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span>
<span id="cb132-543"><a href="#cb132-543" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-544"><a href="#cb132-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-545"><a href="#cb132-545" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-546"><a href="#cb132-546" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-547"><a href="#cb132-547" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb132-548"><a href="#cb132-548" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-549"><a href="#cb132-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-550"><a href="#cb132-550" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-551"><a href="#cb132-551" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb132-552"><a href="#cb132-552" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-553"><a href="#cb132-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-554"><a href="#cb132-554" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-555"><a href="#cb132-555" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb132-556"><a href="#cb132-556" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-557"><a href="#cb132-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-558"><a href="#cb132-558" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-559"><a href="#cb132-559" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb132-560"><a href="#cb132-560" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-561"><a href="#cb132-561" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-562"><a href="#cb132-562" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-563"><a href="#cb132-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-564"><a href="#cb132-564" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-565"><a href="#cb132-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-566"><a href="#cb132-566" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-567"><a href="#cb132-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-568"><a href="#cb132-568" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to only include mice from replicate 2 is <span class="in">`mousezempic_dosage_data %&gt;% filter(replicate == "rep2")`</span>. Option A is incorrect because <span class="in">`2`</span> is not a value of <span class="in">`replicate`</span> (when filtering you need to know what values are actually in your columns! So make sure to <span class="in">`View()`</span> your data first). Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on. Option D is incorrect because <span class="in">`=`</span> is not the correct way to test for equality, you need to use <span class="in">`==`</span>.</span>
<span id="cb132-569"><a href="#cb132-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-570"><a href="#cb132-570" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The invalid option is <span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span>. This is because the second filtering step is missing the name of the filter function, so R doesn't know what to do with <span class="in">`(initial_weight_g &lt; 40)`</span>. The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <span class="in">`filter()`</span> functions in a row to apply multiple conditions, or the <span class="in">`&amp;`</span> operator to combine them into a single <span class="in">`filter()`</span> function. It's just a matter of personal preference.</span>
<span id="cb132-571"><a href="#cb132-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-572"><a href="#cb132-572" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct description of the code is that it filters the data frame to only include mice from the "BALB C" and "Black 6" strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</span>
<span id="cb132-573"><a href="#cb132-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-574"><a href="#cb132-574" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-575"><a href="#cb132-575" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-576"><a href="#cb132-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-577"><a href="#cb132-577" aria-hidden="true" tabindex="-1"></a><span class="fu">### Dealing with missing values {#sec-missing}</span></span>
<span id="cb132-578"><a href="#cb132-578" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-579"><a href="#cb132-579" aria-hidden="true" tabindex="-1"></a>Missing values are a common problem in real-world datasets. In R, missing values are represented by <span class="in">`NA`</span>. In fact, if you look at the <span class="in">`mousezempic_dosage_data`</span> data frame we've been using, you'll see that some of the cells contain <span class="in">`NA`</span>: try spotting them with the <span class="in">`View()`</span> function.</span>
<span id="cb132-580"><a href="#cb132-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-581"><a href="#cb132-581" aria-hidden="true" tabindex="-1"></a>You can also find missing values in a data frame using the <span class="in">`is.na()`</span> function in combination with <span class="in">`filter()`</span>. For example, to find all the rows in the <span class="in">`mousezempic_dosage_data`</span> data frame that have a missing value for the <span class="in">`drug_dose_g`</span> column:</span>
<span id="cb132-582"><a href="#cb132-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-585"><a href="#cb132-585" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-586"><a href="#cb132-586" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-587"><a href="#cb132-587" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb132-588"><a href="#cb132-588" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-589"><a href="#cb132-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-590"><a href="#cb132-590" aria-hidden="true" tabindex="-1"></a>The problem with missing values is that they can cause problems when you try to perform calculations on your data. For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <span class="in">`NA`</span>:</span>
<span id="cb132-591"><a href="#cb132-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-594"><a href="#cb132-594" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-595"><a href="#cb132-595" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb132-596"><a href="#cb132-596" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb132-597"><a href="#cb132-597" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span>
<span id="cb132-598"><a href="#cb132-598" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-599"><a href="#cb132-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-600"><a href="#cb132-600" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span> values in R are therefore referred to as 'contagious': if you put an <span class="in">`NA`</span> in you usually get an <span class="in">`NA`</span> out. If you think about it, that makes sense— when we don't know the value of a particular mouse's drug dose, how can we calculate the average? That missing value could be anything.</span>
<span id="cb132-601"><a href="#cb132-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-602"><a href="#cb132-602" aria-hidden="true" tabindex="-1"></a>For this reason, it's important to deal with missing values before performing calculations. Many functions in R will have an argument called <span class="in">`na.rm`</span> that you can set to <span class="in">`TRUE`</span> to remove missing values before performing the calculation. For example, to calculate the mean of the <span class="in">`drug_dose_g`</span> column with the missing values excluded:</span>
<span id="cb132-603"><a href="#cb132-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-606"><a href="#cb132-606" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-607"><a href="#cb132-607" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb132-608"><a href="#cb132-608" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from session 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb132-609"><a href="#cb132-609" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-610"><a href="#cb132-610" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-611"><a href="#cb132-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-612"><a href="#cb132-612" aria-hidden="true" tabindex="-1"></a>This time, the result is a number, because the missing values have been removed before the calculation.</span>
<span id="cb132-613"><a href="#cb132-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-614"><a href="#cb132-614" aria-hidden="true" tabindex="-1"></a>But not all functions have an <span class="in">`na.rm`</span> argument. In these cases, you can remove rows with missing values. This can be done for a single column, using the <span class="in">`filter()`</span> function together with <span class="in">`is.na()`</span>:</span>
<span id="cb132-615"><a href="#cb132-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-618"><a href="#cb132-618" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-619"><a href="#cb132-619" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb132-620"><a href="#cb132-620" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-621"><a href="#cb132-621" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb132-622"><a href="#cb132-622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb132-623"><a href="#cb132-623" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-624"><a href="#cb132-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-625"><a href="#cb132-625" aria-hidden="true" tabindex="-1"></a>Or, you can remove rows with missing values in any column using the <span class="in">`na.omit()`</span> function:</span>
<span id="cb132-626"><a href="#cb132-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-629"><a href="#cb132-629" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-630"><a href="#cb132-630" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb132-631"><a href="#cb132-631" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-632"><a href="#cb132-632" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb132-633"><a href="#cb132-633" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-634"><a href="#cb132-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-635"><a href="#cb132-635" aria-hidden="true" tabindex="-1"></a>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value. This can be done using the <span class="in">`replace_na()`</span> function from the <span class="in">`tidyr`</span> package. <span class="in">`replace_na()`</span> takes a <span class="in">`list()`</span> which contains each of the column names you want to edit, and the value that should be used.</span>
<span id="cb132-636"><a href="#cb132-636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-637"><a href="#cb132-637" aria-hidden="true" tabindex="-1"></a>For example, to replace missing values in the <span class="in">`weight_lost_g`</span> columns with 0, replace missing values in the <span class="in">`sex`</span> column with 'unknown' and leave the rest of the data frame unchanged:</span>
<span id="cb132-638"><a href="#cb132-638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-641"><a href="#cb132-641" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-642"><a href="#cb132-642" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb132-643"><a href="#cb132-643" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-644"><a href="#cb132-644" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb132-645"><a href="#cb132-645" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb132-646"><a href="#cb132-646" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span>
<span id="cb132-647"><a href="#cb132-647" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-648"><a href="#cb132-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-649"><a href="#cb132-649" aria-hidden="true" tabindex="-1"></a>When deciding how to handle missing values, you might have prior knowledge that <span class="in">`NA`</span> should be replaced with a specific value, or you might decide that removing rows with <span class="in">`NA`</span> is the best approach for your analysis.</span>
<span id="cb132-650"><a href="#cb132-650" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-651"><a href="#cb132-651" aria-hidden="true" tabindex="-1"></a>For example, maybe we knew that the mice were given a <span class="in">`weight_lost_g`</span> of <span class="in">`NA`</span> if they didn't lose any weight, it would then make sense to replace those with 0 (as we did in the code above). However, if the <span class="in">`drug_dose_g`</span> column was missing simply because the data was lost, we might choose to remove those rows entirely.</span>
<span id="cb132-652"><a href="#cb132-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-653"><a href="#cb132-653" aria-hidden="true" tabindex="-1"></a>It's important to think carefully about how missing values should be handled in your analysis.</span>
<span id="cb132-654"><a href="#cb132-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-655"><a href="#cb132-655" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-656"><a href="#cb132-656" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-657"><a href="#cb132-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-658"><a href="#cb132-658" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-659"><a href="#cb132-659" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What would be the result of running this R code: <span class="in">`mean(c(1, 2, 4, NA))`</span></span>
<span id="cb132-660"><a href="#cb132-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-661"><a href="#cb132-661" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-662"><a href="#cb132-662" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-663"><a href="#cb132-663" aria-hidden="true" tabindex="-1"></a>2.333333</span>
<span id="cb132-664"><a href="#cb132-664" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-665"><a href="#cb132-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-666"><a href="#cb132-666" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-667"><a href="#cb132-667" aria-hidden="true" tabindex="-1"></a>0</span>
<span id="cb132-668"><a href="#cb132-668" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-669"><a href="#cb132-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-670"><a href="#cb132-670" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-671"><a href="#cb132-671" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span></span>
<span id="cb132-672"><a href="#cb132-672" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-673"><a href="#cb132-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-674"><a href="#cb132-674" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-675"><a href="#cb132-675" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb132-676"><a href="#cb132-676" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-677"><a href="#cb132-677" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-678"><a href="#cb132-678" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-679"><a href="#cb132-679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-680"><a href="#cb132-680" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-681"><a href="#cb132-681" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Which line of code would you use to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column?</span>
<span id="cb132-682"><a href="#cb132-682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-683"><a href="#cb132-683" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-684"><a href="#cb132-684" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-685"><a href="#cb132-685" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(tail_length_mm != NA)`</span></span>
<span id="cb132-686"><a href="#cb132-686" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-687"><a href="#cb132-687" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-688"><a href="#cb132-688" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-689"><a href="#cb132-689" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(is.na(tail_length_mm))`</span></span>
<span id="cb132-690"><a href="#cb132-690" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-691"><a href="#cb132-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-692"><a href="#cb132-692" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-693"><a href="#cb132-693" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% na.omit()`</span></span>
<span id="cb132-694"><a href="#cb132-694" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-695"><a href="#cb132-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-696"><a href="#cb132-696" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-697"><a href="#cb132-697" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))`</span></span>
<span id="cb132-698"><a href="#cb132-698" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-699"><a href="#cb132-699" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-700"><a href="#cb132-700" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-701"><a href="#cb132-701" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-702"><a href="#cb132-702" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-703"><a href="#cb132-703" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> How would you replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35?</span>
<span id="cb132-704"><a href="#cb132-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-705"><a href="#cb132-705" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-706"><a href="#cb132-706" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-707"><a href="#cb132-707" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))`</span></span>
<span id="cb132-708"><a href="#cb132-708" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-709"><a href="#cb132-709" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-710"><a href="#cb132-710" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-711"><a href="#cb132-711" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(initial_weight_g = 35)`</span></span>
<span id="cb132-712"><a href="#cb132-712" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-713"><a href="#cb132-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-714"><a href="#cb132-714" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-715"><a href="#cb132-715" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g == 35))`</span></span>
<span id="cb132-716"><a href="#cb132-716" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-717"><a href="#cb132-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-718"><a href="#cb132-718" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-719"><a href="#cb132-719" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% replace_na(35)`</span></span>
<span id="cb132-720"><a href="#cb132-720" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-721"><a href="#cb132-721" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-722"><a href="#cb132-722" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-723"><a href="#cb132-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-724"><a href="#cb132-724" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-725"><a href="#cb132-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-726"><a href="#cb132-726" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-727"><a href="#cb132-727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-728"><a href="#cb132-728" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb132-729"><a href="#cb132-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-730"><a href="#cb132-730" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The result of running the code <span class="in">`mean(c(1, 2, 4, NA))`</span> is <span class="in">`NA`</span>. This is because the <span class="in">`NA`</span> value is 'contagious', so when you try to calculate the mean of a vector that contains an <span class="in">`NA`</span>, the result will also be <span class="in">`NA`</span>. If we wanted to calculate the mean of the vector without the <span class="in">`NA`</span>, we would need to use the <span class="in">`na.rm = TRUE`</span> argument.</span>
<span id="cb132-731"><a href="#cb132-731" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct line of code to filter the <span class="in">`mousezempic_dosage_data`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column is <span class="in">`mousezempic_dosage_data %&gt;% filter(!is.na(tail_length_mm))`</span>. The <span class="in">`!`</span> symbol is used to negate the result of <span class="in">`is.na()`</span>, so we are filtering to keep the rows where <span class="in">`tail_length_mm`</span> is not <span class="in">`NA`</span>. We can't use the first option with the <span class="in">`!= NA`</span> because <span class="in">`NA`</span> is a special value in R that represents missing data, and it can't be compared to anything, and the third option is incorrect because <span class="in">`na.omit()`</span> removes entire rows with missing values, rather than just filtering based on a single column.</span>
<span id="cb132-732"><a href="#cb132-732" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct line of code to replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35 is <span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(initial_weight_g = 35))`</span>. The <span class="in">`replace_na()`</span> function takes a <span class="in">`list()`</span> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we're not testing for equality, we're assigning a value.</span>
<span id="cb132-733"><a href="#cb132-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-734"><a href="#cb132-734" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb132-735"><a href="#cb132-735" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-736"><a href="#cb132-736" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-737"><a href="#cb132-737" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-738"><a href="#cb132-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-739"><a href="#cb132-739" aria-hidden="true" tabindex="-1"></a><span class="fu">### Selecting columns {#sec-select}</span></span>
<span id="cb132-740"><a href="#cb132-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-741"><a href="#cb132-741" aria-hidden="true" tabindex="-1"></a><span class="al">![Select allows you to select only certain columns](images/S02F04_select.png)</span></span>
<span id="cb132-742"><a href="#cb132-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-743"><a href="#cb132-743" aria-hidden="true" tabindex="-1"></a>While <span class="in">`filter()`</span> is used to subset rows, <span class="in">`select()`</span> is used to subset columns. You can use <span class="in">`select()`</span> to keep only the columns you're interested in, or to drop columns you don't need.</span>
<span id="cb132-744"><a href="#cb132-744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-745"><a href="#cb132-745" aria-hidden="true" tabindex="-1"></a>The <span class="in">`select()`</span> function takes the names of the columns that you want to keep/remove (no vector notation <span class="in">`c()`</span> or quotation marks <span class="in">`""`</span> necessary). For example, to select only the <span class="in">`mouse_strain`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame:</span>
<span id="cb132-746"><a href="#cb132-746" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-749"><a href="#cb132-749" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-750"><a href="#cb132-750" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-751"><a href="#cb132-751" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span>
<span id="cb132-752"><a href="#cb132-752" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-753"><a href="#cb132-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-754"><a href="#cb132-754" aria-hidden="true" tabindex="-1"></a>We can see that all the other columns have been removed from the data frame.</span>
<span id="cb132-755"><a href="#cb132-755" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-756"><a href="#cb132-756" aria-hidden="true" tabindex="-1"></a>If you want to keep all columns except for a few, you can use <span class="in">`-`</span> to drop columns. For example, to keep all columns except for <span class="in">`cage_number`</span> and <span class="in">`sex`</span>:</span>
<span id="cb132-757"><a href="#cb132-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-760"><a href="#cb132-760" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-761"><a href="#cb132-761" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-762"><a href="#cb132-762" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span>
<span id="cb132-763"><a href="#cb132-763" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-764"><a href="#cb132-764" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-765"><a href="#cb132-765" aria-hidden="true" tabindex="-1"></a>There are also some helper functions that can be used to select columns based on their names :</span>
<span id="cb132-766"><a href="#cb132-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-767"><a href="#cb132-767" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb132-768"><a href="#cb132-768" aria-hidden="true" tabindex="-1"></a>| Function        | Description                                       | Example                                         |</span>
<span id="cb132-769"><a href="#cb132-769" aria-hidden="true" tabindex="-1"></a>+=================+===================================================+=================================================+</span>
<span id="cb132-770"><a href="#cb132-770" aria-hidden="true" tabindex="-1"></a>| <span class="in">`starts_with()`</span> | select column(s) that start with a certain string | select all columns starting with the letter i   |</span>
<span id="cb132-771"><a href="#cb132-771" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb132-772"><a href="#cb132-772" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(starts_with("i"))`</span>                      |</span>
<span id="cb132-773"><a href="#cb132-773" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb132-774"><a href="#cb132-774" aria-hidden="true" tabindex="-1"></a>| <span class="in">`ends_with()`</span>   | select column(s) that end with a certain string   | select all columns ending with <span class="sc">\_</span>g              |</span>
<span id="cb132-775"><a href="#cb132-775" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb132-776"><a href="#cb132-776" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(ends_with("_g"))`</span>                       |</span>
<span id="cb132-777"><a href="#cb132-777" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb132-778"><a href="#cb132-778" aria-hidden="true" tabindex="-1"></a>| <span class="in">`contains()`</span>    | select column(s) that contain a certain string    | select all columns containing the word 'weight' |</span>
<span id="cb132-779"><a href="#cb132-779" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb132-780"><a href="#cb132-780" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(contains("weight"))`</span>                    |</span>
<span id="cb132-781"><a href="#cb132-781" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb132-782"><a href="#cb132-782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-783"><a href="#cb132-783" aria-hidden="true" tabindex="-1"></a>: There are several helper functions that can be used with the select function</span>
<span id="cb132-784"><a href="#cb132-784" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-785"><a href="#cb132-785" aria-hidden="true" tabindex="-1"></a>You need to use quotation marks around the arguments in these helper functions, as they aren't full column names, just strings of characters.</span>
<span id="cb132-786"><a href="#cb132-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-787"><a href="#cb132-787" aria-hidden="true" tabindex="-1"></a>Try using these helper functions to select columns from the <span class="in">`mousezempic_dosage_data`</span> data frame!</span>
<span id="cb132-788"><a href="#cb132-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-789"><a href="#cb132-789" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Reordering columns"}</span>
<span id="cb132-790"><a href="#cb132-790" aria-hidden="true" tabindex="-1"></a><span class="al">![Relocate allows you to move columns around](images/S02F05_reorder.png)</span> We can reorder columns using the <span class="in">`relocate()`</span> function, which works similarly to <span class="in">`select()`</span> (except it just moves columns around rather than dropping/keeping them). For example, to move the <span class="in">`sex`</span> column to before the <span class="in">`cage_number`</span> column:</span>
<span id="cb132-791"><a href="#cb132-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-794"><a href="#cb132-794" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-795"><a href="#cb132-795" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-796"><a href="#cb132-796" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb132-797"><a href="#cb132-797" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span>
<span id="cb132-798"><a href="#cb132-798" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-799"><a href="#cb132-799" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-800"><a href="#cb132-800" aria-hidden="true" tabindex="-1"></a>Two useful helper functions here are the <span class="in">`everything()`</span> and <span class="in">`last_col()`</span> functions, which can be used to move columns to the start/end of the data frame.</span>
<span id="cb132-801"><a href="#cb132-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-804"><a href="#cb132-804" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-805"><a href="#cb132-805" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb132-806"><a href="#cb132-806" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-807"><a href="#cb132-807" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span>
<span id="cb132-808"><a href="#cb132-808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-809"><a href="#cb132-809" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb132-810"><a href="#cb132-810" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-811"><a href="#cb132-811" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span>
<span id="cb132-812"><a href="#cb132-812" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-813"><a href="#cb132-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-814"><a href="#cb132-814" aria-hidden="true" tabindex="-1"></a>Re-ordering columns isn't necessary, but it makes it easier to see the data you're most interested in within the console (since often not all of the columns will fit on the screen at once). For example, if we are doing a lot of computation on the <span class="in">`initial_weight_g`</span> column, we'd probably like to have that near the start so we can easily check it.</span>
<span id="cb132-815"><a href="#cb132-815" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-816"><a href="#cb132-816" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-817"><a href="#cb132-817" aria-hidden="true" tabindex="-1"></a>Note that the output of the <span class="in">`select()`</span> function is a new data frame, even if you only select a single column:</span>
<span id="cb132-818"><a href="#cb132-818" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-821"><a href="#cb132-821" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-822"><a href="#cb132-822" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb132-823"><a href="#cb132-823" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-824"><a href="#cb132-824" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb132-825"><a href="#cb132-825" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from session 1 that class() tells us the type of an object</span></span>
<span id="cb132-826"><a href="#cb132-826" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb132-827"><a href="#cb132-827" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-828"><a href="#cb132-828" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-829"><a href="#cb132-829" aria-hidden="true" tabindex="-1"></a>Sometimes, we instead want to get the values of a column as a vector.</span>
<span id="cb132-830"><a href="#cb132-830" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-831"><a href="#cb132-831" aria-hidden="true" tabindex="-1"></a><span class="al">![Pull allows you to pull acolumn out of a data frame as a vector](images/S02F06_pull.png)</span></span>
<span id="cb132-832"><a href="#cb132-832" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-833"><a href="#cb132-833" aria-hidden="true" tabindex="-1"></a>We can do this by using the <span class="in">`pull()`</span> function, which extracts a single column from a data frame as a vector:</span>
<span id="cb132-834"><a href="#cb132-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-837"><a href="#cb132-837" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-838"><a href="#cb132-838" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb132-839"><a href="#cb132-839" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-840"><a href="#cb132-840" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb132-841"><a href="#cb132-841" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb132-842"><a href="#cb132-842" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-843"><a href="#cb132-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-844"><a href="#cb132-844" aria-hidden="true" tabindex="-1"></a>We can see that the class of the output is now a vector, rather than a data frame. This is important because some functions only accept vectors, not data frames, like <span class="in">`mean()`</span> for example:</span>
<span id="cb132-845"><a href="#cb132-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-848"><a href="#cb132-848" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-849"><a href="#cb132-849" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb132-850"><a href="#cb132-850" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-851"><a href="#cb132-851" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-852"><a href="#cb132-852" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb132-853"><a href="#cb132-853" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-854"><a href="#cb132-854" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-855"><a href="#cb132-855" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-856"><a href="#cb132-856" aria-hidden="true" tabindex="-1"></a>Note how both times we used <span class="in">`na.rm = TRUE`</span> to remove missing values before calculating the mean.</span>
<span id="cb132-857"><a href="#cb132-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-858"><a href="#cb132-858" aria-hidden="true" tabindex="-1"></a>You might remember that we used the <span class="in">`$`</span> operator in the previous session to extract a single column from a data frame, so why use <span class="in">`pull()`</span> instead? The main reason is that <span class="in">`pull()`</span> works within a chain of pipes, whereas <span class="in">`$`</span> doesn't.</span>
<span id="cb132-859"><a href="#cb132-859" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-860"><a href="#cb132-860" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to know the average initial weight of mice that lost at least 4g. We can do this by chaining <span class="in">`filter()`</span> and <span class="in">`pull()`</span> together:</span>
<span id="cb132-861"><a href="#cb132-861" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-864"><a href="#cb132-864" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-865"><a href="#cb132-865" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-866"><a href="#cb132-866" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb132-867"><a href="#cb132-867" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-868"><a href="#cb132-868" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb132-869"><a href="#cb132-869" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb132-870"><a href="#cb132-870" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb132-871"><a href="#cb132-871" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-872"><a href="#cb132-872" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-873"><a href="#cb132-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-874"><a href="#cb132-874" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-875"><a href="#cb132-875" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-876"><a href="#cb132-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-877"><a href="#cb132-877" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-878"><a href="#cb132-878" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which line of code would NOT be a valid way to select the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb132-879"><a href="#cb132-879" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-880"><a href="#cb132-880" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-881"><a href="#cb132-881" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-882"><a href="#cb132-882" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)`</span></span>
<span id="cb132-883"><a href="#cb132-883" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-884"><a href="#cb132-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-885"><a href="#cb132-885" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-886"><a href="#cb132-886" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(contains("g"))`</span></span>
<span id="cb132-887"><a href="#cb132-887" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-888"><a href="#cb132-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-889"><a href="#cb132-889" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-890"><a href="#cb132-890" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(ends_with("_g"))`</span></span>
<span id="cb132-891"><a href="#cb132-891" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-892"><a href="#cb132-892" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-893"><a href="#cb132-893" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-894"><a href="#cb132-894" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)`</span></span>
<span id="cb132-895"><a href="#cb132-895" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-896"><a href="#cb132-896" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-897"><a href="#cb132-897" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-898"><a href="#cb132-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-899"><a href="#cb132-899" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-900"><a href="#cb132-900" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> How would I extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`mousezempic_dosage_data`</span> data frame as a vector?</span>
<span id="cb132-901"><a href="#cb132-901" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-902"><a href="#cb132-902" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-903"><a href="#cb132-903" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-904"><a href="#cb132-904" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% filter(initial_weight_g)`</span></span>
<span id="cb132-905"><a href="#cb132-905" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-906"><a href="#cb132-906" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-907"><a href="#cb132-907" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-908"><a href="#cb132-908" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% $initial_weight_g`</span></span>
<span id="cb132-909"><a href="#cb132-909" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-910"><a href="#cb132-910" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-911"><a href="#cb132-911" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-912"><a href="#cb132-912" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% select(initial_weight_g)`</span></span>
<span id="cb132-913"><a href="#cb132-913" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-914"><a href="#cb132-914" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-915"><a href="#cb132-915" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-916"><a href="#cb132-916" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% pull(initial_weight_g)`</span></span>
<span id="cb132-917"><a href="#cb132-917" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-918"><a href="#cb132-918" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-919"><a href="#cb132-919" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-920"><a href="#cb132-920" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-921"><a href="#cb132-921" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-922"><a href="#cb132-922" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> How would you move the <span class="in">`sex`</span> column to the end of the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb132-923"><a href="#cb132-923" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-924"><a href="#cb132-924" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-925"><a href="#cb132-925" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-926"><a href="#cb132-926" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex)`</span></span>
<span id="cb132-927"><a href="#cb132-927" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-928"><a href="#cb132-928" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-929"><a href="#cb132-929" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-930"><a href="#cb132-930" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col)`</span></span>
<span id="cb132-931"><a href="#cb132-931" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-932"><a href="#cb132-932" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-933"><a href="#cb132-933" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-934"><a href="#cb132-934" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())`</span></span>
<span id="cb132-935"><a href="#cb132-935" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-936"><a href="#cb132-936" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-937"><a href="#cb132-937" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-938"><a href="#cb132-938" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% reorder(sex, .after = last_col())`</span></span>
<span id="cb132-939"><a href="#cb132-939" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-940"><a href="#cb132-940" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-941"><a href="#cb132-941" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-942"><a href="#cb132-942" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-943"><a href="#cb132-943" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-944"><a href="#cb132-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-945"><a href="#cb132-945" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-946"><a href="#cb132-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-947"><a href="#cb132-947" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb132-948"><a href="#cb132-948" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-949"><a href="#cb132-949" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The line of code that would NOT be a valid way to select the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% select(contains("g"))`</span>. This line of code would select all columns that contain the letter 'g', which would include columns like <span class="in">`cage_number`</span> and <span class="in">`tail_length_mm`</span>. We need to specify either <span class="in">`ends_with("g")`</span> or <span class="in">`contains("_g")`</span> to only get those with <span class="in">`_g`</span> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</span>
<span id="cb132-950"><a href="#cb132-950" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct way to extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`mousezempic_dosage_data`</span> data frame as a vector is <span class="in">`mousezempic_dosage_data %&gt;% pull(initial_weight_g)`</span>. The <span class="in">`pull()`</span> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <span class="in">`filter()`</span> is used to subset rows, <span class="in">`$`</span> is not used in a pipe chain, and <span class="in">`select()`</span> is outputs a data frame, not extract them as vectors.</span>
<span id="cb132-951"><a href="#cb132-951" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct way to move the <span class="in">`sex`</span> column to the end of the <span class="in">`mousezempic_dosage_data`</span> data frame is using the <span class="in">`relocate()`</span> function like this: <span class="in">`mousezempic_dosage_data %&gt;% relocate(sex, .after = last_col())`</span>. The <span class="in">`last_col()`</span> function is used to refer to the last column in the data frame. The other options are incorrect because <span class="in">`reorder()`</span> is not a valid function, and you need to remember to include the brackets <span class="in">`()`</span> when using <span class="in">`last_col()`</span>.</span>
<span id="cb132-952"><a href="#cb132-952" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-953"><a href="#cb132-953" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb132-954"><a href="#cb132-954" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-955"><a href="#cb132-955" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-956"><a href="#cb132-956" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-957"><a href="#cb132-957" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-958"><a href="#cb132-958" aria-hidden="true" tabindex="-1"></a><span class="fu">### Modifying data {#sec-mutate}</span></span>
<span id="cb132-959"><a href="#cb132-959" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-960"><a href="#cb132-960" aria-hidden="true" tabindex="-1"></a>So far, we've learned how to filter rows and select columns from a data frame. But what if we want to change the data itself? This is where the <span class="in">`mutate()`</span> function comes in.</span>
<span id="cb132-961"><a href="#cb132-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-962"><a href="#cb132-962" aria-hidden="true" tabindex="-1"></a><span class="al">![Mutate allows you to add new columns to a data frame](images/S02F07_mutate.png)</span></span>
<span id="cb132-963"><a href="#cb132-963" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-964"><a href="#cb132-964" aria-hidden="true" tabindex="-1"></a>The <span class="in">`mutate()`</span> function is used to add new columns to a data frame, or modify existing columns, often by performing some sort of calculation. For example, we can add a new column to <span class="in">`mousezempic_dosage_data`</span> that contains the drug dose in mg (rather than g):</span>
<span id="cb132-965"><a href="#cb132-965" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-968"><a href="#cb132-968" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-969"><a href="#cb132-969" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-970"><a href="#cb132-970" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a new column called drug_dose_mg</span></span>
<span id="cb132-971"><a href="#cb132-971" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert drug_dose_g to mg by multiplying by 1000</span></span>
<span id="cb132-972"><a href="#cb132-972" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug_dose_mg =</span> drug_dose_g <span class="sc">*</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-973"><a href="#cb132-973" aria-hidden="true" tabindex="-1"></a>  <span class="co"># just select the drug dose columns so we can compare them</span></span>
<span id="cb132-974"><a href="#cb132-974" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(drug_dose_g, drug_dose_mg)</span>
<span id="cb132-975"><a href="#cb132-975" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-976"><a href="#cb132-976" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-977"><a href="#cb132-977" aria-hidden="true" tabindex="-1"></a>You can see that the <span class="in">`drug_dose_mg`</span> column has been added to the data frame, and it contains, for each row, the value of the <span class="in">`drug_dose_g`</span> column multiplied by 1000 (<span class="in">`NA`</span> values are preserved).</span>
<span id="cb132-978"><a href="#cb132-978" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-979"><a href="#cb132-979" aria-hidden="true" tabindex="-1"></a>These calculations can be as complex as you like, and involve multiple different columns. For example, to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that calculates the weight lost as a percentage of the initial weight:</span>
<span id="cb132-980"><a href="#cb132-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-983"><a href="#cb132-983" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-984"><a href="#cb132-984" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-985"><a href="#cb132-985" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate the % of initial weight that was lost</span></span>
<span id="cb132-986"><a href="#cb132-986" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_percent =</span> (weight_lost_g <span class="sc">/</span> initial_weight_g) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb132-987"><a href="#cb132-987" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-988"><a href="#cb132-988" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-989"><a href="#cb132-989" aria-hidden="true" tabindex="-1"></a>A useful helper function for <span class="in">`mutate()`</span> is the <span class="in">`case_when()`</span> function, which allows you to create new columns based on multiple conditions. We do this with the notation <span class="in">`case_when(condition1 ~ value1, condition2 ~ value2, ...)`</span>.</span>
<span id="cb132-990"><a href="#cb132-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-991"><a href="#cb132-991" aria-hidden="true" tabindex="-1"></a>For example, to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that categorises the mice based on how much weight they lost:</span>
<span id="cb132-992"><a href="#cb132-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-995"><a href="#cb132-995" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-996"><a href="#cb132-996" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-997"><a href="#cb132-997" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb132-998"><a href="#cb132-998" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb132-999"><a href="#cb132-999" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb132-1000"><a href="#cb132-1000" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb132-1001"><a href="#cb132-1001" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb132-1002"><a href="#cb132-1002" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb132-1003"><a href="#cb132-1003" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span>
<span id="cb132-1004"><a href="#cb132-1004" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1005"><a href="#cb132-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1006"><a href="#cb132-1006" aria-hidden="true" tabindex="-1"></a>Note that the conditions are evaluated in order, and the first condition that is <span class="in">`TRUE`</span> is the one that is used. So if a mouse lost 4.5g, it <span class="in">`case_when()`</span> would first test if it fits the 'Low' category (by checking if 4.5 is less than 4, which it isn't), and then if it fits the 'Medium' category (by checking if 4.5 is less than or equal to 5). Since it is, the mouse would be categorised as 'Medium'.</span>
<span id="cb132-1007"><a href="#cb132-1007" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1008"><a href="#cb132-1008" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Fallback with default value(s)"}</span>
<span id="cb132-1009"><a href="#cb132-1009" aria-hidden="true" tabindex="-1"></a>In the above example, what would happen if a mouse lost -1g (gained weight)? It wouldn't fit any of the conditions, so it would get an <span class="in">`NA`</span> in the <span class="in">`weight_loss_category`</span> column. Sometimes you might want this behaviour, but other times you would prefer to specify a 'fallback' category that will be assigned to everything that doesn't fit in the other categories. You can do this by including a <span class="in">` .default = `</span> argument at the end of the <span class="in">`case_when()`</span> function. For example:</span>
<span id="cb132-1010"><a href="#cb132-1010" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1013"><a href="#cb132-1013" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1014"><a href="#cb132-1014" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1015"><a href="#cb132-1015" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a new column called weight_loss_category</span></span>
<span id="cb132-1016"><a href="#cb132-1016" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_loss_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb132-1017"><a href="#cb132-1017" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span>, <span class="co"># separate conditions with a comma</span></span>
<span id="cb132-1018"><a href="#cb132-1018" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb132-1019"><a href="#cb132-1019" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span>,</span>
<span id="cb132-1020"><a href="#cb132-1020" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"Unknown"</span></span>
<span id="cb132-1021"><a href="#cb132-1021" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb132-1022"><a href="#cb132-1022" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight_lost_g, weight_loss_category)</span>
<span id="cb132-1023"><a href="#cb132-1023" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1024"><a href="#cb132-1024" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1025"><a href="#cb132-1025" aria-hidden="true" tabindex="-1"></a>Notice how the <span class="in">`NA`</span> value in the fourth row is now categorised as 'Unknown'.</span>
<span id="cb132-1026"><a href="#cb132-1026" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1027"><a href="#cb132-1027" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1028"><a href="#cb132-1028" aria-hidden="true" tabindex="-1"></a>One final thing to note is that <span class="in">`mutate()`</span> can be used to modify existing columns as well as add new ones. To do this, just use the name of the existing column as the 'new' one.</span>
<span id="cb132-1029"><a href="#cb132-1029" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1030"><a href="#cb132-1030" aria-hidden="true" tabindex="-1"></a>For example, let's use <span class="in">`mutate()`</span> together with <span class="in">`case_when()`</span> to modify the <span class="in">`sex`</span> column so that it uses <span class="in">`M`</span> and <span class="in">`F`</span> instead <span class="in">`male`</span> and <span class="in">`female`</span>:</span>
<span id="cb132-1031"><a href="#cb132-1031" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1034"><a href="#cb132-1034" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1035"><a href="#cb132-1035" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1036"><a href="#cb132-1036" aria-hidden="true" tabindex="-1"></a>  <span class="co"># modify sex column</span></span>
<span id="cb132-1037"><a href="#cb132-1037" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">case_when</span>(</span>
<span id="cb132-1038"><a href="#cb132-1038" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">~</span> <span class="st">"F"</span>,</span>
<span id="cb132-1039"><a href="#cb132-1039" aria-hidden="true" tabindex="-1"></a>    sex <span class="sc">==</span> <span class="st">"male"</span> <span class="sc">~</span> <span class="st">"M"</span>,</span>
<span id="cb132-1040"><a href="#cb132-1040" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if neither, code it as 'X'</span></span>
<span id="cb132-1041"><a href="#cb132-1041" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">"X"</span>))</span>
<span id="cb132-1042"><a href="#cb132-1042" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1043"><a href="#cb132-1043" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1044"><a href="#cb132-1044" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-1045"><a href="#cb132-1045" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-1046"><a href="#cb132-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1047"><a href="#cb132-1047" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1048"><a href="#cb132-1048" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What line of code would you use to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that converts the <span class="in">`tail_length_mm`</span> column to cm?</span>
<span id="cb132-1049"><a href="#cb132-1049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1050"><a href="#cb132-1050" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1051"><a href="#cb132-1051" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1052"><a href="#cb132-1052" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% create(tail_length_cm = tail_length_mm / 10)`</span></span>
<span id="cb132-1053"><a href="#cb132-1053" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1054"><a href="#cb132-1054" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1055"><a href="#cb132-1055" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1056"><a href="#cb132-1056" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm == tail_length_mm / 10)`</span></span>
<span id="cb132-1057"><a href="#cb132-1057" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1058"><a href="#cb132-1058" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1059"><a href="#cb132-1059" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1060"><a href="#cb132-1060" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)`</span></span>
<span id="cb132-1061"><a href="#cb132-1061" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1062"><a href="#cb132-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1063"><a href="#cb132-1063" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1064"><a href="#cb132-1064" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% tail_length_cm = tail_length_mm / 10`</span></span>
<span id="cb132-1065"><a href="#cb132-1065" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1066"><a href="#cb132-1066" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1067"><a href="#cb132-1067" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1068"><a href="#cb132-1068" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1069"><a href="#cb132-1069" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1070"><a href="#cb132-1070" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb132-1071"><a href="#cb132-1071" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1074"><a href="#cb132-1074" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1075"><a href="#cb132-1075" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1076"><a href="#cb132-1076" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1077"><a href="#cb132-1077" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1078"><a href="#cb132-1078" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_rank =</span> <span class="fu">row_number</span>())</span>
<span id="cb132-1079"><a href="#cb132-1079" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1080"><a href="#cb132-1080" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1081"><a href="#cb132-1081" aria-hidden="true" tabindex="-1"></a>Hint: the row_number() function returns the number of each row in the data frame (1 being the first row and so on).</span>
<span id="cb132-1082"><a href="#cb132-1082" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1083"><a href="#cb132-1083" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1084"><a href="#cb132-1084" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1085"><a href="#cb132-1085" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the least weight.</span>
<span id="cb132-1086"><a href="#cb132-1086" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1087"><a href="#cb132-1087" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1088"><a href="#cb132-1088" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1089"><a href="#cb132-1089" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight.</span>
<span id="cb132-1090"><a href="#cb132-1090" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1091"><a href="#cb132-1091" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1092"><a href="#cb132-1092" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1093"><a href="#cb132-1093" aria-hidden="true" tabindex="-1"></a>Adds a new column to the data frame that ranks the mice</span>
<span id="cb132-1094"><a href="#cb132-1094" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1095"><a href="#cb132-1095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1096"><a href="#cb132-1096" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1097"><a href="#cb132-1097" aria-hidden="true" tabindex="-1"></a>Does nothing, because the <span class="in">`row_number()`</span> function has no arguments</span>
<span id="cb132-1098"><a href="#cb132-1098" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1099"><a href="#cb132-1099" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1100"><a href="#cb132-1100" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1101"><a href="#cb132-1101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1102"><a href="#cb132-1102" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1103"><a href="#cb132-1103" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> What is wrong with this R code?</span>
<span id="cb132-1104"><a href="#cb132-1104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1107"><a href="#cb132-1107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1108"><a href="#cb132-1108" aria-hidden="true" tabindex="-1"></a><span class="co">#| error: true</span></span>
<span id="cb132-1109"><a href="#cb132-1109" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1110"><a href="#cb132-1110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_category =</span> <span class="fu">case_when</span>(</span>
<span id="cb132-1111"><a href="#cb132-1111" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Low"</span></span>
<span id="cb132-1112"><a href="#cb132-1112" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&lt;=</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"Medium"</span></span>
<span id="cb132-1113"><a href="#cb132-1113" aria-hidden="true" tabindex="-1"></a>    weight_lost_g <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">"High"</span></span>
<span id="cb132-1114"><a href="#cb132-1114" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb132-1115"><a href="#cb132-1115" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1116"><a href="#cb132-1116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1117"><a href="#cb132-1117" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1118"><a href="#cb132-1118" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1119"><a href="#cb132-1119" aria-hidden="true" tabindex="-1"></a>You didn't include a <span class="in">`.default =`</span> condition at the end of the <span class="in">`case_when()`</span> function to act as a fallback</span>
<span id="cb132-1120"><a href="#cb132-1120" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1121"><a href="#cb132-1121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1122"><a href="#cb132-1122" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1123"><a href="#cb132-1123" aria-hidden="true" tabindex="-1"></a>You can't use the <span class="in">`case_when()`</span> function with the <span class="in">`mutate()`</span> function</span>
<span id="cb132-1124"><a href="#cb132-1124" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1125"><a href="#cb132-1125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1126"><a href="#cb132-1126" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1127"><a href="#cb132-1127" aria-hidden="true" tabindex="-1"></a><span class="in">`weight_lost_g`</span> is not a valid column name</span>
<span id="cb132-1128"><a href="#cb132-1128" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1129"><a href="#cb132-1129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1130"><a href="#cb132-1130" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1131"><a href="#cb132-1131" aria-hidden="true" tabindex="-1"></a>You need to separate the conditions in the <span class="in">`case_when()`</span> function with a comma</span>
<span id="cb132-1132"><a href="#cb132-1132" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1133"><a href="#cb132-1133" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1134"><a href="#cb132-1134" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1135"><a href="#cb132-1135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1136"><a href="#cb132-1136" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1137"><a href="#cb132-1137" aria-hidden="true" tabindex="-1"></a>4<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb132-1138"><a href="#cb132-1138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1141"><a href="#cb132-1141" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1142"><a href="#cb132-1142" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1143"><a href="#cb132-1143" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1144"><a href="#cb132-1144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_strain =</span> <span class="fu">case_when</span>(</span>
<span id="cb132-1145"><a href="#cb132-1145" aria-hidden="true" tabindex="-1"></a>    mouse_strain <span class="sc">==</span> <span class="st">"Black 6"</span> <span class="sc">~</span> <span class="st">"B6"</span>,</span>
<span id="cb132-1146"><a href="#cb132-1146" aria-hidden="true" tabindex="-1"></a>    <span class="at">.default =</span> mouse_strain</span>
<span id="cb132-1147"><a href="#cb132-1147" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb132-1148"><a href="#cb132-1148" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1149"><a href="#cb132-1149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1150"><a href="#cb132-1150" aria-hidden="true" tabindex="-1"></a>Hint: if you're not sure, try running the code, but pipe it into <span class="in">`View()`</span> so that you can take a good look at what's happening in the <span class="in">`mouse_strain`</span> column.</span>
<span id="cb132-1151"><a href="#cb132-1151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1152"><a href="#cb132-1152" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1153"><a href="#cb132-1153" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1154"><a href="#cb132-1154" aria-hidden="true" tabindex="-1"></a>Renames the strains of all the mice to "B6", regardless of their original strain</span>
<span id="cb132-1155"><a href="#cb132-1155" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1156"><a href="#cb132-1156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1157"><a href="#cb132-1157" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1158"><a href="#cb132-1158" aria-hidden="true" tabindex="-1"></a>This code will produce an error</span>
<span id="cb132-1159"><a href="#cb132-1159" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1160"><a href="#cb132-1160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1161"><a href="#cb132-1161" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1162"><a href="#cb132-1162" aria-hidden="true" tabindex="-1"></a>Adds a new column that categorises the mice based on their strain, so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged.</span>
<span id="cb132-1163"><a href="#cb132-1163" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1164"><a href="#cb132-1164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1165"><a href="#cb132-1165" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1166"><a href="#cb132-1166" aria-hidden="true" tabindex="-1"></a>Modifies the <span class="in">`mouse_strain`</span> column so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged.</span>
<span id="cb132-1167"><a href="#cb132-1167" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1168"><a href="#cb132-1168" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1169"><a href="#cb132-1169" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1170"><a href="#cb132-1170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1171"><a href="#cb132-1171" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-1172"><a href="#cb132-1172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1173"><a href="#cb132-1173" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-1174"><a href="#cb132-1174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1175"><a href="#cb132-1175" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb132-1176"><a href="#cb132-1176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1177"><a href="#cb132-1177" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct line of code to add a new column to the <span class="in">`mousezempic_dosage_data`</span> data frame that converts the <span class="in">`tail_length_mm`</span> column to cm is <span class="in">`mousezempic_dosage_data %&gt;% mutate(tail_length_cm = tail_length_mm / 10)`</span>.</span>
<span id="cb132-1178"><a href="#cb132-1178" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% arrange(desc(weight_lost_g)) %&gt;% mutate(weight_lost_rank = row_number())`</span> adds a new column to the data frame that ranks the mice based on how much weight they lost, with 1 being the mouse that lost the most weight. First, the <span class="in">`arrange(desc(weight_lost_g))`</span> function sorts the data frame by the <span class="in">`weight_lost_g`</span> column in descending order, and then the <span class="in">`mutate(weight_lost_rank = row_number())`</span> function adds a new column that assigns a rank to each row based on its position (row number) in the sorted data frame.</span>
<span id="cb132-1179"><a href="#cb132-1179" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The error is that the conditions in the <span class="in">`case_when()`</span> function are not separated by commas. Each condition should be followed by a comma because these are like the arguments in a function. Remeber that it's optional to include the <span class="in">`.default =`</span> condition at the end of the <span class="in">`case_when()`</span> function.</span>
<span id="cb132-1180"><a href="#cb132-1180" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% mutate(mouse_strain = case_when(mouse_strain == "Black 6" ~ "B6", .default = mouse_strain))`</span> modifies the <span class="in">`mouse_strain`</span> column so that any mice from the "Black 6" strain are now called "B6", and all other strains are left unchanged. As we are calling our column <span class="in">`mouse_strain`</span>, no new column is being created (we are modifying the existing one) and the <span class="in">`.default = mouse_strain`</span> condition acts as a fallback to keep the original values (that already exist in the <span class="in">`mouse_strain`</span> column) for any rows that don't match our first condition (strain being "Black 6").</span>
<span id="cb132-1181"><a href="#cb132-1181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1182"><a href="#cb132-1182" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb132-1183"><a href="#cb132-1183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1184"><a href="#cb132-1184" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-1185"><a href="#cb132-1185" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1186"><a href="#cb132-1186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1187"><a href="#cb132-1187" aria-hidden="true" tabindex="-1"></a><span class="fu">### Summarising data {#sec-summarise}</span></span>
<span id="cb132-1188"><a href="#cb132-1188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1189"><a href="#cb132-1189" aria-hidden="true" tabindex="-1"></a><span class="al">![Summarise allows you to calculate summary statistics that collapse many rows into one](images/S02F08_summarise.png)</span></span>
<span id="cb132-1190"><a href="#cb132-1190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1191"><a href="#cb132-1191" aria-hidden="true" tabindex="-1"></a>The <span class="in">`summarise()`</span> (or <span class="in">`summarize()`</span>, if you prefer US spelling) function is used to calculate summary statistics on your data. It takes similar arguments to <span class="in">`mutate()`</span>, but instead of adding a new column to the data frame, it returns a new data frame with a single row and one column for each summary statistic you calculate.</span>
<span id="cb132-1192"><a href="#cb132-1192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1193"><a href="#cb132-1193" aria-hidden="true" tabindex="-1"></a>For example, to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame:</span>
<span id="cb132-1194"><a href="#cb132-1194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1197"><a href="#cb132-1197" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1198"><a href="#cb132-1198" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1199"><a href="#cb132-1199" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb132-1200"><a href="#cb132-1200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1201"><a href="#cb132-1201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1202"><a href="#cb132-1202" aria-hidden="true" tabindex="-1"></a>We can also calculate multiple summary statistics at once. For example, to calculate the mean, median, and standard deviation of the weight lost by the mice:</span>
<span id="cb132-1203"><a href="#cb132-1203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1206"><a href="#cb132-1206" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1207"><a href="#cb132-1207" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1208"><a href="#cb132-1208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1209"><a href="#cb132-1209" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1210"><a href="#cb132-1210" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1211"><a href="#cb132-1211" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-1212"><a href="#cb132-1212" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb132-1213"><a href="#cb132-1213" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1214"><a href="#cb132-1214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1215"><a href="#cb132-1215" aria-hidden="true" tabindex="-1"></a>The power of summarising data is really seen when combined with grouping, which we will cover in the next section.</span>
<span id="cb132-1216"><a href="#cb132-1216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1217"><a href="#cb132-1217" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-1218"><a href="#cb132-1218" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-1219"><a href="#cb132-1219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1220"><a href="#cb132-1220" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1221"><a href="#cb132-1221" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb132-1222"><a href="#cb132-1222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1225"><a href="#cb132-1225" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1226"><a href="#cb132-1226" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1227"><a href="#cb132-1227" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1228"><a href="#cb132-1228" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_tail =</span> <span class="fu">mean</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1229"><a href="#cb132-1229" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_tail =</span> <span class="fu">min</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1230"><a href="#cb132-1230" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_tail =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb132-1231"><a href="#cb132-1231" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1232"><a href="#cb132-1232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1233"><a href="#cb132-1233" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1234"><a href="#cb132-1234" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1235"><a href="#cb132-1235" aria-hidden="true" tabindex="-1"></a>Calculates the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1236"><a href="#cb132-1236" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1237"><a href="#cb132-1237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1238"><a href="#cb132-1238" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1239"><a href="#cb132-1239" aria-hidden="true" tabindex="-1"></a>Produces a data frame containing one column for each of the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1240"><a href="#cb132-1240" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1241"><a href="#cb132-1241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1242"><a href="#cb132-1242" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1243"><a href="#cb132-1243" aria-hidden="true" tabindex="-1"></a>Finds the average tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1244"><a href="#cb132-1244" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1245"><a href="#cb132-1245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1246"><a href="#cb132-1246" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1247"><a href="#cb132-1247" aria-hidden="true" tabindex="-1"></a>Produces a vector containing the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1248"><a href="#cb132-1248" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1249"><a href="#cb132-1249" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1250"><a href="#cb132-1250" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1251"><a href="#cb132-1251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1252"><a href="#cb132-1252" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1253"><a href="#cb132-1253" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> What is NOT a valid way to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb132-1254"><a href="#cb132-1254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1255"><a href="#cb132-1255" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1256"><a href="#cb132-1256" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1257"><a href="#cb132-1257" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb132-1258"><a href="#cb132-1258" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1259"><a href="#cb132-1259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1260"><a href="#cb132-1260" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1261"><a href="#cb132-1261" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% pull(weight_lost_g) %&gt;% mean(na.rm = TRUE)`</span></span>
<span id="cb132-1262"><a href="#cb132-1262" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1263"><a href="#cb132-1263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1264"><a href="#cb132-1264" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1265"><a href="#cb132-1265" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarize(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb132-1266"><a href="#cb132-1266" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1267"><a href="#cb132-1267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1268"><a href="#cb132-1268" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1269"><a href="#cb132-1269" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)`</span></span>
<span id="cb132-1270"><a href="#cb132-1270" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1271"><a href="#cb132-1271" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1272"><a href="#cb132-1272" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1273"><a href="#cb132-1273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1274"><a href="#cb132-1274" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-1275"><a href="#cb132-1275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1276"><a href="#cb132-1276" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-1277"><a href="#cb132-1277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1278"><a href="#cb132-1278" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The code **produces a data frame** containing one column for each of the average, minimum, and maximum tail length of the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1279"><a href="#cb132-1279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1280"><a href="#cb132-1280" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> The line of code that is NOT a valid way to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% mean(weight_lost_g, na.rm = TRUE)`</span>. This line of code is incorrect because the <span class="in">`mean()`</span> function is being used directly on the data frame, rather than within a <span class="in">`summarise()`</span> function. The other options are valid ways to calculate the mean weight lost by the mice in the <span class="in">`mousezempic_dosage_data`</span> data frame (although note that the second option uses <span class="in">`pull()`</span> to extract the <span class="in">`weight_lost_g`</span> column as a vector before calculating the mean, so the mean value is stored in a vector rather than in a data frame).</span>
<span id="cb132-1281"><a href="#cb132-1281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1282"><a href="#cb132-1282" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-1283"><a href="#cb132-1283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1284"><a href="#cb132-1284" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1285"><a href="#cb132-1285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1286"><a href="#cb132-1286" aria-hidden="true" tabindex="-1"></a><span class="fu">## Grouping {#sec-grouping}</span></span>
<span id="cb132-1287"><a href="#cb132-1287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1288"><a href="#cb132-1288" aria-hidden="true" tabindex="-1"></a><span class="al">![Using summarise with the .by option allows you to create separate summaries for different 'groups' of data](images/S02F09_grouped_summarise.png)</span></span>
<span id="cb132-1289"><a href="#cb132-1289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1290"><a href="#cb132-1290" aria-hidden="true" tabindex="-1"></a>Grouping is a powerful concept in in <span class="in">`dplyr`</span> that allows you to perform operations on subsets of your data. For example, you might want to calculate the mean weight lost by mice in each cage, or find the mouse with the longest tail in each strain.</span>
<span id="cb132-1291"><a href="#cb132-1291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1292"><a href="#cb132-1292" aria-hidden="true" tabindex="-1"></a>We can group data using the <span class="in">`.by`</span> argument that exists in many dplyr functions, like <span class="in">`summarise()`</span> and <span class="in">`mutate()`</span>, and passing it the name(s) of column(s) to group by. For example, to calculate the mean weight lost by mice in each cage:</span>
<span id="cb132-1293"><a href="#cb132-1293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1296"><a href="#cb132-1296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1297"><a href="#cb132-1297" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1298"><a href="#cb132-1298" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1299"><a href="#cb132-1299" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1300"><a href="#cb132-1300" aria-hidden="true" tabindex="-1"></a>    <span class="co"># don't forget it's .by, not by!</span></span>
<span id="cb132-1301"><a href="#cb132-1301" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span>
<span id="cb132-1302"><a href="#cb132-1302" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1303"><a href="#cb132-1303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1304"><a href="#cb132-1304" aria-hidden="true" tabindex="-1"></a>Like when we first learned the summarise function above, we give our new column a name (<span class="in">`mean_weight_lost`</span>), and then we assign its value to be the mean of the <span class="in">`weight_lost_g`</span> column (with <span class="in">`NA`</span>s removed). But this time, we also added the <span class="in">`.by`</span> argument to specify the column we want to group by (<span class="in">`cage_number`</span>, in this case). This will return a data frame with the mean weight lost by mice in each cage.</span>
<span id="cb132-1305"><a href="#cb132-1305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1306"><a href="#cb132-1306" aria-hidden="true" tabindex="-1"></a>Grouping is a powerful tool for exploring your data and can help you identify patterns that might not be obvious when looking at the data as a whole. For example, notice how this grouped summary reveals that mice in cage 3E lost more weight than those in the other two cages.</span>
<span id="cb132-1307"><a href="#cb132-1307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1308"><a href="#cb132-1308" aria-hidden="true" tabindex="-1"></a>It's also possible to group by multiple columns by passing a vector of column names to the <span class="in">`.by`</span> argument. For example, to calculate the mean weight lost by mice in each cage and strain:</span>
<span id="cb132-1309"><a href="#cb132-1309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1312"><a href="#cb132-1312" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1313"><a href="#cb132-1313" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1314"><a href="#cb132-1314" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1315"><a href="#cb132-1315" aria-hidden="true" tabindex="-1"></a>  <span class="co"># group by both cage_number and mouse_strain</span></span>
<span id="cb132-1316"><a href="#cb132-1316" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span>
<span id="cb132-1317"><a href="#cb132-1317" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1318"><a href="#cb132-1318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1319"><a href="#cb132-1319" aria-hidden="true" tabindex="-1"></a>Of course, <span class="in">`mean()`</span> is not the only function that we can use within <span class="in">`summarise()`</span>. We can use any function that takes a vector of values and returns a single value, like <span class="in">`median()`</span>, <span class="in">`sd()`</span>, or <span class="in">`max()`</span>. We can also use multiple functions at once, by giving each column a name and specifying the function we want to use:</span>
<span id="cb132-1320"><a href="#cb132-1320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1323"><a href="#cb132-1323" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1324"><a href="#cb132-1324" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1325"><a href="#cb132-1325" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1326"><a href="#cb132-1326" aria-hidden="true" tabindex="-1"></a>    <span class="at">n =</span> <span class="fu">n</span>(),</span>
<span id="cb132-1327"><a href="#cb132-1327" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1328"><a href="#cb132-1328" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_weight_lost =</span> <span class="fu">median</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1329"><a href="#cb132-1329" aria-hidden="true" tabindex="-1"></a>    <span class="at">sd_weight_lost =</span> <span class="fu">sd</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1330"><a href="#cb132-1330" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_weight_lost =</span> <span class="fu">max</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1331"><a href="#cb132-1331" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_weight_lost =</span> <span class="fu">min</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1332"><a href="#cb132-1332" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> cage_number)</span>
<span id="cb132-1333"><a href="#cb132-1333" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1334"><a href="#cb132-1334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1335"><a href="#cb132-1335" aria-hidden="true" tabindex="-1"></a>Here, we also used the <span class="in">`n()`</span> function to calculate the number of mice in each cage. This is a special helper function that works within <span class="in">`summarise`</span> to count the number of rows in each group.</span>
<span id="cb132-1336"><a href="#cb132-1336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1337"><a href="#cb132-1337" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="To <span class="in">`.by`</span> or not to <span class="in">`.by`</span>?"}</span>
<span id="cb132-1338"><a href="#cb132-1338" aria-hidden="true" tabindex="-1"></a>In the <span class="in">`dplyr`</span> package, there are two ways to group data: using the <span class="in">`.by`</span> argument within various functions (as we have covered so far), or using the <span class="in">`group_by()`</span> function, then performing your operations and ungrouping with <span class="in">`ungroup()`</span>.</span>
<span id="cb132-1339"><a href="#cb132-1339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1340"><a href="#cb132-1340" aria-hidden="true" tabindex="-1"></a>For example, we've seen above how to calculate the mean weight lost by mice in each cage using the <span class="in">`.by`</span> argument:</span>
<span id="cb132-1341"><a href="#cb132-1341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1344"><a href="#cb132-1344" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1345"><a href="#cb132-1345" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1346"><a href="#cb132-1346" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> cage_number)</span>
<span id="cb132-1347"><a href="#cb132-1347" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1348"><a href="#cb132-1348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1349"><a href="#cb132-1349" aria-hidden="true" tabindex="-1"></a>But we can also do the same using <span class="in">`group_by()`</span> and <span class="in">`ungroup()`</span>:</span>
<span id="cb132-1352"><a href="#cb132-1352" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1353"><a href="#cb132-1353" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1354"><a href="#cb132-1354" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cage_number) <span class="sc">%&gt;%</span></span>
<span id="cb132-1355"><a href="#cb132-1355" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_weight_lost =</span> <span class="fu">mean</span>(weight_lost_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1356"><a href="#cb132-1356" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb132-1357"><a href="#cb132-1357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1358"><a href="#cb132-1358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1359"><a href="#cb132-1359" aria-hidden="true" tabindex="-1"></a>The two methods are equivalent, but using the <span class="in">`.by`</span> argument within functions can be more concise and easier to read. Still, it's good to be aware of <span class="in">`group_by()`</span> and <span class="in">`ungroup()`</span> as they are widely used, particularly in older code.</span>
<span id="cb132-1360"><a href="#cb132-1360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1361"><a href="#cb132-1361" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1362"><a href="#cb132-1362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1363"><a href="#cb132-1363" aria-hidden="true" tabindex="-1"></a>Although grouping is most often used with <span class="in">`summarise()`</span>, it can be used with <span class="in">`dplyr`</span> functions too. For example mutate() function can also be used with grouping to add new columns to the data frame based on group-specific calculations. Let's say we wanted to calculate the Z-score (also known as the <span class="co">[</span><span class="ot">standard score</span><span class="co">](https://en.wikipedia.org/wiki/Standard_score)</span>) to standardise the weight lost by each mouse within each strain.</span>
<span id="cb132-1364"><a href="#cb132-1364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1365"><a href="#cb132-1365" aria-hidden="true" tabindex="-1"></a>As a reminder, the formula for calculating the Z-score is $\frac{x - \mu}{\sigma}$, where $x$ is the value (in our case the <span class="in">`weight_lost_g`</span> column), $\mu$ is the mean, and $\sigma$ is the standard deviation.</span>
<span id="cb132-1366"><a href="#cb132-1366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1367"><a href="#cb132-1367" aria-hidden="true" tabindex="-1"></a>We can calculate this for each mouse in each strain using the following code:</span>
<span id="cb132-1368"><a href="#cb132-1368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1371"><a href="#cb132-1371" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1372"><a href="#cb132-1372" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1373"><a href="#cb132-1373" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove NAs before calculating the mean and SD</span></span>
<span id="cb132-1374"><a href="#cb132-1374" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(weight_lost_g)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1375"><a href="#cb132-1375" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight_lost_z =</span> (weight_lost_g <span class="sc">-</span> <span class="fu">mean</span>(weight_lost_g)) <span class="sc">/</span> <span class="fu">sd</span>(weight_lost_g), <span class="at">.by =</span> mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb132-1376"><a href="#cb132-1376" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns</span></span>
<span id="cb132-1377"><a href="#cb132-1377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, weight_lost_g, weight_lost_z)</span>
<span id="cb132-1378"><a href="#cb132-1378" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1379"><a href="#cb132-1379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1380"><a href="#cb132-1380" aria-hidden="true" tabindex="-1"></a>Unlike when we used <span class="in">`.by`</span> with <span class="in">`summarise()`</span>, we still get the same number of rows as the original data frame, but now we have a new column <span class="in">`weight_lost_z`</span> that contains the Z-score for each mouse within each strain. This could be useful for identifying outliers or comparing the weight lost by each mouse to the average for its strain.</span>
<span id="cb132-1381"><a href="#cb132-1381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1382"><a href="#cb132-1382" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-1383"><a href="#cb132-1383" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-1384"><a href="#cb132-1384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1385"><a href="#cb132-1385" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1386"><a href="#cb132-1386" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> Which line of code would you use to calculate the median tail length of mice belonging to each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame?</span>
<span id="cb132-1387"><a href="#cb132-1387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1388"><a href="#cb132-1388" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1389"><a href="#cb132-1389" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1390"><a href="#cb132-1390" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm), .by = mouse_strain)`</span></span>
<span id="cb132-1391"><a href="#cb132-1391" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1392"><a href="#cb132-1392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1393"><a href="#cb132-1393" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1394"><a href="#cb132-1394" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span></span>
<span id="cb132-1395"><a href="#cb132-1395" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1396"><a href="#cb132-1396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1397"><a href="#cb132-1397" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1398"><a href="#cb132-1398" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), by = mouse_strain)`</span></span>
<span id="cb132-1399"><a href="#cb132-1399" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1400"><a href="#cb132-1400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1401"><a href="#cb132-1401" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1402"><a href="#cb132-1402" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span></span>
<span id="cb132-1403"><a href="#cb132-1403" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1404"><a href="#cb132-1404" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1405"><a href="#cb132-1405" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1406"><a href="#cb132-1406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1407"><a href="#cb132-1407" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1408"><a href="#cb132-1408" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> Explain in words what the following code does:</span>
<span id="cb132-1409"><a href="#cb132-1409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1412"><a href="#cb132-1412" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1413"><a href="#cb132-1413" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1414"><a href="#cb132-1414" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1415"><a href="#cb132-1415" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">max_tail_len =</span> <span class="fu">max</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, replicate))</span>
<span id="cb132-1416"><a href="#cb132-1416" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1417"><a href="#cb132-1417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1418"><a href="#cb132-1418" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1419"><a href="#cb132-1419" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1420"><a href="#cb132-1420" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb132-1421"><a href="#cb132-1421" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1422"><a href="#cb132-1422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1423"><a href="#cb132-1423" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1424"><a href="#cb132-1424" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice for each replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb132-1425"><a href="#cb132-1425" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1426"><a href="#cb132-1426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1427"><a href="#cb132-1427" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1428"><a href="#cb132-1428" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of all mice in the <span class="in">`mousezempic_dosage_data`</span> data frame</span>
<span id="cb132-1429"><a href="#cb132-1429" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1430"><a href="#cb132-1430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1431"><a href="#cb132-1431" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1432"><a href="#cb132-1432" aria-hidden="true" tabindex="-1"></a>Calculates the maximum tail length of mice in each unique combination of strain and replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1433"><a href="#cb132-1433" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1434"><a href="#cb132-1434" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1435"><a href="#cb132-1435" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1436"><a href="#cb132-1436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1437"><a href="#cb132-1437" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1438"><a href="#cb132-1438" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> I want to count how many male and how many female mice there are for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame. Which line of code would I use?</span>
<span id="cb132-1439"><a href="#cb132-1439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1440"><a href="#cb132-1440" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1441"><a href="#cb132-1441" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1442"><a href="#cb132-1442" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = sex)`</span></span>
<span id="cb132-1443"><a href="#cb132-1443" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1444"><a href="#cb132-1444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1445"><a href="#cb132-1445" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1446"><a href="#cb132-1446" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain)`</span></span>
<span id="cb132-1447"><a href="#cb132-1447" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1448"><a href="#cb132-1448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1449"><a href="#cb132-1449" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1450"><a href="#cb132-1450" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))`</span></span>
<span id="cb132-1451"><a href="#cb132-1451" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1452"><a href="#cb132-1452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1453"><a href="#cb132-1453" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1454"><a href="#cb132-1454" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = mouse_strain, sex)`</span></span>
<span id="cb132-1455"><a href="#cb132-1455" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1456"><a href="#cb132-1456" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1457"><a href="#cb132-1457" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1458"><a href="#cb132-1458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1459"><a href="#cb132-1459" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1460"><a href="#cb132-1460" aria-hidden="true" tabindex="-1"></a>4<span class="sc">\.</span> I want to find the proportion of weight lost **by each mouse** in each cage in the <span class="in">`mousezempic_dosage_data`</span> data frame. Which line of code would I use?</span>
<span id="cb132-1461"><a href="#cb132-1461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1462"><a href="#cb132-1462" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1463"><a href="#cb132-1463" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1464"><a href="#cb132-1464" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% summarise(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span></span>
<span id="cb132-1465"><a href="#cb132-1465" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1466"><a href="#cb132-1466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1467"><a href="#cb132-1467" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1468"><a href="#cb132-1468" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span></span>
<span id="cb132-1469"><a href="#cb132-1469" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1470"><a href="#cb132-1470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1471"><a href="#cb132-1471" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1472"><a href="#cb132-1472" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE, .by = cage_number))`</span></span>
<span id="cb132-1473"><a href="#cb132-1473" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1474"><a href="#cb132-1474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1475"><a href="#cb132-1475" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1476"><a href="#cb132-1476" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb132-1477"><a href="#cb132-1477" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1478"><a href="#cb132-1478" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1479"><a href="#cb132-1479" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1480"><a href="#cb132-1480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1481"><a href="#cb132-1481" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-1482"><a href="#cb132-1482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1483"><a href="#cb132-1483" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-1484"><a href="#cb132-1484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1485"><a href="#cb132-1485" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb132-1486"><a href="#cb132-1486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1487"><a href="#cb132-1487" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct line of code to calculate the median tail length of mice belonging to each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% summarise(median_tail_length = median(tail_length_mm, na.rm = TRUE), .by = mouse_strain)`</span>. Remember to use <span class="in">`na.rm = TRUE`</span> to remove any missing values before calculating the median, and to use <span class="in">`.by`</span> to specify the column to group by (not <span class="in">`by`</span>). Seeing as we want to calculate the median (collapse down to a single value per group), we need to use <span class="in">`summarise()`</span> rather than <span class="in">`mutate()`</span>.</span>
<span id="cb132-1488"><a href="#cb132-1488" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The code <span class="in">`mousezempic_dosage_data %&gt;% summarise(max_tail_len = max(tail_length_mm, na.rm = TRUE), .by = c(mouse_strain, replicate))`</span> calculates the maximum tail length of mice in each **unique combination** of strain and replicate in the <span class="in">`mousezempic_dosage_data`</span> data frame.</span>
<span id="cb132-1489"><a href="#cb132-1489" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct line of code to count how many male and how many female mice there are for each strain in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% summarise(count = n(), .by = c(mouse_strain, sex))`</span>. We need to group by both <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> to get the count for each unique combination of strain and sex. Don't forget that we specify the column names as a vector when grouping by multiple columns.</span>
<span id="cb132-1490"><a href="#cb132-1490" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The correct line of code to find the proportion of weight lost **by each mouse** in each cage in the <span class="in">`mousezempic_dosage_data`</span> data frame is <span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_proportion = weight_lost_g / sum(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span>. We use <span class="in">`mutate()`</span> because we want a value for each mouse (each row in our data), rather than to collapse down to a single value for each group (cage number in this case). Be careful that you use the <span class="in">`.by`</span> argument within the <span class="in">`mutate()`</span> function call, not within the <span class="in">`sum()`</span> function by mistake (this is what is wrong with the third option).</span>
<span id="cb132-1491"><a href="#cb132-1491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1492"><a href="#cb132-1492" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb132-1493"><a href="#cb132-1493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1494"><a href="#cb132-1494" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-1495"><a href="#cb132-1495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1496"><a href="#cb132-1496" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1497"><a href="#cb132-1497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1498"><a href="#cb132-1498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1499"><a href="#cb132-1499" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reshaping and combining data {#sec-reshaping}</span></span>
<span id="cb132-1500"><a href="#cb132-1500" aria-hidden="true" tabindex="-1"></a>In the previous section, we learned how to filter, select, mutate, and summarise data. These are the most common operations you'll perform on your data, and all generally revolve around working with specific columns or rows of your data frame. But what if you need to change the structure of your entire data frame? Re-shaping (with <span class="in">`pivot_`</span> functions) and combining (with <span class="in">`join`</span> functions) data are two more advanced operations that can help you do this.</span>
<span id="cb132-1501"><a href="#cb132-1501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1502"><a href="#cb132-1502" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reshaping data with pivot functions {#sec-pivot}</span></span>
<span id="cb132-1503"><a href="#cb132-1503" aria-hidden="true" tabindex="-1"></a>Pivoting is a way to change the structure of your data frame by rotating it so that rows become columns and vice versa. This can be useful when you want to change the way your data is organised, for example, to make it easier to plot or analyse.</span>
<span id="cb132-1504"><a href="#cb132-1504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1505"><a href="#cb132-1505" aria-hidden="true" tabindex="-1"></a><span class="al">![Pivot functions allow you to change the structure of your data frame](images/S02F10_pivot.png)</span>{width=50%}</span>
<span id="cb132-1506"><a href="#cb132-1506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1507"><a href="#cb132-1507" aria-hidden="true" tabindex="-1"></a>The <span class="in">`pivot_longer()`</span> function is used to pivot data from wide to long format, and the <span class="in">`pivot_wider()`</span> function is used to pivot data from long to wide format.</span>
<span id="cb132-1508"><a href="#cb132-1508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1509"><a href="#cb132-1509" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot wider</span></span>
<span id="cb132-1510"><a href="#cb132-1510" aria-hidden="true" tabindex="-1"></a>A common use case for <span class="in">`pivot_wider()`</span> is to make a <span class="co">[</span><span class="ot">contingency table</span><span class="co">](https://en.wikipedia.org/wiki/Contingency_table)</span>,</span>
<span id="cb132-1511"><a href="#cb132-1511" aria-hidden="true" tabindex="-1"></a>which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</span>
<span id="cb132-1512"><a href="#cb132-1512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1513"><a href="#cb132-1513" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to create a table that shows how many mice there are of each strain, in each cage number. We can achieve this in a long format using <span class="in">`summarise()`</span> as we learned in the previous section:</span>
<span id="cb132-1514"><a href="#cb132-1514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1517"><a href="#cb132-1517" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1518"><a href="#cb132-1518" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1519"><a href="#cb132-1519" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1520"><a href="#cb132-1520" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb132-1521"><a href="#cb132-1521" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span>
<span id="cb132-1522"><a href="#cb132-1522" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1523"><a href="#cb132-1523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1524"><a href="#cb132-1524" aria-hidden="true" tabindex="-1"></a>Notice how we have a row for each unique combination of <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span>. Instead, we can use <span class="in">`pivot_wider()`</span> to create a table where each row represents a unique <span class="in">`cage_number`</span> and each column represents a unique <span class="in">`mouse_strain`</span>.</span>
<span id="cb132-1525"><a href="#cb132-1525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1526"><a href="#cb132-1526" aria-hidden="true" tabindex="-1"></a>To use a <span class="in">`pivot_wider()`</span> you need to tell R two bits of information:</span>
<span id="cb132-1527"><a href="#cb132-1527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1528"><a href="#cb132-1528" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>What the new columns should be (with <span class="in">`names_from`</span>)</span>
<span id="cb132-1529"><a href="#cb132-1529" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Where the values in these columns should come from (with <span class="in">`values_from`</span>).</span>
<span id="cb132-1530"><a href="#cb132-1530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1531"><a href="#cb132-1531" aria-hidden="true" tabindex="-1"></a>In this case, we want the mouse strains to be our new columns (so <span class="in">`names_from = mouse_strain`</span>), and the average weight lost to be the values in the table (<span class="in">`values_from = n_mice`</span>). Putting it all together, here is the code to make our summary table:</span>
<span id="cb132-1532"><a href="#cb132-1532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1535"><a href="#cb132-1535" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1536"><a href="#cb132-1536" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1537"><a href="#cb132-1537" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1538"><a href="#cb132-1538" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb132-1539"><a href="#cb132-1539" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1540"><a href="#cb132-1540" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb132-1541"><a href="#cb132-1541" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1542"><a href="#cb132-1542" aria-hidden="true" tabindex="-1"></a>Notice how <span class="in">`NA`</span> values are used where there is no data for a particular combination of <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span>.</span>
<span id="cb132-1543"><a href="#cb132-1543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1544"><a href="#cb132-1544" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot longer</span></span>
<span id="cb132-1545"><a href="#cb132-1545" aria-hidden="true" tabindex="-1"></a>While wide data can be easier to read, long data is often easier to work with. That's because a lot of the tidyverse functions we've learned so far (like <span class="in">`filter()`</span>, <span class="in">`mutate()`</span>, and <span class="in">`summarise()`</span>) work in a row-wise fashion, which is more natural with long data.</span>
<span id="cb132-1546"><a href="#cb132-1546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1547"><a href="#cb132-1547" aria-hidden="true" tabindex="-1"></a>To demonstrate <span class="in">`pivot_longer()`</span>, we will introduce a new data frame called <span class="in">`mousezempic_expression_data`</span>. This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate. You can download it here:</span>
<span id="cb132-1548"><a href="#cb132-1548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1549"><a href="#cb132-1549" aria-hidden="true" tabindex="-1"></a>{{&lt; downloadthis data/mousezempic_expression_data.tsv dname=mousezempic_expression_data label="Download mousezempic_expression_data.tsv" icon=file-earmark-arrow-down type=primary class=data-button id=expr_data &gt;}}</span>
<span id="cb132-1550"><a href="#cb132-1550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1551"><a href="#cb132-1551" aria-hidden="true" tabindex="-1"></a>Or run the following code in your R console to download it:</span>
<span id="cb132-1554"><a href="#cb132-1554" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1555"><a href="#cb132-1555" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1556"><a href="#cb132-1556" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://raw.githubusercontent.com/kzeglinski/new_wehi_r_course/refs/heads/main/data/mousezempic_expression_data.tsv"</span>, <span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb132-1557"><a href="#cb132-1557" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1558"><a href="#cb132-1558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1559"><a href="#cb132-1559" aria-hidden="true" tabindex="-1"></a>Once you've done that, read it into R (notice that it's a <span class="in">`.tsv`</span> file, so we need to use either <span class="in">`read_tsv()`</span> or <span class="in">`read_delim()`</span>, not <span class="in">`read_csv()`</span>)</span>
<span id="cb132-1560"><a href="#cb132-1560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1563"><a href="#cb132-1563" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1564"><a href="#cb132-1564" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data</span></span>
<span id="cb132-1565"><a href="#cb132-1565" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb132-1566"><a href="#cb132-1566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1567"><a href="#cb132-1567" aria-hidden="true" tabindex="-1"></a><span class="co"># print it out</span></span>
<span id="cb132-1568"><a href="#cb132-1568" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data</span>
<span id="cb132-1569"><a href="#cb132-1569" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1570"><a href="#cb132-1570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1571"><a href="#cb132-1571" aria-hidden="true" tabindex="-1"></a>As you can see, the data is in wide format, with each row representing a different mouse (identified by its <span class="in">`id_num`</span>) and each column representing a different measurement of a gene. To put this data into a long format (where each measurement is contained on a separate row), we can use <span class="in">`pivot_longer()`</span>, specifying three arguments:</span>
<span id="cb132-1572"><a href="#cb132-1572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1573"><a href="#cb132-1573" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="in">`cols`</span>: the columns to pivot from. You can use selection helpers like <span class="in">`contains()`</span> or <span class="in">`starts_with()`</span> to easily select multiple columns at once.</span>
<span id="cb132-1574"><a href="#cb132-1574" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="in">`names_to`</span>: the name of a new column that will contain the old column names.</span>
<span id="cb132-1575"><a href="#cb132-1575" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="in">`values_to`</span>: the name of a new column that will contain the values from the old columns.</span>
<span id="cb132-1576"><a href="#cb132-1576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1577"><a href="#cb132-1577" aria-hidden="true" tabindex="-1"></a>In this particular case here's what the code would look like:</span>
<span id="cb132-1578"><a href="#cb132-1578" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1581"><a href="#cb132-1581" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1582"><a href="#cb132-1582" aria-hidden="true" tabindex="-1"></a>mousezempic_expression_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1583"><a href="#cb132-1583" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>), <span class="at">names_to =</span> <span class="st">"measurement"</span>, <span class="at">values_to =</span> <span class="st">"expression_level"</span>)</span>
<span id="cb132-1584"><a href="#cb132-1584" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1585"><a href="#cb132-1585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1586"><a href="#cb132-1586" aria-hidden="true" tabindex="-1"></a>Let's go through those three arguments:</span>
<span id="cb132-1587"><a href="#cb132-1587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1588"><a href="#cb132-1588" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="in">`cols = contains("_rep")`</span>: This tells R to pivot all columns that contain the string "_rep". In this case, that's the columns <span class="in">`TH_rep1`</span>, <span class="in">`TH_rep2`</span>, <span class="in">`TH_rep3`</span>, <span class="in">`PRLH_rep1`</span>, <span class="in">`PRLH_rep2`</span>, <span class="in">`PRLH_rep3`</span>, <span class="in">`HPRT1_rep1`</span>, <span class="in">`HPRT1_rep2`</span>, and <span class="in">`HPRT1_rep3`</span>.</span>
<span id="cb132-1589"><a href="#cb132-1589" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="in">`names_to = "measurement"`</span>: This tells R to create a new column called 'measurement' that contains the names of the old columns. So, for example, the value of the 'measurement' column for the first row will be 'TH_rep1'.</span>
<span id="cb132-1590"><a href="#cb132-1590" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="in">`values_to = "expression_level"`</span>: This tells R to create a new column called 'expression_level' that contains the values from the old columns. So, for example, the value of the 'expression_level' column for the first row will be the value of the <span class="in">`TH_rep1`</span> column.</span>
<span id="cb132-1591"><a href="#cb132-1591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1592"><a href="#cb132-1592" aria-hidden="true" tabindex="-1"></a>After running this code, you'll see that the data frame is now in long format, with each row representing a different measurement of a gene for a different mouse.</span>
<span id="cb132-1593"><a href="#cb132-1593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1594"><a href="#cb132-1594" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Perplexed by pivoting?"}</span>
<span id="cb132-1595"><a href="#cb132-1595" aria-hidden="true" tabindex="-1"></a>Pivoting can be a bit tricky to get your head around! Often when you're doing analysis, you'll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <span class="co">[</span><span class="ot">like those in the R for Data Science book</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html#sec-pivoting)</span>, or to just experiment with different arguments until you get the result you want.</span>
<span id="cb132-1596"><a href="#cb132-1596" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1597"><a href="#cb132-1597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1598"><a href="#cb132-1598" aria-hidden="true" tabindex="-1"></a><span class="fu">### Combining data with join functions {#sec-join}</span></span>
<span id="cb132-1599"><a href="#cb132-1599" aria-hidden="true" tabindex="-1"></a>Another way you might need to change the structure of your data is by combining it with other data frames. This is where the <span class="in">`join`</span> functions come in. There are many different types of joins in <span class="in">`dplyr`</span> (which you can read about here if you're interested), but the most useful one is the <span class="in">`left_join()`</span> function.</span>
<span id="cb132-1600"><a href="#cb132-1600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1601"><a href="#cb132-1601" aria-hidden="true" tabindex="-1"></a><span class="al">![The left_join() function allows you to combine two data frames](images/S02F11_left_join.png)</span></span>
<span id="cb132-1602"><a href="#cb132-1602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1603"><a href="#cb132-1603" aria-hidden="true" tabindex="-1"></a><span class="in">`left_join()`</span> is used to combine two data frames based on a common column. It takes three arguments - the first data frame, the second data frame, and the common column to join on. The reason it's called a 'left join' is because it keeps all the rows from the first data frame (the 'left' data frame), and only adds rows from the second data frame (the 'right' data frame) if they match the common column.</span>
<span id="cb132-1604"><a href="#cb132-1604" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1605"><a href="#cb132-1605" aria-hidden="true" tabindex="-1"></a>For example, we can join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column, which is common to both data frames:</span>
<span id="cb132-1606"><a href="#cb132-1606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1609"><a href="#cb132-1609" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1610"><a href="#cb132-1610" aria-hidden="true" tabindex="-1"></a><span class="co"># example of using it in a pipe</span></span>
<span id="cb132-1611"><a href="#cb132-1611" aria-hidden="true" tabindex="-1"></a><span class="co"># here mousezempic_dosage_data is the left data frame (first) data frame passed on by the pipe</span></span>
<span id="cb132-1612"><a href="#cb132-1612" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1613"><a href="#cb132-1613" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="st">"id_num"</span>)</span>
<span id="cb132-1614"><a href="#cb132-1614" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1615"><a href="#cb132-1615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1616"><a href="#cb132-1616" aria-hidden="true" tabindex="-1"></a>This code will return a new data frame that contains all the columns from <span class="in">`mousezempic_dosage_data`</span> and all the columns from <span class="in">`mousezempic_expression_data`</span>, with the rows matched based on the <span class="in">`id_num`</span> column. If there is no match for a particular <span class="in">`id_num`</span> in the <span class="in">`mousezempic_expression_data`</span> data frame, the columns from that data frame will be filled with <span class="in">`NA`</span>. Rows that only appear in the <span class="in">`mousezempic_expression_data`</span> data frame will not be included in the final data frame.</span>
<span id="cb132-1617"><a href="#cb132-1617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1618"><a href="#cb132-1618" aria-hidden="true" tabindex="-1"></a>We can also use <span class="in">`left_join()`</span> when our columns have different names (but contain the same value) using a named vector the format <span class="in">`c("name1" = "name2")`</span>, where 'name1' is the name of the column in the left/first data frame (the one being piped in) and 'name2' is the name of the corresponding column in the right/second data frame.</span>
<span id="cb132-1619"><a href="#cb132-1619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1620"><a href="#cb132-1620" aria-hidden="true" tabindex="-1"></a>For example, if the <span class="in">`id_num`</span> column in the <span class="in">`mousezempic_dosage_data`</span> data frame was called <span class="in">`mouse_id`</span> instead, we could still join the two data frames using the following code:</span>
<span id="cb132-1621"><a href="#cb132-1621" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1624"><a href="#cb132-1624" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1625"><a href="#cb132-1625" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1626"><a href="#cb132-1626" aria-hidden="true" tabindex="-1"></a>  <span class="co"># for the sake of this example, make a new column called mouse_id</span></span>
<span id="cb132-1627"><a href="#cb132-1627" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mouse_id =</span> id_num) <span class="sc">%&gt;%</span></span>
<span id="cb132-1628"><a href="#cb132-1628" aria-hidden="true" tabindex="-1"></a>  <span class="co"># join the two data frames where mouse_id in the first data frame matches id_num in the second data frame</span></span>
<span id="cb132-1629"><a href="#cb132-1629" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(mousezempic_expression_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"mouse_id"</span> <span class="ot">=</span> <span class="st">"id_num"</span>))</span>
<span id="cb132-1630"><a href="#cb132-1630" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1631"><a href="#cb132-1631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1632"><a href="#cb132-1632" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Practice exercises"}</span>
<span id="cb132-1633"><a href="#cb132-1633" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb132-1634"><a href="#cb132-1634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1635"><a href="#cb132-1635" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1636"><a href="#cb132-1636" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> What does the following code do?</span>
<span id="cb132-1637"><a href="#cb132-1637" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1640"><a href="#cb132-1640" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1641"><a href="#cb132-1641" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1642"><a href="#cb132-1642" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1643"><a href="#cb132-1643" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1644"><a href="#cb132-1644" aria-hidden="true" tabindex="-1"></a>    <span class="at">med_tail =</span> <span class="fu">median</span>(tail_length_mm, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-1645"><a href="#cb132-1645" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(mouse_strain, sex)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1646"><a href="#cb132-1646" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> sex, <span class="at">values_from =</span> med_tail)</span>
<span id="cb132-1647"><a href="#cb132-1647" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1648"><a href="#cb132-1648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1649"><a href="#cb132-1649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1650"><a href="#cb132-1650" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1651"><a href="#cb132-1651" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1652"><a href="#cb132-1652" aria-hidden="true" tabindex="-1"></a>Pivots data into a wide format where there is a column for each sex.</span>
<span id="cb132-1653"><a href="#cb132-1653" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1654"><a href="#cb132-1654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1655"><a href="#cb132-1655" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1656"><a href="#cb132-1656" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots into a wide format where there is a column for each sex.</span>
<span id="cb132-1657"><a href="#cb132-1657" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1658"><a href="#cb132-1658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1659"><a href="#cb132-1659" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1660"><a href="#cb132-1660" aria-hidden="true" tabindex="-1"></a>Calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots into a wide format where there is a column for each mouse strain.</span>
<span id="cb132-1661"><a href="#cb132-1661" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1662"><a href="#cb132-1662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1663"><a href="#cb132-1663" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1664"><a href="#cb132-1664" aria-hidden="true" tabindex="-1"></a>It just gives an error</span>
<span id="cb132-1665"><a href="#cb132-1665" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1666"><a href="#cb132-1666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1667"><a href="#cb132-1667" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1668"><a href="#cb132-1668" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1669"><a href="#cb132-1669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1670"><a href="#cb132-1670" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> I have run the following code to create a new column in the <span class="in">`mousezempic_dosage_data`</span> data frame that gives the weight of the mice at the end of the experiment.</span>
<span id="cb132-1671"><a href="#cb132-1671" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1674"><a href="#cb132-1674" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1675"><a href="#cb132-1675" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1676"><a href="#cb132-1676" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a column for the weight at the end of the experiment</span></span>
<span id="cb132-1677"><a href="#cb132-1677" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">final_weight_g =</span> initial_weight_g <span class="sc">-</span> weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb132-1678"><a href="#cb132-1678" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select the relevant columns only</span></span>
<span id="cb132-1679"><a href="#cb132-1679" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(id_num, initial_weight_g, final_weight_g)</span>
<span id="cb132-1680"><a href="#cb132-1680" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1681"><a href="#cb132-1681" aria-hidden="true" tabindex="-1"></a>Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)?</span>
<span id="cb132-1682"><a href="#cb132-1682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1683"><a href="#cb132-1683" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1684"><a href="#cb132-1684" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1685"><a href="#cb132-1685" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")`</span></span>
<span id="cb132-1686"><a href="#cb132-1686" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1687"><a href="#cb132-1687" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1688"><a href="#cb132-1688" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1689"><a href="#cb132-1689" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")`</span></span>
<span id="cb132-1690"><a href="#cb132-1690" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1691"><a href="#cb132-1691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1692"><a href="#cb132-1692" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1693"><a href="#cb132-1693" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)`</span></span>
<span id="cb132-1694"><a href="#cb132-1694" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1695"><a href="#cb132-1695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1696"><a href="#cb132-1696" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1697"><a href="#cb132-1697" aria-hidden="true" tabindex="-1"></a><span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span></span>
<span id="cb132-1698"><a href="#cb132-1698" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1699"><a href="#cb132-1699" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1700"><a href="#cb132-1700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1701"><a href="#cb132-1701" aria-hidden="true" tabindex="-1"></a>::: question</span>
<span id="cb132-1702"><a href="#cb132-1702" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> Which of the following is NOT a valid way to join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column?</span>
<span id="cb132-1703"><a href="#cb132-1703" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1704"><a href="#cb132-1704" aria-hidden="true" tabindex="-1"></a>::: choices</span>
<span id="cb132-1705"><a href="#cb132-1705" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1706"><a href="#cb132-1706" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = "id_num")`</span></span>
<span id="cb132-1707"><a href="#cb132-1707" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1708"><a href="#cb132-1708" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1709"><a href="#cb132-1709" aria-hidden="true" tabindex="-1"></a><span class="in">`left_join(mousezempic_dosage_data, mousezempic_expression_data, by = "id_num")`</span></span>
<span id="cb132-1710"><a href="#cb132-1710" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1711"><a href="#cb132-1711" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb132-1712"><a href="#cb132-1712" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")`</span></span>
<span id="cb132-1713"><a href="#cb132-1713" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1714"><a href="#cb132-1714" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb132-1715"><a href="#cb132-1715" aria-hidden="true" tabindex="-1"></a><span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, by = ("id_num" = "id_num"))`</span></span>
<span id="cb132-1716"><a href="#cb132-1716" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1717"><a href="#cb132-1717" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1718"><a href="#cb132-1718" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1719"><a href="#cb132-1719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1720"><a href="#cb132-1720" aria-hidden="true" tabindex="-1"></a>&lt;details&gt; &lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-1721"><a href="#cb132-1721" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span> The code first calculates the median tail length for each unique combination of <span class="in">`mouse_strain`</span> and <span class="in">`sex`</span> in the <span class="in">`mousezempic_dosage_data`</span> data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument <span class="in">`names_from = sex`</span> )</span>
<span id="cb132-1722"><a href="#cb132-1722" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1723"><a href="#cb132-1723" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span> The correct pivot function call to take the data from a wide format to a long format is <span class="in">`pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`</span>. This code tells R to pivot the <span class="in">`initial_weight_g`</span> and <span class="in">`final_weight_g`</span> columns into a long format, where there is a row for each mouse and each weight measurement. The <span class="in">`names_to`</span> argument specifies to make a column called 'timepoint' that tells us whether the measurement is initial or final, and the <span class="in">`values_to`</span> argument specifies the name of the new column that will contain these measurements.</span>
<span id="cb132-1724"><a href="#cb132-1724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1725"><a href="#cb132-1725" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span> The line of code that is NOT a valid way to join the <span class="in">`mousezempic_dosage_data`</span> data frame with the <span class="in">`mousezempic_expression_data`</span> data frame based on the <span class="in">`id_num`</span> column is <span class="in">`mousezempic_dosage_data %&gt;% left_join(mousezempic_expression_data, .by = "id_num")`</span>. This line of code is incorrect because the <span class="in">`.by`</span> argument is not used in the <span class="in">`left_join()`</span> function (this can be confusing! it's <span class="in">`.by`</span> when grouping by <span class="in">`by`</span> when joining). The other options are valid ways to join the two data frames based on the <span class="in">`id_num`</span> column: remember that we don't have to use pipes to join data frames, we can use the <span class="in">`left_join()`</span> function directly, and we can use a named vector to specify the columns to join on (although here it's a bit redundant as the columns have the same name).</span>
<span id="cb132-1726"><a href="#cb132-1726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1727"><a href="#cb132-1727" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb132-1728"><a href="#cb132-1728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1729"><a href="#cb132-1729" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1730"><a href="#cb132-1730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1731"><a href="#cb132-1731" aria-hidden="true" tabindex="-1"></a><span class="fu">## Saving data to a file {#sec-saving}</span></span>
<span id="cb132-1732"><a href="#cb132-1732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1733"><a href="#cb132-1733" aria-hidden="true" tabindex="-1"></a>Once you've cleaned and transformed your data, you'll often want to save it to a file so that you can use it in other programs or share it with others. The <span class="in">`write_csv()`</span> and <span class="in">`write_tsv()`</span> functions from the <span class="in">`readr`</span> package are a great way to do this. They take two arguments - the data frame you want to save and the file path where you want to save it.</span>
<span id="cb132-1734"><a href="#cb132-1734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1735"><a href="#cb132-1735" aria-hidden="true" tabindex="-1"></a>For example, let's say I want to save my contingency table of the number of mice in each cage and strain to a CSV file called <span class="in">`cage_strain_counts.csv`</span>:</span>
<span id="cb132-1736"><a href="#cb132-1736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1739"><a href="#cb132-1739" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1740"><a href="#cb132-1740" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1741"><a href="#cb132-1741" aria-hidden="true" tabindex="-1"></a><span class="co"># create the contingency table</span></span>
<span id="cb132-1742"><a href="#cb132-1742" aria-hidden="true" tabindex="-1"></a>cage_strain_counts <span class="ot">&lt;-</span> mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1743"><a href="#cb132-1743" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1744"><a href="#cb132-1744" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb132-1745"><a href="#cb132-1745" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1746"><a href="#cb132-1746" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb132-1747"><a href="#cb132-1747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1748"><a href="#cb132-1748" aria-hidden="true" tabindex="-1"></a><span class="co"># save the data to a CSV file</span></span>
<span id="cb132-1749"><a href="#cb132-1749" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(cage_strain_counts, <span class="st">"cage_strain_counts.csv"</span>)</span>
<span id="cb132-1750"><a href="#cb132-1750" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1751"><a href="#cb132-1751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1752"><a href="#cb132-1752" aria-hidden="true" tabindex="-1"></a>CSV files are particularly great because they can be easily read into other software, like Excel.</span>
<span id="cb132-1753"><a href="#cb132-1753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1754"><a href="#cb132-1754" aria-hidden="true" tabindex="-1"></a>It's also possible to use the <span class="in">`write_*()`</span> functions along with a pipe:</span>
<span id="cb132-1755"><a href="#cb132-1755" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1758"><a href="#cb132-1758" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb132-1759"><a href="#cb132-1759" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb132-1760"><a href="#cb132-1760" aria-hidden="true" tabindex="-1"></a>mousezempic_dosage_data <span class="sc">%&gt;%</span></span>
<span id="cb132-1761"><a href="#cb132-1761" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb132-1762"><a href="#cb132-1762" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb132-1763"><a href="#cb132-1763" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain)) <span class="sc">%&gt;%</span></span>
<span id="cb132-1764"><a href="#cb132-1764" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice) <span class="sc">%&gt;%</span></span>
<span id="cb132-1765"><a href="#cb132-1765" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_csv</span>(<span class="st">"cage_strain_counts.csv"</span>)</span>
<span id="cb132-1766"><a href="#cb132-1766" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb132-1767"><a href="#cb132-1767" aria-hidden="true" tabindex="-1"></a>Remember here that the first argument (the data frame to save) is passed on by the pipe, so the only argument in the brackets is the second one: the file path.</span>
<span id="cb132-1768"><a href="#cb132-1768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1769"><a href="#cb132-1769" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb132-1770"><a href="#cb132-1770" aria-hidden="true" tabindex="-1"></a>Here's what we've covered in this session:</span>
<span id="cb132-1771"><a href="#cb132-1771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1772"><a href="#cb132-1772" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The pipe operator <span class="in">`%&gt;%`</span> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</span>
<span id="cb132-1773"><a href="#cb132-1773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1774"><a href="#cb132-1774" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The basic dplyr verbs <span class="in">`filter()`</span>, <span class="in">`select()`</span>, <span class="in">`mutate()`</span>, and <span class="in">`arrange()`</span> and how they can be used to tidy and analyse data.</span>
<span id="cb132-1775"><a href="#cb132-1775" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1776"><a href="#cb132-1776" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Missing values (<span class="in">`NA`</span>) and how to remove or replace them</span>
<span id="cb132-1777"><a href="#cb132-1777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1778"><a href="#cb132-1778" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The <span class="in">`summarise()`</span> function and how it can be used to calculate summary statistics on your data, as well as the power of grouping data with the <span class="in">`.by`</span> argument.</span>
<span id="cb132-1779"><a href="#cb132-1779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1780"><a href="#cb132-1780" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Reshaping data with <span class="in">`pivot_longer()`</span> and <span class="in">`pivot_wider()`</span> to change the structure of your data frame.</span>
<span id="cb132-1781"><a href="#cb132-1781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1782"><a href="#cb132-1782" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Combining data with <span class="in">`left_join()`</span> to merge two data frames based on a common column.</span>
<span id="cb132-1783"><a href="#cb132-1783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1784"><a href="#cb132-1784" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Why does data need to be tidy anyway?"}</span>
<span id="cb132-1785"><a href="#cb132-1785" aria-hidden="true" tabindex="-1"></a>In this session, we've been focusing on making our data 'tidy': that is, structured in a consistent way that makes it easy to work with. A nice visual illustration of tidy data and its importance can be <span class="co">[</span><span class="ot">found here</span><span class="co">](https://allisonhorst.com/other-r-fun)</span>.</span>
<span id="cb132-1786"><a href="#cb132-1786" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb132-1787"><a href="#cb132-1787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1788"><a href="#cb132-1788" aria-hidden="true" tabindex="-1"></a><span class="fu">### Practice questions</span></span>
<span id="cb132-1789"><a href="#cb132-1789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1790"><a href="#cb132-1790" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>What is the purpose of the pipe operator <span class="in">`%&gt;%`</span>? Keeping this in mind, re-write the following code to use the pipe.</span>
<span id="cb132-1791"><a href="#cb132-1791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1792"><a href="#cb132-1792" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`round(mean(c(1, 2, 3, 4, 5)))`</span></span>
<span id="cb132-1793"><a href="#cb132-1793" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`print(as.character(1 + 10))`</span></span>
<span id="cb132-1794"><a href="#cb132-1794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1795"><a href="#cb132-1795" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>What would be the result of evaluating the following expressions? You don't need to know these off the top of your head, use R to help! (Hint: some expressions might give an error. Try to think about why)</span>
<span id="cb132-1796"><a href="#cb132-1796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1797"><a href="#cb132-1797" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`mousezempic_dosage_data %&gt;% filter(weight_lost_g &gt; 10)`</span></span>
<span id="cb132-1798"><a href="#cb132-1798" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`mousezempic_dosage_data %&gt;% select(tail_length_mm, weight_lost_g)`</span></span>
<span id="cb132-1799"><a href="#cb132-1799" aria-hidden="true" tabindex="-1"></a>    c.  <span class="in">`mousezempic_dosage_data %&gt;% mutate(weight_lost_kg = weight_lost_g / 1000)`</span></span>
<span id="cb132-1800"><a href="#cb132-1800" aria-hidden="true" tabindex="-1"></a>    d.  <span class="in">`mousezempic_dosage_data %&gt;% arrange(tail_length_mm)`</span></span>
<span id="cb132-1801"><a href="#cb132-1801" aria-hidden="true" tabindex="-1"></a>    e.  <span class="in">`mousezempic_dosage_data %&gt;% filter(initial_Weight_g &gt; 10) %&gt;% arrange(mouse_strain)`</span></span>
<span id="cb132-1802"><a href="#cb132-1802" aria-hidden="true" tabindex="-1"></a>    f.  <span class="in">`mousezempic_dosage_data %&gt;% relocate(mouse_strain, .after = cage_number)`</span></span>
<span id="cb132-1803"><a href="#cb132-1803" aria-hidden="true" tabindex="-1"></a>    g.  <span class="in">`mousezempic_dosage_data %&gt;% pull(weight_lost_g)`</span></span>
<span id="cb132-1804"><a href="#cb132-1804" aria-hidden="true" tabindex="-1"></a>    h.  <span class="in">`mousezempic_dosage_data %&gt;% filter(!is.na(weight_lost_g))`</span></span>
<span id="cb132-1805"><a href="#cb132-1805" aria-hidden="true" tabindex="-1"></a>    i.  <span class="in">`mousezempic_dosage_data %&gt;% replace_na(list(weight_lost_g = 0))`</span></span>
<span id="cb132-1806"><a href="#cb132-1806" aria-hidden="true" tabindex="-1"></a>    j.  <span class="in">`mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE))`</span></span>
<span id="cb132-1807"><a href="#cb132-1807" aria-hidden="true" tabindex="-1"></a>    k.  <span class="in">`mousezempic_dosage_data %&gt;% summarise(mean_weight_lost = mean(weight_lost_g, na.rm = TRUE), .by = cage_number)`</span></span>
<span id="cb132-1808"><a href="#cb132-1808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1809"><a href="#cb132-1809" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>What is a missing value in R? What are two ways to deal with missing values in a data frame?</span>
<span id="cb132-1810"><a href="#cb132-1810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1811"><a href="#cb132-1811" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Using the <span class="in">`mousezempic_dosage_data`</span> data frame, write R code to:</span>
<span id="cb132-1812"><a href="#cb132-1812" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1813"><a href="#cb132-1813" aria-hidden="true" tabindex="-1"></a>      a.  Make a data frame that shows the number of mice of each strain, in each replicate.</span>
<span id="cb132-1814"><a href="#cb132-1814" aria-hidden="true" tabindex="-1"></a>      b.  Pivot this data frame into a wide format to create a contingency table.</span>
<span id="cb132-1815"><a href="#cb132-1815" aria-hidden="true" tabindex="-1"></a>      c.  Pivot the wide data frame from (b) back into a long format.</span>
<span id="cb132-1816"><a href="#cb132-1816" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1817"><a href="#cb132-1817" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Let's say I have two data frames, <span class="in">`df1`</span> and <span class="in">`df2`</span>, that I want to join based a shared 'key' column, that is called 'key' in <span class="in">`df1`</span> and 'item_key' in <span class="in">`df2`</span>. Write R code to join these two data frames using the <span class="in">`left_join()`</span> function.</span>
<span id="cb132-1818"><a href="#cb132-1818" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1819"><a href="#cb132-1819" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb132-1820"><a href="#cb132-1820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1821"><a href="#cb132-1821" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb132-1822"><a href="#cb132-1822" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1823"><a href="#cb132-1823" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>The pipe operator <span class="in">`%&gt;%`</span> is used to chain together multiple function calls, passing the result of one function to the next. Here's how you could re-write the code to use the pipe:</span>
<span id="cb132-1824"><a href="#cb132-1824" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1825"><a href="#cb132-1825" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`c(1, 2, 3, 4, 5) %&gt;% mean() %&gt;% round()`</span></span>
<span id="cb132-1826"><a href="#cb132-1826" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`as.character(1 + 10) %&gt;% print()`</span></span>
<span id="cb132-1827"><a href="#cb132-1827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1828"><a href="#cb132-1828" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>The result of evaluating the expressions would be:</span>
<span id="cb132-1829"><a href="#cb132-1829" aria-hidden="true" tabindex="-1"></a>    a. A data frame containing only the rows where <span class="in">`weight_lost_g`</span> is greater than 10.</span>
<span id="cb132-1830"><a href="#cb132-1830" aria-hidden="true" tabindex="-1"></a>    b. A data frame containing only the <span class="in">`tail_length_mm`</span> and <span class="in">`weight_lost_g`</span> columns.</span>
<span id="cb132-1831"><a href="#cb132-1831" aria-hidden="true" tabindex="-1"></a>    c. A data frame with an additional column <span class="in">`weight_lost_kg`</span> that contains the weight lost in kilograms.</span>
<span id="cb132-1832"><a href="#cb132-1832" aria-hidden="true" tabindex="-1"></a>    d. A data frame sorted by <span class="in">`tail_length_mm`</span>, in ascending order.</span>
<span id="cb132-1833"><a href="#cb132-1833" aria-hidden="true" tabindex="-1"></a>    e. An error because <span class="in">`initial_Weight_g`</span> is not a column in the data frame.</span>
<span id="cb132-1834"><a href="#cb132-1834" aria-hidden="true" tabindex="-1"></a>    f. A data frame with the <span class="in">`mouse_strain`</span> column moved to be after the <span class="in">`cage_number`</span> column.</span>
<span id="cb132-1835"><a href="#cb132-1835" aria-hidden="true" tabindex="-1"></a>    g. A **vector** containing the values of the <span class="in">`weight_lost_g`</span> column.</span>
<span id="cb132-1836"><a href="#cb132-1836" aria-hidden="true" tabindex="-1"></a>    h. A data frame containing only the rows where <span class="in">`weight_lost_g`</span> is not <span class="in">`NA`</span>.</span>
<span id="cb132-1837"><a href="#cb132-1837" aria-hidden="true" tabindex="-1"></a>    i. A data frame with missing values in the <span class="in">`weight_lost_g`</span> column replaced with 0.</span>
<span id="cb132-1838"><a href="#cb132-1838" aria-hidden="true" tabindex="-1"></a>    j. A data frame with the mean weight lost by all mice.</span>
<span id="cb132-1839"><a href="#cb132-1839" aria-hidden="true" tabindex="-1"></a>    k. A data frame with the mean weight lost by mice in each cage.</span>
<span id="cb132-1840"><a href="#cb132-1840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1841"><a href="#cb132-1841" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>A missing value in R is represented by <span class="in">`NA`</span>. Two ways to deal with missing values in a data frame are to remove them using <span class="in">`filter(!is.na(column_name))`</span> or to replace them with a specific value using <span class="in">`replace_na(list(column_name = value))`</span>.</span>
<span id="cb132-1842"><a href="#cb132-1842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1843"><a href="#cb132-1843" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Here's how you could write R code to achieve the tasks:</span>
<span id="cb132-1844"><a href="#cb132-1844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1845"><a href="#cb132-1845" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate))`</span></span>
<span id="cb132-1846"><a href="#cb132-1846" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice)`</span></span>
<span id="cb132-1847"><a href="#cb132-1847" aria-hidden="true" tabindex="-1"></a>    c.  <span class="in">`mousezempic_dosage_data %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice) %&gt;% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")`</span></span>
<span id="cb132-1848"><a href="#cb132-1848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1849"><a href="#cb132-1849" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>To join the two data frames you could use <span class="in">`df1 %&gt;% left_join(df2, by = c("key" = "item_key"))`</span> (with pipe) or <span class="in">`left_join(df1, df2, by = c("key" = "item_key"))`</span> (without pipe).</span>
<span id="cb132-1850"><a href="#cb132-1850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-1851"><a href="#cb132-1851" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>